---
title: "LPS Mus feces 16S XieJ"
author: "Jinxin Meng"
date: "2025-02-14"
output:
  prettydoc::html_pretty:
    theme: cayman
    highlight: github
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = F, message = F, eval = F)
```

## pipeline
```{bash}
realpath raw_fq/*gz | filelist -2 | csvtk add-header -t -n sample-id,absolute-filepath > sample_list
cut -f1 sample_list | sed 1d | perl -ne 'chomp;$x=substr $_, 0, 1;print "$_\t$x\n"' | csvtk add-header -t -n sample-id,group > sample_group

qiime tools import --type 'SampleData[SequencesWithQuality]' --input-format SingleEndFastqManifestPhred33V2 --input-path sample_list --output-path seqs/seqs.qza
qiime demux summarize --i-data seqs/seqs.qza --o-visualization seqs/seqs.qzv

qiime dada2 denoise-single --i-demultiplexed-seqs seqs/seqs.qza --p-trunc-len 0 --p-trim-left 0 --o-representative-sequences dada2/rep_seqs.qza --o-table dada2/table.qza --o-denoising-stats dada2/stats.qza --p-n-threads 110
qiime feature-table summarize --i-table dada2/table.qza --o-visualization dada2/table.qzv --m-sample-metadata-file sample_group
qiime tools export --input-path dada2/table.qza --output-path dada2 && biom convert -i dada2/feature-table.biom -o dada2/table.tsv --to-tsv
qiime tools export --input-path dada2/rep_seqs.qza --output-path dada2

qiime phylogeny align-to-tree-mafft-fasttree --i-sequences dada2/rep_seqs.qza --o-alignment tree/aligned.qza --o-masked-alignment tree/masked-aligned.qza --o-tree tree/unrooted-tree.qza --o-rooted-tree tree/rooted-tree.qza --p-n-threads 100
qiime tools export --input-path tree/rooted-tree.qza --output-path tree/ && mv tree/tree.nwk tree/rooted_tree.nwk
qiime tools export --input-path tree/unrooted-tree.qza --output-path tree/ && mv tree/tree.nwk tree/unrooted_tree.nwk

qiime diversity core-metrics --i-table dada2/table.qza --p-sampling-depth 17100 --m-metadata-file sample_group --p-n-jobs 30 --o-rarefied-table div/table_rarefied.qza --output-dir div/core
qiime tools export --input-path div/table_rarefied.qza --output-path div/ && biom convert -i div/feature-table.biom -o div/table_rarefied.tsv --to-tsv

export TMPDIR=/data/mengjx/tmp/qiime2/
qiime feature-classifier classify-sklearn --p-n-jobs 56 --i-classifier /data/database/silva/27F_1492R/q2-138-99-27F-1492R-classifier.qza --i-reads dada2/rep_seqs.qza --o-classification taxa/taxa.qza
```

## 16S amplicon analysis

```{r}
#### info ####
# group_level <- c('survive', 'dead_24h','dead_48h')
group_level <- c('survive', 'dead')
# group_color <- structure(c('#e64b35','#4dbbd5','#00a087'), names = group_level)
group_color <- structure(c('#e64b35','#4dbbd5'), names = group_level)

group <- read.delim('../data/group.tsv') %>% 
  select(sample, group = group2)
  
profile <- read.delim('../data/table.tsv', row.names = 1, ) %>% 
  t %>% 
  rrarefy(15086) %>% 
  t %>% 
  data.frame() %>% 
  profile_transRA()
taxa <- read.delim('../data/taxonomy.tsv')
tr <- treeio::read.tree('../data/rooted_tree.nwk')

#### alpha ####
source('/code/R_func/diversity.R')

map(c('pd', 'shannon', 'richness'), ~
  calcu_alpha(profile, method = .x, tree = tr) %>% 
    left_join(group, by = 'sample') %>% 
    mutate(group = factor(group, group_level)) %>% 
    ggboxplot('group', 'value', fill = 'group', palette = group_color, 
            legend = 'none', outlier.size = 1, xlab = '', x.text.angle = 45, 
            ylab = paste0(.x, ' index'), add = 'jitter', outlier.shape = F) +
    stat_compare_means(comparisons = combn(group_level, 2, simplify = F), 
                       step.increase = .1, vjust = .1, tip.length = .02,
                       size = 3) +
    theme(aspect.ratio = 1.5,
          axis.ticks.length = unit(2, 'mm')) ) %>% 
  cowplot::plot_grid(plotlist = ., nrow = 1, align = 'v')
ggsave('taxa.div.alpha.pdf', width = 8, height = 4)

#### PCoA ####
source('/code/R_func/plot_PCoA.R')
source('/code/R_func/plot_NMDS.R')

plot_PCoA(profile, group, group_level = group_level, group_color = group_color, 
          add_group_label = T, label_size = 3, ellipse_level = .9, 
          show_legend = F, show_grid = T, show_line = F, conf_type = 'encircle')
ggsave('taxa.div.PCoA.group.pdf', width = 5, height = 4)

group_x <- read.delim('../data/group.tsv') %>% 
  select(sample, group = sex)
plot_PCoA(profile, group_x, add_group_label = T, label_size = 3, ellipse_level = .9, 
          show_legend = F, show_grid = T, show_line = F, conf_type = 'encircle')
ggsave('taxa.div.PCoA.sex.pdf', width = 5, height = 4)

data <- calcu_PCoA(profile, group = group, prefix = 'PC')

left_join(data$point, group, by = 'sample') %>% 
  ggscatter('PC1', 'PC2', color = 'group', shape = 'sex', palette = group_color,
            legend = 'right', xlab = data$eig_lab[1], ylab = data$eig_lab[2],
            size = 2) +
  theme(aspect.ratio = 1)
ggsave('taxa.div.PCoA.pdf', width = 5, height = 3.5)

#### compos ####
source('F:/code/R_func/taxa.R')

colors <- c('#8dd3c7','#ffffb3','#80b1d3','#b3de69','#fdb462','#bc80bd',
            '#fb8072','#ffed6f','#fccde5','#bebada','#e5c494','#ccebc5')

plot_compos_multiple(profile, taxa, group, display = 'group', top_n = 12, 
                     taxa_color = colors)
ggsave('taxa.compos.multiple.pdf', width = 17, height = 10)

plot_compos_multiple(profile, taxa, group, display = 'sample', top_n = 12, 
                     taxa_color = colors, 
                     sample_level = arrange(group, desc(group)) %>% pull(sample))
ggsave('taxa.compos.multiple2.pdf', width = 25, height = 10)

plot_compos(profile, taxa, group, to = 'species', top_n = 15, display = 'sample', 
            sample_level = arrange(group, desc(group)) %>% pull(sample))

taxa_trans(profile, taxa, to = 'species', top_n = 40) %>% 
  profile_transLOG10() %>% 
  plot_taxa_boxplot(group) %>% 
  cowplot::plot_grid(plotlist = ., nrow = 4, align = 'v')
ggsave('taxa.diff.species.pdf', width = 21, height = 12)

# diff data
# comparisons <- list(c('dead_24h', 'survive'),
#                     c('dead_48h', 'survive'))
# map(c('phylum', 'family', 'genus', 'species'), ~ {
#   .data <- taxa_trans(profile, taxa, group, to = .x, out_all = T) %>%
#     profile_transRA()
#   map_dfr(comparisons, ~ 
#             difference_analysis(.data, group, comparison = .x, method = 'wilcox', 
#                                 add_enriched = T, prefix = c('group1', 'group2'))) } ) %>% 
#   set_names(c('phylum', 'family', 'genus', 'species')) %>% 
#   write.xlsx('compos.diffs.xlsx')

#### massalin2 ####
library(Maaslin2)

group <- read.delim('../data/group.tsv') %>% 
  column_to_rownames('sample')

data <- taxa_trans(profile, taxa, to = 'genus', out_all = T, transRA = T)
Maaslin2(data, group, output = 'masslin2_genus_out', analysis_method = "LM",
         normalization = 'NONE', transform = 'LOG', correction = 'bonferroni', 
         fixed_effects = c('group', 'sex'), save_models = T, save_scatter = T,
         reference = 'group,survive')

data <- taxa_trans(profile, taxa, to = 'species', out_all = T, transRA = T)
Maaslin2(data, group, output = 'masslin2_species_out', analysis_method = "LM",
         normalization = 'NONE', transform = 'LOG', correction = 'bonferroni', 
         fixed_effects = c('group', 'sex'), save_models = T, save_scatter = T,
         reference = 'group,survive')

#### func alpha ####
source('/code/R_func/diversity.R')

group <- read.delim('../data/group.tsv')
profile <- read.delim('../data/pred_metagenome_unstrat.tsv', row.names = 1, ) %>% 
  profile_transRA()

map(c('shannon', 'richness'), ~
      calcu_alpha(profile, method = .x) %>% 
      left_join(group, by = 'sample') %>% 
      mutate(group = factor(group, group_level)) %>% 
      ggboxplot('group', 'value', fill = 'group', palette = group_color, 
                legend = 'none', outlier.size = 1, xlab = '', x.text.angle = 45, 
                ylab = paste0(.x, ' index'), add = 'jitter', outlier.shape = F) +
      stat_compare_means(comparisons = combn(group_level, 2, simplify = F), 
                         step.increase = .1, vjust = .1, tip.length = .02,
                         size = 3) +
      theme(aspect.ratio = 1.5,
            axis.ticks.length = unit(2, 'mm')) ) %>% 
  cowplot::plot_grid(plotlist = ., nrow = 1, align = 'v')
ggsave('func.div.alpha.pdf', width = 8, height = 4)

#### func PCoA ####
source('/code/R_func/plot_PCoA.R')

plot_PCoA(profile, group, group_level = group_level, group_color = group_color, 
          add_group_label = T, label_size = 3, ellipse_level = .9, 
          show_legend = F, show_grid = T, show_line = F, conf_type = 'encircle')
ggsave('func.div.PCoA.group.pdf', width = 5, height = 4)

group_x <- read.delim('../data/group.tsv') %>% 
  select(sample, group = sex)
plot_PCoA(profile, group_x, add_group_label = T, label_size = 3, ellipse_level = .9, 
          show_legend = F, show_grid = T, show_line = F, conf_type = 'encircle')
ggsave('func.div.PCoA.sex.pdf', width = 5, height = 4)

data <- calcu_PCoA(profile, group = group, prefix = 'PC')

left_join(data$point, group, by = 'sample') %>% 
  ggscatter('PC1', 'PC2', color = 'group', shape = 'sex', palette = group_color,
            legend = 'right', xlab = data$eig_lab[1], ylab = data$eig_lab[2],
            size = 2) +
  theme(aspect.ratio = 1)
ggsave('func.div.PCoA.pdf', width = 5, height = 3.5)

#### func massalin2 ####
source('F:/code/R_func/microbiome_utilities.R')

group <- read.delim('../data/group.tsv') %>% 
  column_to_rownames('sample')

data <- profile_KEGG_convert(profile, to = 'C', transRA = T, rownames_fmt = 'both')
Maaslin2(data, group, output = 'func.masslin2_lvC_out', analysis_method = "LM",
         normalization = 'NONE', transform = 'LOG', correction = 'bonferroni', 
         fixed_effects = c('group', 'sex'), save_models = T, save_scatter = T,
         reference = 'group,survive')

data <- read.delim('../data/path_abun_unstrat.tsv', row.names = 1)
Maaslin2(data, group, output = 'func.masslin2_path_out', analysis_method = "LM",
         normalization = 'NONE', transform = 'LOG', correction = 'bonferroni', 
         fixed_effects = c('group', 'sex'), save_models = T, save_scatter = T,
         reference = 'group,survive')
```
