---
title: "MGX 16S on PDCoV Sus feces ZhangYQ"
author: "Jinxin Meng"
date: "2025-03-25"
output:
  prettydoc::html_pretty:
    theme: cayman
    highlight: github
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = F, message = F, eval = F)
```

## pipeline

```{bash}
source activate qiime2-2024.10
qiime tools import --type 'SampleData[PairedEndSequencesWithQuality]' --input-format PairedEndFastqManifestPhred33V2 --input-path sample_list --output-path seqs/seqs.qza
qiime demux summarize --i-data seqs/seqs.qza --o-visualization seqs/seqs.qzv

qiime dada2 denoise-paired --i-demultiplexed-seqs seqs/seqs.qza --p-trim-left-f 26 --p-trim-left-r 26 --p-trunc-len-f 296 --p-trunc-len-r 295 \
    --o-representative-sequences dada2/rep_seqs.qza --o-table dada2/table.qza --o-denoising-stats dada2/stats.qza --p-n-threads 100
qiime tools export --input-path dada2/table.qza --output-path dada2 && biom convert -i dada2/feature-table.biom -o dada2/table.tsv --to-tsv

qiime feature-table filter-features --i-table dada2/table.qza --p-min-frequency 1 --p-min-samples 2 --o-filtered-table dada2/table_f.qza
qiime tools export --input-path dada2/table_f.qza --output-path dada2/ && biom convert -i dada2/feature-table.biom -o dada2/table_f.tsv --to-tsv
qiime feature-table summarize --i-table dada2/table_f.qza --o-visualization dada2/table_f.qzv --m-sample-metadata-file sample_group
qiime feature-table filter-seqs --i-data dada2/rep_seqs.qza --i-table dada2/table_f.qza --o-filtered-data dada2/rep_seqs_f.qza
qiime tools export --input-path dada2/rep_seqs_f.qza --output-path dada2/

qiime phylogeny align-to-tree-mafft-fasttree --i-sequences dada2/rep_seqs_f.qza --o-alignment tree/aligned.qza \
    --o-masked-alignment tree/masked-aligned.qza --o-tree tree/unrooted-tree.qza --o-rooted-tree tree/rooted-tree.qza --p-n-threads 56
qiime tools export --input-path tree/rooted-tree.qza  --output-path tree/ && mv tree/tree.nwk tree/rooted_tree.nwk

qiime diversity core-metrics --i-table dada2/table_f.qza --p-sampling-depth 27279 --m-metadata-file sample_group --p-n-jobs 56 \
    --o-rarefied-table div/table_rarefied.qza --output-dir div/core_div
qiime tools export --input-path div/table_rarefied.qza --output-path div/ && biom convert -i div/feature-table.biom -o div/table_rarefied.tsv --to-tsv

qiime feature-classifier classify-sklearn --i-classifier /data/database/silva/338F_806R/silva-138-99-338F-806R-classifier.qza --i-reads dada2/rep_seqs_f.qza \
    --o-classification tax/taxonomy.qza

picrust2_pipeline_oldIMG.py -s dada2/dna-sequences.fasta -i div/feature-table.biom -o picrust2 -p 100
```

## 16S amplicon downstream analysis 

```{R}
#### metadata 20250325 ####
grp <- c('Mock','PDCoV')
grp_col <- structure(c('#a0a0a4','#d7b0b0'), names = grp)

group <- read.delim("../data/group.tsv")
profile <- read.delim("../data/profile.tsv", row.names = 1, check.names = F) %>% 
  profile_transRA()
taxa <- read.delim("../data/taxa.tsv")

#### alpha 20250325 ####
source("/code/R_func/diversity.R")

map(c('pd', 'shannon', 'richness', 'ace'), ~
  calcu_alpha(profile, method = .x, tree = tr) %>% 
    left_join(group, by = "sample") %>% 
    mutate(group = factor(group, grp)) %>% 
    ggboxplot("group", "value", fill = "group", palette = grp_col, 
            legend = "none", outlier.size = 1, xlab = "", 
            ylab = paste0(.x, ' index')) +
    stat_compare_means(comparisons = list(grp), label = "p.signif", 
                       step.increase = .06, vjust = .7, tip.length = .02) +
    theme(aspect.ratio = 2) ) %>% 
  cowplot::plot_grid(plotlist = ., nrow = 1, align = 'v')
ggsave("div.alpha.pdf", width = 8, height = 3)

#### PCoA 20250325 ####
source("/code/R_func/plot_PCoA.R")
source("/code/R_func/plot_NMDS.R")
plot_NMDS(profile, group, group_order = grp, group_color = grp_col)

plot_PCoA(profile, group, group_order = grp, group_color = grp_col, 
          add_group_label = T, lab_size = 3, ellipse_level = .9, 
          show_legend = F, show_grid = T, show_line = F)
ggsave("div.PCoA.pdf", width = 4, height = 3)

#### compos 20250325 ####
source("F:/code/R_func/taxa.R")

colors <- c("#8dd3c7","#ffffb3","#80b1d3","#b3de69","#fdb462","#bc80bd","#fb8072",
            "#ffed6f","#fccde5","#bebada","#e5c494","#ccebc5","#d9d9d9")

# phylum
data <- taxa_trans(profile, taxa, group, to = 'phylum', out_all = T) %>% 
  filter(rownames(.) != 'p__Unknown')
plot_compos_manual(data, group, taxa_color = colors, group_order = grp,
                   plot_title = 'Composition analysis at phylum level') +
  theme(aspect.ratio = 2.4)
ggsave(file = "compos.phylum.pdf", width = 5, height = 4)

# family
data <- taxa_trans(profile, taxa, group, to = 'family', top_n = 12,
                   other_name = 'f__Other') %>% 
  filter(rownames(.) != 'f__Unknown')
plot_compos_manual(data, group, taxa_color = colors, group_order = grp, out_all = T,
                   plot_title = 'Composition analysis at family level') +
  theme(aspect.ratio = 2.4)
ggsave(file = "compos.family.pdf", width = 5, height = 4)

# genus
data <- taxa_trans(profile, taxa, group, to = 'genus', top_n = 12,
                   other_name = 'g__Other') %>% 
  filter(rownames(.) != 'g__Unknown')
plot_compos_manual(data, group, taxa_color = colors, group_order = grp, out_all = T,
                   plot_title = 'Composition analysis at genus level') +
  theme(aspect.ratio = 2.4)
ggsave(file = "compos.genus.pdf", width = 5, height = 4)

# abundance
data <- map(c('phylum', 'family', 'genus'), ~
              taxa_trans(profile, taxa, group, to = .x, out_all = T, transRA = T) %>% 
              profile_smp2grp(group) %>% 
              rownames_to_column('name')) %>% 
  set_names(c('phylum', 'family', 'genus'))
write.xlsx(data, 'compos.real_ab.xlsx')

# diff abundance
# phylum
data <- taxa_trans(profile, taxa, group, to = 'phylum', out_all = T)
plot_taxa_boxplot(data, group, group_order = grp, group_color = grp_col, 
                  show_legend = F, legend_title = '', aspect_ratio = 2) %>% 
  cowplot::plot_grid(plotlist = ., nrow = 3, align = 'v')
ggsave("compos.phylum.diff.pdf", width = 6, height = 9)

# family
data <- taxa_trans(profile, taxa, group, to = 'family', out_all = T)
plot_taxa_boxplot(data, group, group_order = grp, group_color = grp_col, 
                  show_legend = F, legend_title = '', aspect_ratio = 2) %>% 
  cowplot::plot_grid(plotlist = ., nrow = 7, align = 'v')
ggsave("compos.family.diff.pdf", width = 18, height = 21)

# genus
data <- taxa_trans(profile, taxa, group, to = 'genus', out_all = T)
plot_taxa_boxplot(data, group, group_order = grp, group_color = grp_col, 
                  show_legend = F, legend_title = '', aspect_ratio = 2) %>% 
  cowplot::plot_grid(plotlist = ., nrow = 11, align = 'v')
ggsave("compos.genus.diff.pdf", width = 24, height = 33)

# diff data
map(c('phylum', 'family', 'genus', 'species'), ~
  taxa_trans(profile, taxa, group, to = .x, out_all = T) %>%
    profile_transRA() %>% rownames_to_column('name') %>% 
    gather('sample', 'value', -name) %>% 
    left_join(group, by = 'sample') %>% 
    group_by(name) %>% 
    wilcox_test(value ~ group, comparisons = c('PDCoV', 'Mock'), 
                p.adjust.method = 'BH', detailed = T) %>% 
    select(-.y.) ) %>% 
  set_names(c('phylum', 'family', 'genus', 'species')) %>% 
  write.xlsx('compos.diffs.xlsx')

#### lefse 20250325 ####
library(microeco)
library(magrittr)

set.seed(2025)
group <- read.delim("../data/group.tsv") %>% 
  column_to_rownames('sample') %>% 
  mutate(group = factor(group))
profile <- read.delim("../data/profile.tsv", row.names = 1) %>% 
  profile_transRA()
taxa <- read.delim("../data/taxa.tsv") %>% 
  column_to_rownames('name') %>% 
  tidy_taxonomy() %>% 
  rename_all(~ str_to_title(.x))
tr <- treeio::read.tree('../data/tree.nwk') # 有根树

dataset <- microtable$new(sample_table = group, otu_table = profile, 
                          tax_table = taxa, phylo_tree = tr)

trans_abund$new(dataset)

lefse <- trans_diff$new(dataset = dataset, method = 'lefse', group = 'group')

lefse$plot_diff_bar(use_number = 1:30, width = 0.6, group_order = grp, 
                    color_values = grp_col) + 
  theme_pubr()
ggsave('lefse.barplot.pdf', width = 5, height = 6)

lefse$plot_diff_cladogram(use_taxa_num = 160, use_feature_num = 40, 
                          clade_label_level = 6, group_order = grp, 
                          color = grp_col)
ggsave('lefse.cladogram.pdf', width = 13, height = 6)

write.xlsx(lefse$res_diff, 'lefse.result.xlsx')

#### lefse 20250411 ####
library(microeco)
library(magrittr)

set.seed(2025)
group <- read.delim("../data/group.tsv") %>% 
  column_to_rownames('sample') %>% 
  mutate(group = factor(group))
profile <- read.delim("../data/profile.tsv", row.names = 1) %>% 
  profile_transRA()
taxa <- read.delim("../data/taxa.tsv") %>% 
  column_to_rownames('name') %>% 
  tidy_taxonomy() %>% 
  rename_all(~ str_to_title(.x))
tr <- treeio::read.tree('../data/tree.nwk') # 有根树

dataset <- microtable$new(sample_table = group, otu_table = profile, 
                          tax_table = taxa, phylo_tree = tr)

trans_abund$new(dataset)

lefse <- trans_diff$new(dataset = dataset, method = 'lefse', group = 'group',
                        alpha = 1)

write.xlsx(lefse$res_diff, 'lefse.result.v2.xlsx')

#### path 20250325 ####
source("/code/R_func/difference_analysis.R")

group <- read.delim("../data/group.tsv")
profile <- read.delim("../data/path_abun_unstrat.tsv.gz", row.names = 1)

data <- profile %>% 
  rownames_to_column('name') %>% 
  gather('sample', 'value', -name) %>% 
  left_join(group, by = 'sample') %>% 
  group_by(name) %>% 
  wilcox_test(value ~ group, comparisons = c('PDCoV', 'Mock'), 
              p.adjust.method = 'BH', detailed = T) %>% 
      select(-.y.)
write.xlsx(data, 'path.diffs.v1.xlsx')

data <- difference_analysis(profile, group, comparison = c('PDCoV', 'Mock'))
write.xlsx(data, 'path.diffs.v2.xlsx')

#### KO 20250325 ####
group <- read.delim("../data/group.tsv")

profile <- read.delim("../data/ko_metagenome_unstrat.tsv.gz", row.names = 1) %>% 
  profile_KEGG_convert(to = 'C')

ko_info <- read.delim('/database/KEGG/v20230401/KO_level_A_B_C_D_Description') %>% 
  select(lvC, lvCdes) %>% 
  distinct()

data <- profile %>% 
  rownames_to_column('name') %>% 
  gather('sample', 'value', -name) %>% 
  left_join(group, by = 'sample') %>% 
  group_by(name) %>% 
  wilcox_test(value ~ group, comparisons = c('PDCoV', 'Mock'), 
              p.adjust.method = 'BH', detailed = T) %>% 
  select(-.y.) %>% 
  left_join(ko_info, by = c('name' = 'lvC'))
write.xlsx(data, 'koC.diffs.v1.xlsx')

data <- difference_analysis(profile, group, comparison = c('PDCoV', 'Mock')) %>% 
  left_join(ko_info, by = c('name' = 'lvC'))
write.xlsx(data, 'koC.diffs.v2.xlsx')

#### correlation 20250330 ####
# virus_load / fecal virus_load
metadata <- cbind(
  read.delim('../../MGX_WGS_PDCoV_Sus_feces/data/metadata_virus_load.tsv') %>% 
    filter(group %in% c('Mock', 'PDCoV')) %>% 
    group_by(group) %>% 
    slice_tail(n = 6) %>% 
    ungroup() %>% 
    select(-group, -type), 
  read.delim('../../MGX_WGS_PDCoV_Sus_feces/data/metadata_fecal_virus_load.tsv') %>% 
    filter(group %in% c('Mock', 'PDCoV')) %>% 
    group_by(group) %>% 
    slice_tail(n = 6) %>% 
    ungroup() %>% 
    select(feces = D3)) %>% 
  select(-MLN)

data <- taxa_trans(profile, taxa, to = 'genus', out_all = T) %>% 
  profile_filter(group, by_group = T, all_group = T, min_n = 1)

corr <- psych::corr.test(t(data), metadata)
r_data <- data.frame(t(corr$r), check.names = F)
p_data <- data.frame(t(corr$p), check.names = F)

r_data[abs(r_data) < 0.1] <- 0
p_data[p_data >= 0.05] <- -1
p_data[p_data < 0.05 & p_data >= 0] <- 1
p_data[p_data == -1] <- 0
cor_out <- r_data * p_data
cor_out <- cor_out[apply(cor_out, 1, \(x) sum(x != 0) >= 1), 
                   apply(cor_out, 2, \(x) sum(x != 0) >= 1)]

plot_data <- data.frame(t(corr$r), check.names = F)[rownames(cor_out), colnames(cor_out)]

plot_label <- data.frame(t(corr$p), check.names = F)[rownames(cor_out), colnames(cor_out)] %>%
  apply(2, \(x) ifelse(x < 0.001, "+", ifelse(x < 0.01, "**", ifelse(x < 0.05, "*", ""))))

pdf("cor_spearman.pdf", height = 4, width = 4)
pheatmap(plot_data,
         cluster_rows = T, cluster_cols = T,
         border = "#ffffff", border_gp = gpar(col = "#000000"),
         show_rownames = T, show_colnames = T,
         cellwidth = 10,cellheight = 10,
         treeheight_col = 0, treeheight_row = 0,
         fontsize_col = 8, fontsize_row = 8,
         display_numbers = plot_label,
         fontface_col = "italic", angle_col = "45")
dev.off()

r_data <- data.frame(t(corr$r), check.names = F) %>% 
  rownames_to_column('name') %>% 
  gather('taxa', 'rval', -name)
p_data <- data.frame(t(corr$p), check.names = F) %>% 
  rownames_to_column('name') %>% 
  gather('taxa', 'pval', -name)

plot_data <- cbind(r_data, select(p_data, pval)) %>% 
  filter(pval < 0.05, taxa != 'g__Unknown')
write.xlsx(plot_data, 'cor_spearman.xlsx')

# graph
edges <- select(plot_data, from = name, to = taxa, rval) %>% 
  mutate(direct = ifelse(rval < 0, 'neg', 'pos'))
nodes <- data.frame(node = unique(c(edges$from, edges$to))) %>% 
  mutate(type = ifelse(grepl('g__', node), 'taxa', 'virus'),
         phylum = ifelse(type == 'taxa', taxa$phylum[match(node, taxa$genus)], ''))

g <- tbl_graph(nodes = nodes, edges = edges)

colors <- c("#8dd3c7","#bebada","#fb8072","#80b1d3",
            "#fdb462","#b3de69","#fccde5","#bc80bd","#ffed6f",
            '#1b9e77','#d95f02','#7570b3','#e7298a')

ggraph(g, layout = 'linear', circular = T) +
  geom_edge_arc(aes(edge_width = rval, edge_linetype = direct, 
                    edge_colour = direct), strength = .2) +
  scale_edge_linetype_manual(values = c(2, 1)) +
  scale_edge_width(range = c(.3, 1.2)) +
  scale_edge_color_manual(values = c("#7fc97f","#fdc086")) +
  geom_node_point(aes(color = phylum, shape = type), size = 3) +
  scale_shape_manual(values = c(16, 15)) +
  scale_color_manual(values = colors) +
  geom_node_text(aes(label = node), size = 2) +
  theme_void() +
  theme(aspect.ratio = 1)
ggsave('cor_network.pdf', width = 5, height = 4)

#### correlation 20250411 ####
# virus_load / fecal virus_load
metadata <- cbind(
  read.delim('../../MGX_WGS_PDCoV_Sus_feces/data/metadata_virus_load.tsv') %>% 
    filter(group %in% c('Mock', 'PDCoV')) %>% 
    group_by(group) %>% 
    slice_tail(n = 6) %>% 
    ungroup() %>% 
    select(-group, -type), 
  read.delim('../../MGX_WGS_PDCoV_Sus_feces/data/metadata_fecal_virus_load.tsv') %>% 
    filter(group %in% c('Mock', 'PDCoV')) %>% 
    group_by(group) %>% 
    slice_tail(n = 6) %>% 
    ungroup() %>% 
    select(feces = D3))

data <- taxa_trans(profile, taxa, to = 'genus', out_all = T) %>% 
  profile_top_n(n = 30)

corr <- psych::corr.test(t(data), metadata)
r_data <- data.frame(t(corr$r), check.names = F)
p_data <- data.frame(t(corr$p), check.names = F)

r_data[abs(r_data) < 0.1] <- 0
p_data[p_data >= 0.05] <- -1
p_data[p_data < 0.05 & p_data >= 0] <- 1
p_data[p_data == -1] <- 0
cor_out <- r_data * p_data
cor_out <- cor_out[apply(cor_out, 1, \(x) sum(x != 0) >= 1), 
                   apply(cor_out, 2, \(x) sum(x != 0) >= 1)]

plot_data <- data.frame(t(corr$r), check.names = F)[rownames(cor_out), colnames(cor_out)]

plot_label <- data.frame(t(corr$p), check.names = F)[rownames(cor_out), colnames(cor_out)] %>%
  apply(2, \(x) ifelse(x < 0.001, "+", ifelse(x < 0.01, "**", ifelse(x < 0.05, "*", ""))))

pdf("cor_spearman.v2.pdf", height = 4, width = 4)
pheatmap(plot_data,
         cluster_rows = T, cluster_cols = T,
         border = "#ffffff", border_gp = gpar(col = "#000000"),
         show_rownames = T, show_colnames = T,
         cellwidth = 10,cellheight = 10,
         treeheight_col = 0, treeheight_row = 0,
         fontsize_col = 8, fontsize_row = 8,
         display_numbers = plot_label,
         fontface_col = "italic", angle_col = "45")
dev.off()

r_data <- data.frame(t(corr$r), check.names = F) %>% 
  rownames_to_column('name') %>% 
  gather('taxa', 'rval', -name)
p_data <- data.frame(t(corr$p), check.names = F) %>% 
  rownames_to_column('name') %>% 
  gather('taxa', 'pval', -name)

plot_data <- cbind(r_data, select(p_data, pval)) %>% 
  filter(taxa != 'g__Unknown')
write.xlsx(plot_data, 'cor_spearman.v2.xlsx')

# graph
edges <- select(plot_data, from = name, to = taxa, rval) %>% 
  mutate(direct = ifelse(rval < 0, 'neg', 'pos'),
         rval = abs(rval))
nodes <- data.frame(node = unique(c(edges$from, edges$to))) %>% 
  mutate(type = ifelse(grepl('g__', node), 'taxa', 'virus'),
         phylum = ifelse(type == 'taxa', taxa$phylum[match(node, taxa$genus)], ''))

g <- tbl_graph(nodes = nodes, edges = edges)

colors <- c("#8dd3c7","#bebada","#fb8072","#80b1d3",
            "#fdb462","#b3de69","#fccde5","#bc80bd","#ffed6f",
            '#1b9e77','#d95f02','#7570b3','#e7298a')

ggraph(g, layout = 'linear', circular = T) +
  geom_edge_arc(aes(edge_width = rval, edge_linetype = direct, 
                    edge_colour = direct), strength = .2) +
  scale_edge_linetype_manual(values = c(2, 1)) +
  scale_edge_width(range = c(.5, .8)) +
  scale_edge_color_manual(values = c("#7fc97f","#fdc086")) +
  geom_node_point(aes(color = phylum, shape = type), size = 3) +
  scale_shape_manual(values = c(16, 15)) +
  scale_color_manual(values = colors) +
  geom_node_text(aes(label = node), size = 2) +
  theme_void() +
  theme(aspect.ratio = 1)
ggsave('cor_network.v2.pdf', width = 5, height = 4)
```
