<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Jinxin Meng" />

<meta name="date" content="2024-09-01" />

<title>Pig gut metagenome collection and mining microbial genetical resource (No-published)</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 

a.sourceLine {
pointer-events: auto;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">@font-face{font-family:"Open Sans";font-style:normal;font-weight:400;src:local("Open Sans"),local("OpenSans"),url(data:font/woff;base64,d09GRgABAAAAAE8YABIAAAAAhWwAAQABAAAAAAAAAAAAAAAAAAAAAAAAAABHREVGAAABlAAAABYAAAAWABAA3UdQT1MAAAGsAAAADAAAAAwAFQAKR1NVQgAAAbgAAABZAAAAdN3O3ptPUy8yAAACFAAAAF8AAABgoT6eyWNtYXAAAAJ0AAAAmAAAAMyvDbOdY3Z0IAAAAwwAAABZAAAAog9NGKRmcGdtAAADaAAABJsAAAe0fmG2EWdhc3AAAAgEAAAAEAAAABAAFQAjZ2x5ZgAACBQAADWFAABReBn1yj5oZWFkAAA9nAAAADYAAAA293bipmhoZWEAAD3UAAAAHwAAACQNzAapaG10eAAAPfQAAAIIAAADbLTLWYhrZXJuAAA//AAAChcAAB6Qo+uk42xvY2EAAEoUAAABuQAAAbz3ewp/bWF4cAAAS9AAAAAgAAAAIAJ2AgpuYW1lAABL8AAAAKwAAAEyFNwvSnBvc3QAAEycAAABhgAAAiiYDmoRcHJlcAAATiQAAADyAAABCUO3lqQAAQAAAAwAAAAAAAAAAgABAAAA3AABAAAAAQAAAAoACgAKAAB4AR3HNcJBAQDA8d+rLzDatEXOrqDd4S2ayUX1beTyDwEyyrqCbXrY+xPD8ylAsF0tUn/4nlj89Z9A7+tETl5RXdNNZGDm+vXYXWjgLDRzEhoLBAYv0/0NHAAAAHgBY2Bm2cY4gYGVgYN1FqsxAwOjPIRmvsiQxviRg4mJm42NmZWFiYnlAQPTewcGhWgGBgYNBiAwdAx2ZgAK/P/LJv9PhKGFo5cpQoGBcT5IjsWDdRuQUmBgBgD40BA5AHgBY2BgYGRgBmIGBh4GFoYDQFqHQYGBBcjzYPBkqGM4zXCe4T+jIWMw0zGmW0x3FEQUpBTkFJQU1BSsFFwUShTWKAn9/w/UpQBU7cWwgOEMwwWg6iCoamEFCQUZsGpLhOr/jxn6/z/6f5CB9//e/z3/c/7++vv877MHGx6sfbDmwcoHyx5MedD9IOGByr39QHeRAABARzfieAFjE2EQZ/Bj3QYkS1m3sZ5lQAEsHgwiDBMZGP6/AfEQ5D8REAnUJfxnyv+3/1r/v/q3Eigi8W8PA1mAA0J1MzQy3GWYwdDP0Mcwk6GDoZGRn6ELAE09H/8AAAB4AXVUR3fbxhPfhRqr/6Cr3h8pi4wpN9K9V4QEYCrq7b2F0gC1R+XkS3rjKWXlfJeBfaF88jH1M6TfoqNzdWaXxZ0NM7/ftJ2ZpXfzzeVILi0uzM/NzkxPTU68Md64GQZ+vfa6d+P6tatXLl+6eOH8uVMnTxyvVg4fGisfhNfcV0f3luz/7Srmc9nMyPDQ4IDFWUUgjwMcKItSmEAASaNaEcFo069WAghjFIlAegyOQaNhIEhQxALHEqIeg2P0yHLjKUuvY+n1LbktrrKrOgUI/MUH0ebLc5Lk73yIBO4YeUrL5GGUIimuSx6mKl2tCDD8oKmCmGrkaT5Xh/p6rlphaS5PYp4kPAy3Un74OjeCdTi4nFosU6Qg+qRBsoazczLwHdeNqpVx3AW+oVjdhMThOo6YkGJTl862RFq5r263bbYSHyuswVrylsSBhHzVQKDU11g6hkfAxyOf/DVKJ1/HCvgBHtNRJ+b7eSYepeQ4VLZBqAeMjgM7/zyJJF1kuGw/YFpEq458Xrr65YTUa6VCEKGKVdJ+2FoBYYNKCwV1K6B2s1mJnPB7Ww6GtyO04ya/HHWPHs5P4J65NyVa5VA0E0LocwPci45b6tvMvohm1BYc1h12Xd2GrbbHVkjB1pzs6IKtOHeYd+JYhFasmfs9Zt+SZlo9pu8eg0utWZAKB8vjaxBQx7cSbK3Qdr2nBwM27vrXcUHtLolLJyJjK3CAbDcFDo3hsPZ63IH2RrsoWyskdB47jiKitFtcAgqj4wQQxN3PB81RCiCo0Y1jnUVYlOj5JHhJd2JBevIEeSQxDWzTN8PEE3AL90KtP11dVrC5II1L1w331pHFq10vPBGYeyUCFRvB7PAEzMltdubhb+lZ4dw9w86yyNfG++u0ZWOBkmsb+GrsrKGIN4R0XPQimnAEcj3CI6ZDR35zzHJEZlcW5cQCTMwty4umkB5B4ajHwVNhQDqdMLSAmClnhLScgYgMbQJESALUrtIvjpQz9LVxuIPSiYgQkjusZ01l4BERrPtdO9KfDErKQLne6EUbJlXHqTccNzL163tuES26ickjo5va6FIkCyIyaFEYA+lejuqlFxLWIYKmQG9W0tlMe0yXu80wPe/OavEJrd8srSFziSal30wMj5H2mH7T6H218RQ93qOFysDEgtLBoRuQUeXjyPQKexdLjoa4vtAQJiBsEXYutEo9T1/m5mUdBMbXFCzIq8Z6Yl5+7nyic+1mE3xisVatpBarpcC/mUs9/s3Csty2GRPfLMo7FrfqcS1KDxIntwVjnkEtjRJoFKEVHWmelIyxd7Y9xlqGHTSA0VfbnBks08M4W21bHczuJBrTiYixiBnsMF7PepCwTAdrGcy8UqZb5uWGvIyX9QpW0XJSrqE7hNzjjGU5u1vgRe6k5DVv4DZvpVnP6Vi0yMKLOhUvPUq9tCzvFhi5mV9KVNMvWpfRJg1bggjEml6Uz6KmiiN92dh+Gg19OHK4TmOC61TIcAFzsF7DPNQ0fkPjNzr4sMZHaEX5fk7uLZr9LHK9AW9KF2wU///BUfaOnlREfyrK/rv6Hyn3ISkAAAEAAwAIAAoADQAH//8AD3gBhXwHfFRV1vg5974yvZdMQspkSIYkQkgmhdAyIIQQWsSADCLSpajUiMgiAkuJNGmhKyJGDCyybCiyiGBHRGQtyLIuf2UX19UPy7oWyFz+972ZBxOE72N+L2+Yd+be0+5p99wBAscBBIN4ACjI4D4oUJEIVAbIL8wPYX4oP1TQ3um3+0v5dZz2bj44nsyKLhYPXKkaL1wCAhuuXcQ69dsWyAu7qF5PBMFqQzQRkzQgYvIQCuXleXYHlCXl2x1YZg+F7HxMDNAQLQoVetwuKZCZjRUTQqc/f7RjebisqAeuEQJXmpZUdA/3KgcgsJA2kL1xDNPDZqCyQAWdXiIy5YOHThUq4/KB1XFpgPr5heVtJuSQvJzxOeKB6HfEplzKWCEA4Sc+Vgqkw8bwIF16K7fg0ttNJr3DajEKBqfT5UlNkwXJKyD4hCRRlFySwU+TvTTJkJTh1wkms6l/pBWa08Fmt/WP+Nz2AWYcYEez3WwXvU5qECE/VB5ylJXl5993Hyc3zw6hkHaPoerldxVjh7eMX/F3hYWxu0KF382pcKpXsV+9QlS93Mj/Sz/ujinsVE1dDTszcEk1u4LpPdjXmDdw6UAsqFlUg7rmf2J+d3aGLmC757GBuEe55mHNXGxifZVrLtuNNUBhwbU6wSQ5IAOyoS2MCxcH7VmpXkHIdZlFP4BPtOvFdvlZZsncL0Kl1pZcS99Iam5eK1erfhFvrkviL9HDKc5X6OV/ChUq7aGEvw5U6QuFVCbEhOSSZHegODM7WOzxhOzZ2cVFJaXFIbfHK2cH7WlELuK3EnR5vHZJEkzvHZw35S933n0ucur5ky/MO7SraN2mrVuqGiNPnIt+NnTy6HF4fMkfvf+6EEjfkpWPh7rtXrJgp+NAk9hzQScj6194/+yxlZE72Ow0KvcdloMLbPcBiDD+2jdSW/Ek6MENfk55AfQMtwabaPC0aZWZ2a6Nob1NKgxRc3qemb/aF0jtk3xZPtkpc4Xjr3KVXE7WDfpi+sfVJ1RotwUyJVFVbE4ZV3JUPi0pLsq++XMM4A9Vd+/YcXcVvrtx7bLN61av2oINVTU11dU1NVV4cuPaFRvXrV7xDGPNH6+heQJpbMQaHLiz8R9fXb5w8dLl5vO7XnzhD7uef37Xxa8u//3ipa9pxpUqrt5AYeq1b8QPxVNg5BQWw13h9k4PpEqB3Lx2eW0DlmxfqkdfUhoy9Y6EnNZgW0t7MZ/6smlubka+I0NfFckQoDwPkjih+d4yrpTleTdRqoinJE6Ts7AULcTt8mRxQbYjMeLcXMpYwucgMgaCkrrMn668Z97YBwZHJm/+/hnWZ/KwOzazl5c2DerS+o2Xth9eshXXd7jTu7NHHeb98+VHfqw/+z/Cmp5zhvSZe3e/kSOubt2EO3tExnWrrbsy/51x94+aWFa/84V1k/bfx2Z1fWE0+2It+2zfxGEfAaBiMbBctRiug0CpIBLFUpyK2R+OumYgYrZB+cZAdoT4+TfM0CpsksEggGCxGoNUsV4J5sVpc5SGJE6pwxvIJgM3r97+1Kq1S7et2UQKUI/v7znOCn/8jpW80ohvKaN24aOatFEFAx8XLFYDFYItR0UbkQMljuIiEgx5HMS0efW2pWtXPbVdGZb9yjruPIInv/sR3z/+EisAhMFkrmCRXGCB9uEUKgoomw16o95qEwxoJiaT2cDtl84CUP5G4XWJOTBmWLK8olOmNOjMKhUpWZWHK5LZgl9279229we2OBUX50kuVjv5QDo7PBwnsvrhWJF+YDIuVagZDxeFHOF1MEKbsBMEQS+KJjOVdXJ1BKw61EH+feqSTzTz3I7ZA3Zuv+whshy3sDFL2TjctJR6n2SDsfFJ3A0I5ewXfAgugw7s+0XQG0SAfFVWHOEsr6TyphSHW5NHFc9J6Wa+7B3Dfp42HguHAUINniPlZCpQ/l0CogDIrW/8u85iv7sGv8ZzGzYAxjwV/MCxTwobJQCTWU8HRPQeruaaXpRqestVdUOXso7dupeF7px4Z8+ed3arKFc44AIg51W9ch4kIIiUEocmSk4sBpCcj15oUDRJXYYExl37RmirrkIv55rLASYJJF+S3t0nopeptU+E+mLrLK+lPgQyid3mCBU6UP1rVz8R2n770zc/Xf7x8s/Nn9fvaFi3rmFHPfmMLWRP4lycho/jNPY4W82Os88wiJ34K4tdAIQjAOQkx8YArcM2PaAOjSZBL8uolzAJFFvGDXd8ej67P2AvKpUkOYghcnK7zl300RBcsExwzJ/hbrd7GuYBwhgAIYtbTx/3+d4klJ3gtKCQnGIz9InYZEzqG8EkjSzNavCB/cXYlcQshhyMsZrI6PYLWc3lOG/vlA4rHr/3uTFD3r38/r+3fMKOke9W4oJ9G566u7au84CpOz/ct5R99wF7W6dIYjjnawrHIAh3hlungFOWgXoyzVKbHOr1eD19Il6vISsrrU8kSzbY+0QMGpdjgYh60zDTHJKHoyP4404pw27zB4o1o62gq+BLL299am8j+zv774zj995/dgTOZsOfWr3rnTWPj2h8qGbo1/M//kYYvmxfms7TtPrM54E7ns4vwBw0rFy/aNJjRRVTet31OgCBPABhongUDOCAzuE0h6gnxChToCJ1ulB0iH0jeqvscFBZotflk+hMQ5oJDqhrC/l//FxmAUlGYeK5Z6Jl5MDec2yJQdc+l5ViNduL1avoZ805eGll04jy6COKheT8S+U6kQwdw+lW6nPpXF4qtEoBziwAye3mMnRLkqlPRLqZdQlsKxTcLghkqhzjrLL5M+WgUwldSkjbL1HPLrCf51d8MHbv66zu/mcGl5Kz0YNZ0+mcf759kbEB29qGGrZiYWop2b2R9fYqnKnlWOVzqXqgNfQIB5LtRr8fQLLT7CyT0ZLaL2K0WFzU5e0TcfmojkckcgvcyhJ4pNlr8Bd63VyEhIbiGhfIBFGTq8R9lqcWB2Dl1G79Rn/9i8n08OU3L/760UX2E369YuvqVUPrI9VryFR8CXc5V/rYefbW7svv/YNdxUHv/OnFVQ1V8yse2Dde0UcAIY/zU4L0sA1FEQg3jJT0jVAJFBlqbOOrALk1dCOmkuHNF+mpaKOYunHhldNAlZhEyFGpz4R20C+c47Vmu+6gqXo9lewuq5TfXrLnZORk9Ink5JjAlNwvYvJBoF8E5N8qd9nN3jrmj7mOx8OPLDXqolpgwv0zZkpuzaeTynf+vWjNvnr22b+bsfDJR7+e+cL6dQ1bXlu3CDvOWfHIMytnrhJPHt7x4L7eg/48+8C5U0euLuu/f8ozr1xteHTRssdGru8V3kwfeHTMsN937/zksLEzFdlO5NQpNsMLWdAtnJlizzQYAAQu26AljUvWZbEQlyuJi1Ymcr8Iaal2jjKNg5qJ9Ctqx02jMyDFKHJw8TpUIvjHKhXZQlZ0/Iwe1eO++6/RVHpg2mv/uPbBuguPMtfKLU+tuXfjkIFraEVzg2tlMuZg6O57/vXBP1C3kZ3H9od2PPV81RMVE/aNAy3HEcaokRS34Ta+LAA8XotzQMRiizkRDVfN87X0JXae6NzkVR6Znehb6J8XL+Y3IKovXMjn0oEDMrkmmc2iXu9yGm0DIkab6hgTZklwj/T6FDccpXsmn6Rjlxv+knyrTFMR8+U/cF9+DiRwh/UCiChwdeXD58cDhSwsRjeikNNcTo83/0AtP2DDKLywji1nhxSezMTjgo9eVHOy3LBbJgIQ0OsEsToiIFRHrIjI4wHOlfxEz6a4ZOTXTLq9eTjdTofW1bEH6up+g5GIBDhGEr2BkRNVlMZTa/P3HKVyrMMKrF3H/KPYUAWjlGsXaRnXrxTIhrJwqp/bMtnphFYWIdgGoLWtddqASGuPzdA7YhNaqFZLvVJSEa48LZwUd4YSN4mJ+aq/ctSSXgtmD6gf2emV91/9KNj38bHd9l3PX0tq19dMnzFw3OSsgsWjj+zqPXn0w4On3e9nZ+NJLYFZ1yqkQ2ITFEM5zzwyA+1KLJ1kVwpAjsvSTgx3S+rQQeiisxv5Ky+9kGbnqUmllmSFEhOP6/G4ug6C2nJQUPdSt0td36R1IFMgbsUalrqlQAbw4KK1v1BwIH/udKqm8NCQbeMHP2LUtVk3rv7Fb4712N3Tt/DeaWvZt3+8wA7swe6Y/5cvjv3I1rHJn+AyhLM44ODVn14/7bBUDpq/hpxb8c388XfdM+rU3veu+Tws17Pv7O79aFvzMnvxc3aaHRq8sAZX4jgUsP7CfvYntoNhGYquJiAAAKJNPAIyWLjk0ojFqENR0SwqyILNaiG9I0bRYhFECoKD518xh6iplZYz+5W8H0OIlBsz/tURB6IHmnaT7itJORvb6A94cnbjGZYvHrnSg0zENwfPGTGddQIKJwCEo9xyW8ALGdA7nO0UUg1Wn89iEGQLjwd01iRrUlXEarWAxVcVsTjAWxUBevt4QnM9/gxBMbluwe4SAjxpj/mcgN0ef3cCt2IAhVVLsR/7+TIjjZjU9PTeY1ew4I9/Ovhn8cCeI/Nf9BnK2Pk3/kZ7TF00+6HoquhndauXPAGAMIdb09Oqr8gOu6jFpbdQb5IDekccglHi/HK2DL+4emRymUNIE3+Ro3WokKfbtNP37Cs0/7rxjQ0X2Cvs2Rex/NNLuysbxBB7lX3FPmdvl64rwyU44QusOVSzuj8AUTgmDuEc04FdsYcWQQ8COJyiuSoiUsFSFREct4ppwc9rSBlA+ZuAPZTBx2Az2Uo2CY/hIHysic/1z59PI/dU5CtWz+aJB9gi9gKmYebVKZgHgMq89Bc+r1GJWSSDAQXQoWAyS/reEUlCQsTeEUKRr3B03DZmUZBwxy/6S/MZmh+dTYZHt5OF4oH1LKc+eilhJj0UhpMlAKQ6pAbjTRPxSW45Q0CbAac3asPzwaNfrY9LTuyi2ilOhUvnI8SSohNapUJK7wiAaDLZe0dMgujtHRGdt4+8/HaphRyV9+rq5lT1xe9nfPc0a2IrDuKQL//9bve3DrL/so/Qj0kbVrGXCYuWZWXjUhzzD7xn/+D6GvYau8Q+Ze8H8LUY7WK6yuVQ2KdHBJ0giCCaTTraO6LTiQaJoshJV81RgnG/Qbydi5f/DYnpjc2ssZGSRrI3Ws1z7dXkYQC8NoLNxfFqVpwaNht1OotVT4GzFDJj9GrpGI15+JJiPpxLMg0v6dVv9AONx9jclFWuR6fyFGvI0TNxvRC+UjHmnkjBViRGg4Ix0Yn6RGzLWkgJZRVRDKHw1TvRrzc2NpL1J6JN5M0l0dc5snnk4+jCBF0QIT1soQCCJCMFzgtw3EBXxTekkO0+0aio0pV/bIp9V+KIgpPrUZJOFCUev/JSmsuNBjuVjDK1gKQgp2DnLbuZlRjwuJUAn2MY4nce4COtZjadZSsCntbhh6zRomMm0bbpo+bh4oGrVQLPOume7Uev/BCXo1IDsUG7sFsvcaytVpDB7jBS2aqjKCdypaUI4xPzabNJKZdj+WvNn+tsW4/RVB2xkGeEk582NR/nE3ZMwaxy2guAqFp99FZ5bu+IXqDW3hHqvLVNiOltBiTmueJRtpW9oZgjHIE9sBOOujo9+v1/fvn5h/9Eeb77LHuYa+94HIt1bArbxs6yU1iIuRjEAnYqZp+E8erqdUBRONnA+c75DE6XQaiKGAySLDuqIjKVEtavhpXmSgW/mlplYChutYXx7Ay7tLsRZ5PWUePGL949euKoYPr7t1HOh2jK6mdXrVC5wHaoXLBCCp+Zp8MeAIEa+OqmZtns6x0xC7KTL2yZM+MtlRs3J6I2pViG8q258sX7OOxndrH0tpz5ki3rzuqxivyf/DnN+WMCN1SGs8yIxKS3y0aDQdYTwePVm8EMVRGzmVDK5UepkSi6cntnp2Ku8ktw20SOf5bGNm4BcRXyGdhfcfkJ9jQ7/VXTzl2vfEZGRLeJB94/zf4+LjqZjFi9cuWqJwDVHIFw29ha4V6a0wSQ5BSFrGxTGvV4uH30CFSfoEoJiY4mt0CGlozy8D+o5jgx+6jmBbwy4BEI+9d3rHnZ0I/GN+7usnL1ey+xM389WLx/1+INHRbWXfoDLjz+6Z07su+YN73vyIFFvd959sV3qtf2nfFA35F3FQw8AoDgABCGcv7JvJ7iABSRUp1epgK3CYLmFeJ5qGYSi7k3IEsbWYFQyQrE9PWqJzjM14yPj2OHrLDdhgYZZafDrqOCmQ8UpzGUuFzsLkUnVHMYs4uij/2F/cJfFxrfee3ld8QDzf2vsC8wo5nuaa44+Mabh+ghQAAA4XW1/pMcNqJgMuooCJQqiPLlrxWvQhjgF8//SgXTwej3O6M/NmF1x8zWHdVaFh/5uU3bnwXkmg1yXz6aT6km+QwpyW6LRdQn2Q0U9TGTotqUGOKqNclWAjJldKcyenwSZ0h8cyc75y5CT3v2xU42u+nL9p6UYpSa0Nne7yy+1EQ/7PaW6/dbm0N88llHNx18ic5qnrv59RXv0YUK93QAQr1q9QNhhyCJ3ORLiskXFJMvtDT5KhocAz63Yu7rj/PIY0oTXmKdjuAkfHg/60QWROeQZnI4+gq5M9oX4lybrUY5GWGrIBJRpnoDiChTUeOcJmE+qKL+GCJdcNEhlrSb+Q6T8+R887zoCZJPFyv1ZQBBscZ6pWKmQyqDLKBgMIoCNwcUdUrMcuuKmVot8AvlzU6qi9roq82/0LSFwoaNC69OAIQGdoRMVnSRY2mRUFAYoxcJlTDIOdBSfeJRD5nMSvEEu4B+dkS6svyKX6HWC0A+i1c2Kd5c2XRy3h0mgYbo/4spg/KNEDuCzdrMFFACSacHOUgFevPMXj5rMb9CfMoLfOrSA+KF5b9KyigFJCgExOMgQVJYD1TWiQQEwrO+G5rpVFUTC3DfaPxsA1vG9pEg3dQ8jnwV9QJea2Zv0k3XKtUKsJLHIlEqwBgjmU/LQUfRp9mbCwCxTjhHHZIf9OA8AILRID2BkJ+s1ZoxwDW1OMStBHU83G1fm5MZ0+4QzhUdK3f33F8MRKk50lPCUEXzoVc4K1NnTEvz+Rw6yqMpYkzrFSFGI7jd1ooIt4LJFRHRA24o/98LVH4tX7NllapJZ7zS6LZn8QVeLKsVKjrQrxv43GPPvUychyc/VveH0F3HR77xCrNs/mPDWy89tOWB3js3Y1+b1GPe7Jq5dxTuORZ11TZuHC3LD00fOhwI7OVWtVZygRPSeVUt0+D1Wq2mVGqiGX4zmNwOu8HOhccRljzgqoiArYV5DSXF1SDB1sddEk825YBijeRQiVcrvHAqyJ5Pv/3+k0l/7GwKzGzQ6Wa811i/qXFjfb0wlJ1jP/DXxwMGLpdcbNHcsTuWvv7ll29fOPPJXwAQpnMOLxWGxbIaK6VuPU3ySmaOmQ0cHDPPzVmNGM9qlJ1DHgNzu6hmOGTcZXYV9f8d8HTbUOn8QrbvuW11Tz3swiw0oRPvyPQu96Sywe9+2mlNGRBlVqGU88fB+dM97E+VvGCx2CV7ht/htgIgmqhez9mjt1FnRYR6bscerSYTkLTqvTcUDPLPA6osi+JOiG7ST//n2W+/++TCTLMsNCxmTzdu3Ny4evOmNS9gNlr5647tA/rh0V+/mfny+4Gv3r54+i+fxLF0cN44IRk6hdOTDF4jpdzqtkrxGit4uRskyaUyyqIw6paZQyiRZQ632++JsUuivNbh53Kb+x/2JYp/e/+7qFl8eecf/zBk65bfb7WQLstc2AZl1GMH9v3fJxx/p2pttp/+c/eGrS8oUksFoBYpHVxK3cVlMjkJ4UaSuj0GvhQMgKIsVkScspUqq0GtY98IAxWmOZS1p2QNgeJSXkPW3DX3mE+zrxreeANH3lObN6LH8KHopW83l9G3+3TugmsDC9PnPNkLgEKQuYQCzplcKIVu8HC4a56vQ5YpvYtY4ESnSHIzW6Vn+Qzd72xlLbYWV0R0nXpFDJm6XKvOqvPk5pJekVxrm/JekTY2T7teEU9KnHUa+zj/8pXd+rzbxD1uragaVBdAqDC+jaAUkrJv/OXKcGMXmJOnbhQXF/F3QsHJVnf87VhB3sSqoa/te5X9jf3r7FdPzMgtC/ccNOnTtwb3ZPb6ZWdOPLzh7amPD50/4z8/1T4uVE5ICkzt9ewxXYdBbfPqVx54ddvqMauTndXFnYfmBnY+2PS66ypEhs2ZFOn5IO08/ZFvfn4cEPYCCD24nnuUzM5i0nFz7dF7vEkWvcMhVEQcNgOA3q0Y7xjlCatesVT2mALbtRUfM1P06cfm/+GZhgadoWD/jBMnyJuLfn/kk+jrfHXnDOow4N5XP4gWAxDYDoDjxAtAwcr9tZ3PJCDa7Ga5MmImVlQ04/3EwqZSIqAJJVQc3NDQ1CG3TceObXI7CJWYU1Zc0qFDaSkAubaKudSxTZAEd4Q9TqPRrNP5kj22yognrLcC1z6ISzW5xSTOhATTljhb3v2det7Zv/eNGZnLt9g16B6h+aqNHZHv0yaP8TSV89QGJTzetxgMRqNOEkSdYHeYAGw2nY7KRje1xiKGfD5zeUyFyuJsRTUiQi0bdclYkzcER73JeuD5E2zOnB07dKSgy2icydpGlxLpQTZOcjW/XTo9NjcO5nNT4GQCoiASQHfca2tMVBjHYVRo6SRfJQGoCAfcdruDiz+gdwRo66xWHrfb4RPMPm5p0302p1UPDkUPuCLEt534Igi1bHVIVIgEzfAqepHh1bRDypryyOa1DVNmblnVsDhFl79rIuIAXcHhmYdfJicWLNj3cnSLcv/zx9HjQmV99dDDg8e8+heuMZq2cnxdUBBOApeiri69x23S22xcWW02g/V2ytpSV72Jmrp7m4JG6NDUt95RNPXwJ+q8d0XUSWM2dhSfU9EknsU6wSyDnOwzeLgds1GbYvxvmcVylSHFilGFxE4PYRT74fKaf/wOTZcvobX5lZ3PPffii88/10Cy2I/swyeR/AFNmMfeZ1f/8rfzH545p1j5vdyW1apU+6E8nOEzCrKsS3foHJkBwQhWq7siYrXprboUaHXDzMdZ0GLBqpaeO2hPAhMUr62Y+gRHrThpU8Niry7c+PBf/+f7yzvryabGFc8+6xowcMRg1kUqqh9azT5h/1GcNr14+GTWl29fevfUeYVXHNNSlVexqMKW6qHJyT6bL8OfnOK1pqalecxOp8wtv80MFRHz/+Y2VT5yJ1l63Ul6r3vQ0njtQyL9GzaIW15cvXnjnI8uf/fJ57P0SQsajObpM/d9mHXp3YunT59birloRDO2a6z/9T38eEzFCzE9okGOpw1ywy6zXm8wEF4DsZrB4FYtg03rc2nRkaE5IY15ZEfvjt4eRQtfaahz6rrsFoaZNlk/fTbaJFSenDQjlrnS6XyW1twOtIplrqLzeuZaEfHYJKq/rj/5t8pdueG5kbsG25Hfpq50+j/e/+tjA/bXzF82+dmN88r/evSPL3Z6ftEjj7Yds+J13jSzsaHnpjbt7h4Uvrdr2aAH+yzaXLm4R1W3O7p2KO71FCCkX/uG7BQrwKPWJlwu3jPioEKS1+C0OXtFLGGbVeaCkj1xU3kqIVjV5ONWqo52xVGXhtxKNuHyEMcdA5NSJuSy17ZurRiBXdlrw2vN8lyzHQeQZdU9/83mRWePngiAsIOvrjKhElx8fh86ZZPJ4DS4PSaz2aZzWdVV7TFqEbMS/4daVmW0rJcrhBY127EvX9TPNNQl6UP7Z7zztlAZLeMO6GMSvnpozV2Dj54hp7RcjgiVau+HAQ0ms6hHK6jhiJZl+NX0NFTicIYQt7ER+76ptuiMte/tYyP4oI/8o0cx9iPtrx6K5UpSgI/Winsblz4lNc3rsZipYBZ0yQ7ubnTuxCyYK7c2A1U2Z2Rlk8LhUHSq1BmbsoRPKeSfcBbp2qSdPsY+3jNxsk5nLHCcaHqjg0snBF7dzc6QBZ3OvHR/dK5QyUaz6j5l+4tJbXTp7trW9eRvHClACAIIOpXGzLBdFiVAUWlxQZ3RLaD1pnQ4ngmjmhUfYgteQT9m/JktwFVH2Cn27hFSQLxsGO6IfhU9jUdYD0AgfL1LfHw3z/sVMqnHK5jB7OBLO0UHfIJCVam1GRJo46KKOdrSUrLvuwFOnfnuS/tYTsWfl/StKu2xq3cXzuCVn9wf+pn87mrGy5vtC03HtkAsZ6YPCZW3yJl7RUQr6npF0P2/5cz0oeZ/ksHR0+TL6D5y31Q6eN685sPxrixetlPl5/YlJxu9AFbZRbmnpqlpTq09K3F7TdV/bpXcPJZTfEtxCddDvj7d3EK4ZLfHjedrpx794PFH58/49MClCxdM44aRZaRxE+aPjywnw0Zg4ebdS6Xj7NzZoCl4FhAvMxuZrfluorSo0RSABN+tlHzx8nKeJv3cDAiV7Ijaw5Oq4OwWDQ4H8UFqqsXiE2laujso0QScEzYFFXSDxYr7U7DPVNCV5Dj2pcRw4eKhDx+Z/9jjp45OnvHwVFIePIvB49LSPRvZ+yPvJcsjvOq5cRenZNg4zJn2qEvdpyXVQg6tAS/XAzu1JvkcpuoIdVglCaojEuTngS3pjfw38rSkOlOZT8nQVNOmbD9lKoU5HFg8t2TMUz2mRrqPyi95omTcisrHK/sMJSfuLFn/UKvsVinhsvqH/RkZSeoOPFuKdcJwrcuYCALV8343AGpSu4xtNPOWXcZcCQNO1/Xt0PNKk/Gszp3Ly0IVZPfVC2Lfxb3C5ZVhQDjK7fd5dVemazjNozNTahCARxo62irVJxKnwUz4SzDKgg+07k9ljt9sw2apra1KOJCldLR6NAOuqD89OWHNwpPHcdniPisKChY+tHv7My8sX/FdifTO+xlov4LNXXfvoH7vstCH5z462QkQypUYSDzBpV4Zzk5y6s3mZI+dGD1OMS3dlORL6h/R+3xOcNr6RpxJIPa5uRWkRdPQzZ6Nm29lf5Lfinl2ypuduEqQxqONXTatnD0HG9jQblU05erVU2+99f/EEzUL+/1uGTs397MxS+7YtDz/xwtzsfO+U4psZqMkeIVtnHNByAibW0GmBSxtctLd7iwZeNSYn1gJchaVBku9il8r9co82Ja9clCxDnKwNLs0IXQ6VLV4+OLx8+eOq7t/UVXVgmF14+YuGrN42MKqeVtnzHh627QZW8mHj01aNmxh794Lhz059ZEFD/CHvfj7JZN+N2XbM1Onbd8BiscDEJT9Fw8MDrdzWGSj0WYS9URPTS6LW/YmGSwW2So5HBScbqsz3UmsTqvThG7JlATlWg+33RHrzL7lpjuGUOGj1uaovjBEKnH2HjYCJfY6dmGv72BvYGd+ARu7j1wgZ5vZ3Ma57Ec08RslQBKsgaxUVYkkUR726QUqUDlmFjgmiYqtbgjFLYRiI5p/YebmnxVpXPuF1kupUABdeGdcdiE4pdy0Dj5fmkmCgNS13E07lbRqK/n1/mCviN+tt/WK6OGGznh/s4t9I39VVFmLztSUlwuwZdCiRC2l/Kk33lG0dHD/qprTbw5/ZmTxqMV9Z8yYvelw/cCqjf/+6K9P9H9t4KLl7R+cvmJR99W/f6Ggbs3LPQbRnMF1WW0mD5q1NDW4IJjSKdy5prTH+klDl+fctXrZxm5rs9r27dWuY8e8oqHTRvWb0MVZPfnuKWXOMUCwWLTQ8eKH6u5TWpiTanKAI8lnpW495N90QCAhzctKeI/FxVnZpaXZWcU4pzgrq7Q0K6tYnFrUrl1RYUFBYfwOQGEM7xzvEdt5hxKeSwWDXmrNT0936a1esbSDZAKH1ZRuIuCwOYjJYXKk5AWcoRQByhNPBdhblgFRMxHuG90bnN2obu8KDjc3eYHM1py5DiFU2NqhNXTQOXMWz10weE77sRWvffDZq0880vHB5vXv4PB3les1tv2D02z76xP2YNvdezD3pT3s7N497JOXhMCeTTu3t/2dq9X3n575qfMjIXZI/Q7b/u6brOGD0zj0rT+wD/+wB3P2xr8GQKCCushU8W1OdzqUhlt5pRQDokeJazP8rQwGh88D1EYJNTvSOakf3feGku9qVGpqG4xTV8ojfbXWGSt18iYUtdZJXEnDlt0/edPztWvHjM+btnB+HauecmLUlAeov2bk6HHjJkhCcGFoRIcJs1jnI2OaCgRBqd8NhFraSI+CBGbICTupxI21YNTrBbMkWKwmUYegHGS5WbPRiyhjVuw2EAfPVEriM1kjLsUhtexzTK9lO0kQ1/dk29mzvXB9yo23qh9EHfeDXhAhJWwiKKAki0J1RCSQr20nattixUJOXfM71Bv9Hhc+CdeuaV3LRAIbAAjXdUoX16r7wqGgF3iOLui5Zpn1JodXKu1gsnFoi9Pi0DmtjnQHAR63E4fT4bythikCCP22ZKVVoUS+hp0Bqm51Fnr+L2UjHz5YPXLwfRNx36B+l3eeXrwWxYbNVy/8n+pGrtwd7tNtSfXsNFaLo9jTdPZ89ub/pXB47YrkEiRpzW3r+oJ09UfBJLnmAoG5dBi5LJ5U83Z/2GIGp7L7nGwzHPNQhS3J7yWaAKe27LkytvA6c/fPn39g4Oqa+fun195VPX3qwLunC2vmH9i/oGZlTdOCgdOm3l0zdZoiv/GASic8yQYLAMhwBiA6Q93NqCLLub9OUmpcstOLaHGCwAsItnQvZqjyadHEUVx6cz+0JMt+sjy645vIQH91edGont0XbPj9msiaPXiIVI2/NHhk35IePbMLh0yeP6V6/ZPPA4KflKlzBqAsnGkVRaCONIPUOstxn/MhJ+nrRKMzxUmcTl2yP92s88eVhKvIfTe2KDHRmKtlyd/2PpPpA3vsPbRzw4w1sz/8snbmA6Or7+w+pUPP8mXDl2wVvqx+wJu//YmVHWb32L5q0oAeXXrkBYa2LZl5056LnkfvwhP6xD0X5YAIN3pyAOvaT85494494cnCD133dnN3O1oEqNZDegiV4IHicLJoMOhs4HS6dC6+LeC2ulLMRKks6LWkMWHX6XqfaELKyMnTOhsGs13PNCxJNkz+Z/0Qg6GhAeewK698pKaNLwyr2caOScrsU1mzMEJygRWCYYcgIoBopDa7TidSq4jaQa/8RJkG7MortqVTEvILI6Z9PL1rzacn//ov0pY1S3t/raYhx5WrKDBA2ED6Yh0dqvitsEECMJuofkCEQsyAJOqq2jzatUOseZR82L1nz+7xMwlZzIVNAOBQIge7xQhgUfrILXa7jtog/71CzQq3qDNoZYbSkOzBpo31obZtOw24a8BDQx4ubWIXRk7UT9S1Kckrtu+bHgSEvqQKP1d3kPleHwFKDSZuX2mGBGlK3sc5EGO7FpnEzw8MXLlQ8pQsvpNv4K4ld9471NP2/hFAoDt1kaPi26q3zgo7lONnEnBvHfMfbr3iP964r4XTTjgzJSYsWHJ0V/3qF3eu3/B8lN07fsKwYRMeGCZM3nHw8LPP7T+w/TH+b/YjjwCBau4hdsY9BF+ZRr1AgMrEoJdu5R/4fBhELEUxdqM72c5aTGef1+IQVnvjPTGxCb3wfhzek01IufGW24c+AOIZzq8gnCYLACAbHrsGKMNHNDV6EPR/osTBA8ziYuCw7Tjs+ThseQz2CwV2Ou3PYeV9xMZBVchkAMkvnuAQM34FFf4CxEZ9KD5qXmxUIBBiM2mNMBxSoY3Sba1zpQWwlbVVwCXk5EIqmmhqKj93lzEgkm2zG3tH7IEWecP9w+9rGZ4ohslCYnXDUm9MGF2J0ihbnJBfkf59Rs7q4vv9Y9X1ozq9+dbRTwPhSMnYbk2zOnXtXqqkXKHH1tZM7NOvw5ip2e0XjzjcWDEhMjB/yIz70jFvcU/eGRvmVKrdoPJ0bltbq9R1v/YaDgTdn4hNzIa84ltA1MLCGETS7SCOQSAGkdoSIv86xGsg3HKMrOsQE6CUQxiaKGmtgtyAkWIwIMNxKIN5QK4xAIk3MIIVnNA/fAdPM+wIOhPaRNEtuvROycm7kHm7iMHM7wabASUqOtByowkglmHm5an5G8bOiYau9y/SAF7vYVQ2zqR5UUeUXdxLDtMT0SMkNXqR9Lhag0cfURpetbZG/AvZr2jRHOZSOkc5ztkqzrMIAf55rM9N5VmbON8PqhxBs8aRmyFqoTwG4b4dxLFrV2MQyS0hsq5DTACHylWC/hhXgUA+gFip9id54Z5wod3t1glmAKcgCUk+rogS11erXC6/JJ+WL8jcIsuyoNfbqiJ6Kri17tNEXW55EDWhHZV7uVhLarxnM5QhVqpNqbM3bcJ9eBf+bn/07S9xNlt4lIyKtaWSunqyntWxHSQcba5nhhhNYrmqS+3jurSmJdWx7jiVLwUx3sKsmLb5bgdRi4YYhP92EMegKQaR3RIiX4PgeGy65RhZ1yEmwMdxnW4b5z7CQrQJJmEDGMEX1st6ino0mXXgy0+0x2rMHLeOu0ewbTh8BHua7RiLw9m2MThS2DCa/3fbaLyfPTsaR+CIsWwrAOXzv877434CJ6RAQFkZnnRvmsAPExtcAA6rqFMCF0+a32f2945YHTpRoDazQHnjnES1lrm3+Fq4+YgL/ygm0lglwc7fxSoM1BZEj3qKzovZ1zsLv1479tEH9ykddGe2jnx04rGmh6Mjpu/9zy/NwbFk68SdWpPhmOUDNr2FDyl9dMMXV699l61D26bmvgOVZjp2ZRN9qTc7xVdOrI9LlUxpXLoVMfk7Nb7fDFELp2MQKbeDOAZzYhAZLSGyrkNMgA3xlRNMtEfCbHWUTvF5CmKjOFSQeO/frHjvH9+pMOtFUbKDBB6vWeALiC8fs96sl2LdkZoVarkRrHVH8v9lCDcaJGexM+zzQ42NZ9GHnuYrO3mL5LvvUdvFy4zXWq/B6ei/V+5Y9yQAqv0oW6R0aK94ppxcMTUAXpMJUu25YkGhw5Hbrl12RaQd5LrV3S5tj+vm0xpaZCBL2vZIQjWCo6Q2/2lnOTKUqE/1UYJv5ZAOKb36Lxv32p+OTCrfUnn27ofnjujZq094yVz2TcPf/v7+58IPi6dX3OnPyC0L3b917LZdPTcF8w/0mVQxcHZN+cTisqHF1YMuXO0r7Nv3562c52pXkOTnPL8TACXovgLUVWlXOH6L57V56vN2t3t+7FP1eajFc/Gz689fe+UW3xc/vP58whegruiOKsCNGRZehzj+cwyiTQwCqAIhKbtXOVDENWdkOJQLre3tedlIaF+WlJTe3ghi5y4pbYNtKyK+AqGgV6RD66BdECyZQU+xzqKriLgsNtBaO9R97viBxZsNL1corarUot3Jy/+qHSkOv7bLFExMz5TiAMaaVIb/wg7NmPnUc0VVb4+a/3xO8a6Hj/0reqcOO967tWbwurHswpy73lz03Mt7Jg1ZtfPpwzvoK7OWGon8BOY/+yddrEUqp/ie+4eMYP/9+yRWGwjyVpav5k5sXH9/5MVNo2XdQ6Sw4ektO5V1zXc4lW4kzreeMU+JFaqnVDtxVIn1ikl8vyqRVppEbn5e21993vp2z4/9rD7PafGcS1R7PsEQk1d7TaLX/gqAo9URXolZHHYXKGOgqI3xIgApTICovZYRgzDHIa79iUMMSoA4xl6IQTg0iG84RDrHQ4OYwA4CqBbHZ9d89VRlx1zyq6euqsJ5fsnUqhXwYN5jsTttkj7YRp9eETFSj91nsfLIR0+9LqSttY3QmLJw6/3b430QyITiIlAqxdlBMcj/lHpUk+6gRVqnV4kwil39+e/sK5T/9sUYXdkp9n3vr4YN77ll3OW+pzc8v7NpC3vppe0vPUtC7Ev2FzR/cQmlWcInr25+cGHXgtrefZ6cNHMlm8b+taaRbXjh4Aku21jXgbraqmOrzaLyJC1RNqNUrt0Vk/1HquySb/e8drD6PPN2z4+p45Ngi+d8fu35a9/f4vtcJtrzCSkx3Wh3fS2Ph2YhR9gJVO1CD4WTPAaDTSACKjsZTifKZjMqJ/QQ8tX1yhOfG8nPjUN6iccXE96Pp8ejezqVFHXsFCrqot3J8iefZP/q3KW8Y1m4nPwYfwOUY3tEGCUsjvv7PvxEa3orl8vQ6iZn76u47uxt1M+b2Kjnf3P2ZWVxBdGcfXw7QXSpTl4Si1SnX6L2X2yaUjNt+Dw0Xd40o6Z25NzmV4rxTJ9pvAljfYjl95r63Iuxboyetf0XbEBQGjL6zuy7cMOvu8aRRcWffLRjTHRO6DzXjNjutSq5e2KSf0PVDI8mmZuf107VNOfWz4851OeBFs+5ZLXnE/yxtZarrfrYDqw6wr2xGWIjpKsAWu+I2t+VyXex0jOkFJfNZpfsrQMOsKeYPHqqT+NdjB7q5euvRZPnb3oYUWsXUUomXo/W9JUVbx7J4HugOKR748Sz333/yd8fMwk63mSElTs38OYRzF9LmyID2Efsvwpjn83sV86KdcDaFQ1NOXQi58u3ce/ZMxo1nF6Nmgn7Y/TmxejV+puEyuv9TaJArLfsb+Iw6gkU6UvxFLggHe4Ot0uSrE5nKpjtqZKY4bc6eDxpBaOR51hGGj+Vwg8UUAc4b5zk4det2ia1fWVJO2TlvZF9aafq7NnSl1EYN4y9zJ7BYRgeN5RaonxdR8+Rfs09fmXXEH+ecs89LqzDiTgeF3ljSZmwlZ1m55QTGn6hNi32qy1yujAU0iAXCmBQuG26zkI8nqx8t7tVlk4oDOW1Mbbh0RHvSCKixdiunWg32pIyxcyKCIieFj7YoVjVRAeseV9R9a0q5rdyvYktTFkxnyvWs/Nzup6pu8B+ROnrBae6djz2+InL0aAOq4Y/e8+QDVf9G154buPm5xvWCb3mrjKRjN+7vp4xEwtQh3q8Y+a0KbPYz19MYDO5tw1mkLIPz3985rOPP/10x9NP7wBEE68Q7pH8YFF6wGWwWXmN0KJs3CSfKkwsE/Igzx1QzhIE0DR3nLfB89CcmUMWLuFF2u+WPJGTu3C+t3TBoiIAgpP5iG2lhdp+kEMyxSpMejflw753u9KSrHUfcfpp29njxj46a8zY3z3YPRTq3rmsqJu4b9TM2lGjps8c3qFLlw78AkQdn+k78TN1N5wPn+Szg2gC/nKrZc73En4mKLYb3o4vKU6BwvQ0olRTQpJEXXkDB/TOLAxZRpmn39tucP/KjIL21tHmqcL5rLZZnbvMquO3Tl1n1aldEci5Ff/FEyCCePMvngykw+K/eMIh5f8VUtYgffQ49lB7+R0HUNTpQenhP6WBBkscHEs5y+QZ1WF29yx63DMUTVyicNM3RdTpRZly061Rq55Od5RisXIk/bGKDPGARzmLjqmfcouq/e4LkcAKAEQZizSpY1khOWwS0KwXbHbQUZP2M1+x3pUgbyrhA/vjeGG9tcNjs9M6maNnb2B4FnXTeR1Tw7TF6DZldL0ZRcHuMIs2WRn9LW10DWe/ei9JQJ4ELUkjOsxJ7m6+QYbnXvbTY2Ow6D6FHh/7lTTBZZSVLOtqB8g4iCCHzeZK+dC1Y38ymWJ3vb5SBnteXszG7cAfyXB6EYzgPBD/URrIP3Wr6u+OqQ9OmDF94qRp5JtZj/9u9sx5C/icym8TiHvgB8gGOwAEwU4c/M4nELJA1RaoJelK5ZPTbBAIlYikk0WuCInpvPM3e2CJ+16ASv2UpGqjUBAIkMRRWhRNSeqtK6QAyGYBkJXxUyYgEkE7ZYLxAQJIVjbPWkkXx4+ZIJRzr1gnnuT0TQ2Xp3rTPZ5kI5Hl5NZ2wZDslYJtjN4kb/+ILklMTUvtHyFp1rT0tPw0qqdJaUlpzsxM6BvJlJ0W3iDhg5ZN3bwwdMsfKruRW2ZQbuRlt9evdcorVpPyolGwuJT/dUDsCHUKOz4AWfRHQvA065Z1snHLxtW7/oddaNewgZANO4LY+n9OPN+rQSxmD80rC7ed1/Rm9/puaEacl3tH9TwUsfXIpYPVzprl6o4iBXdYT0AUtDAtYc3y+EuJtrjkUwGEVlI650ylKvE+5ABA/HNTwuf9lc+BgItUcf0/AgZwQedwuks0ypTyaYjSqY+iqLe60l3E5aIWOZ1mxPuV70toergeGwR4g0v8V2eKi0otVJZJ05xV7GHcsHQO+0ESk9LSjDup6913x/KzVKdeX9THFGzb1v5TDDfpQ45bECoJ9+43cBcf0nCXXr/F8/43notvxJ6rVEnqc1TWG05X9cp+AAQRKWiHl2Knck80KgqljCAC4Aq1QvJpPHP6XaxCImp1FiUv6pwAUXstt2Ud9NrbHGJCAsQx9ufEKktsFtJBzroOMYF9EK/V+GK1mv8PflNJUQAAAAABAAAAARmahXJJOF8PPPUACQgAAAAAAMk1MYsAAAAAyehMTPua/dUJoghiAAAACQACAAAAAAAAeAFjYGRg4Oj9u4KBgXPN71n/qjkXAUVQwU0Ap6sHhAB4AW2SA6wYQRRF786+2d3atm3b9ldQ27atsG6D2mFt2zaC2ra2d/YbSU7u6C3OG7mIowAgGQFlKIBldiXM1CVQQRZiurMEffRtDLVOYqbqhBBSS/ohgnt9rG+ooxYiTOXDMvUBGbnWixwgPUgnUoLMJCOj5n1IP3Oe1ImajzZpD0YOtxzG6rSALoOzOiUm6ps4K8NJPs6vc/4cZ1UBv4u85FoRnHWr4azjkRqYKFej8hP3eqCfDER61uyT44DbBzlkBTwZD8h8/sMabOD3ZmFWkAiUs5f4f2SFNZfv6iTPscW+jOHynEzEcLULuaQbivCdW5SDNcrx50uFYLzFHYotZl1umvNM1tgNWX+V/3gdebi3ThTgVEMWKYci4kHZhxBie3TYx3rHbGr+Pdo7x4dIHTKe5DFn+O/j+W2VnE3ooW6isf0LIUENvZs1gf/LHojJwdpplCP5gn/5gi26FoYa19ZVFOJ6Sxuoz/q2Ti20IKVJdnqvYJwnhfPH/2f6YHoQF30aZaK9J8T026RxH5fA/WPW/8IW4zkpnIfoFLifGB86v0ffm5nbyRs5iaHR3hNBD0HSfTzoPugRM+hdN0x052KoHLBS0tdgpidAiEesDsgWYO73RWQz2LWIwjqnMe/uYISQtlbyf2NlT9Q9PoBcBnrO6I5ELoMeyHkNnIXGdv809H/DXNOTeAEc0jWMJFcQxvFnto/5LjEvHrdbmh2Kji9aPL4839TcKPNAa6mlZUyOmZk6lzbPJ3bo56//Cz+Vaqqrat5rY8x7xnzxl3nvo+27jFnz8c/mI9Nmh2XBdMsilrBitsnD9rI8aiN5DI/jSftC9mIf9pMfIB4kHiI+hWfQY5aPAYYYYYwpcyfpMMX0aZzBWZzDeVygchGXcBlX8ApexWt4HW/gLbzNbnfwLt7DJ/p0TX4+Uucji1hCnY/U+cijVB7D46jzkb3Yh/3kB4gHiYeIT+EZ9JjlY4AhRhhjytxJOkwxfRpncBbncB4XqFzEJVzGFbyCV/EaXscbeAtvs9sdvIv3cjmftWavuWs2mg6byt3ooIsFOyx77Kos2kiWsIK/UVPDOjawiQmO4CgdxnAcJzClz2PVbNKsy2ZzvoncjQ66qE2kNpHaRJawgr9RU8M6NrCJCY6gNpFjOI4TmNIn36TNfGSH5RrssKtyN+59b410iF0sUFO0l2UJtY/8jU9rWMcGNjHBEUypf0z8mm7vZLvZaC/LzdhmV2XBvpBF25IlLJOvEFfRI+NjgCFGGGNK5Rs6Z7Ij/45yNzro4m9Ywzo2sIkJjuBj2ZnvLDdjGxntLLWzLGGZfIW4ih4ZHwMMMcIYUyq1s8xkl97bH0y3JkZyM36j/+58rvTQxwBDjDDGNzyVyX35Ccjd6KCLv2EN69jAJiY4go/lfr05F+Ua7CCzGx10sYA9tiWLxCWs2BfyN+Ia1rGBTUxwBEfpMIbjOIEpfdjHvGaTd9LJb0duRp2S1O1I3Y4sYZl8hbiKHhkfAwwxwhhTKt/QOZPfmY3//Ss3Y5tNpTpL9ZQeGR8DDDHCGN/wbCbdfHO5GbW51OZSm8sSlslXiKvokfExwBAjjDGlUpvLTBY0K5KbiDcT672SbXZY6k7lbnTQxQI1h+1FeZTKY3gcT2KvTWUf9pMZIB4kHiI+xcQzxGfpfA7P4wW8yG4eT/kYYIgRxvgb9TWsYwObmOAITlI/xf7TOIOzOIfzuEDlIi7hMq7gFbyK1/A63sBbeJtvdwfv4j28zyaP8QmVL/imL/ENJ5PJHt3RqtyMbbYlPfQxwBAjjPEN9ZksqkMqN6PuV7bZy7LDtuRudNDFwzx1FI/hcTzJp73Yh/3kB4gHiYeIT+EZ9JjlY4AhRhjjb1TWsI4NbGKCIzjJlCmcxhmcxTmcxwVcxCVcxhW8glfxGl7HG3gLbzPxDt7Fe/gY/+egvq0YCAEoCNa1n+KVyTUl3Q0uIhoe+3DnRfV7nXGOc5zjHOc4xznOcY5znOMc5zjHOc5xjnOc4xznOMc5znGOc5zjHOc4xznOcY5znOMc5zjHOc5xjnOc4xznOMc5znGOc5zjHOc4xznOcY5znOM8XZouTZemS1OAKcAUYAowBZgCTAHm3x31O7p3vNf5c1iXeBkEAQDFcbsJX0IqFBwK7tyEgkPC3R0K7hrXzsIhePPK/7c77jPM1yxSPua0WmuDzNcuNmuLtmq7sbyfsUu7De/xu9fvvvDNfN3ioN9j5pq0ximd1hmd1TmlX7iky7qiq7qmG3pgXYd6pMd6oqd6pud6oZd6pdd6p/f6oI/6pC/KSxvf9F0/1LFl1naRcwwzrAu7AHNarbW6oEu6rCu6qmu6ob9Y7xu+kbfHH1ZopCk25RVrhXKn4LCO6KiOGfvpd+R3is15xXmVWKGRptgaysQKpUwc1hEdVcpEysTI7xTbKHMcKzTSFDtCmVihkab4z0FdI0QQBAEUbRz6XLh3Lc7VcI/WN54IuxXFS97oH58+MBoclE1usbHHW77wlW985wcHHHLEMSecsUuPXMNRqfzib3pcllj5xd+0lSVW5nNIL3nF6389h+Y5NG3Thja0oQ1taEMb2tCGNrQn+QwjrcwxM93gJre4Y89mvsdb3vGeD3zkE5/5wle+8Z0fHHDIEceccMaOX67wNz3747gObCQAQhCKdjlRzBVD5be7rwAmfOMQsUvPLj279OzSYBks49Ibl97In/HCuNDGO+NOW6qlWqqlWqqlWqqlWqqYUkwpphTzifnEfII92IM92IM92IM92IM92IM92I/D4/A4PA6Pw+PwODwOj8M/f7kaaDXQyt7K3mqglcCVwNVAq4FWA60GWglZCVkJWQlZCVkJWQlZDbQyqhpoNdAPh3NAwCAAwwDM+7b2sg8kCjIO4zAO4zAO4zAO4zAO4zAO4zAO4zAO4zAO4zAO4zAO47AO67AO67AO67AO67AO67AO67AO67AO67AO67AO67AO63AO53AO53AO53AO53AO53AO53AO53AO53AO53AO53AO5xCHOMQhDnGIQxziEIc4xCEOcYhDHOIQhzjEIQ5xiEMd6lCHOtShDnWoQx3qUIc61KEOdahDHepQhzrUoQ6/h+P6RpIjiKEoyOPvCARUoK9LctP5ZqXTop7q/6H/0H+4P9yfPz82bdm2Y9ee/T355bS3/divDW9reFtDb4beDL0ZejP0ZujN0JuhN0Nvht4MvRl6M/Rm6M3w1of3PVnJSlaykpWsZCUrWclKVrKSlaxkJStZySpWsYpVrGIVq1jFKlaxilWsYhWrWMUqVrGa1axmNatZzWpWs5rVrGY1q1nNalazmtWsYQ1rWMMa1rCGNaxhDWtYwxrWsIY1rGENa1nLWtaylrWsZS1rWcta1rKWtaxlLWtZyzrWsY51rGMd61jHOtaxjnWsYx3rWMc61rEeTf1o6kdTP/84rpMqCKAYhmH8Cfy2JjuLCPiYPDH1Y+rH1I+pH1M/pn5M/Zh6FEZhFEZhFEZhFEZhFEZhFFZhFVZhFVZhFVZhFVZhFVbhFE7hFE7hFE7hFE7hFE7hFCKgCChPHQFlc7I52ZxsTgQUAUVAEVAEFAFFQBFQBBQBRUARUAQUAUVAEVAEFAFFQBFQti5bl63L1mXrsnXZuggoAoqAIqAIKAKKgCKgCCgCioAioAgoAoqAIqAIKAKKgCKgCCgCyt5GQBFQBPTlwD7OEIaBKAxSOrmJVZa2TsJcwJ6r0/+9sBOGnTDshOF+DndyXG7k7vfh9+n35fft978Thp2wKuqqqKtarmq58cYbb7zzzjvvfPDBBx988sknn3zxxRdfPHnyVPip8FPhp8JPhZ8KP78czLdxBDAMAMFc/bdAk4AERoMS5CpQOW82uWyPHexkJzvZyU52spOd7GQnu9jFLnaxi13sYhe72MVudrOb3exmN7vZzW52s8EGG2ywwQYbbLDBBnvZy172spe97GUve9nLJptssskmm2yyySabbLHFFltsscUWW2yxxX6+7P+rH/qtf6+2Z3u2Z3u2Z3u2Z3u2Z3s+O66jKoYBGASA/iUFeLO2tqfgvhIgVkOshvj/8f/jF8VqiL8dqyG+d4klllhiiSWWWGKJJY444ogjjjjiiCOO+Pua0gPv7paRAHgBLcEDFOsGAADAurFtJw/bt23btm3btm3btm3btq27UCik/1sq1CH0I9wl/DTSONInsjxyKcpGc0VrRNtGx0dXRF/FpFiV2KbYl3j++Jz4vkTaxKjEgcSXpJzMm6yb3ALkAnoCV0ARLAcOBjdCAJQJqgWNhJZDT2EbbgTPhz8h+ZFJyDbkFSqgVdGh6Br0BhbFFCwHVhNrj43DXuH58V74WcIkahHvyDRkLXIGeY18SxWl+lMHaIVuSc+h3zHpmNbMJOYuy7DF2E7sFvYMJ3Clf+3DHecNvjm/m38g1BYmioxYS5wqbhZ3S0Wl2tJkab50U04pl5CHy9vlmwqlZFJaK4uVnco55YlaUK2kNla7qEPV6epi9aMW01jN0zJohbRZ2mptj3ZWu6e91wE9vT5LX63v0c/q9/UPRiZjprHS2GmcNG4ar8yIOcycZC4yN5mHzMvmE/OrhVq6NcCaYC2wNlgHrAvWQ/t/e6w9115r77XP2fecrE4xp65zwM3lNnZnuBfdZ17E071sXj6vrTfP2+Hd8F74lJ/eL+Hv86/6D/23Qfogf1A+qB10CAYGk4LFwdaf2C+JfQAAAAABAAAA3QCKABYAVgAFAAIAEAAvAFwAAAEOAPgAAwABeAFljgNuBEAUhr/ajBr3AHVY27btds0L7MH3Wysz897PZIAO7mihqbWLJoahiJvpl+Wxc4HRIm6tyrQxwkMRtzNIooj7uSDDMRE+Cdk859Ud50z+TZKAPMaqyjsm+HDGzI37GlqiNTu/tj7E00x5rrBBXDWMWdUJdMrtUveHhCfCHJOeNB4m9CK+d91PWZgY37oBfov/iTvjKgfsss4mR5w7x5kxPZUFNtEoQ3gBbMEDjJYBAADQ9/3nu2zbtm3b5p9t17JdQ7Zt21zmvGXXvJrZe0LA37Cw/3lDEBISIVKUaDFixYmXIJHEkkgqmeRSSCmV1NJIK530Msgok8yyyCqb7HLIKZfc8sgrn/wKKKiwIooqprgSSiqltDLKKqe8CiqqpLIqqqqmuhpqqqW2Ouqqp74GGmqksSaaaqa5FlpqpbU22mqnvQ466qSzLrrqprs9NpthprNWeWeWReZba6ctQYR5QaTplvvhp4VWm+Oyt75bZ5fffvljk71uum6fHnpaopfbervhlvfCHnngof36+Gappx57oq+PPpurv34GGGSgwTYYYpihhhthlJFGG+ODscYbZ4JJJjphoykmm2qaT7445ZkDDnrujRcOOeyY46444qirZtvtnPPOBFG+BtFBTBAbxAXxQYJC7rvjrnv/xpJXmpPDXpqXaWDg6MKZX5ZaVJycX5TK4lpalA8SdnMyMITSRjxp+aVFxaUFqUWZ+UVQQWMobcKUlgYAHQ14sAAAeAFNSzVaxFAQfhP9tprgntWkeR2PGvd1GRwqaiyhxd1bTpGXbm/BPdAbrFaMzy+T75H4YoxiYFN0UaWoDWhP2IGtZtNuNJMW0fS8E3XHLHJEiga66lFTq0cNtR5dXhLRpSbXJTpJB5U00XSrgOqEGqjqwvxA9GsekiJBw2KIekUPdQCSJZAQ86hE8QMVxDoqhgKMQDDaZ6csYH9Msxic9YIOVXgLK2XO01WzXkrLSGFTwp10yq05WdyQxp1ktLG5FgK8rF8/P7PpkbQcLa/J2Mh6Wu42D2sk7GXT657H+Y7nH/NW+Nzz+f9ov/07DXE7QQYAAA==) format("woff")}@font-face{font-family:"Open Sans";font-style:normal;font-weight:700;src:local("Open Sans Bold"),local("OpenSans-Bold"),url(data:font/woff;base64,d09GRgABAAAAAFIkABIAAAAAjFQAAQABAAAAAAAAAAAAAAAAAAAAAAAAAABHREVGAAABlAAAABYAAAAWABAA3UdQT1MAAAGsAAAADAAAAAwAFQAKR1NVQgAAAbgAAABZAAAAdN3O3ptPUy8yAAACFAAAAGAAAABgonWhGGNtYXAAAAJ0AAAAmAAAAMyvDbOdY3Z0IAAAAwwAAABdAAAAqhMtGpRmcGdtAAADbAAABKQAAAfgu3OkdWdhc3AAAAgQAAAADAAAAAwACAAbZ2x5ZgAACBwAADiOAABYHAyUF61oZWFkAABArAAAADYAAAA29+HHDmhoZWEAAEDkAAAAHwAAACQOKQeIaG10eAAAQQQAAAICAAADbOuUTaVrZXJuAABDCAAAChcAAB6Qo+uk42xvY2EAAE0gAAABugAAAbyyH8b/bWF4cAAATtwAAAAgAAAAIAJoAh9uYW1lAABO/AAAALcAAAFcGJAzWHBvc3QAAE+0AAABhgAAAiiYDmoRcHJlcAAAUTwAAADnAAAA+MgJ/GsAAQAAAAwAAAAAAAAAAgABAAAA3AABAAAAAQAAAAoACgAKAAB4AR3HNcJBAQDA8d+rLzDatEXOrqDd4S2ayUX1beTyDwEyyrqCbXrY+xPD8ylAsF0tUn/4nlj89Z9A7+tETl5RXdNNZGDm+vXYXWjgLDRzEhoLBAYv0/0NHAAAAAADBQ8CvAAFAAgFmgUzAAABHwWaBTMAAAPRAGYB/AgCAgsIBgMFBAICBOAAAu9AACBbAAAAKAAAAAAxQVNDACAAIP/9Bh/+FACECI0CWCAAAZ8AAAAABF4FtgAAACAAA3gBY2BgYGRgBmIGBh4GFoYDQFqHQYGBBcjzYPBkqGM4zXCe4T+jIWMw0zGmW0x3FEQUpBTkFJQU1BSsFFwUShTWKAn9/w/UpQBU7cWwgOEMwwWg6iCoamEFCQUZsGpLhOr/jxn6/z/6f5CB9//e/z3/c/7++vv877MHGx6sfbDmwcoHyx5MedD9IOGByr39QHeRAABARzfieAFjE2EQZ2Bg3QYkS1m3sZ5lQAEscUDxagaG/29APAT5TwRIgnSJ/pny//W//v8P/u0Bigj9C2MgC3BAqKcM3xgZGLUZLjNsYmQCsoGY4S3DfYZNDAyMIQAKyCHTAAAAeAGNVEd320YQ3oUaqwO66gUpi6wpN9K9V4QEYCquKnxvoTRA7VE5+ZLemEvKyvkvA+tC+eRj6m9Iv0VH5+rMLEiml1XhzPdNn3n0rj6/EKn2/NzszO1bN29cv/bcdOtqGPjNxrPelcuXLl44f+7smdOnjh09crhe279vqrpXPuM+PbmzYj+2rVws5HMT42OjIxZnNQE8DmCkKiphIgOZtOo1EUx2/HotkGEMIhGAH6NTstUykExAxAKmEqSGMFl6aLn6J0svs/SGltwWF9lFSiEFfO1L0eMLMwrlT30ZCdgy8g2S0cMoZVRcFz1MVVStCCB8raOD2Md4abHQlM2VQr3G0kIRxSJKsF/eSfn+y9wI1v7gfGqxXBmDUKdBsgy3Z1TgO64b1WvTsE36hmJNExLGmzBhQoo1Kp2ti7T2QN/t2WwxPlRalsvJCwpGEvTVI4HWH0HlEByQPhx468dJ7HwFatIP4BBFvTY7zHPtt5Qcxqq2FPohw3bk1s9/RJI+Ml61HzISwWoCn1UuPSfEWWsdShHqWCe9R91FKWyp01JJ3wlw3Oy2Ao74/XUHwrsR2HGHn4/6rYez12DHzPMKrGooOgki+HtFumcdtzK0uf1PNMOxwDhN2HVpDOs9jy2iAt0ZlemCLTr3mHfkUARWTMyDAbOrTUx3wAzdY+niaOaUhtHq9LIMcOLrCXQXQSSv0GKkDdt+cVypt1fEuSORsRUwgrZrAsamYJy8fu+Ad0Mu2iYFhexjy9FIVLaLcxLDUJxABnH/97XOJAYQOOjWoewQ5hV4Pgpe0t9YkB49gh5JjAtb880y4Yi8AztlY7hdKitYm1PGpe8GO5vA4qW+FxwJfMosAk2X9n9X2cVVfnA36pzHNHJGbbITj75NTwpn4wQ7ySKfAu9u4kVOBVotr8LTsbMMIl4VynHBizBEJNVKBAfMNA9867j0InNX8+ranLw2s6DOmqIHBIbDfQR/CiOVk4XBY4VcNSeU5YxEaGgjIEIUZOMi/oeJag4mEB3PUOweCaG4wwbWWAYcEMGKn9mR/segY3R6zdYg2jipGKfZctzINQ/vxkJa9BOjR44W0OpTKAskcnjLTcKyuU/SVIWSKzKSHQHebYW9mfGYjfSHYfbT3+v877XhsIwGzEUaleEwITyE2u/0q0Yfqq0/0dMDWuicvDanKbjsB2RY+TQwOnfvbMUhiNPFyDCRwhZhdjE69Ty6FjoOoeX0spZz6qKxxu+ed523KNd2do1fm2/Ua6nFGqnkH8+kHv94bkFt2oyJj+fVPYtbzbgRpXuRU5uCMc+gFqEIGkWQQpFmUckZe2fTY6xr2FEDGH2px5nBcgOMs6WelWF2lmiKEiFjITOaMd7AehSxXIZ1DWZeymhkXmHMy3l5r2SVLSflBN1D5D5nLM/ZRomXuZOi16yBe7yb5j0ns+iihRdlFbd/S91eUBslhm7mPyZq0MNzmezgspUUgVimQ3kn6ug48mntu3E1+MuBy8u4JnkZCxkvQUGuNKAoG4RfIfxKho8TPoEnyndzdO/i7m8Dpwt4XrnSBvH45462t2hTEX4Bafun+q8jIzK/AAEAAgAIAAr//wAPeAF8egd8lFXW9zn3PmX6PNMnPZNJMRRDMkzmDYgZMRRDCEmMMUPJIgZEepHlRYyIiNhRUdYuS4ksy9reLDYsdOmLLC/Ly7L2CgKrrCJkLt+9T2YyYPl+D8804J5zT/n/zznPBQKbACSTvAEoqJAdtUhUJpQYjBJVAUrKSkIOJ1ZUOEKOUGkfV8ARiPB7E72m87WJZF58ibzhXPVE6QsAAnMufI4H9XXsUBh1UpOJSJLmQNWqNsasLkKhsrKnA/T1HCF9PQzSAPYtD5V5PW4lmFeIK86EcCRbObLp2lGjGxpH4+f0wLkjjU3NDSNGxYSMxbSdDkzomhE1SypQalCISvniob1lDuTL7injC1O+Mr/xmeJtxeRt/iJviJ8mmrjFOr0BJCZ3QAbkQFu0ypCZ45HcRqNJQkiT/LKsOO02s2Ryudze7CxVUnw+v9+tmKTcgEEymzPRlgN2e5rHaeOXyeeiisnJFagMOSsqSkr45kL8Tr450SfM5/y1V66pGvBwTV1BcYcDEX67QjQkbo8cigTplyVI2OHh/6zdXHO4+iR6SjoxMPzo8O21h2tPx7O2lmylNV/tY5Nwubj3fXUA/8BuFveBr74CoNB84V6pSnFCLhRCL7g7OijfR7Oy3FalR49AcXYRFBnsQUcgkAYO6H15j6wiAGu+I+Ao6pleFDAWKJZMX+aImNunWOpiskIVH796ewAqEzvV9gqX9nQ4Qd8S/1V/ScSM/rmsTP9FfNUNIvzuVlRPMFxY5PB6fY6iwsJw3/JIOOTx+lT+WzaR+xYWecrR7fWFFanqi/33nnn9+v+MvXr7mk933/v5Gy3PrN6yZjg7WFV1D5s2oGoh7nx+k2vvTrkeDT0HKlieXvvakkfecj/5uKnhm6iNHRk27a6bevTL+clH3ulVkX3cBTJUXjip/CDvBiO4wQ95PB6qo/len0+WTRpofo8nLa04mB3UgpeX5PbMLEzzKz4/tapOlXt5a1llpXhN7FF7r8zJ37o/iN15Q2XhvsE8RdajOqwFyrwFGETXr/0F9u9dNnZsWW9869X1azow9qe/kpc7D52mPRf//HcJFrR1npvf9sWX336EO7/9x7lqeUMn6frt8y+//ZD/JjzecOGEAnxvWdzjpTAzWtHbGjRhlhdMXqvLVZSWnl5kpSoChLJVtcwXSPea8vNLSrT0dEnTegyPaZIUqIlJLnSKhAV/pfBuhb9EbE53bYVIM/3S45hfiZ+7th8IFPHN5QuXcscms1vF8kiAZ2qBsEEEFQX7FnJDeNy+8nIF2JLZ7/77DPtk3rJhVV9vefPD+57CzCF98cr82+s631s4/vbxrKPf1XjT0Iqrh/+uafTMxR+9e++mxqZnxzzx5l8embstxo7PeX0Ju3DjoqYJA7C611hyd3hAtH/zpD5jAAVm4DM6Zjj5C5WIAIu9DuxCIB0kuvEBAKGBbSTz+L+3Qm7UZjaZqCSBqtrN+VQgmAMTua3joeaMhBTicTt9wULS8PSj5x58eNk9Z5c9RUrRiPte3MTKzvyHRd5Yh9vFygP4yq3JlfmyfHG+so1LyP/5yqgRNVjuDPclRSGvk7Q+/ejZJY89/OA5sTT7ifVb+zru/OEM7tv0EisFhErSJGUpbrBBOOo3ms0ypVZUVc0umUyqilarYrDxpN1aJrKQuykJwvwz/yPMUOCTXSqlRa6CiEzJy8U4J8DWf/jpM/eeOMZeLMKpxYqbPTyx088Oz8MKtnMuFqefm4gzAKEZPpUqpG1g5qivGRSjkSKAxWo2giJRKOFCysqS4vjNhQXCAa4Bxz1HEI+yNlx0FBextqOk9SjezW49yhaIHbGzuBtOggKe1wgFWVapDCXbdSNt5ghfoNCgMxLA3X1v++dV+eg/vIsdR9MJYWVcS5rISqDg+CuVQQLkSiTc7QoHPANIGq49dw6wi7GwgmvujZoUrrSRNsaMLqjsmfjnkYu4aU6SlJZ28xECNyqt0mMrM2pBricBidueiNS5iDcRA0ir4h+y4yQgGJP/DwLVF05IQ+W9XLoPLou6LYoTFPCnGT0jYkaV2kfEaBok8y+1kkYCeeDQnIEyQI2nUrlDE3kkDT3PzsfZhXMoxZHGw2OmTRl7w+SpLeQoW8gexttwNi7C6ewO9hD7/usTaELr8eOAMA+A1nJtTNAj6jJKAAZEs8WgqihJRgX9wJHOkYoXkf8iwR2RiKKqRRiitWw3lYdnr30cDzNae/8Tw/1L3sS5gFALINXpKDQgmp1pQxW86M3O8aoqMTlNtTGnSjATM2tjXEgCYfS3hKyuCkFHkzBeScI6WKhFVxLuD+EQLt4TkOo6CU5f1drrhvrrVly/dspDayfe+8EtQx7fuJG0HcbZLyyc1r+5qXbojtE1xa0dt4x/5c31r9hA6MYtP5DrVgijoiV5Po6KKs3MBOCVStFlgez8bG57v8/vq4tZ/Gilfr8pX7VqJm1EzJQGeg3j5/xX8ruWMbrG4oduFyXxMEFyQlkpkMeJTvhKbCMY1j/o2ykPlEmSr335KxvYPvbZydev29P65KNrX58+c92zfxv6+Kil76PnU1Sl6fe+l694//zIweMjUO1ZPnH2TU3fxqa09+l/6OHXAQgEAaSZuhddMDiaZ1epkRAzpTKAxyVzrnGh7JLreGi7qF1VqO5WvoGQ0DwF584uo3cpz4sCBzc9T9SAQPKgoqI082X2QfxhshCzXmZ5Jmoo6MvOYAk7gCWH6cudN5+98oSroZZNBoRWbuEw1ygDmqI9OZ36aJrbbTPYqIFmZrldRpdFA27ONADF4/HXxjyKYhkRU9LgYsIJ6e+pgHAkGUjkgUhLSBg2N9w3IMwpylMaKScT/n6efcC+PLN8xActmMGOhu+4bH6EpsV/yAgOoO0n9/+HnR2B5h7hr455LAPJ1+wc+1i1AYGhXOs6eQf4IR+uigYUp8WSlweZTnAWFNpz6mJ2u4d60kbEPGnUwENEvUTbVJbqTCjIAQJlPo8IXEUNdQEJcCAhMvd/gvy8Q3E6TmsbErv++Z2tRuuN/7f1X+zsNyv/vYhoN066sbVlcRuZiq/iWvuP7rEb/7LuhyPfsFPLMffdxfMnz7+1fu5qEc0RPdM6QIHLo14FgCDKRFYNMiWU1MaoAsLfupYpQwobhpDby4OfkoJ4iZQWPyy9jNLm8wLSdEtUyzvBB3lwOVwbLXYqnl6U+o3+Qo/Hnp1ttBtL+ihOZyBQXGwBS0Z9zJIGwfoYXGwTYYlLnVeWdKFwoCSqAj0/LqoW8qk7kShFiku3kK9cfCPVHyDedt/qpeyLL06zk4uXtU1DyfXfE2fPmrng0Ccjbhg+flxtq7zz3ZUzXhrU/O6sjqN73mrbXD2iY/Kzm89vbBp7Y/3VcwaOI3vqq674XdnlYysH1Ym8GajvcgekQQFURnOzZJfFEgyCCwqLtNy6mKZRrzd9RMyrUkMdR+Nfdbfu7DIBzCIaw0J5kS16edcXuNOdBXwbyU1J1ewxtvTOqxtHP/3+JIOl3xOz3v0nmr9Y+f2d8VNjp4xrbbm7jQ5mdazJdtYzasufW2r+83/H0fEE+3DTXbdNum1+Hfd4stOSZuvMURh1OXnyAPjtnsaYXeumMPAnaOwXTOb4NVYT72PqU+xG7xcf6mPNQAQX6/IUcHKmcllV1UUlBRXFZdIaYyZNUjgzJ6Rpm8u6mKrApzM0vUgYbrTrbF2SFHbS18Xa5GhSmF5P7JYqZODSiqKajIK/VYNEqQIEZRigFxShVFwJURhGD6JU0ZlDP443kvW7ccNSPH2abWFfCns140peoYDeNeZHHSqlRgkMcp00ViJSV30QKhkjagSue7JMQH4304/FkrTgKC9Tjh69VLueUScBrhFPNVAUJJTKEur6Ce0u1dCFuorNZH28UayJb2IaDjjNtKWsWmioXPicrpB365FYFc3LTU9PA+B2dlqdhUV2QCMFCAazGmNBl900ImaXkg7mVCR4KJVkyfpRJFR5F86oRckaXOFoe0m/7W6YevPVY5uWvzf1w3P7vm99YGyIHU4139VjH6ob1tLvqqpxR9u2r5m2onVI9RVXsHUX9eMTLkxQdnCc6AuVEIv2VCsq3G5XOGzt77rMZaWBtEDvNOgN0au8hkhEMg3QTPzqkVUq5feAklS7rOucMleiPU7ivc6kQtuiYCqrfNTdlVF8fxLxCKgtj3iUQC44+jrzOa06UfyDSESH3x2j106vnpWmTXnhlT1o+UfT/qt9NdGau79/Zhf73+exCP2T2Pz/ZefZXez6I/gIyv/EkRs7Yf3IFpM1FG27n5x++NQ9Q/otPPTGQSQBH/Pd/9Yf/vjjne1sx152gh0p6f3eKHwYW3/EZZ93sA627uCCpcfMzwj7AIC8WN4IKljh6miAWKkBQZHNZgqip6CSZLOSmpjVSs0yBZocIpTouZRiZWGortKL8gsDiITjI5Uik+LHJ7FXiYTziRJnywoMgWdwNFstbzxXRcbikdvy72CqiPvXAaQznI/t4Idczsm9VLdbktKzzeY83vfZ7QGDlqalDY9ZNLRSTbODPb0mZneCvyYG9BLcSxY9KQVDSTe5ArmSp7voCQYwWfE4HPqnwOu4AyOYNn/C/fPZh2fjx7C84/aZ8xev2nXHraxT3vDKpkVrHaacdQ++/xGdXTuy8Zr4NrZo3PgNgDCXI/UBnh9eKI36VZeLN+NWnxscUBNzSKpskmtiJleyNBOvSfVEKuQRD2+0Iw4l2BUdoTI+ZiikBS+9h9OfOtrxL7aJvdiOkQOHDrc2tEs72U/HmW846xyGi3DSZ3j9azd1FvUDImwoz+E2NIBd1OtGAIdVkjTZUhOTqWTlLbMzaamUcEELnGVzAbVA0BHKleew8ew2Ng534wR8gL3Dxq5ZjO/xGuQP7A55A7ubrcHDnUMBdY8RLs0Mg6L5BgnAqphMiBbFWBOzKNxLAnII3zehaKqJofOXXkp5iCsitPAkbol0bqDV8RN4ijmIm4tl7zK2BLqkUsalGqFvNN1AqVkBQDQJoSl5QlZS0MVSLhaCX7P9dHD8OHKMEwKWxLu8KBdxL6ZDTbQo3e8nNquVEFemy2DIsGlmjQdbOr9BNkt+r+zlsmTu1FB3wd0z5VlnstgW8BBwKLpv9YJL5RlPdMKNOALkU1L14E93sr+yVfg43vTxgZtW/GXnd1vevKGVHafhuOnyAlyMU3AcPjDybB377rOT591Y2mUHeYJu/Ug004jIzW+QJFm2GGhNrMaABoNsUijK3QmbMnfKFN2XPIHtjr/NdmE5uRrDZG78Xj5t2EIGAOCFiawBT+ozgRw+bSAGXiPLwM0MRsr79e4NCw4Rxa5IJL6kRnJurq0bOKEZy79hDV4k7gVL5JHn1l4AdgYS+tfxVS0wMJpjIcRkNiOAzUBl2cq/UrNZoXwP3VtwpgBXF1eWAOXEQAdVfSMRDKBcx1awhYvEZm7FB7CZETKxJf4D39CN6/Hf8XkJ6VIlly6LPUkqBVCQArccJKJUl6GXoPq6r3PD1MsbzldfSPxvRcyR3dAvmukGo9nI1bbxUPHKisdJjEQxq9QGilBcN36X0mUp6hA6Y9DpEYujXuXykscVRBpkK4wudhzbcaSC07GdfUgtRrZEms9Wzok3cw1WSi3nqklH6R3oPr8kYcedOm6WR9NMYETFagVwUFlRVM1MVW5RVLtHv11adI/EnAKwL1KEcM/JO9nv43fpSiwh81U7+qQGdrQtXseFv4FZvycdQPQ8+VKfDHgE0jgAfBZF8RpdNTGjRO01Mer6daQROSBexQQy16Hxpkj+kj3BXubXE3gz1vNr/PlDb76Bs9nSNzaSY+xxdivejVP5tZCj0mP/OYvf4smfoAvtpHU62rkEFkhGowdsNrvdbQXBV3ZNM9TENGr/TSzoRn/ZLXHoEyAo4ckJSx+au+BBspEdYacX8yA6iCb0UGXmlKkTd504Fz8rb/gchAXYat0CdkjjEZynUFmSCDVIJg9AhmYypVOVEwBXRFK5UWSV22N7Ev4uHU92T9OQe+LX7PPaKziWzWZnfL9pJMZW1bO5OPS3LSUP1S3lg9poocvnk0ySppm8njQw8cTzu4wWMA6PAZgtFm40C/WaRcikzJbSWfPzuXKqQ0sxKLdfgl3BF0A82brsgaXLW7gB12EPzH7oTqxuZWvZKtp73M0Tm+Pz4vvlDUeOLdxZwVwPk1KRVS2cQX0ce4s4n+RlpKcHICC7LeCGy4rdAbAELNlGX3ZNzCdRYyq+uhvwVHHWrRpn+IvGGoVFl/MhDadWMcJP9LZen9cr+din7JuOx/ZeN2FqnzFL7767DtWvZu2f2TrnyermlsJrn977BC7f/lkz5g4srx3e8+orqypveeqmzf8qL/13n8KGgcUDKqrHbRP6FwNIYiqrimdLCgBFNBhVKlHOuxSdv3y2lARgcoLtYrOlOn53IGEMEF7k+dXC13JCQdThQHSbDQaX08hRhsdSYuuXVBAOtyLx4BHI6+6CYLnlEXbyLfYFex/D9zz7BAf0ztqVZ+7EwHn6YufCPz33/DraBqjXfyHBI2K+RonRKAOiVZYkC3BDJ+q9VNpUJOaj+sXtVx6h57CC2dmLTMMKdPlKFXO0a4DY+dTwvZeN/qJLhrqRy8gSsx+T0e52yQh+v2ynlszMrKwci9mcnemSzdRvt6NJiOSi+EtCbgo1UyM3WkiKOMKJUtMlGvCIi78nPihD2fPbzWFJ6WPdxqngfix9q9Sr9HQdwoJDth5mUy/nm1hKoRixV/mpUJxwVT85trLi1EAa6twb+aS+9uuhNBsStmnSbVMVzTXLnPpUo6oYTYpJ0C2VLGYDkWXJqFCUkhDL9evG+ooUZ3VpjZj8Izex59h6fnXg56wfNmF/DGMtC5Pi+GHyHdka/47Y4j27dJCYyF2B7wZVlZEQEERvNFFF4QqiSgVDdslOjEH5Z65AarLLowIDZAGWchEZbA/LwDo6mozsXBTfQUqoXleVJiZ0RugfzTJISFUVEExmlYuSRP1I0IAGUcZdOgxNpl1qFqqPbALSzPPvkbfjTVJ6vIrs30m/RXi/0ykkLWUbyWw9T7KjVgXRIIFRJlTBfN2EuvH0BNZX4iUpmc0y8bOPPmIblXMHz60Xa1gA6MDkVFt/ZIKYnGpfnBa6sUmAHY9/mJhqI4S4fJ+QL55xoKIY+VYNoOZTiaaCvQtCfCFHMMy1CH34IX7GMmfKjQd/UoR8AzFIA+R3QIHeUTdBWVYkSTznFd6SVJko0DW+xLKLeyTRZYcwiGjADQ/jqVO8uP6KGOiGzmqyKN4maq1OtpHWXhja9SRIRonoRhEaJZ5K0NrOFyl//vMAAGKNdIQ+qATAwK1gBjVKRVTIdwCUpB/rioP0XWLww7EvHPD6PGRL5ZkqbKpcLx3ptW2gZ/z7GYIdmjju9pfm6E8Zq6OFTovBQvLy/P78LIMhaEkbFrNYZLfbPjjm5jWdnDM4JnvBk0Az/y+ZVYSeXlcUJWdMvMcN9+1u8h0omny9N6YT+huGr1r0xzd+Or/5xbv/On7T8Y9PswO/X3znY5MWPHHDsNfXvfono1K6rn7f+K3vx32E27h55MJbxwOBFVznDsUNTsjh7BvIojRg1Mw2n89szrWA2WPUFFDSh8QUL7iGxEC7mCz83SHi7H5mUeZ0aISzRVANCgTlw1AfH9d2D8WobftHX+7YNsMT+hpLLZbJM2ZOJJNvaZk+Q5rNdrPv2XH2t6XzFTdbPuiJ9jP3rwh0PPOXNWvWAMLoCyfoMWk2eDi6esRYymclxCubh8RkDexcM++lZZJuOTk32SdwmnJoYkjgUBQyIf4DZqJx81Mjh9525cmTzcuHVf/BTQZgFvauOZFVwBH49ZIydr4kH4iQK81M2CcaDRi9Gi+obTZhqFy7xwIOIyi6fTTdPt5ft4+oT4Q+ecShOXlPGioU/BLkji3iOnVPiAnZ9vHnOw9ON/mw7Jv+1omT5kyVp7dNmDnLjWVoRx7zq9vG4YSfTjyy5vt7ViWNk9BynD61y+DMEKROSUpzOLKcJlOm3+OkzuoYFVUUVMesmuoZHFNTel5aloiry3bI3RbgrbNeR4XKwOMJ6AVAxMMtOP2GaQZcT2aVs+/Y3zDt7LdoiJfID985vmNc3Qb61PyZM+d3NmAPdGAahth3Jx+789Eel5+4rCjB7nSOkgMeuCKa7SZElSn1+qwAPhndyHVz283akJgZqJ4bgp8v7QVDiRwWFgxH9KfOeieocBWpiZ1l+9eu3bj/ufm1o2uv6ocGOq9zCZ23rKHh3ZdLPsoafsVgoKAwtzSV26sYyiEKd0SrzFlZAwZIfRwOUqzmSkGUpIHpPXr4fJFg8Kp0K1jRqlj7qv2GxYy5Eke5wr7FpDpWXFxYWDksVqi5e1fH3BkXz+n4pxIOWz79gRHv0LneqJs2FQ76ewKfPao+pSsqEvmsj+ykQFfCF6ZeRcGFyUQK8v26El/4WGzqS33OfxjpXbL2ndc3sTfYvm9+vP3WksHVg5tvOnmsZKGTFc2buvrNabOfa5w5/drrmura10otT/ceNqZjJ5Xzew187smt/1i1bPw9We5Roeh1xYVrZ732vkM6L1UOHVlb2WcEHT5q0qRRuwBhBYC0lmeDB8LRdATw2Y0Wg8Fo9Nolp1MaEnNqJkCjR6D/JfU5336yUOPaKqJJEuCQeFQirWX7O+6YxfZjqapqE/61bQ958LsXt8S/40CwpeDekav/vh0ILAPAD7lsA1jEZFcyGsFksprtJg9Rr4kR6DJ/ZWoO7uobKtNnnyJUlrW3X3ttO14phMgLHn98yIjzPqkFgFxoY259XSt4oSTqd/L0JgaDT/NcE9PAaBctOk/sjOTEKYEwCRGJxwB6tajQpMDBcxoHXzN8CJbum6GLZe60066mRmnd+eJXN6mThXRIWPMH/Un+NdGgxLmTUKrIsmYzWa0Gg8lkN4P41WCzUcXkofbu2oTf3cjSZdpuokXRuGOyi1dx22KswGZWhYd5AffOIrF9jYxdh40sI74Et93MVivueDXr0gYPcG0ouF4DRIkAevQioLvExgPivyvuhO7qQJ5BQRgeLXS7XPrsKDMzI6PAajSaTPkuq9WRKzu46XwOzWzPRJNH7+G7krl7+OC8ePqbjJDCRIiEfKFykdziVfBd8q+ke9n++uvnTGL7vy529F437Xwso/dL097ZwvbVXz9jOnlw3rz12+LfSS1Lh1+/urZpy+F4kfhtxYuQjGCut1tMFxHAq6vrscoOoatQFU0Xx29SyV/XLRG8TS0ierkyof+ZtWWXEPbn7boC9dce3JHE5yf0pzhpostXLJYMcLnSvcYhMa9mp0Nidu8vu/xUrvPeVQMOCCQs6MzrxGVT5986ecr8W6dQmX3ELvzxh7swGyl/I6Xt6/70Qnv7mhfYKbbnQTS8jE7s8wA7B4LrOep1cC1ckMMn1Hl+RVFNlKpZmqrlcuQEq9U9hBOEwa5mQEaKzBKmSBWoSQVlTvPepDFCnPndRKFJtuemosq2GZrG9p/taZv8wfaPbt58TGf7vePdSx/wsv5K9SPtbB87/T/s7H10mU722JDgM67pTN1euaIq8dIsyh+TpOUZ+fg6PcNnz/ZanE5V4I0FhsQsv8m6iSfIBUmS5S2dL8HBXl8ook+LIkFBaLdMkafPPzxZ2v7R5zsmPXeFIQMJ22e1lq48uri9oOMZ9uLa9lNYiho3Z9+6xqU/bcBDAybXN3ZFFJ3LddVEh0mcejw5BCxZZVnUS7wGFxqlMrTMRy+JIqpdWewrCD+6iu3/sre97yvSbCP7xLR8SXyH1LKxZTYkqp/1XIZ4dpmjpLktAEU5bnchWNw5lhxTli9rcMynUdPgGPX+vJ2/2BgiqPTHK2HB5clePsGgXCkPt082oetPnbx1/bDrDtW395oycuG8yJd/3/Xu6MZHa5Zcv2zRrf2wZn1HILfzsvKx+b0rCstHz73+8VXN/8y//JriK/qHR/+30LeE6xuRa8AjToRYDHa7y2UyEIfB4fWZnHbn4JjVYrfL3HVyQt3QpktOVnRhgnBcxKOXvoLpIyFPwCO6cjK3bsas9tdeeHRt8xasYDuu+TD4aeiNN0jGwgknTn4e//yqK4UOT/Gc4zM+cENZ1E8cDrfby3t/j9NoJ7JNtumyPcmJ1sVDgItr7tQYgH+grxdrpR2zt72PpSLjsXRp7XUHt5Mj8dki4Ynt/EpI9JkPcrlm6BV1m0GWiYgIK0G0GNEuC5llKWndDU1X/x0SbTfiOtaElf/INyryZYexkjVJLfFF86aMXUzaumS4AZRtXEaWOMsoSyaOIVng81ETVTMyMjNzVEXJ9plMVLbbMxQ7yDqidR3RdPz2LIDSIO1WQ8wBsin/pGskRZpuUfew19lm7LMwJ1eRcrT7sG6R5NCsqBgvN92NPdk7uARPdt4vtTDH4m9q1lxH/PGvvE03jMkcer4XnuKKI5gApOW6bWqi+YoMaKSUSAQlGWWzQVWtfIZmMSoUAA1mj4T2S2cBqaROkYZeq3KlhdkClOu/mD2BI48cxZHsMWxja46fYO2kPwmyZ7A1fiy+DRewhcJLzK17ycs1KTC73ZrXK0koahm/Jgob/pNT8no0p9XJMTHDAFyVskQJkKKvhBlTUzxHyokifvTqgNsSaw9mmBRz7n4cwoqu+vcfR9RErqqfl+fkfr2/YcZNo8ic866XXnR8Z72xNZI450HXce2MIn+oKqkIYDYgmvQhAm8c7YR/MwyOoefSIULSSMJGySlCWEwR6LrOB4nC0uhAZiCmDrLp6+3xekDI4T38Id7D54ipCHUbcnIcfn+uNTMzIFGXy8qjKd9qSbTzYosp2hbbF7bnuBrm+REWRw08Coc18VTQ4xFQ6+EJhDmL2m6/c/OZG4cpn31T3XpmM9quH32qucGAVz7Z9jEdXMUObcyzBF8xskNVg+knbU8BIO5gJWSlYgMK7tcIpZJMAaCyhONDYlbqCOKOo0cV29lA1ylOauB7yBN7yOHlOmgGQ75bkoI52TabW3Z7qCzl/3/2IIuHzuFynuSi2BZnlftyiBSnzxyCyzwcrImh4e0Xbhz2+9mfKtWtL7xTP39x26LeM2aFPyFVQ7CnuWmyw5K3EXsOrqIfh2dPY5tNjY2nGm7QTxGQIqmCtoEHIlG/Ag4zmKnd7qNeu82mSJSaHQ5QoCRU1lYi9ElBdqqp5pwa1sv/RAMmELwQB0baym968pqFwxaOC99ePv7pgf89chFZcXX5l1NzcyPRii+nphf8lzhBwpbiQanl0rP6Dg26zurbad4v56mukCugE0Wi7Vh7JsTasSV5lIO0dJbKBcljHAhLOdJqfN6cwad7QYchPV3OyCA+n4mYMrPSXCNiBtuIGMiGNH4pGWmKygXqpwH4S8+ePzvOII575nOCTh4R15lS69q26gmSEBt94OCr7YtF6z7vlm8b7mpdcN+rL/fHcyhjZk77c8arjmflv/Bn9kZObzbAuFFEB4A0ST+d2BztZXeaidFqTfd6iV/zO51ado7Fn+avjxnT0sDFqcleG3P6QR7xs+NNXUfUIJTSVqjbjT+pBpRfbpXXFSKawsFwiBuQbNyyZcyzs2sbcS679w9k3/mvbhr+6qufy7sbvojGrt10dOm6WtZ5ttes1keObtl5BAjMBCYFpHXcnkW8R87TLC6j7EsnBrDZ8jIhM/OyYp9LSycWo2xQPZ4ctYBHz/YyHc11H2qb9S+iA4oURXyC3SM+0WGqPrVIoJJaFCmMXFRdbixfuGzBqEk3j1qwfGE43Pbogt+Nn93Y9siC8v1T6+qnzxxRO50cnPC7BcsWhCMLly6MTZs8uu2RtlBo/iNtYyYOnz6ttm7aDBHpCoDEp+PghZnR/7I53U6Plce2UaYyMYkJqxeRED/HBp/idDkbYkCRuuwmm93WEFPtdgt6FMsl5xX9mtiW3kNfypcpEhAfkgPKkCfoEXdAGF7cGCBD0YAVbOGWH374gX38448/vsOW4BViZBv3vHrfq8eO8RdyHMhFiKNCMGoniiKGmUaJSlTVsUcEbCpFdAhyJGBIAFHnAbag8wAAgUm89lnw/0o5D7g2jvTvPzOzu9KCJNSFaAKEBMYHAokSuQpiY04OODjYsWxCcjbkNaluuPdyiXuaS0jHpPfeE0N68fVO/ObSe+8uy39mVlqEzr76oeyi+bG7U3bK83yfkUZBGZwCMyKlaRaXRRTLC6E4JyfkAld4DKmpsbkrK0ttpSafxzc15nHqTVNjepQycUvmivi5NiuyMYtA0qyNo3NOVr9OFfZJmt75WUW7VMhOWtE4fsubj9zRP33SzuaW6LxFB3rWTJj4xSuvXdHyYsOAb/bpj257c+OS5s4tvmrim7appHXPputbn8kPlVdURssit194/xklXdGr7p3261Hh7uKKUGH0uu2nzi8Pxya1V5qmAUYu4UfygiRwVi0/YrQaWIvIdGcQ4pBB7dzU9snCdpLZJF/SOXJNjdRPPa0uMhVd2TKurqk5Mq5FXFPXEB0/7ucNExvqGieOb6wDIIw7lSbR99oBPqhmvm9ikm0mm7/c7yzPc+bV1IrpYEmnX1mlhbZglpActKMVbEo36zBrHWyifBGnSASrw44ZvIhr6bwgFCxiuH4R45HIul+c91p4c3j55tf/fvilPddGFx5b8zJqf5X9DCi9v/m10vvcrj6U09uHsg/0Ke/29invHSBfX7VJ+TAv99nwkcNvfNd82xjlI/4/Su+rLyi3/ObXaPaLTJb0b6xlBfCX+DHKMLqgAOoieZk65HLlmXXU56PLK/RmGI2e9HQbys4GEGweShSEA0F1mAtak3BQbR1SPGxVVo3K6irbp3YM1ToJV3pGr452r7n58XnrWi6tr79h3tY9yqTy/KbYvMvxsYvGRLrPu/BCWegef0l+cNcmpeGP/qIz6oqkNPas06Fd6BEEkMAIbZHRaUaDTKd2RMKCgERqGDdkGNkrBpBGCE4XBIMoIpOMsR4lWko4kLBqJI+K5j8Faab66Q897w8yR4ALIR3yqYfpaPGg8hFyDSo70RG06A12/oayC49HL1E/s9K3DL2QNXzKGb8fhTCZCCJkRZgzSkcQkogAAdYJoQTf6LXQWZQQHjx2hLz1I7pgEIaGErEHWAIzAAhaezTEW+S5kUqBYFHUgcViJEbamxB9uT/ROLFE8QLBIegdsp5+naSN8spKbara53ErgY4FlFnoIwadmhP5X7VaYcvuz5QHAu8h/cO3K+s89eFTJuceP+dft9utd0xUFqDpyj3kqh3K1+H6uhrlzX/ZctHQEckuSNLhJG8MjPTGCNLRbwWDZH+Fr/6Jm7D5hAmyIDMiQ0ZGTrbVkMkqRQ3FUq17vL06HSowmDyctbXd2N5201ln3XjW5a88G6uvnz2nLjJHWMg+7W0766bZL10emd02YWJ7G+NFAYSwiCGdcx+ZGTqdRB35BoSomd9sMRrSZYQkAYOKeoYC8S5MM5WnxriwyfZwnAs9I2/h3kG0RVlFY12UNylYiiCAo/gZTriVRKwOA5LAgiyuTNnkwQ4Hyucer4lJXb96j39EPHUF+JnjK/5+briipGXeqiuf3np9+4YudA6O3jbYEQv6S2bt37Cle8be7rMBwVgcxo+Ir4APJkRy7enY7QbIl/LTzVK65C8mdrvDIed4PSa5IIE5pbQ8dlABTRX6S6xu1DgHrezj3QjuuaN9/n1P7N541ards5oXtJ3REgwFWsOdE/b9v3W9wlu7a432i6at2N7wzOzzq6tvrAr76ePuDExYn+qLI0JEDyCnCdwXdyjui3uFjR/VNMjMIUk6ao6YiGZWHZ0i/DX75U5H1aEgAOK2LmrkhkxmMUmXJFnOsjrBQR/drXNlOGl7yiCq4Y2Z+zTTkbYwT8qwtv73xo0CxS6XhZtDZ7WvpVaAD0ZnlC6fNWF+vigy+yj67YoVdz/PrAF7Z8wo/9mM65SDUhQQLFSOCbslO2RAIOJINwsiAoTMFr0emUykKWYSWc8XiHtk4gMlbe5qgAb7UsMIa0IFwu6bbumd0PqX1/72IW5Tjkmn/3QfCVmPHEWCwiKd8Cj0e7KGEUURmUU6Ebk1RiCQCHSypSLhfEr/+2Eqe2hQsaNeALBCVcRlNjI7Fh1Y7Gaz0W60ySYW9pXNXt9QQI0EXB1/3PjAIiZPQYprQ3RWgnr3Xd88KXuOu/GW5v7s6Kwj6xc5btOZJpzh7hmf2cktXDiKGxPRSYI8MjopD+WfMDoJeePRSb4QbvyciNkVzReismdxFD2z4Oyi0vHr6MwOwnTUfEt8ic9KPBFjIvYqgzhkDw/xTGK3kxc9YlKPgt969IarH3/wwP4nFG9dY+PEiY2NdULbnf0v3Hr7wAu3dHR2dnTMm5cy6s2OlKZTy49OL2AW1Ib01FNiGh70BD7YIdHEB79/Oej1B9UBL+6NL0aoFonqQehRdg4ip/LxIFqsSMPn2KuMXYbaUNsyJZw1fMrGrnIA6Qpa2n5Y+TuAYvg1fgUA6eAP5Nrjj4L8IMFW+uJUVye0D51Au5h8T7W6B7CZSZlyNlXeJ75ClUs8XEnM8as+Eb9qmXpVwDBeWUH+LLTzNU5DpKiQug4YJk0jh0pMoyDbnI1lQp0JPk9rzJdhoRy8xZvKwaN4g9Cm5HHsnddbrUub3bCVWHLF4ldiF1wYPjM27aFzzp37w3lvHP3F7rOrUcnw6jY6d1dT86yJ4eiY0sOnTO6//YLru+j0cyyamXhHhoZU2lu3GPuhiOexHiQ0HfQPYqfoh9HVJ1B0w2//heIgzFQV2SMV52iKgYTCOlIxU1N0cUXaQwR7uWRYkxbXSNDfPYvXhpfEa4MpdD7OPtrg4sg4yUbMNmIRLCjNZEJsvgbgEETRbiYUvqb4syENGQkj/JFkkzkxTAQrMmlscsKiQLvUAAeUNb8G7yQ062PCs0QKkEYsI9rR6nzH9imOvcoLeLew9/ghbKIUT+hoLlq5jiPvcYqZDnXNrC6WKXZGjNP8+VlGYAXOBfY556p5+ZaodTT0KC89ZE+UXqqiG9pSFPdShT1JcXDoO1XhHnmNmZqia+gnXgMYFag1wGbucZ7cAJnQGCmivUCW3ep0GlBamtthAIqVWwGovcRJi9eKLYy8TgmP0+BgddahWmkscQqUlpiPo4MhBwPPA1tV5FzFz7cKwm9+d+CzzzahATIdd1Du/G5GoOPWnR9+ofQoyl1qHsRXeDuriLez36eUA+dUeTlUxtt7N1fgvJMpulHDv1AchOdUhXek4hxNMZBQZI1UzNQUXVzB2vvoeGkj2IAMglnogXTIjaRLBGTZYORGZXcgqMUn8260FqnLBlSM7lL+uB+Vocqr6Rhetkf5tfL7vfj3qKxH+SMavZf++VuaSiUAhD7DLeIHkgA2yIZCCEdyXJ4cuz0tB9LAW+TMK3Ab3QxXJQWpdOWImbyK8arGGFaJqpEG2V2IO/yqihEFV1Wm94Xts3tnv8iA1RevaL1x1sDRP56CjrR2UWL1/ZBiOG0+WqzyvXWXXHDpANrEwNWGNfM3DSi/fHYJ/rbsp+8e6j5uKR4aUmlIXgO18Vocrdaz1uOkKrqR6V8oDkKPqsgfqZipKbq4gr0RJcl9kqDwq4yNv3kb1KtYuCSJSmbrqZpIDiOjjbIoSpJTMDbFZEdTTJAFWdIRyZowKGrdjOZBjePIDroW0tZGwh2UUz1yNcPaH1CQ4fikjst3rbt0NcHv/agMUij5c2Vc18rz5/NZJM3JfMkD1dAaGU3tegXFxQDlWSZTbXkgUGPKKtBBcbEui2SWhkqnxEIQcFgyozFLwnGq7ZUx0g03TH/aTYLqcnOkuuX8iaFL8zhXsVAn4a3SSDRSWl1/RVfoo3fmXTau+ubIbfnTo2vnNjQ0TVjXsWQjbb4+hL9FfuGvkV+cNqai1JldVTJn7srmu+7JLfy6KLhqVGhcaeOylsh5lbWnl49r6TrnKPVMv/LO/azH5ASbVEBr5VQ+UtQfAPb2jbbEazY1vfvCE6Xna+kHfxhi6RUj001a+kAasPTikemClt4lAX+3T+GCYcUDmqJ/lKrwqwogTCEpQjeUQBBOgS2RydU1JDM/P2g3GoNBuabG7/GMKZPlsC/fW50fjVVXsyDp7OxQNJZtNo6aSoF3p+S0NFDHPHgbYiBJgQZGv/ERLZmZ0t5q6wkJKnqMhzBz8MufZG0ZXsZRzHYYrWJk1TDShwoZfiVWbn2rce4L19/03NdfPRtr2nHzvKc/emdx/d3LDyM4XkaJq+cfm/bY8bqFq1fv6FyOvX+1oHvwefbOru7Y0zcz5q91cn3Tq52bInXKZx9RCGvWp8UlOEsQzpxD6T/05acLVrNap952xtZhP0xWx0+0iY+fnCrjtT1FbQ2389oqStRWanr34n+eflDP00eNTBe09C6rWpeVidoeugYAvcGv8LTaXynTgF0DGRLXuBwA/y5J0T00eaRi6JdU8UmS4qDyuqqwJBTvUMXlkqApuriC9Vdu9UkSBIfk5fPVpZGx4MYuV46oJ+kEY0tOTnr6qEKLpcQNmZh+SJ2ImdjppB56CnnSKS02+RpiJifBU2MEnYC8izsQ2clwI9I+1YYLf3Gtkw8SVgdtm4XAwyNdtX46hDAvXCL2GCmnN3ZetuitjjuuvUr5/0PfKX9DwuFDDfpT17zfga0rz19x8fIFq84TXdXF99Wdtr1n/m5lz4fKh8pLyPrJR8gyV+hdtuva4/Mv2Lj1ih27+lg74MwMf2tPV9/aEPAZUHI97ucl3KK2k5t4PReeOJ319ZfAyRW8pRiS+gUt3aSlD6jpeSPTBS29y6C2pIDWK8yCw0JYeIl7wbKhNGJ1pqWZBQEIyYUcNwVKAXHz0vPBYdBQiw8WTxJRTWOGj2+K1tf/PFpXNzVaf2ojO+KOwcEvTpva/POG6c1EmNrUMqWhpRkIfcaHKAN0OZ81eEfOGnzxWQOjb0jBFAZx/C+zhmCNsJ9hQWsvOLVn0n5GBm1eUrt/zK5jR21o/OiJKy9AhwzKa/6alefjSoYJlXV2dVyL7IwUqpp+Qes1ytH2RjTouvnWlnFKMOP2oSGVpeD1c2ZST4ByefGmpvMavgVOruA1XMnTC0emC1p6V0B9A0u1np977PkV5qi9zXh+BQ8XJOgmziYWsLhqD+1vHQZzli2Dxi8VWsCcbXDIRM6dEpOdxEnL+CQocxLLTDtnDWdWTT4Wyh0nAU7ot8Herhf//uZLf5xv0ulUfvGjOONEDrXMYEgzK+CtE9qVsXpQVixvbB7mnLQ8CVqeut5Qc/0zNdcJKk9oH6byMk5M5VGJGk2mO108BE7wQmekxuJwGFF+vs6WAeDL0umKLHa6drMgI7HQX0YznaWSNBddcwhCLotpRQ5tBcd+ThplmiAy+BMMx2M6XcOLuERnVGvx+3WnH9vn31Wm9Cv3oTPQhPGbvaRDW9Q9dstdd/XVrfR7t8jpaBvqQuejTSZZXeCR145+8+1PDivZbnPyN+hT3SphMXhgNARhQWRMoMKEHQ6/X19RkWu3V+Xr9aEchzvgiMYCATCbfxaNmc3YJNDOmfLEZnDT4VwQvFNiQupwHj45Cp00iOdT56kG4bniI7dDo6KTeT2fSk+Ltyhf7dl5pPfHLSgb4QUvT7nsi2+R+bhTt2fL+U90tDx99FwN5Pu4fbWMBnC3/ZprdiD9/ciByqY1XcvYaf26naXlbOCeHGf7BhavuJhFHD0h/FXwSAVgZP0Zi5ozAMh6jE0ZWF4vsh39sg5pyx2NKqQzEZ2XGU+dFNAgrdc1Ne977elTUafn6kbhr2ed0XJ29tMLqh5sYBENqFX4M4lKD8Q9ehmS1eqmkUWyR8ay7CDxvRTYHVKNZ7qk8YhEdy1YcOklCy+67Pqa0tKaiorSGvGlCzavv+iCDZu7ykKhsrKqKkDwa+HPgkEygQuqIm4KNEUEQjLdBhvobPTrYvM6MzavFyCQ9fpZmoNENQebXw6qkISXvbF5mNVHiE23yjF6xRM27knfvXTUtKZoET+/fAk7F+uray7vKyjOr+KHAr4bGHqI3IN7+G5S+AS7SU0nbeih999Xlbp/qtQllG7Sj/p4jIw7kiaIOqTTySBou5KZB5gLq7jGWhvCumKTs7N6sN5L+p1zkG2h8t3HkHQFCVwRmQhIknSCRC8wvD8WUrffQHtNwbWDkz3iI84XlPdRySFI3luLeVIwEfnuWhIEtNuffHstwOzeZBl/+gzwRczUIGsiggSSZNFlkHRtI0Z+oT8E+bOoWSnwxY/oUzVPdILhSZyRP8ezp2Vz+E4SGJn/ndpNDXwrMFMaMYjsRi+qN9Luoz60qB5QH885cqO31JNM8Ua1DBJFgVlJkOt5SRihMGIaeQcIpN7Ap91gROGgt0eWkkvbi2wunXrfKIyCdLA9wszuRplAgHssUq3uc6/avnXvvku37cGf9hzou3r/LbcAELbTizQXhfm75mXsYF6m6kEvys4gbKuXAofMQuS5LUhtbJnmP9AJy8gdX3yp56m7v+Aps89kZzPacGPqPmctKUf+VkA7vpHbtCsijrgDV9RLQAg9pa0JI9VZmsxW0W/VN5vqlE12xKZeO24nRzp2bfoHPRPEf7z2SBs4vvHEBm8ApCxj83oe25YVSSeAEcaCFtqW8B8j5EX48mN//IKMjge2AeK7BW0S+6EYdkQaJaL3+XI8RW5ntmywWIrSafaLika5cnP12dklBpdLzpRy83Knx0heRt66PJxOMvMy82yFPiiEabFCndlkMzXHbNp2YiNNoxZenyxzKUghO/CtQOhvro/H5DgKdA420DrVfS4oWELdb/7qWvq7BuL7XXhXXu9CVyrtGKN5yj0hZNq9ecn93ynPj9q6VMBLtvjQpG+e6ps7ebnwys5f3ucNFDzwTXgIxqK0Tx5wFVff9zVyT//Q4+XsWgfzjp+0n6MTYDbdHRriMbs/Sh7wQyNfQ04lboD45x8nfd7MPgcMBhzF34tPQRpYGbthFXUmWnBEBixim90k62TJikTRaiW6PJLPDTwBLSYu4RpNwn+8DhpfWI1CfA+zWrZnHP5+zefKBrTh0zXKHkmuzliH39q3rwfXHT/UN3Nu1gWuZ9Wn05u0pyuGRuJWn14KAMTT4QTpzcPp0q6k3PF0dS8BvtMDAcsjIIiIQGKXQLYPAt8FgTU2uvZ8EQDruB3sL/EV7krVDmZIWNNupYoPkxTdQ3NGKoYYgS4mKQ4q76sKS0JxHADfqZupKbq4gq9wuaT6/wCVeR0IAAAAAQAAAAEZmiehT9dfDzz1AAkIAAAAAADJQhegAAAAAMnoSqH7DP2oCo0IjQABAAkAAgAAAAAAAHgBY2BkYODo/buCgYGr9zfPv0quXqAIKrgJAJZXBsIAeAFtkQOsGEEQhv/bnd272rZtG0Ft27ZtW1G9dYMiamrbZlgrqN17M89K8uVfTna/oRs4AwCUGVBCU0zQl7DAlEIZWoPOfhXUs0BbVQAL1CG0ZepQd9STPdUW9dQ61FGN+U5LpOW1pswUpmU0hZj+TGOmWnQ2lPNyV2rEoO/A+mUw0CwATG8cNjkwyXzEYZrG9Of5NUyy+XBY7Q4Hm9a8tgCH/WU4bOcwPfmsjc7GvDcYPWk7StjU2G8qAf5xwHQE6D+zHRXUbqzi96bmrEQNEeim4V965jWnB+ho0sNRHnTn7E5H0V3nQAlaAGsawqkxWKfGhDPoO2Ts/Gdwsk5fIecd011vh9O/OaegHO9toBWAfYLM5JBSxvoNquliyEeDvUucbeXvMd55vIqRtTGMJTnzAkP5bdnsXvTX6VGOPkbfYe+yRgh/6xHoLms6QDmmlvyFPThTB2PEtbczfMbr3XUu1JD7fmqUjaYre68jzpPD3wJIH6QH0RyQ5L6Ui/GeGFqDOZLiPj7iXnpkDsKJ5+TwO3LmEe8JYecb2fcazoXMC/Ed4z0J7EFS3MdH3EuPJJX07gom+ff4/DMcpS1ee85bBLQNGO84cgiqPerpVcghUBEeK/S1jzBBfUZbwUv5X/7bkOlslqCEwJ5TBw4lBFsBJdRuHA4vYk/own8RLYvLrQAAeAEc0jWMJFcQxvFnto/5LjEvHrdbmh2Kji9aPL4839TcKPNAa6mlZUyOmZk6lzbPJ3bo56//Cz+Vaqqrat5rY8x7xnzxl3nvo+27jFnz8c/mI9Nmh2XBdMsilrBitsnD9rI8aiN5DI/jSftC9mIf9pMfIB4kHiI+hWfQY5aPAYYYYYwpcyfpMMX0aZzBWZzDeVygchGXcBlX8ApexWt4HW/gLbzNbnfwLt7DJ/p0TX4+Uucji1hCnY/U+cijVB7D46jzkb3Yh/3kB4gHiYeIT+EZ9JjlY4AhRhhjytxJOkwxfRpncBbncB4XqFzEJVzGFbyCV/EaXscbeAtvs9sdvIv3cjmftWavuWs2mg6byt3ooIsFOyx77Kos2kiWsIK/UVPDOjawiQmO4CgdxnAcJzClz2PVbNKsy2ZzvoncjQ66qE2kNpHaRJawgr9RU8M6NrCJCY6gNpFjOI4TmNIn36TNfGSH5RrssKtyN+59b410iF0sUFO0l2UJtY/8jU9rWMcGNjHBEUypf0z8mm7vZLvZaC/LzdhmV2XBvpBF25IlLJOvEFfRI+NjgCFGGGNK5Rs6Z7Ij/45yNzro4m9Ywzo2sIkJjuBj2ZnvLDdjGxntLLWzLGGZfIW4ih4ZHwMMMcIYUyq1s8xkl97bH0y3JkZyM36j/+58rvTQxwBDjDDGNzyVyX35Ccjd6KCLv2EN69jAJiY4go/lfr05F+Ua7CCzGx10sYA9tiWLxCWs2BfyN+Ia1rGBTUxwBEfpMIbjOIEpfdjHvGaTd9LJb0duRp2S1O1I3Y4sYZl8hbiKHhkfAwwxwhhTKt/QOZPfmY3//Ss3Y5tNpTpL9ZQeGR8DDDHCGN/wbCbdfHO5GbW51OZSm8sSlslXiKvokfExwBAjjDGlUpvLTBY0K5KbiDcT672SbXZY6k7lbnTQxQI1h+1FeZTKY3gcT2KvTWUf9pMZIB4kHiI+xcQzxGfpfA7P4wW8yG4eT/kYYIgRxvgb9TWsYwObmOAITlI/xf7TOIOzOIfzuEDlIi7hMq7gFbyK1/A63sBbeJtvdwfv4j28zyaP8QmVL/imL/ENJ5PJHt3RqtyMbbYlPfQxwBAjjPEN9ZksqkMqN6PuV7bZy7LDtuRudNDFwzx1FI/hcTzJp73Yh/3kB4gHiYeIT+EZ9JjlY4AhRhjjb1TWsI4NbGKCIzjJlCmcxhmcxTmcxwVcxCVcxhW8glfxGl7HG3gLbzPxDt7Fe/gY/+egvq0YCAEoCNa1n+KVyTUl3Q0uIhoe+3DnRfV7nXGOc5zjHOc4xznOcY5znOMc5zjHOc5xjnOc4xznOMc5znGOc5zjHOc4xznOcY5znOMc5zjHOc5xjnOc4xznOMc5znGOc5zjHOc4xznOcY5znOM8XZouTZemS1OAKcAUYAowBZgCTAHm3x31O7p3vNf5c1iXeBkEAQDFcbsJX0IqFBwK7tyEgkPC3R0K7hrXzsIhePPK/7c77jPM1yxSPua0WmuDzNcuNmuLtmq7sbyfsUu7De/xu9fvvvDNfN3ioN9j5pq0ximd1hmd1TmlX7iky7qiq7qmG3pgXYd6pMd6oqd6pud6oZd6pdd6p/f6oI/6pC/KSxvf9F0/1LFl1naRcwwzrAu7AHNarbW6oEu6rCu6qmu6ob9Y7xu+kbfHH1ZopCk25RVrhXKn4LCO6KiOGfvpd+R3is15xXmVWKGRptgaysQKpUwc1hEdVcpEysTI7xTbKHMcKzTSFDtCmVihkab4z0FdI0QQBAEUbRz6XLh3Lc7VcI/WN54IuxXFS97oH58+MBoclE1usbHHW77wlW985wcHHHLEMSecsUuPXMNRqfzib3pcllj5xd+0lSVW5nNIL3nF6389h+Y5NG3Thja0oQ1taEMb2tCGNrQn+QwjrcwxM93gJre4Y89mvsdb3vGeD3zkE5/5wle+8Z0fHHDIEceccMaOX67wNz3747gObCQAQhCKdjlRzBVD5be7rwAmfOMQsUvPLj279OzSYBks49Ibl97In/HCuNDGO+NOW6qlWqqlWqqlWqqlWqqYUkwpphTzifnEfII92IM92IM92IM92IM92IM92I/D4/A4PA6Pw+PwODwOj8M/f7kaaDXQyt7K3mqglcCVwNVAq4FWA60GWglZCVkJWQlZCVkJWQlZDbQyqhpoNdAPh3NAwCAAwwDM+7b2sg8kCjIO4zAO4zAO4zAO4zAO4zAO4zAO4zAO4zAO4zAO4zAO47AO67AO67AO67AO67AO67AO67AO67AO67AO67AO67AO63AO53AO53AO53AO53AO53AO53AO53AO53AO53AO53AO5xCHOMQhDnGIQxziEIc4xCEOcYhDHOIQhzjEIQ5xiEMd6lCHOtShDnWoQx3qUIc61KEOdahDHepQhzrUoQ6/h+P6RpIjiKEoyOPvCARUoK9LctP5ZqXTop7q/6H/0H+4P9yfPz82bdm2Y9ee/T355bS3/divDW9reFtDb4beDL0ZejP0ZujN0JuhN0Nvht4MvRl6M/Rm6M3w1of3PVnJSlaykpWsZCUrWclKVrKSlaxkJStZySpWsYpVrGIVq1jFKlaxilWsYhWrWMUqVrGa1axmNatZzWpWs5rVrGY1q1nNalazmtWsYQ1rWMMa1rCGNaxhDWtYwxrWsIY1rGENa1nLWtaylrWsZS1rWcta1rKWtaxlLWtZyzrWsY51rGMd61jHOtaxjnWsYx3rWMc61rEeTf1o6kdTP/84rpMqCKAYhmH8Cfy2JjuLCPiYPDH1Y+rH1I+pH1M/pn5M/Zh6FEZhFEZhFEZhFEZhFEZhFFZhFVZhFVZhFVZhFVZhFVbhFE7hFE7hFE7hFE7hFE7hFCKgCChPHQFlc7I52ZxsTgQUAUVAEVAEFAFFQBFQBBQBRUARUAQUAUVAEVAEFAFFQBFQti5bl63L1mXrsnXZuggoAoqAIqAIKAKKgCKgCCgCioAioAgoAoqAIqAIKAKKgCKgCCgCyt5GQBFQBPTlwD7OEIaBKAxSOrmJVZa2TsJcwJ6r0/+9sBOGnTDshOF+DndyXG7k7vfh9+n35fft978Thp2wKuqqqKtarmq58cYbb7zzzjvvfPDBBx988sknn3zxxRdfPHnyVPip8FPhp8JPhZ8KP78czLdxBDAMAMFc/bdAk4AERoMS5CpQOW82uWyPHexkJzvZyU52spOd7GQnu9jFLnaxi13sYhe72MVudrOb3exmN7vZzW52s8EGG2ywwQYbbLDBBnvZy172spe97GUve9nLJptssskmm2yyySabbLHFFltsscUWW2yxxX6+7P+rH/qtf6+2Z3u2Z3u2Z3u2Z3u2Z3s+O66jKoYBGASA/iUFeLO2tqfgvhIgVkOshvj/8f/jF8VqiL8dqyG+d4klllhiiSWWWGKJJY444ogjjjjiiCOO+Pua0gPv7paRAHgBLcEDlNxQAADArI3Ydv7Vtm3btm3btm3btm3bD7VvBoIgLXVVqCf0ztXT9dzd3j3cvcX90CN5Snmae/p45np2e356gbeH94HP8Q3x3feH/X38NwJwoHigQ2Ba4GBQCK4NfgxVDE0OnQr7w1nCI8P7wi8jdqR4ZGzkRDQSLRmdH/0UqxTrEVsbux/PHe8b3xh/lgglzESJRJfE6MS6ZChZJzkj+RouCA9GJKQuMhI5hsZRHR2A7kZ/YZWxldhtPDPeFd+IPybyE0OIy2SIrEy2IneSX8mvFKB6UpfodPQYeiOTjmnK3GOzsCPYpexaLjdXiRvBHeJ+8BX5Lvxe/qOACmWEnsJ60SsyYjqxiLhE3CoeE6+LL8RvUlRqJXWThkszpJXSbjkq83JaOZ9cXm4gd5IXKZACK4qSSSmiVFWmq0lVUtOr+dXyagO1oxbRSM3UsmnFtOpaC62nNkqbo7M60HPppfXaemu9j77X4IwUI49RxqhrtDWOGzeM92Y985lFWWWtcdZia4d10/piU3YZu6+91j7rME5xp5szGVAgDcgBioDhYDpYDjaDE+AmeAW+p8R/A5ajfCcAAAABAAAA3QCKABYAWAAFAAIAEAAvAFwAAAEAAQsAAwABeAF9jgNuRAEYhL/aDGoc4DluVNtug5pr8xh7jj3jTpK18pszwBDP9NHTP0IPs1DOexlmtpz3sc9iOe9nmddyPsA8+XI+qI1COZ/kliIXhPkiyDo3vCnG2CaEn0+2lH+gmfIvotowZa3769ULZST4K+cujqTb/j36S4w/QmgDF0tWvalemNWLX+KSMBvYkhQSLG2FZR+afmERIsqPpn7+yvxjfMlsTjlihz3OuZE38bTtlAAa/TAFAHgBbMEDjJYBAADQ9/3nu2zbtm3b5p9t17JdQ7Zt21zmvGXXvJrZe0LA37Cw/3lDEBISIVKUaDFixYmXIJHEkkgqmeRSSCmV1NJIK530Msgok8yyyCqb7HLIKZfc8sgrn/wKKKiwIooqprgSSiqltDLKKqe8CiqqpLIqqqqmuhpqqqW2Ouqqp74GGmqksSaaaqa5FlpqpbU22mqnvQ466qSzLrrqprs9NpthprNWeWeWReZba6ctQYR5QaTplvvhp4VWm+Oyt75bZ5fffvljk71uum6fHnpaopfbervhlvfCHnngof36+Gappx57oq+PPpurv34GGGSgwTYYYpihhhthlJFGG+ODscYbZ4JJJjphoykmm2qaT7445ZkDDnrujRcOOeyY46444qirZtvtnPPOBFG+BtFBTBAbxAXxQYJC7rvjrnv/xpJXmpPDXpqXaWDg6MKZX5ZaVJycX5TK4lpalA8SdnMyMITSRjxp+aVFxaUFqUWZ+UVQQWMobcKUlgYAHQ14sAAAeAFFSzVCLEEQ7fpjH113V1ybGPd1KRyiibEhxt1vsj3ZngE9AIfgBmMR5fVk8qElsRjHOHAYW+Qwyumxct4bKxXkWDEvx7JjdszQNAZcekzi9Zho8oV8NCbnIT/fEXNRJwqmlaemnQMbN8E1OE7Mzb/P/8xzKZrEMA2hl3rQATa0Uxs2bN+2f8M2AEpwj5yQBvklvJ3AqRcEaMKrWq/19eWakl7NsZbyJoNblqlZc7KywcRbRnBjc00FeF6/enoi05EcG62tsXhkPcdk87BHVC+ZXleUPrOsUHaUI2tb4y/8OwbsTEAJAA==) format("woff")}*{box-sizing:border-box}body{padding:0;margin:0;font-family:"Open Sans","Helvetica Neue",Helvetica,Arial,sans-serif;font-size:16px;line-height:1.5;color:#606c71}a{color:#1e6bb8;text-decoration:none}a:hover{text-decoration:underline}.page-header{color:#fff;text-align:center;background-color:#159957;background-image:linear-gradient(120deg,#155799,#159957);padding:1.5rem 2rem}.page-header :last-child{margin-bottom:.5rem}@media screen and (max-width:42em){.page-header{padding:1rem 1rem}}.project-name{margin-top:0;margin-bottom:.1rem;font-size:2rem}@media screen and (max-width:42em){.project-name{font-size:1.75rem}}.project-tagline{margin-bottom:2rem;font-weight:400;opacity:.7;font-size:1.5rem}@media screen and (max-width:42em){.project-tagline{font-size:1.2rem}}.project-author,.project-date{font-weight:400;opacity:.7;font-size:1.2rem}@media screen and (max-width:42em){.project-author,.project-date{font-size:1rem}}.main-content,.toc{max-width:64rem;padding:2rem 4rem;margin:0 auto;font-size:1.1rem}.toc{padding-bottom:0}.toc .toc-box{padding:1.5rem;background-color:#f3f6fa;border:solid 1px #dce6f0;border-radius:.3rem}.toc .toc-box .toc-title{margin:0 0 .5rem;text-align:center}.toc .toc-box>ul{margin:0;padding-left:1.5rem}@media screen and (min-width:42em) and (max-width:64em){.toc{padding:2rem 2rem 0}}@media screen and (max-width:42em){.toc{padding:2rem 1rem 0;font-size:1rem}}.main-content :first-child{margin-top:0}@media screen and (min-width:42em) and (max-width:64em){.main-content{padding:2rem}}@media screen and (max-width:42em){.main-content{padding:2rem 1rem;font-size:1rem}}.main-content img{max-width:100%}.main-content h1,.main-content h2,.main-content h3,.main-content h4,.main-content h5,.main-content h6{margin-top:2rem;margin-bottom:1rem;font-weight:400;color:#159957}.main-content p{margin-bottom:1em}.main-content code{padding:2px 4px;font-family:Consolas,"Liberation Mono",Menlo,Courier,monospace;color:#567482;background-color:#f3f6fa;border-radius:.3rem}.main-content pre{padding:.8rem;margin-top:0;margin-bottom:1rem;font:1rem Consolas,"Liberation Mono",Menlo,Courier,monospace;color:#567482;word-wrap:normal;background-color:#f3f6fa;border:solid 1px #dce6f0;border-radius:.3rem;line-height:1.45;overflow:auto}@media screen and (max-width:42em){.main-content pre{font-size:.9rem}}.main-content pre>code{padding:0;margin:0;color:#567482;word-break:normal;white-space:pre;background:0 0;border:0}@media screen and (max-width:42em){.main-content pre>code{font-size:.9rem}}.main-content pre code,.main-content pre tt{display:inline;max-width:initial;padding:0;margin:0;overflow:initial;line-height:inherit;word-wrap:normal;background-color:transparent;border:0}.main-content pre code:after,.main-content pre code:before,.main-content pre tt:after,.main-content pre tt:before{content:normal}.main-content ol,.main-content ul{margin-top:0}.main-content blockquote{padding:0 1rem;margin-left:0;color:#819198;border-left:.3rem solid #dce6f0;font-size:1.2rem}.main-content blockquote>:first-child{margin-top:0}.main-content blockquote>:last-child{margin-bottom:0}@media screen and (max-width:42em){.main-content blockquote{font-size:1.1rem}}.main-content table{width:100%;overflow:auto;word-break:normal;word-break:keep-all;-webkit-overflow-scrolling:touch;border-collapse:collapse;border-spacing:0;margin:1rem 0}.main-content table th{font-weight:700;background-color:#159957;color:#fff}.main-content table td,.main-content table th{padding:.5rem 1rem;border-bottom:1px solid #e9ebec;text-align:left}.main-content table tr:nth-child(odd){background-color:#f2f2f2}.main-content dl{padding:0}.main-content dl dt{padding:0;margin-top:1rem;font-size:1rem;font-weight:700}.main-content dl dd{padding:0;margin-bottom:1rem}.main-content hr{height:2px;padding:0;margin:1rem 0;background-color:#eff0f1;border:0}code span.kw { color: #a71d5d; font-weight: normal; } 
code span.dt { color: #795da3; } 
code span.dv { color: #0086b3; } 
code span.bn { color: #0086b3; } 
code span.fl { color: #0086b3; } 
code span.ch { color: #4070a0; } 
code span.st { color: #183691; } 
code span.co { color: #969896; font-style: italic; } 
code span.ot { color: #007020; } 
</style>





</head>

<body>




<section class="page-header">
<h1 class="title toc-ignore project-name">Pig gut metagenome collection
and mining microbial genetical resource (No-published)</h1>
<h4 class="author project-author">Jinxin Meng</h4>
<h4 class="date project-date">2024-09-01</h4>
</section>



<section class="main-content">
<div id="mash_s1" class="section level2">
<h2>00.mash_s1</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">le</span> pigs_gut_metagenome_4333_file_list <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 2 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> seqkit seq <span class="at">-g</span> <span class="at">-m</span> 1500 <span class="at">-o</span> contigs/{2}.m1500.fa {3}</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">find</span> /share/data1/mjx/proj/07.pigs_binning_mcvg_20230929/00.mash_s1/contigs/ <span class="at">-size</span> <span class="st">&quot;+1M&quot;</span> <span class="kw">|</span> <span class="ex">filelist</span> <span class="op">&gt;</span> contigs_3997.filelist</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">cat</span> contigs_3997.filelist <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 2 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> seqkit seq <span class="at">-g</span> <span class="at">-m</span> 5000 {2} <span class="dt">\|</span> seqkit replace <span class="at">-p</span> <span class="dt">\&quot;</span> <span class="dt">\.\*\&quot;</span> <span class="at">-r</span> <span class="dt">\&quot;\&quot;</span> <span class="dt">\|</span> seqkit replace <span class="at">-p</span> <span class="dt">\&quot;\^\&quot;</span> <span class="at">-r</span> <span class="dt">\&quot;</span>{1}<span class="dt">\|\&quot;</span> <span class="at">-o</span> contigs_m5k/{1}.m5k.fa</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="fu">cat</span> contigs_m5k/<span class="pp">*</span> <span class="kw">|</span> <span class="ex">seqkit</span> split2 <span class="at">--by-part</span> 25 <span class="at">--by-part-prefix</span> contigs_m5k_split/contigs.m5k.part_ <span class="at">-O</span> contigs_m5k_split</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="fu">cat</span> pigs_gut_metagenome_4333_file_list <span class="kw">|</span> <span class="ex">csvtk</span> join <span class="at">-H</span> <span class="at">-t</span> <span class="at">-f</span> <span class="st">&#39;2;1&#39;</span> <span class="at">-</span> contigs_3997.filelist <span class="kw">|</span> <span class="ex">csvtk</span> cut <span class="at">-t</span> <span class="at">-H</span> <span class="at">-f</span> 1,2,5,4 <span class="op">&gt;</span> samples_3997.filelist</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="ex">mash</span> sketch <span class="at">-s</span> 100000 <span class="at">-k</span> 32 <span class="at">-p</span> 32 <span class="at">-l</span> contigs_3997.filepath <span class="at">-o</span> contigs_3997</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="fu">cat</span> contigs_3997.filelist <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 3 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> mash dist contigs_3997.msh {2} <span class="dt">\|</span> sort <span class="at">-nk3</span> <span class="dt">\&gt;</span> dist_sort/{1}.dist</span></code></pre></div>
</div>
<div id="depth_s1" class="section level2">
<h2>01.depth_s1</h2>
<div class="sourceCode" id="cb2"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f1</span> ../00.mash_s1/contigs_3997.filelist <span class="op">&gt;</span> samples_3997</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">split</span> <span class="at">-l</span> 1000 <span class="at">-d</span> 2 samples_3997 samples_3997.part_</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="fu">cat</span> samples_3997.part_00 <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 5 run_mapping.sh {} depth_00</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="fu">cat</span> samples_3997.part_01 <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 5 run_mapping.sh {} depth_01</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="fu">cat</span> samples_3997.part_02 <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 5 run_mapping.sh {} depth_02</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="fu">cat</span> samples_3997.part_03 <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 5 run_mapping.sh {} depth_03</span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> 00 01 02 03<span class="kw">;</span><span class="cf">do</span> <span class="fu">realpath</span> depth_<span class="va">$i</span>/<span class="pp">*</span>depth <span class="kw">|</span> <span class="ex">filelist</span> <span class="op">&gt;</span> depth_<span class="va">$i</span>.filelist <span class="kw">;</span><span class="cf">done</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> 00 01 02 03<span class="kw">;</span><span class="cf">do</span> <span class="fu">cat</span> samples_3997.part_<span class="va">$i</span> <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 2 run_depth_combine.pl depth_<span class="va">$i</span>.filelist {} 15 depth_combine/{} <span class="kw">;</span><span class="cf">done</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="fu">realpath</span> depth_combine/<span class="pp">*</span>depth <span class="kw">|</span> <span class="ex">filelist</span> <span class="op">&gt;</span> depth_combine.filelist</span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f2,3</span> ../00.mash_s1/samples_3997.filelist <span class="kw">|</span> <span class="ex">csvtk</span> join <span class="at">-H</span> <span class="at">-t</span> <span class="at">-f</span> 1 <span class="at">-</span> depth_combine.filelist <span class="op">&gt;</span> mcvg_config.filelist</span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a></span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a><span class="fu">cat</span> depth_<span class="pp">*</span>filelist <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-lne</span> <span class="st">&#39;@s=split;@s2=split/_/, $s[0];print &quot;$s2[0]\t$s[1]&quot; if $s2[0] eq $s2[1]&#39;</span> <span class="op">&gt;</span> depth_single.filelist</span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f2,3</span> ../00.mash_s1/samples_3997.filelist <span class="kw">|</span> <span class="ex">csvtk</span> join <span class="at">-H</span> <span class="at">-t</span> <span class="at">-f</span> 1 <span class="at">-</span> depth_single.filelist <span class="op">&gt;</span> single_config.filelist</span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a></span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a><span class="co">#metabat2 -t 24 -m 1500 -s 200000 --saveCls --unbinned --seed 2024 -i /share/data1/mjx/proj/07.pigs_binning_mcvg_20230929/00.data/contigs/CNR0176253.m1500.fa -a CNR0176253.mcvg.depth -o tmp_mcvg/CNR0176253.bin </span></span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a><span class="co">#metabat2 -t 24 -m 1500 -s 200000 --saveCls --unbinned --seed 2024 -i /share/data1/mjx/proj/07.pigs_binning_mcvg_20230929/00.data/contigs/CNR0176253.m1500.fa -a CNR0176253.single.depth -o tmp_single/CNR0176253.bin</span></span></code></pre></div>
</div>
<div id="bin_s1" class="section level2">
<h2>02.bin_s1/</h2>
<div class="sourceCode" id="cb3"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># mcvg binning</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="fu">split</span> <span class="at">-l</span> 200 <span class="at">-d</span> mcvg_config.filelist mcvg_config.filelist_part_</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="fu">seq</span> <span class="at">-w</span> 00 1 19 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">cat</span> mcvg_config.filelist_part_<span class="va">$i</span> <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 30 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> echo run_metabat2.sh {2} {3} bins_<span class="va">$i</span>/{1}.mtb2.mcvg <span class="kw">;</span><span class="cf">done</span> <span class="op">&gt;</span> run.sh</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="fu">seq</span> <span class="at">-w</span> 00 1 19 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">ls</span> <span class="at">-l</span> bins_<span class="va">$i</span>/<span class="pp">*</span> <span class="op">&gt;</span> bins_<span class="va">$i</span>.list <span class="kw">;</span><span class="cf">done</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="co"># bins 保留大于500K的基因组</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="fu">cat</span> bins<span class="pp">*</span>list <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-lne</span> <span class="st">&#39;$_=~/mjx\s+(\d+)\s.*\s(\S+)/; $x=$1; next if $x/1024 &lt; 500; $x=$2; @s=split/\//, $x; $s[-1]=~s/.fa//; @s2=split/\./, $s[-1]; print &quot;$s2[0]\t$s[-1]\t/share/data1/mjx/proj/07.pigs_binning_mcvg_20230929/02.bin_s1/metabat2_mcvg/$x&quot;&#39;</span> <span class="op">&gt;</span> bins_m500k.filelist</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="co"># single binning</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="fu">split</span> <span class="at">-l</span> 200 <span class="at">-d</span> single_config.filelist single_config.filelist_part_</span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="fu">seq</span> <span class="at">-w</span> 00 1 19 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">mkdir</span> bins_<span class="va">$i</span> <span class="kw">;</span><span class="cf">done</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a><span class="fu">seq</span> <span class="at">-w</span> 00 1 19 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">cat</span> single_config.filelist_part_<span class="va">$i</span> <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 30 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> echo run_metabat2.sh {2} {3} bins_<span class="va">$i</span>/{1}.mtb2.single <span class="kw">;</span><span class="cf">done</span> <span class="op">&gt;</span> run.sh</span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a><span class="fu">seq</span> <span class="at">-w</span> 00 1 19 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">ls</span> <span class="at">-l</span> bins_<span class="va">$i</span>/<span class="pp">*</span> <span class="op">&gt;</span> bins_<span class="va">$i</span>.list <span class="kw">;</span><span class="cf">done</span></span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a><span class="fu">cat</span> bins<span class="pp">*</span>list <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-lne</span> <span class="st">&#39;$_=~/mjx\s+(\d+)\s.*\s(\S+)/; $x=$1; next if $x/1024 &lt; 500; $x=$2; @s=split/\//, $x; $s[-1]=~s/.fa//; @s2=split/\./, $s[-1]; print &quot;$s2[0]\t$s[-1]\t/share/data1/mjx/proj/07.pigs_binning_mcvg_20230929/02.bin_s1/metabat2_single/$x&quot;&#39;</span> <span class="op">&gt;</span> bins_m500k.filelist</span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a></span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a><span class="co"># semibin</span></span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a><span class="fu">split</span> <span class="at">-l</span> 200 <span class="at">-d</span> single_config.filelist single_config.filelist_part_</span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a><span class="fu">seq</span> <span class="at">-w</span> 00 1 19 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">mkdir</span> bins_<span class="va">$i</span> <span class="kw">;</span><span class="cf">done</span></span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a><span class="fu">seq</span> <span class="at">-w</span> 00 1 19 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">cat</span> single_config.filelist_part_<span class="va">$i</span> <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 30 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> echo run_semibin2.sh {2} {3} bins_<span class="va">$i</span>/{1} <span class="kw">;</span><span class="cf">done</span> <span class="op">&gt;</span> run.sh</span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a><span class="fu">seq</span> <span class="at">-w</span> 00 1 19 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">ls</span> <span class="at">-l</span> bins_<span class="va">$i</span>/<span class="pp">*</span> <span class="op">&gt;</span> bins_<span class="va">$i</span>.list <span class="kw">;</span><span class="cf">done</span></span>
<span id="cb3-20"><a href="#cb3-20" tabindex="-1"></a><span class="fu">cat</span> bins<span class="pp">*</span>list <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-lne</span> <span class="st">&#39;$_=~/mjx\s+(\d+)\s.*\s(\S+)/; $x=$1; next if $x/1024 &lt; 500; $x=$2; @s=split/\//, $x; $s[-1]=~s/.fa//; @s2=split/\./, $s[-1]; print &quot;$s2[0]\t$s[-1]\t/share/data1/mjx/proj/07.pigs_binning_mcvg_20230929/02.bin_s1/semibin2_single/$x&quot;&#39;</span> <span class="op">&gt;</span> bins_m500k.filelist</span>
<span id="cb3-21"><a href="#cb3-21" tabindex="-1"></a></span>
<span id="cb3-22"><a href="#cb3-22" tabindex="-1"></a><span class="co"># dRep</span></span>
<span id="cb3-23"><a href="#cb3-23" tabindex="-1"></a><span class="fu">cat</span> samples_3997 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">grep</span> <span class="at">-w</span> <span class="va">$i</span> metabat2_mcvg.filelist metabat2_single.filelist semibin2_single.filelist <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-f3</span> <span class="op">&gt;</span> bins_samples/<span class="va">$i</span>.filepath <span class="kw">;</span><span class="cf">done</span></span>
<span id="cb3-24"><a href="#cb3-24" tabindex="-1"></a><span class="fu">cat</span> bins_undRep.samples <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">grep</span> <span class="at">-w</span> <span class="va">$i</span> metabat2_mcvg.filelist metabat2_single.filelist semibin2_single.filelist <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-f3</span> <span class="op">&gt;</span> bins_undRep/<span class="va">$i</span>.filepath <span class="kw">;</span><span class="cf">done</span></span>
<span id="cb3-25"><a href="#cb3-25" tabindex="-1"></a><span class="fu">find</span> bins_samples/ <span class="at">-size</span> <span class="st">&#39;0&#39;</span> <span class="op">&gt;</span> unbinning.tsv </span>
<span id="cb3-26"><a href="#cb3-26" tabindex="-1"></a><span class="fu">ls</span> bins_samples/<span class="pp">*</span> <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 10 echo run_dereplicate.sh {} bins_dRep/{/.} <span class="op">&gt;</span> run.sh</span>
<span id="cb3-27"><a href="#cb3-27" tabindex="-1"></a></span>
<span id="cb3-28"><a href="#cb3-28" tabindex="-1"></a><span class="fu">cat</span> bins_dRep/<span class="pp">*</span>info.tsv <span class="kw">|</span> <span class="ex">csvtk</span> cut <span class="at">-t</span> <span class="at">-f4,1,2</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;$1!=&quot;genome&quot;&#39;</span> <span class="op">&gt;</span> bins_final.len</span></code></pre></div>
</div>
<div id="mash_s2" class="section level2">
<h2>03.mash_s2</h2>
<div class="sourceCode" id="cb4"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">le</span> pigs_gut_metagenome_2133_file_list <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 2 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> seqkit seq <span class="at">-g</span> <span class="at">-m</span> 1500 <span class="at">-o</span> contigs/{2}.m1500.fa {3}</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="fu">find</span> /share/data1/mjx/proj/07.pigs_binning_mcvg_20230929/03.mash_s2/contigs <span class="at">-size</span> <span class="st">&quot;+1M&quot;</span> <span class="kw">|</span> <span class="ex">filelist</span> <span class="op">&gt;</span> contigs_2121.filelist</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="co">#cat contigs_2121.filelist | parallel -j 2 --colsep=&quot;\t&quot; seqkit seq -g -m 5000 {2} \| seqkit replace -p \&quot; \.\*\&quot; -r \&quot;\&quot; \| seqkit replace -p \&quot;\^\&quot; -r \&quot;{1}\|\&quot; -o contigs_m5k/{1}.m5k.fa</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co">#cat contigs_m5k/* | seqkit split2 --by-part 25 --by-part-prefix contigs_m5k_split/contigs.m5k.part_ -O contigs_m5k_split</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="fu">cat</span> pigs_gut_metagenome_2133_file_list <span class="kw">|</span> <span class="ex">csvtk</span> join <span class="at">-H</span> <span class="at">-t</span> <span class="at">-f</span> <span class="st">&#39;2;1&#39;</span> <span class="at">-</span> contigs_2121.filelist <span class="kw">|</span> <span class="ex">csvtk</span> cut <span class="at">-t</span> <span class="at">-H</span> <span class="at">-f</span> 1,2,5,4 <span class="op">&gt;</span> samples_2121.filelist</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="ex">mash</span> sketch <span class="at">-s</span> 100000 <span class="at">-k</span> 32 <span class="at">-p</span> 32 <span class="at">-l</span> contigs_2121.filepath <span class="at">-o</span> contigs_2121.msh</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="fu">cat</span> contigs_2121.filelist <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 10 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> mash dist contigs_2121.msh {2} <span class="dt">\|</span> sort <span class="at">-nk3</span> <span class="dt">\&gt;</span> dist_sort/{1}.dist</span></code></pre></div>
</div>
<div id="depth_s2" class="section level2">
<h2>04.depth_s2</h2>
<div class="sourceCode" id="cb5"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f1</span> ../03.mash_s2/contigs_2121.filelist <span class="op">&gt;</span> samples_2121</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="fu">split</span> <span class="at">-d</span> <span class="at">-l</span> 1000 samples_2121 samples_2121.part_</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="fu">cat</span> samples_2121.part_00 <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 5 run_mapping.sh {} depth_00</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="fu">cat</span> samples_2121.part_01 <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 5 run_mapping.sh {} depth_01</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="fu">cat</span> samples_2121.part_02 <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 5 run_mapping.sh {} depth_02</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="co"># 提取 m1500 contig的丰度</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="fu">perl</span> <span class="at">-e</span> <span class="st">&#39;%h;open I, &quot;samples_2121.filelist&quot;;while(&lt;I&gt;){chomp;@s=split/\t/;$h{$s[1]}=$s[2]} while(&lt;&gt;){chomp;@s=split/_/;print&quot;$s[0]\t$_\t$h{$s[0]}\n&quot;}&#39;</span> depth_00.filelist <span class="op">&gt;</span> filelist.00</span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="fu">perl</span> <span class="at">-e</span> <span class="st">&#39;%h;open I, &quot;samples_2121.filelist&quot;;while(&lt;I&gt;){chomp;@s=split/\t/;$h{$s[1]}=$s[2]} while(&lt;&gt;){chomp;@s=split/_/;print&quot;$s[0]\t$_\t$h{$s[0]}\n&quot;}&#39;</span> depth_01.filelist <span class="op">&gt;</span> filelist.01</span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="fu">perl</span> <span class="at">-e</span> <span class="st">&#39;%h;open I, &quot;samples_2121.filelist&quot;;while(&lt;I&gt;){chomp;@s=split/\t/;$h{$s[1]}=$s[2]} while(&lt;&gt;){chomp;@s=split/_/;print&quot;$s[0]\t$_\t$h{$s[0]}\n&quot;}&#39;</span> depth_02.filelist <span class="op">&gt;</span> filelist.02</span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="fu">cat</span> filelist.00 <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 5 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> <span class="at">-q</span> perl <span class="at">-e</span> <span class="st">&#39;%h; open I, &quot;$ARGV[0]&quot;; $head=&lt;I&gt;; while(&lt;I&gt;){chomp;@s=split/\t/;$h{$s[0]}=$_}; open I, &quot;$ARGV[1]&quot;; open O, &quot;&gt;$ARGV[2]&quot;; print O $head; while(&lt;I&gt;){chomp;if(/&gt;(\S+)/){print O &quot;$h{$1}\n&quot;} } close O&#39;</span> {3} {4} ../depth_00/{2}.depth</span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a><span class="fu">cat</span> filelist.01 <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 5 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> <span class="at">-q</span> perl <span class="at">-e</span> <span class="st">&#39;%h; open I, &quot;$ARGV[0]&quot;; $head=&lt;I&gt;; while(&lt;I&gt;){chomp;@s=split/\t/;$h{$s[0]}=$_}; open I, &quot;$ARGV[1]&quot;; open O, &quot;&gt;$ARGV[2]&quot;; print O $head; while(&lt;I&gt;){chomp;if(/&gt;(\S+)/){print O &quot;$h{$1}\n&quot;} } close O&#39;</span> {3} {4} ../depth_01/{2}.depth</span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a><span class="fu">cat</span> filelist.02 <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 5 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> <span class="at">-q</span> perl <span class="at">-e</span> <span class="st">&#39;%h; open I, &quot;$ARGV[0]&quot;; $head=&lt;I&gt;; while(&lt;I&gt;){chomp;@s=split/\t/;$h{$s[0]}=$_}; open I, &quot;$ARGV[1]&quot;; open O, &quot;&gt;$ARGV[2]&quot;; print O $head; while(&lt;I&gt;){chomp;if(/&gt;(\S+)/){print O &quot;$h{$1}\n&quot;} } close O&#39;</span> {3} {4} ../depth_02/{2}.depth </span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a></span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a><span class="co"># 合并depth</span></span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> 00 01 02<span class="kw">;</span><span class="cf">do</span> <span class="fu">realpath</span> depth_<span class="va">$i</span>/<span class="pp">*</span>depth <span class="kw">|</span> <span class="ex">filelist</span> <span class="op">&gt;</span> depth_<span class="va">$i</span>.filelist <span class="kw">;</span><span class="cf">done</span></span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> 00 01 02<span class="kw">;</span><span class="cf">do</span> <span class="fu">cat</span> samples_2121.part_<span class="va">$i</span> <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 2 run_depth_combine.pl depth_<span class="va">$i</span>.filelist {} 15 depth_combine/{} <span class="kw">;</span><span class="cf">done</span></span>
<span id="cb5-19"><a href="#cb5-19" tabindex="-1"></a></span>
<span id="cb5-20"><a href="#cb5-20" tabindex="-1"></a><span class="fu">realpath</span> depth_combine/<span class="pp">*</span>depth <span class="kw">|</span> <span class="ex">filelist</span> <span class="op">&gt;</span> depth_combine.filelist</span>
<span id="cb5-21"><a href="#cb5-21" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f2,3</span> ../03.mash_s2/samples_2121.filelist <span class="kw">|</span> <span class="ex">csvtk</span> join <span class="at">-H</span> <span class="at">-t</span> <span class="at">-f</span> 1 <span class="at">-</span> depth_combine.filelist <span class="op">&gt;</span> mcvg_config.filelist</span>
<span id="cb5-22"><a href="#cb5-22" tabindex="-1"></a></span>
<span id="cb5-23"><a href="#cb5-23" tabindex="-1"></a><span class="fu">cat</span> depth_<span class="pp">*</span>filelist <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-lne</span> <span class="st">&#39;@s=split;@s2=split/_/, $s[0];print &quot;$s2[0]\t$s[1]&quot; if $s2[0] eq $s2[1]&#39;</span> <span class="op">&gt;</span> depth_single.filelist</span>
<span id="cb5-24"><a href="#cb5-24" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f2,3</span> ../03.mash_s2/samples_2121.filelist <span class="kw">|</span> <span class="ex">csvtk</span> join <span class="at">-H</span> <span class="at">-t</span> <span class="at">-f</span> 1 <span class="at">-</span> depth_single.filelist <span class="op">&gt;</span> single_config.filelist</span></code></pre></div>
</div>
<div id="bin_s2" class="section level2">
<h2>05.bin_s2</h2>
<div class="sourceCode" id="cb6"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># mcvg binning</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="fu">split</span> <span class="at">-l</span> 200 <span class="at">-d</span> mcvg_config.filelist mcvg_config.filelist_part_</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="fu">seq</span> <span class="at">-w</span> 00 1 10 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">cat</span> mcvg_config.filelist_part_<span class="va">$i</span> <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 30 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> echo run_metabat2.sh {2} {3} bins_<span class="va">$i</span>/{1}.mtb2.mcvg <span class="kw">;</span><span class="cf">done</span> <span class="op">&gt;</span> run.sh</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="fu">seq</span> <span class="at">-w</span> 00 1 10 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">ls</span> <span class="at">-l</span> bins_<span class="va">$i</span>/<span class="pp">*</span> <span class="op">&gt;</span> bins_<span class="va">$i</span>.list <span class="kw">;</span><span class="cf">done</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="co"># bins 保留大于500K的基因组</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="fu">cat</span> bins<span class="pp">*</span>list <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-lne</span> <span class="st">&#39;$_=~/mjx\s+(\d+)\s.*\s(\S+)/; $x=$1; next if $x/1024 &lt; 500; $x=$2; @s=split/\//, $x; $s[-1]=~s/.fa//; @s2=split/\./, $s[-1]; print &quot;$s2[0]\t$s[-1]\t/share/data1/mjx/proj/07.pigs_binning_mcvg_20230929/05.bin_s2/metabat2_mcvg/$x&quot;&#39;</span> <span class="op">&gt;</span> bins_m500k.filelist</span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="co"># single binning</span></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a><span class="fu">split</span> <span class="at">-l</span> 200 <span class="at">-d</span> single_config.filelist single_config.filelist_part_</span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="fu">seq</span> <span class="at">-w</span> 00 1 10 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">mkdir</span> bins_<span class="va">$i</span> <span class="kw">;</span><span class="cf">done</span></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a><span class="fu">seq</span> <span class="at">-w</span> 00 1 10 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">cat</span> single_config.filelist_part_<span class="va">$i</span> <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 30 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> echo run_metabat2.sh {2} {3} bins_<span class="va">$i</span>/{1}.mtb2.single <span class="kw">;</span><span class="cf">done</span> <span class="op">&gt;</span> run.sh</span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a><span class="fu">seq</span> <span class="at">-w</span> 00 1 10 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">ls</span> <span class="at">-l</span> bins_<span class="va">$i</span>/<span class="pp">*</span> <span class="op">&gt;</span> bins_<span class="va">$i</span>.list <span class="kw">;</span><span class="cf">done</span></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a><span class="fu">cat</span> bins<span class="pp">*</span>list <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-lne</span> <span class="st">&#39;$_=~/mjx\s+(\d+)\s.*\s(\S+)/; $x=$1; next if $x/1024 &lt; 500; $x=$2; @s=split/\//, $x; $s[-1]=~s/.fa//; @s2=split/\./, $s[-1]; print &quot;$s2[0]\t$s[-1]\t/share/data1/mjx/proj/07.pigs_binning_mcvg_20230929/05.bin_s2/metabat2_single/$x&quot;&#39;</span> <span class="op">&gt;</span> bins_m500k.filelist</span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a></span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a><span class="co"># semibin</span></span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a><span class="fu">split</span> <span class="at">-l</span> 200 <span class="at">-d</span> single_config.filelist single_config.filelist_part_</span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a><span class="fu">seq</span> <span class="at">-w</span> 00 1 10 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">mkdir</span> bins_<span class="va">$i</span> <span class="kw">;</span><span class="cf">done</span></span>
<span id="cb6-18"><a href="#cb6-18" tabindex="-1"></a><span class="fu">seq</span> <span class="at">-w</span> 00 1 10 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">cat</span> single_config.filelist_part_<span class="va">$i</span> <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 30 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> echo run_semibin2.sh {2} {3} bins_<span class="va">$i</span>/{1} <span class="kw">;</span><span class="cf">done</span> <span class="op">&gt;</span> run.sh</span>
<span id="cb6-19"><a href="#cb6-19" tabindex="-1"></a><span class="fu">seq</span> <span class="at">-w</span> 00 1 10 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">ls</span> <span class="at">-l</span> bins_<span class="va">$i</span>/<span class="pp">*</span> <span class="op">&gt;</span> bins_<span class="va">$i</span>.list <span class="kw">;</span><span class="cf">done</span></span>
<span id="cb6-20"><a href="#cb6-20" tabindex="-1"></a><span class="fu">cat</span> bins<span class="pp">*</span>list <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-lne</span> <span class="st">&#39;$_=~/mjx\s+(\d+)\s.*\s(\S+)/; $x=$1; next if $x/1024 &lt; 500; $x=$2; @s=split/\//, $x; $s[-1]=~s/.fa//; @s2=split/\./, $s[-1]; print &quot;$s2[0]\t$s[-1]\t/share/data1/mjx/proj/07.pigs_binning_mcvg_20230929/05.bin_s2/semibin2_single/$x&quot;&#39;</span> <span class="op">&gt;</span> bins_m500k.filelist</span>
<span id="cb6-21"><a href="#cb6-21" tabindex="-1"></a></span>
<span id="cb6-22"><a href="#cb6-22" tabindex="-1"></a><span class="co"># dRep</span></span>
<span id="cb6-23"><a href="#cb6-23" tabindex="-1"></a><span class="fu">cat</span> samples_2121 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">grep</span> <span class="at">-w</span> <span class="va">$i</span> metabat2_mcvg.filelist metabat2_single.filelist semibin2_single.filelist <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-f3</span> <span class="op">&gt;</span> bins_samples/<span class="va">$i</span>.filepath <span class="kw">;</span><span class="cf">done</span></span>
<span id="cb6-24"><a href="#cb6-24" tabindex="-1"></a><span class="fu">find</span> bins_samples/ <span class="at">-size</span> <span class="st">&#39;0&#39;</span> <span class="op">&gt;</span> unbinning.tsv </span>
<span id="cb6-25"><a href="#cb6-25" tabindex="-1"></a><span class="fu">ls</span> bins_samples/<span class="pp">*</span> <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 10 echo run_dereplicate.sh {} bins_dRep/{/.} <span class="op">&gt;</span> run.sh</span>
<span id="cb6-26"><a href="#cb6-26" tabindex="-1"></a></span>
<span id="cb6-27"><a href="#cb6-27" tabindex="-1"></a><span class="fu">cat</span> bins_dRep/<span class="pp">*</span>info.tsv <span class="kw">|</span> <span class="ex">csvtk</span> cut <span class="at">-t</span> <span class="at">-f4,1,2</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;$1!=&quot;genome&quot;&#39;</span> <span class="op">&gt;</span> bins_final.len</span></code></pre></div>
</div>
<div id="ckm2" class="section level2">
<h2>06.ckm2</h2>
<div class="sourceCode" id="cb7"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="fu">cat</span> ../../02.bin_s1/dRep/bins_final.len ../../05.bin_s2/dRep/bins_final.len <span class="op">&gt;</span> bins.len</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="fu">cat</span> ../../02.bin_s1/dRep/bins_final.filepath ../../05.bin_s2/dRep/bins_final.filepath <span class="op">&gt;</span> bins.filepath</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="fu">split</span> <span class="at">-d</span> <span class="at">-l</span> 50000 bins.filepath bins.filepath.</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="fu">seq</span> <span class="at">-w</span> 0 1 14 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">mkdir</span> bins_<span class="va">$i</span><span class="kw">;</span><span class="cf">done</span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a><span class="fu">seq</span> <span class="at">-w</span> 0 1 14 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">cat</span> bins.filepath.00 <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 30 ln <span class="at">-s</span> {} bins_<span class="va">$i</span><span class="kw">;</span><span class="cf">done</span> <span class="kw">&amp;</span></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a><span class="co"># checkm2</span></span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a><span class="co">#seq -w 0 1 14 | while read i;do checkm2 predict --input bins/bins_$i --output-directory ckm2_$i --tmpdir /share/data1/mjx/tmp/tmp_ckm2 -x .fa --threads 120 --force;done</span></span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a><span class="fu">cat</span> ckm2_<span class="pp">*</span>/quality_report.tsv <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;NR==1 || $1!=&quot;Name&quot;&#39;</span> <span class="op">&gt;</span> ckm2_quality_report.tsv</span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a></span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a><span class="co"># gunc</span></span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a><span class="co">#seq -w 0 1 14 | while read i;do gunc run -d ckm2_$i/protein_files/ -t 120 -g -e .faa -o gunc_$i/;done</span></span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a><span class="fu">cat</span> gunc_<span class="pp">*</span>/GUNC.progenomes_2.1.maxCSS_level.tsv <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;NR==1 || $1!=&quot;genome&quot;&#39;</span> <span class="op">&gt;</span> gunc_progenomes.tsv</span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a></span>
<span id="cb7-15"><a href="#cb7-15" tabindex="-1"></a><span class="co"># qc</span></span>
<span id="cb7-16"><a href="#cb7-16" tabindex="-1"></a><span class="ex">csvtk</span> join <span class="at">--left-join</span> <span class="at">-t</span> <span class="at">-f</span> 1 ckm2_quality_report.tsv gunc_progenomes.tsv <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;{print $1&quot;\t&quot;$2&quot;\t&quot;$3&quot;\t&quot;$NF}&#39;</span> <span class="op">&gt;</span> total_bins.before_qc.tsv</span>
<span id="cb7-17"><a href="#cb7-17" tabindex="-1"></a><span class="fu">awk</span> <span class="at">-F</span> <span class="st">&#39;\t&#39;</span> <span class="st">&#39;$2&gt;=70 &amp;&amp; $3&lt;=5 &amp;&amp; ($2-$3*5)&gt;55 &amp;&amp; $4==&quot;True&quot;&#39;</span> total_bins.before_qc.tsv <span class="op">&gt;</span> total_bins.after_qc.tsv</span>
<span id="cb7-18"><a href="#cb7-18" tabindex="-1"></a><span class="fu">awk</span> <span class="at">-F</span> <span class="st">&#39;\t&#39;</span> <span class="st">&#39;$2&gt;=90 &amp;&amp; $3&lt;=5 &amp;&amp; $4==&quot;True&quot;&#39;</span> total_bins.before_qc.tsv <span class="op">&gt;</span> total_bins.high-quality.tsv</span>
<span id="cb7-19"><a href="#cb7-19" tabindex="-1"></a></span>
<span id="cb7-20"><a href="#cb7-20" tabindex="-1"></a><span class="ex">csvtk</span> join <span class="at">--left-join</span> <span class="at">-H</span> <span class="at">-t</span> <span class="at">-f</span> 1 total_bins.after_qc.tsv total_bins.filelist <span class="op">&gt;</span> total_bins.after_qc.filepath</span></code></pre></div>
</div>
<div id="isolate" class="section level2">
<h2>07.isolate</h2>
<div class="sourceCode" id="cb8"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="ex">datasets</span> summary genome taxon <span class="at">--as-json-lines</span> 2 <span class="op">&gt;</span> summary_bacteria.json </span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="ex">datasets</span> summary genome taxon <span class="at">--as-json-lines</span> 2157 <span class="op">&gt;</span> summary_archaea.json</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="ex">datasets</span> summary genome taxon <span class="at">--as-json-lines</span> 4751 <span class="op">&gt;</span> summary_fungi.json</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="ex">dataformat</span> tsv genome <span class="at">--inputfile</span> summary_archaea.json <span class="op">&gt;</span> summary_archaea.tsv</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="ex">dataformat</span> tsv genome <span class="at">--inputfile</span> summary_bacteria.json <span class="op">&gt;</span> summary_bacteria.tsv</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="ex">dataformat</span> tsv genome <span class="at">--inputfile</span> summary_fungi.json <span class="op">&gt;</span> summary_fungi.tsv</span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-i</span> <span class="st">&#39;pig\|scrofa\|swine\|boar\|porcine&#39;</span> summary_bacteria.json <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a>    <span class="fu">grep</span> <span class="at">-iv</span> <span class="st">&#39;pigeon\|feed\|product\|Pigeonpea\|Fly\|rpig&#39;</span> <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a>    <span class="fu">grep</span> <span class="at">-i</span> <span class="st">&#39;anal\|fece\|caec\|cace\|colon\|intestin\|gastrointestinal\|duodenum\|enteri\|faec\|fecal\|gastr\|gut\|ileum\|ileal\|jejunum\|rectal\|rectum\|stool\|stomach&#39;</span> <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a>    <span class="fu">grep</span> <span class="at">-iv</span> <span class="st">&#39;binning\|metagenomic assembly&#39;</span> <span class="op">&gt;</span> summary_bacteria.json.f</span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a><span class="ex">get_JsonAttributes.py</span> summary_bacteria.json.f summary_bacteria.json.f.attr</span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f1</span> summary_bacteria.json.f.attr <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-e</span> <span class="st">&#39;%h;while(&lt;&gt;){chomp;@s=split/_/;push @{$h{$s[1]}}, $s[0];}; for (keys %h){$x=join(&quot;,&quot;, @{$h{$_}}); if($x=~/,/){print &quot;GCF_$_\n&quot;}else{print &quot;GCA_$_\n&quot;}}&#39;</span> <span class="op">&gt;</span> summary_bacteria.json.f.attr.name</span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a><span class="fu">perl</span> <span class="at">-e</span> <span class="st">&#39;%h;open I, &quot;summary_bacteria.json.f.attr.name&quot;;while(&lt;I&gt;){chomp;$h{$_}=0} while(&lt;&gt;){$_=~/^\{\&quot;accession\&quot;\:\&quot;(\S+?)\&quot;/;print &quot;$_&quot; if (exists $h{$1})}&#39;</span> summary_bacteria.json.f <span class="op">&gt;</span> summary_bacteria.json.f.attr.name.json</span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a><span class="ex">dataformat</span> tsv genome <span class="at">--fields</span> accession,assminfo-assembly-method,assminfo-biosample-accession,assminfo-biosample-attribute-name,assminfo-biosample-attribute-value,assminfo-biosample-bioproject-accession,assminfo-biosample-collection-date,assminfo-biosample-description-organism-common-name,assminfo-biosample-description-organism-name,assminfo-biosample-description-organism-tax-id,assminfo-biosample-geo-loc-name,assminfo-biosample-isolation-source,assminfo-biosample-isolate,assminfo-biosample-host,assminfo-biosample-geo-loc-name,assminfo-biosample-package,assmstats-contig-n50,assmstats-genome-coverage,checkm-completeness,organism-name,organism-tax-id <span class="at">--inputfile</span> summary_bacteria.json.f.attr.name.json <span class="op">&gt;</span> summary_bacteria.json.f.attr.name.tsv</span>
<span id="cb8-16"><a href="#cb8-16" tabindex="-1"></a><span class="fu">cat</span> summary_bacteria.json.f.attr.name.tsv <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-e</span> <span class="st">&#39;$head=&lt;&gt;;$head=~s/\n//; %h; %l; while(&lt;&gt;){chomp;@s=split/\t/; if($s[0] ne $x){$l{$s[0]}=$_;$x=$s[0]}; $h{$s[0]}-&gt;{$s[3]}=$s[4];} %name;for $i(keys %h){for $j(keys $h{$i}){$name{$j}++}}; @t=sort(keys %name); print &quot;$head\t&quot;.join(&quot;\t&quot;, @t).&quot;\n&quot;; for $i(keys %h){print &quot;$l{$i}\t&quot;; @p=(); for $j(@t){if(exists $h{$i}{$j}){push @p, $h{$i}{$j} }else{push @p, &quot;&quot;;} }; print join(&quot;\t&quot;, @p).&quot;\n&quot;;}&#39;</span> <span class="op">&gt;</span> summary_bacteria.json.f.attr.name.tsv2</span>
<span id="cb8-17"><a href="#cb8-17" tabindex="-1"></a></span>
<span id="cb8-18"><a href="#cb8-18" tabindex="-1"></a><span class="ex">datasets</span> download genome accession <span class="at">--inputfile</span> ncbi_accession.tsv <span class="at">--include</span> genome <span class="at">--filename</span> ncbi_accession.tsv.zip</span>
<span id="cb8-19"><a href="#cb8-19" tabindex="-1"></a></span>
<span id="cb8-20"><a href="#cb8-20" tabindex="-1"></a><span class="co"># checkm2</span></span>
<span id="cb8-21"><a href="#cb8-21" tabindex="-1"></a><span class="ex">checkm2</span> predict <span class="at">--input</span> isolate <span class="at">--output-directory</span> ckm2 <span class="at">--tmpdir</span> /share/data1/mjx/tmp/tmp_ckm2 <span class="at">-x</span> .fa <span class="at">--threads</span> 120 <span class="at">--force</span></span>
<span id="cb8-22"><a href="#cb8-22" tabindex="-1"></a><span class="fu">cat</span> ckm2/quality_report.tsv <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;NR==1 || $1!=&quot;Name&quot;&#39;</span> <span class="op">&gt;</span> ckm2_quality_report.tsv</span>
<span id="cb8-23"><a href="#cb8-23" tabindex="-1"></a></span>
<span id="cb8-24"><a href="#cb8-24" tabindex="-1"></a><span class="co"># gunc</span></span>
<span id="cb8-25"><a href="#cb8-25" tabindex="-1"></a><span class="ex">gunc</span> run <span class="at">-d</span> ckm2/protein_files/ <span class="at">-t</span> 120 <span class="at">-g</span> <span class="at">-e</span> .faa <span class="at">-o</span> gunc</span>
<span id="cb8-26"><a href="#cb8-26" tabindex="-1"></a><span class="fu">cat</span> gunc/GUNC.progenomes_2.1.maxCSS_level.tsv <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;NR==1 || $1!=&quot;genome&quot;&#39;</span> <span class="op">&gt;</span> gunc_progenomes.tsv</span>
<span id="cb8-27"><a href="#cb8-27" tabindex="-1"></a></span>
<span id="cb8-28"><a href="#cb8-28" tabindex="-1"></a><span class="co"># qc</span></span>
<span id="cb8-29"><a href="#cb8-29" tabindex="-1"></a><span class="ex">csvtk</span> join <span class="at">--left-join</span> <span class="at">-t</span> <span class="at">-f</span> 1 ckm2_quality_report.tsv gunc_progenomes.tsv <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;{print $1&quot;\t&quot;$2&quot;\t&quot;$3&quot;\t&quot;$NF}&#39;</span> <span class="op">&gt;</span> isolate.before_qc.tsv</span>
<span id="cb8-30"><a href="#cb8-30" tabindex="-1"></a><span class="fu">awk</span> <span class="at">-F</span> <span class="st">&#39;\t&#39;</span> <span class="st">&#39;$2&gt;=70 &amp;&amp; $3&lt;=5 &amp;&amp; ($2-$3*5)&gt;55 &amp;&amp; $4==&quot;True&quot;&#39;</span> isolate.before_qc.tsv <span class="op">&gt;</span> isolate.after_qc.tsv</span>
<span id="cb8-31"><a href="#cb8-31" tabindex="-1"></a><span class="fu">awk</span> <span class="at">-F</span> <span class="st">&#39;\t&#39;</span> <span class="st">&#39;$2&gt;=90 &amp;&amp; $3&lt;=5 &amp;&amp; $4==&quot;True&quot;&#39;</span> isolate.before_qc.tsv <span class="op">&gt;</span> isolate.high-quality.tsv</span>
<span id="cb8-32"><a href="#cb8-32" tabindex="-1"></a></span>
<span id="cb8-33"><a href="#cb8-33" tabindex="-1"></a><span class="ex">csvtk</span> join <span class="at">--left-join</span> <span class="at">-H</span> <span class="at">-t</span> <span class="at">-f</span> 1 isolate.after_qc.tsv isolate.filelist <span class="op">&gt;</span> isolate.after_qc.filepath</span></code></pre></div>
</div>
<div id="drep" class="section level2">
<h2>08.dRep</h2>
<div class="sourceCode" id="cb9"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co"># 第一次聚类</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="co"># mash</span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f5</span> total_genomes.after_qc.tsv <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 20 mash sketch {} <span class="at">-s</span> 1000 <span class="at">-o</span> mash_out/{/.}</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a><span class="fu">find</span> mash_out/ <span class="at">-name</span> <span class="st">&quot;*msh&quot;</span> <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-k</span> <span class="at">-j</span> 2 <span class="at">--pipe</span> <span class="at">-n</span> 25000 <span class="st">&quot;cat &gt; mash.chunk.{#}.filepath&quot;</span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="ex">parallel</span> <span class="at">-j</span> 10 <span class="at">-k</span> mash paste mash.chunk.{} <span class="at">-l</span> mash.chunk.{}.filepath ::: <span class="dt">{</span><span class="dv">1</span><span class="dt">..</span><span class="dv">9</span><span class="dt">}</span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a><span class="co"># dist</span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a><span class="ex">parallel</span> <span class="at">-j</span> 10 mash dist <span class="at">-p</span> 30 mash.chunk.{}.msh mash.chunk.{}.msh <span class="dt">\&gt;</span> mash.chunk.{}.dist ::: <span class="dt">{</span><span class="dv">1</span><span class="dt">..</span><span class="dv">9</span><span class="dt">}</span></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a><span class="ex">parallel</span> <span class="at">-j</span> 2 tools_mashANI_cluster.py <span class="at">-i</span> mash.chunk.{}.dist <span class="at">-o</span> mash.chunk.{}.dist.clu <span class="at">-ani</span> 0.9 ::: <span class="dt">{</span><span class="dv">1</span><span class="dt">..</span><span class="dv">9</span><span class="dt">}</span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a><span class="co"># skani</span></span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a><span class="fu">seq</span> 1 1 9 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">mkdir</span> mash.chunk.<span class="va">$i</span>.skani<span class="kw">;</span> <span class="cf">done</span></span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a><span class="fu">seq</span> 1 1 9 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">perl</span> <span class="at">-e</span> <span class="st">&#39;open I, &quot;$ARGV[0]&quot;; while(&lt;I&gt;){chomp; next if /cluster/; @s=split/\s+/; if($s[1] ne $x){open O, &quot;&gt;$ARGV[1]/$s[1].filelist&quot;; print O &quot;$s[0]\n&quot;; $x=$s[1] }else{ print O &quot;$s[0]\n&quot; } }&#39;</span> mash.chunk.<span class="va">$i</span>.dist.clu mash.chunk.<span class="va">$i</span>.skani <span class="kw">;</span><span class="cf">done</span></span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a><span class="fu">seq</span> 1 1 9 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">perl</span> <span class="at">-e</span> <span class="st">&#39;open I, &quot;ls $ARGV[0]/*filelist|&quot;; while(&lt;I&gt;){chomp;$_=~/(\d+).filelist/; print &quot;skani dist --ql $_ --rl $_ -o $ARGV[0]/$1.skani --min-af 30 -t 20\n&quot;}&#39;</span> mash.chunk.<span class="va">$i</span>.skani <span class="kw">;</span><span class="cf">done</span> <span class="op">&gt;</span> run_skani.sh</span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a><span class="fu">seq</span> 1 1 9 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">perl</span> <span class="at">-e</span> <span class="st">&#39;open I, &quot;ls $ARGV[0]/*skani|&quot;; while(&lt;I&gt;){chomp;$_=~/\/(\d+).skani/; print &quot;tools_skANI_cluster.py -i $_ -o $ARGV[0]/$1.clu -nc 0.3 -ani 0.95\n&quot;}&#39;</span> mash.chunk.<span class="va">$i</span>.skani <span class="kw">;</span><span class="cf">done</span> <span class="op">&gt;</span> run_clu.sh</span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a><span class="co"># extract clu info</span></span>
<span id="cb9-15"><a href="#cb9-15" tabindex="-1"></a><span class="fu">seq</span> 1 1 9 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">ls</span> mash.chunk.<span class="va">$i</span>.skani/<span class="pp">*</span>clu <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 20 <span class="at">-q</span> perl <span class="at">-e</span> <span class="st">&#39;open I, &quot;$ARGV[0]&quot;;$ARGV[0]=~/(\d+).clu/;$x=$1; readline(I);while(&lt;I&gt;){chomp;@s=split/\t/;print &quot;$s[0]\t$x\t$s[1]\t$x\_$s[1]\n&quot;}&#39;</span> {} <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb9-16"><a href="#cb9-16" tabindex="-1"></a>    <span class="ex">csvtk</span> add-header <span class="at">-t</span> <span class="at">-n</span> <span class="st">&quot;name,primary_cluster,second_cluster,final_cluster&quot;</span> <span class="op">&gt;</span> mash.chunk.<span class="va">$i</span>.skani.Cdb.tsv <span class="kw">;</span><span class="cf">done</span></span>
<span id="cb9-17"><a href="#cb9-17" tabindex="-1"></a><span class="co"># filter</span></span>
<span id="cb9-18"><a href="#cb9-18" tabindex="-1"></a><span class="ex">parallel</span> <span class="at">-j</span> 10 parse_dRep.pl mash.chunk.{}.skani.Cdb.tsv total_genomes.after_qc.tsv mash.chunk.{}.skani.Cdb.info ::: <span class="kw">`</span><span class="fu">seq</span> 1 1 9<span class="kw">`</span></span>
<span id="cb9-19"><a href="#cb9-19" tabindex="-1"></a></span>
<span id="cb9-20"><a href="#cb9-20" tabindex="-1"></a><span class="co"># 第二次聚类</span></span>
<span id="cb9-21"><a href="#cb9-21" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f3</span> <span class="pp">*</span>Cdb.info <span class="kw">|</span> <span class="fu">sed</span> <span class="st">&#39;s/^/mash_out\//g;s/$/.msh/g&#39;</span> <span class="op">&gt;</span> sec.mash.filepath</span>
<span id="cb9-22"><a href="#cb9-22" tabindex="-1"></a><span class="ex">mash</span> paste sec.mash <span class="at">-l</span> sec.mash.filepath</span>
<span id="cb9-23"><a href="#cb9-23" tabindex="-1"></a><span class="ex">mash</span> dist <span class="at">-p</span> 50 sec.mash.msh sec.mash.msh <span class="op">&gt;</span> sec.mash.dist</span>
<span id="cb9-24"><a href="#cb9-24" tabindex="-1"></a><span class="ex">tools_mashANI_cluster.py</span> <span class="at">-i</span> sec.mash.dist <span class="at">-o</span> sec.mash.dist.clu <span class="at">-ani</span> .9</span>
<span id="cb9-25"><a href="#cb9-25" tabindex="-1"></a><span class="fu">perl</span> <span class="at">-ne</span> <span class="st">&#39;next if /cluster/;@s=split/\s+/;if($s[1] ne $x){open O, &quot;&gt;sec.mash.skani/$s[1].filelist&quot;; print O &quot;$s[0]\n&quot;; $x=$s[1] }else{print O &quot;$s[0]\n&quot;}&#39;</span> sec.mash.dist.clu</span>
<span id="cb9-26"><a href="#cb9-26" tabindex="-1"></a><span class="fu">perl</span> <span class="at">-e</span> <span class="st">&#39;open I, &quot;ls sec.mash.skani/*filelist|&quot;; while(&lt;I&gt;){chomp;$_=~/(\d+).filelist/; print &quot;skani dist --ql $_ --rl $_ -o sec.mash.skani/$1.skani --min-af 30 -t 20\n&quot;}&#39;</span> <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 10</span>
<span id="cb9-27"><a href="#cb9-27" tabindex="-1"></a><span class="fu">perl</span> <span class="at">-e</span> <span class="st">&#39;open I, &quot;ls sec.mash.skani/*skani|&quot;; while(&lt;I&gt;){chomp;$_=~/\/(\d+).skani/; print &quot;tools_skANI_cluster.py -i $_ -o sec.mash.skani/$1.clu -nc 0.3 -ani 0.95\n&quot;}&#39;</span> <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 50</span>
<span id="cb9-28"><a href="#cb9-28" tabindex="-1"></a><span class="fu">ls</span> sec.mash.skani/<span class="pp">*</span>clu <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 20 <span class="at">-q</span> perl <span class="at">-e</span> <span class="st">&#39;open I, &quot;$ARGV[0]&quot;;$ARGV[0]=~/(\d+).clu/;$x=$1; readline(I);while(&lt;I&gt;){chomp;@s=split/\t/;print &quot;$s[0]\t$x\t$s[1]\t$x\_$s[1]\n&quot;}&#39;</span> {} <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb9-29"><a href="#cb9-29" tabindex="-1"></a>    <span class="ex">csvtk</span> add-header <span class="at">-t</span> <span class="at">-n</span> <span class="st">&quot;name,primary_cluster,second_cluster,final_cluster&quot;</span> <span class="op">&gt;</span> sec.mash.skani.Cdb.tsv</span>
<span id="cb9-30"><a href="#cb9-30" tabindex="-1"></a><span class="ex">parse_dRep.pl</span> sec.mash.skani.Cdb.tsv total_genomes.after_qc.tsv sec.mash.skani.Cdb.info</span>
<span id="cb9-31"><a href="#cb9-31" tabindex="-1"></a></span>
<span id="cb9-32"><a href="#cb9-32" tabindex="-1"></a><span class="fu">cat</span> mash.chunk.<span class="pp">*</span>.skani.Cdb.info <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-e</span> <span class="st">&#39;%clu;while(&lt;&gt;){chomp;@s=split/\t/;$clu{$s[2]}=$s[3]}; open I, &quot;sec.mash.skani.Cdb.info&quot;; while(&lt;I&gt;){chomp; @s=split/\t/; @s2=split/,/,$s[3]; %h=(); for (@s2){@x=split/,/,$clu{$_}; for (@x){$h{$_}++} }; @out=keys %h; $len=scalar @out;print &quot;$s[0]\t$len\t$s[2]\t&quot;.join(&quot;,&quot;, @out).&quot;\n&quot;}&#39;</span> <span class="op">&gt;</span> final.clu.info</span></code></pre></div>
</div>
<div id="taxa" class="section level2">
<h2>09.taxa</h2>
<div class="sourceCode" id="cb10"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="fu">cat</span> xx2 <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-e</span> <span class="st">&#39;$x=1;while(&lt;&gt;){chomp;printf(&quot;PIG_GENOME%04s\t&quot;, $x);print &quot;$_\n&quot;;$x+=1}&#39;</span> <span class="op">&gt;</span> kraken_build.genome.metadata</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="ex">kraken_build.py</span> <span class="at">-l</span> kraken_build.genome.filelist.tsv <span class="at">-t</span> kraken_build.genome.taxonomy.tsv <span class="at">-d</span> kraken2db_150mers <span class="at">-p</span> 30 <span class="at">-m</span> 150</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a><span class="ex">kraken_build.py</span> <span class="at">-l</span> kraken_build.genome.filelist.tsv <span class="at">-t</span> kraken_build.genome.taxonomy.tsv <span class="at">-d</span> kraken2db_100mers <span class="at">-p</span> 30 <span class="at">-m</span> 100</span></code></pre></div>
</div>
<div id="geneset" class="section level2">
<h2>10.geneset</h2>
<div class="sourceCode" id="cb11"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f1,5</span> ../08.dRep/total_genomes.after_qc.tsv <span class="kw">|</span> <span class="fu">split</span> <span class="at">-d</span> <span class="at">-l</span> 30000 <span class="at">-</span> total_genomes.filelist.chunk_</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f2</span> total_genomes.filelist.chunk_01 <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-e</span> <span class="st">&#39;while(&lt;&gt;){chomp;open I, &quot;$_&quot;; $_=~/.*\/(\S+).fa/;$n=$1;while(&lt;I&gt;){chomp;if(/&gt;(\S+)/){print &quot;&gt;$n\|$1\n&quot;}else{print &quot;$_\n&quot;}}}&#39;</span> <span class="op">&gt;</span> total_genomes.filelist.chunk_01.fa</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="co"># 基因预测，长度过滤，名称修改</span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a><span class="ex">flow_prodigal.sh</span> total_genomes.filelist.chunk_01.fa total_genomes.filelist.chunk_01 100</span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a><span class="ex">seqkit</span> seq <span class="at">-g</span> <span class="at">-m</span> 100 total_genomes.filelist.chunk_01.ffn <span class="kw">|</span> <span class="ex">seqkit</span> replace <span class="at">-p</span> <span class="st">&quot;\s.*&quot;</span> <span class="at">-r</span> <span class="st">&quot;&quot;</span> <span class="at">-o</span> total_genomes.filelist.chunk_01.ffn.f</span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a><span class="co"># 聚类</span></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a><span class="ex">mmseqs</span> easy-cluster gene.ffn cluster . <span class="at">--cluster-mode</span> 2 <span class="at">--cov-mode</span> 1 <span class="at">--min-seq-id</span> 0.95 <span class="at">-c</span> 0.9 <span class="at">--kmer-per-seq-scale</span> 0.8 <span class="at">--threads</span> 112</span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a><span class="co"># 蛋白序列提取</span></span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a><span class="fu">ls</span> /share/data1/mjx/proj/04.black_pig_metagenome_20230529/10.geneset/prodigal/<span class="pp">*</span>faa <span class="op">&gt;</span> gene.faa.list</span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a><span class="fu">perl</span> <span class="at">-e</span> <span class="st">&#39;while(&lt;&gt;){chomp; open I, &quot;$_&quot;; while(&lt;I&gt;){chomp;if(/^(\S+)/){print &quot;$1\n&quot;}else{print &quot;$_\n&quot;}}}&#39;</span> gene.faa.list <span class="op">&gt;</span> gene.faa</span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a><span class="co"># 最终</span></span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a><span class="fu">cp</span> mmseqs/cluster_rep_seq.fasta geneset.ffn</span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a><span class="ex">seqkit</span> fx2tab <span class="at">-n</span> <span class="at">-l</span> geneset.ffn <span class="at">-o</span> geneset.len</span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f1</span> geneset.len <span class="op">&gt;</span> geneset.id</span>
<span id="cb11-15"><a href="#cb11-15" tabindex="-1"></a><span class="ex">seqkit</span> grep <span class="at">-f</span> geneset.id mmseqs/gene.faa <span class="at">-o</span> geneset.faa</span>
<span id="cb11-16"><a href="#cb11-16" tabindex="-1"></a></span>
<span id="cb11-17"><a href="#cb11-17" tabindex="-1"></a><span class="co"># map</span></span>
<span id="cb11-18"><a href="#cb11-18" tabindex="-1"></a><span class="ex">minimap2</span> <span class="at">-I</span> 100g <span class="at">-d</span> geneset_full.mmi geneset_full/cluster_rep_seq.fasta</span>
<span id="cb11-19"><a href="#cb11-19" tabindex="-1"></a><span class="fu">cat</span> samples.6574.filelist <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 6 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> run_minimap2.sh {2} geneset_full.mmi geneset_full_ab/{1}</span>
<span id="cb11-20"><a href="#cb11-20" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f1</span> samples.6574.filelist <span class="kw">|</span> <span class="fu">head</span> <span class="at">-1500</span> <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 5 samtools coverage geneset_full_ab/{}.sort.bam <span class="dt">\|</span> awk <span class="dt">\&#39;\$</span>4<span class="dt">\!\=</span>0<span class="dt">\&#39;</span> <span class="dt">\&gt;</span> geneset_full_ab/{}.cvg</span></code></pre></div>
</div>
<div id="euk" class="section level2">
<h2>12.euk</h2>
<div class="sourceCode" id="cb12"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="co"># bin大小大于5M认为是潜在的真核生物bin，进行busco初级分析</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="fu">cat</span> ../../02.bin_s1/dRep/bins_final.len ../../05.bin_s2/dRep/bins_final.len <span class="op">&gt;</span> bins.len</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="fu">awk</span> <span class="st">&#39;$3&gt;5000000&#39;</span> bins.len <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-rnk3,3</span> <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-f2</span> <span class="op">&gt;</span> bins.len.m5M.filepath</span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a></span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a><span class="fu">cat</span> total_bins.euk.list <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 20 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> busco <span class="at">-i</span> {2} <span class="at">-c</span> 4 <span class="at">-o</span> busco_out/{1} <span class="at">-m</span> geno <span class="at">-l</span> /share/data1/database/busco_v5/busco_downloads/lineages/fungi_odb10/ <span class="at">--offline</span></span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a><span class="fu">ls</span> <span class="pp">*</span>/<span class="pp">*</span>txt <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-ne</span> <span class="st">&#39;chomp;$_=~/odb10.(\S+).txt/;$x=$1;open I, &quot;$_&quot;; while (&lt;I&gt;) {if(/\s+C:(\S+?)%(\S+)/){print &quot;$x\t$1\t$_&quot;}}&#39;</span> <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-rnk2</span> <span class="op">&gt;</span> ../busco_out.tsv</span></code></pre></div>
</div>
<div id="bgc" class="section level2">
<h2>13.BGC</h2>
<div class="sourceCode" id="cb13"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="fu">cat</span> 13.BGC/r1.sh </span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="co"># 对基因组进行BGCs预测</span></span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a><span class="ex">seqkit</span> split <span class="at">-p</span> 100 chunk_00.m5000.fa <span class="at">-O</span> chunk_00</span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a><span class="fu">ls</span> <span class="pp">*</span>fa <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 20 antismash <span class="at">-c</span> 6 <span class="at">--output-dir</span> {.}.BGC_out <span class="at">--genefinding-tool</span> prodigal <span class="at">--allow-long-headers</span> {}</span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-d</span> <span class="pp">*</span>out <span class="kw">|</span> <span class="fu">sed</span> <span class="st">&#39;s/.BGC_out//&#39;</span> <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 40 parse_antismash.py {}.BGC_out/index.html {}.tsv</span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a><span class="fu">ls</span> chunk_<span class="pp">*</span>/<span class="pp">*</span>tsv <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-ne</span> <span class="st">&#39;chomp;$x=$_;$x=~s/.tsv//;open I, &quot;&lt;$_&quot;;while(&lt;I&gt;){chomp;if(!/contigs/){print &quot;$x\t$_\n&quot;}}&#39;</span> <span class="op">&gt;</span> total_BGCs_find.tsv</span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a></span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a><span class="co"># 如果BGCs所在的contigs小于5K，则过滤；和MIBIG比较</span></span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a><span class="fu">find</span> /share/data1/mjx/proj/07.pigs_binning_mcvg_20230929/13.BGC/ <span class="at">-name</span> <span class="st">&quot;*region*.gbk&quot;</span> <span class="op">&gt;</span> total_BGCs.filepath</span>
<span id="cb13-10"><a href="#cb13-10" tabindex="-1"></a><span class="fu">cat</span> total_BGCs.filepath <span class="kw">|</span> <span class="ex">filelist</span> <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-ne</span> <span class="st">&#39;chomp; @s=split/\t/; open I, &quot;$s[1]&quot;; print &quot;$_\t&quot;; while(&lt;I&gt;){chomp;if(/LOCUS\s+(\S+)\s+(\S+)\s+bp/){print &quot;$1\t$2\n&quot;}}&#39;</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;$4&gt;5000&#39;</span> <span class="op">&gt;</span> total_BGCs.info.filter</span>
<span id="cb13-11"><a href="#cb13-11" tabindex="-1"></a><span class="ex">bigscape</span> <span class="at">-i</span> total_BGCs.m5k.gbk <span class="at">-o</span> total_BGCs.bigscape <span class="at">--pfam_dir</span> /share/data1/database/pfam/release36/ <span class="at">-c</span> 100 <span class="at">--mibig</span></span>
<span id="cb13-12"><a href="#cb13-12" tabindex="-1"></a></span>
<span id="cb13-13"><a href="#cb13-13" tabindex="-1"></a><span class="co"># To prevent sampling biases in quantitative analysis (taxonomic and functional compositions of GCCs/GCFs, GCF and GCC distances to reference databases as well as GCF </span></span>
<span id="cb13-14"><a href="#cb13-14" tabindex="-1"></a><span class="co"># metagenomic abundances), the 73,864 BGCs were further dereplicated by retaining only the longest BGC per GCF per species, resulting in a total of 30,182 BGCs.</span></span>
<span id="cb13-15"><a href="#cb13-15" tabindex="-1"></a><span class="co"># 确定新颖性</span></span>
<span id="cb13-16"><a href="#cb13-16" tabindex="-1"></a><span class="ex">bigslice</span> <span class="at">--query</span> BGCs.gbk.m5k.rep <span class="at">--query_name</span> jinxmeng_run0006 <span class="at">--n_ranks</span> 3 /share/data1/database/bigslice/full_run_result/ <span class="at">--run_id</span> 0006</span>
<span id="cb13-17"><a href="#cb13-17" tabindex="-1"></a><span class="ex">extract_data_from_bigscape.py</span> <span class="at">-i</span> /share/data1/database/bigslice/full_run_result/reports/1/data.db <span class="at">-tn</span> bgc <span class="at">-o</span> bigslice.out.bgc</span>
<span id="cb13-18"><a href="#cb13-18" tabindex="-1"></a><span class="ex">extract_data_from_bigscape.py</span> <span class="at">-i</span> /share/data1/database/bigslice/full_run_result/reports/1/data.db <span class="at">-tn</span> gcf_membership <span class="at">-o</span> bigslice.out.gcf_membership</span>
<span id="cb13-19"><a href="#cb13-19" tabindex="-1"></a><span class="ex">csvtk</span> join <span class="at">--left-join</span> <span class="at">-t</span> <span class="at">-f</span> <span class="st">&quot;2;1&quot;</span> bigslice.out.gcf_membership bigslice.out.bgc <span class="op">&gt;</span> bigslice.out.metadata</span>
<span id="cb13-20"><a href="#cb13-20" tabindex="-1"></a><span class="ex">le</span> bigslice.out.metadata <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-e</span> <span class="st">&#39;%h;while(&lt;&gt;){chomp;next if /gcf_id/;@s=split/\t/;$h{$s[1]}{$s[2]}=$_;} for (keys %h){@x=sort keys %{$h{$_}};print &quot;$h{$_}{$x[-1]}\n&quot;}&#39;</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;$3&gt;900&#39;</span> <span class="op">&gt;</span> bigslice.out.metadata.d_m900</span>
<span id="cb13-21"><a href="#cb13-21" tabindex="-1"></a></span>
<span id="cb13-22"><a href="#cb13-22" tabindex="-1"></a><span class="co"># 聚类分析</span></span>
<span id="cb13-23"><a href="#cb13-23" tabindex="-1"></a><span class="co"># cat BGCs.gbk.m5k.rep.list | perl -lne &#39;$_=~/(\S+?)\.re/;print &quot;mkdir -p bigslice.in/dataset_1/$1&quot;&#39; | sort -u | parallel -j 30</span></span>
<span id="cb13-24"><a href="#cb13-24" tabindex="-1"></a><span class="co"># cat BGCs.gbk.m5k.rep.list | perl -lne &#39;$_=~/(\S+?)\.re/;print &quot;cp BGCs.gbk.m5k/$_\.gbk bigslice.in/dataset_1/$1/&quot;&#39; | parallel -j 30</span></span>
<span id="cb13-25"><a href="#cb13-25" tabindex="-1"></a><span class="co"># for i in *.gbk; do sed -i &#39;s/Version      :: False/Version      :: 6.1.1/&#39; $i; done # 修改MIBiG数据库的格式，才能被bigslice读入，进行聚类分析</span></span>
<span id="cb13-26"><a href="#cb13-26" tabindex="-1"></a><span class="co"># bigslice -i bigslice.in bigslice.out # 聚类分析</span></span>
<span id="cb13-27"><a href="#cb13-27" tabindex="-1"></a><span class="co"># bash bigslice.out/start_server.sh 33001</span></span>
<span id="cb13-28"><a href="#cb13-28" tabindex="-1"></a></span>
<span id="cb13-29"><a href="#cb13-29" tabindex="-1"></a><span class="co"># 定量</span></span>
<span id="cb13-30"><a href="#cb13-30" tabindex="-1"></a><span class="fu">cat</span> BGCs.gbk.m5k.rep.filepath <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 60 extract_biosynthetic_cds.py {} {.}.fa <span class="kw">&amp;</span></span>
<span id="cb13-31"><a href="#cb13-31" tabindex="-1"></a><span class="fu">cat</span> BGCs.gbk.m5k.rep/<span class="pp">*</span>fa <span class="kw">|</span> <span class="ex">seqkit</span> replace <span class="at">-p</span> <span class="st">&quot;.*\/&quot;</span> <span class="at">-r</span> <span class="st">&quot;&quot;</span> <span class="op">&gt;</span> BGCs.gbk.m5k.rep.biosyn.fa</span>
<span id="cb13-32"><a href="#cb13-32" tabindex="-1"></a><span class="ex">minimap2</span> <span class="at">-d</span> BGCs.gbk.m5k.rep.biosyn.mmi BGCs.gbk.m5k.rep.biosyn.fa</span>
<span id="cb13-33"><a href="#cb13-33" tabindex="-1"></a><span class="fu">cat</span> ../clean_fq_list <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 8 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> run_minimap2.sh {2} {3} BGCs.gbk.m5k.rep.biosyn.mmi profile/{1} <span class="kw">&amp;</span></span>
<span id="cb13-34"><a href="#cb13-34" tabindex="-1"></a><span class="fu">cat</span> ../17.public_wp/CNP0000824.fq.list <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 8 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> run_minimap2.sh {2} {3} BGCs.gbk.m5k.rep.biosyn.mmi profile/{1} <span class="kw">&amp;</span></span>
<span id="cb13-35"><a href="#cb13-35" tabindex="-1"></a><span class="ex">le</span> clean_fq.list <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 5 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> samtools coverage <span class="at">-d</span> 0 <span class="at">-o</span> profile/{1}.cvg profile/{1}.sort.bam</span>
<span id="cb13-36"><a href="#cb13-36" tabindex="-1"></a><span class="ex">combine_file_zy_folder_allsample.py</span> <span class="at">-D</span> profile/ <span class="at">-suffix</span> .cvg <span class="at">-o</span> profile.rc <span class="at">-t</span> 1 <span class="at">-n</span> 1 <span class="at">-v</span> 4</span>
<span id="cb13-37"><a href="#cb13-37" tabindex="-1"></a><span class="ex">combine_file_zy_folder_allsample.py</span> <span class="at">-D</span> profile/ <span class="at">-suffix</span> .cvg <span class="at">-o</span> profile.cvg <span class="at">-t</span> 1 <span class="at">-n</span> 1 <span class="at">-v</span> 6</span>
<span id="cb13-38"><a href="#cb13-38" tabindex="-1"></a><span class="ex">profile_filter_by_cvg.py</span> profile.rc profile.cvg 50 profile.rc2</span>
<span id="cb13-39"><a href="#cb13-39" tabindex="-1"></a><span class="ex">profile_filter.py</span> profile.rc2 profile.rc3</span>
<span id="cb13-40"><a href="#cb13-40" tabindex="-1"></a></span>
<span id="cb13-41"><a href="#cb13-41" tabindex="-1"></a><span class="co"># seqkit fx2tab -n -l BGCs.gbk.m5k.rep.biosyn.fa -o BGCs.gbk.m5k.rep.biosyn.len</span></span>
<span id="cb13-42"><a href="#cb13-42" tabindex="-1"></a><span class="co"># profile_rc2tpm.pl profile.rc3 BGCs.gbk.m5k.rep.biosyn.len profile.rc3.tpm</span></span>
<span id="cb13-43"><a href="#cb13-43" tabindex="-1"></a></span>
<span id="cb13-44"><a href="#cb13-44" tabindex="-1"></a><span class="co"># perl -ne &#39;chomp;print &quot;$_&quot; if /name/;$_=~/\|(.*)$/;print &quot;$1\n&quot;&#39; PULs.filter.rc &gt; PULs.filter.rc2</span></span>
<span id="cb13-45"><a href="#cb13-45" tabindex="-1"></a><span class="ex">profile_summary_by_group.py</span> PULs.rc.filter PULs.summary.rc</span></code></pre></div>
</div>
<div id="kegg" class="section level2">
<h2>14.KEGG</h2>
<div class="sourceCode" id="cb14"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="co"># gene2KO</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="ex">diamond</span> blastp <span class="at">-d</span> /share/data1/database/KEGG/KEGG20230401.dmnd <span class="at">--outfmt</span> 6 <span class="at">--min-score</span> 60 <span class="at">--query-cover</span> 50 <span class="at">--max-target-seqs</span> 10 <span class="at">-p</span> 112 <span class="at">-q</span> ../10.geneset/cluster_rep_seq.faa <span class="at">-o</span> kegg.btp <span class="op">&gt;</span>/dev/null <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span></span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a><span class="fu">perl</span> <span class="at">-ne</span> <span class="st">&#39;chomp;@s=split/\t/;if($s[0] ne $a){$s[1]=~/\|(.*)/;print &quot;$s[0]\t$1\n&quot;;$a=$s[0]}&#39;</span> kegg.btp <span class="op">&gt;</span> kegg.tsv</span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a></span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a><span class="co"># tpm</span></span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a><span class="fu">cat</span> geneset.cvg.filelist <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 10 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> calcu_tpm_for_KO.py kegg.tsv {2} KO_profile/{1}.tpm</span>
<span id="cb14-7"><a href="#cb14-7" tabindex="-1"></a><span class="ex">combine_file_zy_folder_allsample.py</span> <span class="at">-D</span> KO_profile/ <span class="at">-suffix</span> .tpm <span class="at">-o</span> kegg.tpm</span>
<span id="cb14-8"><a href="#cb14-8" tabindex="-1"></a><span class="ex">kegg_convert_levels.py</span> <span class="at">-i</span> kegg.tpm <span class="at">-mode</span> pathway <span class="at">-o</span> kegg.tpm</span></code></pre></div>
</div>
<div id="cazy" class="section level2">
<h2>15.CAZy</h2>
<div class="sourceCode" id="cb15"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="ex">diamond</span> blastp <span class="at">-d</span> /share/data1/database/run_dbcan4/db_20240122/CAZy.dmnd <span class="at">--outfmt</span> 6 <span class="at">--min-score</span> 60 <span class="at">--query-cover</span> 50 <span class="at">--max-target-seqs</span> 5 <span class="at">-p</span> 32 <span class="at">-q</span> ../10.geneset/cluster_rep_seq.faa <span class="at">-o</span> CAZyme.btp <span class="op">&gt;</span> /dev/null <span class="dv">2</span><span class="op">&gt;</span>/dev/null</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="ex">parse_CAZy_blast_for_geneset.pl</span> CAZyme.btp CAZyme</span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a><span class="co"># tpm</span></span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a><span class="fu">cat</span> geneset.cvg.filelist <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 15 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> calcu_tpm_for_KO.py CAZyme.tsv {2} CAZyme_profile/{1}.tpm</span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a><span class="ex">combine_file_zy_folder_allsample.py</span> <span class="at">-D</span> CAZyme_profile <span class="at">-suffix</span> .tpm <span class="at">-o</span> CAZyme.tpm</span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a></span>
<span id="cb15-8"><a href="#cb15-8" tabindex="-1"></a><span class="co"># strain genomes cgc pul substrate prediction</span></span>
<span id="cb15-9"><a href="#cb15-9" tabindex="-1"></a><span class="fu">cat</span> ../06.taxa/final_strains.list <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 10 run_dbcan {} meta <span class="at">--tools</span> diamond <span class="at">-c</span> cluster <span class="at">--cgc_substrate</span> <span class="at">--out_dir</span> dbcan4.out/{/.} <span class="at">--dia_cpu</span> 10 <span class="at">--hmm_cpu</span> 10 <span class="at">--tf_cpu</span> 10 <span class="at">--stp_cpu</span> 10 <span class="at">--db_dir</span> /share/data1/database/run_dbcan4/db_20240122/</span>
<span id="cb15-10"><a href="#cb15-10" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f2</span> ../06.taxa/add_genome/fa.filepath <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 8 run_dbcan {} meta <span class="at">--tools</span> diamond <span class="at">-c</span> cluster <span class="at">--cgc_substrate</span> <span class="at">--out_dir</span> dbcan4.out/{/.} <span class="at">--dia_cpu</span> 10 <span class="at">--hmm_cpu</span> 10 <span class="at">--tf_cpu</span> 10 <span class="at">--stp_cpu</span> 10 <span class="at">--db_dir</span> /share/data1/database/run_dbcan4/db_20240122/</span>
<span id="cb15-11"><a href="#cb15-11" tabindex="-1"></a></span>
<span id="cb15-12"><a href="#cb15-12" tabindex="-1"></a><span class="co"># CAZyme</span></span>
<span id="cb15-13"><a href="#cb15-13" tabindex="-1"></a><span class="ex">le</span> total_genomes.name <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 30 cut <span class="at">-f1,5</span> dbcan4.out/{}/overview.txt <span class="dt">\|</span> sed <span class="dt">\&#39;</span>1d<span class="dt">\&#39;</span> <span class="dt">\&gt;</span>  dbcan4.CAZyme.count/{}.CAZyme.out <span class="kw">&amp;</span></span>
<span id="cb15-14"><a href="#cb15-14" tabindex="-1"></a><span class="fu">ls</span> <span class="pp">*</span> <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 20 <span class="at">-q</span> perl <span class="at">-e</span> <span class="st">&#39;open I, &quot;$ARGV[0]&quot;;open O, &quot;&gt;$ARGV[1]&quot;;%h;while(&lt;I&gt;){chomp;@s=split/\t/;@s2=split/\+/, $s[1];$len=$#s2+1; for(@s2){$_=~s/_.*//g;$h{$_}+=1/$len}} for (keys %h){print O &quot;$_\t$h{$_}\n&quot;}&#39;</span> {} {}2 </span>
<span id="cb15-15"><a href="#cb15-15" tabindex="-1"></a><span class="ex">combine_file_zy_folder_allsample.py</span> <span class="at">-D</span> dbcan4.CAZyme.count <span class="at">-o</span> dbcan4.CAZyme.count.profile <span class="at">-suffix</span> .CAZyme.out2</span></code></pre></div>
</div>
<div id="mgc" class="section level2">
<h2>16.MGC</h2>
<div class="sourceCode" id="cb16"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="co"># split fasta</span></span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a><span class="ex">pigz</span> <span class="at">-dc</span> total_genomes.filelist.chunk_00.fa.gz <span class="kw">|</span> <span class="ex">seqkit</span> split <span class="at">-p</span> 100 <span class="at">-O</span> chunk_00</span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a><span class="co"># calling mgc</span></span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a><span class="fu">ls</span> <span class="pp">*</span>fa <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 20 run_gutsmash.py <span class="at">-c</span> 6 <span class="at">--output-dir</span> {.}.MGC_out <span class="at">--genefinding-tool</span> prodigal {} <span class="kw">&amp;</span></span>
<span id="cb16-6"><a href="#cb16-6" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-d</span> <span class="pp">*</span>out <span class="kw">|</span> <span class="fu">sed</span> <span class="st">&#39;s/.MGC_out//&#39;</span> <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 40 parse_gutsmash.py {}.MGC_out/index.html {}.tsv</span>
<span id="cb16-7"><a href="#cb16-7" tabindex="-1"></a></span>
<span id="cb16-8"><a href="#cb16-8" tabindex="-1"></a><span class="fu">cat</span> chunk_0<span class="pp">*</span>/<span class="pp">*</span>tsv <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;NR==1 || $1!=&quot;contig&quot;&#39;</span> <span class="op">&gt;</span> total_MGC.tsv</span>
<span id="cb16-9"><a href="#cb16-9" tabindex="-1"></a></span>
<span id="cb16-10"><a href="#cb16-10" tabindex="-1"></a><span class="co"># 汇总gbk，该名称</span></span>
<span id="cb16-11"><a href="#cb16-11" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-d</span> <span class="pp">*</span>out <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">ls</span> <span class="va">$i</span>/c<span class="pp">*</span>region<span class="pp">*</span>gbk <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-ne</span> <span class="st">&#39;chomp;$_=~/\/(c.*)...(region\d+).gbk/;print &quot;$_\t$1\t$2\n&quot;&#39;</span> <span class="op">&gt;</span> <span class="va">$i</span>.filelist<span class="kw">;</span> <span class="cf">done</span></span>
<span id="cb16-12"><a href="#cb16-12" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-d</span> <span class="pp">*</span>out <span class="kw">|</span> <span class="fu">sed</span> <span class="st">&#39;s/.MGC_out//&#39;</span> <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span> <span class="cf">do</span> <span class="fu">sed</span> 1d <span class="va">$i</span>.tsv <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-ne</span> <span class="st">&#39;chomp;$_=~/^(.*?)\..*name was\: (.*?)\)/; $y=$1; $x=$2; $x=~s/\|/_/;print &quot;$y\t$x\n&quot;&#39;</span> <span class="kw">|</span> <span class="ex">csvtk</span> join <span class="at">--left-join</span> <span class="at">-H</span> <span class="at">-t</span> <span class="at">-f</span> <span class="st">&quot;1;2&quot;</span> <span class="at">-</span> <span class="va">$i</span>.MGC_out.filelist <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-u</span> <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 10 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> echo cp {3} gbk/{2}.{4}.gbk <span class="kw">|</span> <span class="fu">sh</span> <span class="kw">;</span><span class="cf">done</span></span></code></pre></div>
</div>
<div id="vir_ident" class="section level2">
<h2>17.vir_ident</h2>
<div class="sourceCode" id="cb17"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="fu">cat</span> contigs.name <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 2 run_virFinder.sh {}.fa 5000 {} <span class="kw">&amp;</span></span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a><span class="fu">cat</span> contigs.name <span class="kw">|</span> <span class="fu">sed</span> 1,2d <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 2 run_virFinder.sh {}.fa 5000 {} <span class="kw">&amp;</span></span></code></pre></div>
</div>
<div id="drep_hq" class="section level2">
<h2>22.dRep_HQ</h2>
<div class="sourceCode" id="cb18"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f5</span> genome_HQ <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 20 mash sketch {} <span class="at">-s</span> 1000 <span class="at">-o</span> mash_out/{/.}</span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a><span class="fu">find</span> mash_out/ <span class="at">-name</span> <span class="st">&quot;*msh&quot;</span> <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-k</span> <span class="at">-j</span> 2 <span class="at">--pipe</span> <span class="at">-n</span> 10000 <span class="st">&quot;cat &gt; mash.chunk.{#}.filepath&quot;</span></span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a><span class="ex">parallel</span> <span class="at">-j</span> 10 <span class="at">-k</span> mash paste mash.chunk.{} <span class="at">-l</span> mash.chunk.{}.filepath ::: <span class="dt">{</span><span class="dv">1</span><span class="dt">..</span><span class="dv">8</span><span class="dt">}</span></span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a><span class="co"># dist</span></span>
<span id="cb18-5"><a href="#cb18-5" tabindex="-1"></a><span class="ex">parallel</span> <span class="at">-j</span> 10 mash dist <span class="at">-p</span> 30 mash.chunk.{}.msh mash.chunk.{}.msh <span class="dt">\&gt;</span> mash.chunk.{}.dist ::: <span class="dt">{</span><span class="dv">1</span><span class="dt">..</span><span class="dv">8</span><span class="dt">}</span></span>
<span id="cb18-6"><a href="#cb18-6" tabindex="-1"></a><span class="ex">parallel</span> <span class="at">-j</span> 2 tools_mashANI_cluster.py <span class="at">-i</span> mash.chunk.{}.dist <span class="at">-o</span> mash.chunk.{}.dist.clu <span class="at">-ani</span> 0.9 ::: <span class="dt">{</span><span class="dv">1</span><span class="dt">..</span><span class="dv">8</span><span class="dt">}</span></span>
<span id="cb18-7"><a href="#cb18-7" tabindex="-1"></a><span class="co"># skani</span></span>
<span id="cb18-8"><a href="#cb18-8" tabindex="-1"></a><span class="fu">seq</span> 1 1 8 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">mkdir</span> mash.chunk.<span class="va">$i</span>.skani<span class="kw">;</span> <span class="cf">done</span></span>
<span id="cb18-9"><a href="#cb18-9" tabindex="-1"></a><span class="fu">seq</span> 1 1 8 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">perl</span> <span class="at">-e</span> <span class="st">&#39;open I, &quot;$ARGV[0]&quot;; while(&lt;I&gt;){chomp; next if /cluster/; @s=split/\s+/; if($s[1] ne $x){open O, &quot;&gt;$ARGV[1]/$s[1].filelist&quot;; print O &quot;$s[0]\n&quot;; $x=$s[1] }else{ print O &quot;$s[0]\n&quot; } }&#39;</span> mash.chunk.<span class="va">$i</span>.dist.clu mash.chunk.<span class="va">$i</span>.skani <span class="kw">;</span><span class="cf">done</span></span>
<span id="cb18-10"><a href="#cb18-10" tabindex="-1"></a><span class="fu">seq</span> 1 1 8 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">perl</span> <span class="at">-e</span> <span class="st">&#39;open I, &quot;ls $ARGV[0]/*filelist|&quot;; while(&lt;I&gt;){chomp;$_=~/(\d+).filelist/; print &quot;skani dist --ql $_ --rl $_ -o $ARGV[0]/$1.skani --min-af 30 -t 20\n&quot;}&#39;</span> mash.chunk.<span class="va">$i</span>.skani <span class="kw">;</span><span class="cf">done</span> <span class="op">&gt;</span> run_skani.sh</span>
<span id="cb18-11"><a href="#cb18-11" tabindex="-1"></a><span class="fu">seq</span> 1 1 8 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">perl</span> <span class="at">-e</span> <span class="st">&#39;open I, &quot;ls $ARGV[0]/*skani|&quot;; while(&lt;I&gt;){chomp;$_=~/\/(\d+).skani/; print &quot;tools_skANI_cluster.py -i $_ -o $ARGV[0]/$1.clu -nc 0.3 -ani 0.95\n&quot;}&#39;</span> mash.chunk.<span class="va">$i</span>.skani <span class="kw">;</span><span class="cf">done</span> <span class="op">&gt;</span> run_clu.sh</span>
<span id="cb18-12"><a href="#cb18-12" tabindex="-1"></a><span class="co"># extract clu info</span></span>
<span id="cb18-13"><a href="#cb18-13" tabindex="-1"></a><span class="fu">seq</span> 1 1 8 <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">i</span><span class="kw">;</span><span class="cf">do</span> <span class="fu">ls</span> mash.chunk.<span class="va">$i</span>.skani/<span class="pp">*</span>clu <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 20 <span class="at">-q</span> perl <span class="at">-e</span> <span class="st">&#39;open I, &quot;$ARGV[0]&quot;;$ARGV[0]=~/(\d+).clu/;$x=$1; readline(I);while(&lt;I&gt;){chomp;@s=split/\t/;print &quot;$s[0]\t$x\t$s[1]\t$x\_$s[1]\n&quot;}&#39;</span> {} <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb18-14"><a href="#cb18-14" tabindex="-1"></a>        <span class="ex">csvtk</span> add-header <span class="at">-t</span> <span class="at">-n</span> <span class="st">&quot;name,primary_cluster,second_cluster,final_cluster&quot;</span> <span class="op">&gt;</span> mash.chunk.<span class="va">$i</span>.skani.Cdb.tsv <span class="kw">;</span><span class="cf">done</span></span>
<span id="cb18-15"><a href="#cb18-15" tabindex="-1"></a><span class="co"># filter</span></span>
<span id="cb18-16"><a href="#cb18-16" tabindex="-1"></a><span class="ex">parallel</span> <span class="at">-j</span> 10 ./parse_dRep.pl mash.chunk.{}.skani.Cdb.tsv ../08.dRep/total_genomes.after_qc.tsv mash.chunk.{}.skani.Cdb.info ::: <span class="kw">`</span><span class="fu">seq</span> 1 1 8<span class="kw">`</span></span>
<span id="cb18-17"><a href="#cb18-17" tabindex="-1"></a></span>
<span id="cb18-18"><a href="#cb18-18" tabindex="-1"></a><span class="co"># 第二次聚类</span></span>
<span id="cb18-19"><a href="#cb18-19" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f3</span> <span class="pp">*</span>Cdb.info <span class="kw">|</span> <span class="fu">sed</span> <span class="st">&#39;s/^/mash_out\//g;s/$/.msh/g&#39;</span> <span class="op">&gt;</span> sec.mash.filepath</span>
<span id="cb18-20"><a href="#cb18-20" tabindex="-1"></a><span class="ex">mash</span> paste sec.mash <span class="at">-l</span> sec.mash.filepath</span>
<span id="cb18-21"><a href="#cb18-21" tabindex="-1"></a><span class="ex">mash</span> dist <span class="at">-p</span> 50 sec.mash.msh sec.mash.msh <span class="op">&gt;</span> sec.mash.dist</span>
<span id="cb18-22"><a href="#cb18-22" tabindex="-1"></a><span class="ex">tools_mashANI_cluster.py</span> <span class="at">-i</span> sec.mash.dist <span class="at">-o</span> sec.mash.dist.clu <span class="at">-ani</span> .9</span>
<span id="cb18-23"><a href="#cb18-23" tabindex="-1"></a><span class="fu">perl</span> <span class="at">-ne</span> <span class="st">&#39;next if /cluster/;@s=split/\s+/;if($s[1] ne $x){open O, &quot;&gt;sec.mash.skani/$s[1].filelist&quot;; print O &quot;$s[0]\n&quot;; $x=$s[1] }else{print O &quot;$s[0]\n&quot;}&#39;</span> sec.mash.dist.clu</span>
<span id="cb18-24"><a href="#cb18-24" tabindex="-1"></a><span class="fu">perl</span> <span class="at">-e</span> <span class="st">&#39;open I, &quot;ls sec.mash.skani/*filelist|&quot;; while(&lt;I&gt;){chomp;$_=~/(\d+).filelist/; print &quot;skani dist --ql $_ --rl $_ -o sec.mash.skani/$1.skani --min-af 30 -t 20\n&quot;}&#39;</span> <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 10</span>
<span id="cb18-25"><a href="#cb18-25" tabindex="-1"></a><span class="fu">perl</span> <span class="at">-e</span> <span class="st">&#39;open I, &quot;ls sec.mash.skani/*skani|&quot;; while(&lt;I&gt;){chomp;$_=~/\/(\d+).skani/; print &quot;tools_skANI_cluster.py -i $_ -o sec.mash.skani/$1.clu -nc 0.3 -ani 0.95\n&quot;}&#39;</span> <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 50</span>
<span id="cb18-26"><a href="#cb18-26" tabindex="-1"></a><span class="fu">ls</span> sec.mash.skani/<span class="pp">*</span>clu <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 20 <span class="at">-q</span> perl <span class="at">-e</span> <span class="st">&#39;open I, &quot;$ARGV[0]&quot;;$ARGV[0]=~/(\d+).clu/;$x=$1; readline(I);while(&lt;I&gt;){chomp;@s=split/\t/;print &quot;$s[0]\t$x\t$s[1]\t$x\_$s[1]\n&quot;}&#39;</span> {} <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb18-27"><a href="#cb18-27" tabindex="-1"></a>        <span class="ex">csvtk</span> add-header <span class="at">-t</span> <span class="at">-n</span> <span class="st">&quot;name,primary_cluster,second_cluster,final_cluster&quot;</span> <span class="op">&gt;</span> sec.mash.skani.Cdb.tsv</span>
<span id="cb18-28"><a href="#cb18-28" tabindex="-1"></a><span class="ex">parse_dRep.pl</span> sec.mash.skani.Cdb.tsv ../08.dRep/total_genomes.after_qc.tsv sec.mash.skani.Cdb.info</span>
<span id="cb18-29"><a href="#cb18-29" tabindex="-1"></a></span>
<span id="cb18-30"><a href="#cb18-30" tabindex="-1"></a><span class="fu">cat</span> mash.chunk.<span class="pp">*</span>.skani.Cdb.info <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-e</span> <span class="st">&#39;%clu;while(&lt;&gt;){chomp;@s=split/\t/;$clu{$s[2]}=$s[3]}; open I, &quot;sec.mash.skani.Cdb.info&quot;; while(&lt;I&gt;){chomp; @s=split/\t/; @s2=split/,/,$s[3]; %h=(); for (@s2){@x=split/,/,$clu{$_}; for (@x){$h{$_}++} }; @out=keys %h; $len=scalar @out;print &quot;$s[0]\t$len\t$s[2]\t&quot;.join(&quot;,&quot;, @out).&quot;\n&quot;}&#39;</span> <span class="op">&gt;</span> final.clu.info</span></code></pre></div>
</div>
<div id="gtdbtk_hq" class="section level2">
<h2>23.gtdbtk_HQ</h2>
<div class="sourceCode" id="cb19"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="fu">cat</span> species.rename <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 4 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> seqkit replace <span class="at">-p</span> <span class="dt">\&quot;</span>.<span class="dt">\*\&quot;</span> <span class="at">-r</span> <span class="dt">\&quot;</span>{1}_<span class="dt">\{</span>nr<span class="dt">\}\&quot;</span> {2} <span class="at">-o</span> species_fna/{1}.fna.gz</span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a><span class="ex">kraken_build.py</span> <span class="at">-l</span> kraken_build.filelist.tsv <span class="at">-t</span> kraken_build.taxonomy.tsv <span class="at">-d</span> kraken2db_150mers <span class="at">-p</span> 30 <span class="at">-m</span> 150</span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a><span class="ex">kraken_build.py</span> <span class="at">-l</span> kraken_build.filelist.tsv <span class="at">-t</span> kraken_build.taxonomy.tsv <span class="at">-d</span> kraken2db_125mers <span class="at">-p</span> 30 <span class="at">-m</span> 125</span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a><span class="ex">kraken_build.py</span> <span class="at">-l</span> kraken_build.filelist.tsv <span class="at">-t</span> kraken_build.taxonomy.tsv <span class="at">-d</span> kraken2db_100mers <span class="at">-p</span> 30 <span class="at">-m</span> 100</span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a><span class="ex">kraken_build.py</span> <span class="at">-l</span> kraken_build.filelist.tsv <span class="at">-t</span> kraken_build.taxonomy.tsv <span class="at">-d</span> kraken2db_75mers <span class="at">-p</span> 30 <span class="at">-m</span> 75</span></code></pre></div>
</div>
<div id="profile_hq" class="section level2">
<h2>24.profile_HQ</h2>
<div class="sourceCode" id="cb20"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="fu">ls</span> kraken2/<span class="pp">*</span>log <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-ne</span> <span class="st">&#39;chomp;open I, &quot;$_&quot;; /\/(\S+).log/;$x=$1; while(&lt;I&gt;){chomp;print &quot;$x\t$1\t$2\t&quot; if /^(\d+) sequences \((.*) Mbp\)/;; print &quot;$1 ($2)\t&quot; if /  (\d+) sequences classified \((.*)\)/; print &quot;$1 ($2)\n&quot; if /  (\d+) sequences unclassified \((.*)\)/;}&#39;</span> <span class="kw">|</span> <span class="ex">csvtk</span> add-header <span class="at">-t</span> <span class="at">-n</span> sample,total_reads,total_bases,classified_reads,unclassified_reads <span class="op">&gt;</span> kraken2.seqs.stats.tsv</span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" tabindex="-1"></a><span class="co"># new est count</span></span>
<span id="cb20-4"><a href="#cb20-4" tabindex="-1"></a><span class="ex">combine_file_zy_folder_allsample.py</span> <span class="at">-D</span> kraken2/ <span class="at">-suffix</span> .bk_out.p <span class="at">-o</span> est_rc.p.profile <span class="at">-n</span> 1 <span class="at">-v</span> 6 <span class="at">-t</span> 1 <span class="at">-s</span> 0</span>
<span id="cb20-5"><a href="#cb20-5" tabindex="-1"></a><span class="ex">combine_file_zy_folder_allsample.py</span> <span class="at">-D</span> kraken2/ <span class="at">-suffix</span> .bk_out.f <span class="at">-o</span> est_rc.f.profile <span class="at">-n</span> 1 <span class="at">-v</span> 6 <span class="at">-t</span> 1 <span class="at">-s</span> 0</span>
<span id="cb20-6"><a href="#cb20-6" tabindex="-1"></a><span class="ex">combine_file_zy_folder_allsample.py</span> <span class="at">-D</span> kraken2/ <span class="at">-suffix</span> .bk_out.g <span class="at">-o</span> est_rc.g.profile <span class="at">-n</span> 1 <span class="at">-v</span> 6 <span class="at">-t</span> 1 <span class="at">-s</span> 0</span>
<span id="cb20-7"><a href="#cb20-7" tabindex="-1"></a><span class="ex">combine_file_zy_folder_allsample.py</span> <span class="at">-D</span> kraken2/ <span class="at">-suffix</span> .bk_out.s <span class="at">-o</span> est_rc.s.profile <span class="at">-n</span> 1 <span class="at">-v</span> 6 <span class="at">-t</span> 1 <span class="at">-s</span> 0 </span>
<span id="cb20-8"><a href="#cb20-8" tabindex="-1"></a><span class="ex">combine_file_zy_folder_allsample.py</span> <span class="at">-D</span> kraken2/ <span class="at">-suffix</span> .bk_out.t <span class="at">-o</span> est_rc.t.profile <span class="at">-n</span> 1 <span class="at">-v</span> 6 <span class="at">-t</span> 1 <span class="at">-s</span> 0</span></code></pre></div>
</div>
<div id="phylophlan" class="section level2">
<h2>25.phylophlan</h2>
<div class="sourceCode" id="cb21"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="ex">phylophlan_write_config_file</span> <span class="at">-d</span> a <span class="at">-o</span> config_file <span class="at">--db_aa</span> diamond <span class="at">--map_aa</span> diamond <span class="at">--msa</span> mafft <span class="at">--trim</span> trimal <span class="at">--tree1</span> iqtree <span class="at">--tree2</span> raxml</span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a><span class="ex">phylophlan</span> <span class="at">-i</span> ../23.gtdbtk_HQ/species_faa <span class="at">-t</span> a <span class="at">-d</span> phylophlan <span class="at">--databases_folder</span> /share/data1/database/phylophlan3/ <span class="at">-f</span> config_file <span class="at">--diversity</span> high <span class="at">--verbose</span> <span class="at">--fast</span> <span class="at">-o</span> . <span class="at">--nproc</span> 30 <span class="op">&gt;</span> log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span></span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a></span>
<span id="cb21-4"><a href="#cb21-4" tabindex="-1"></a><span class="fu">ls</span> <span class="pp">*</span> <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 10 <span class="at">-q</span> perl <span class="at">-e</span> <span class="st">&#39;open I, &quot;$ARGV[0]&quot;;$g=$ARGV[1]; open O, &quot;&gt;$ARGV[2]&quot;; while(&lt;I&gt;){chomp;if(/&gt;(\S+)/){print O &quot;&gt;$g.$1\n&quot;}else{print O &quot;$_\n&quot;}}&#39;</span> {} {/.} /share/data1/software/phylophlan/input/mjx_bp_family.rep_20240529/{}</span>
<span id="cb21-5"><a href="#cb21-5" tabindex="-1"></a><span class="fu">sed</span> <span class="st">&#39;1d&#39;</span> dat.family.rep.genome.list <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 10 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> ln <span class="at">-s</span> /share/data1/mjx/proj/04.black_pig_metagenome_20230529/07.annotation/faa/{2}.faa faa/{1}.faa</span>
<span id="cb21-6"><a href="#cb21-6" tabindex="-1"></a><span class="ex">phylophlan_write_config_file</span> <span class="at">-d</span> a <span class="at">-o</span> cfg <span class="at">--db_aa</span> diamond <span class="at">--map_aa</span> diamond <span class="at">--msa</span> mafft <span class="at">--trim</span> trimal <span class="at">--tree1</span> fasttree</span>
<span id="cb21-7"><a href="#cb21-7" tabindex="-1"></a><span class="co"># 每个diamond线程已经设置为4了，那么总线程数搞一个30就行了</span></span>
<span id="cb21-8"><a href="#cb21-8" tabindex="-1"></a><span class="ex">phylophlan</span> <span class="at">-i</span> faa <span class="at">-t</span> a <span class="at">-d</span> phylophlan <span class="at">--databases_folder</span> /share/data1/database/phylophlan3/ <span class="at">-f</span> cfg <span class="at">--diversity</span> high <span class="at">--verbose</span> <span class="at">--fast</span> <span class="at">-o</span> . <span class="at">--nproc</span> 30 <span class="op">&gt;</span> log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;</span></span>
<span id="cb21-9"><a href="#cb21-9" tabindex="-1"></a></span>
<span id="cb21-10"><a href="#cb21-10" tabindex="-1"></a><span class="fu">ls</span> <span class="pp">*</span>faa <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 10 <span class="at">-q</span> perl <span class="at">-e</span> <span class="st">&#39;open I, &quot;$ARGV[0]&quot;; open O, &quot;&gt;../genomospecies_4019_2/$ARGV[0]&quot;; $x=$ARGV[1]; while(&lt;I&gt;){chomp;if(/&gt;(\S+)/){print O &quot;&gt;$x.$1\n&quot;}else{print O &quot;$_\n&quot;}}&#39;</span> {} {/.} <span class="kw">&amp;</span></span></code></pre></div>
</div>
<div id="arg" class="section level2">
<h2>26.ARG</h2>
<div class="sourceCode" id="cb22"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="ex">diamond</span> blastp <span class="at">-d</span> /share/data1/database/SARG/SARG_v3.2_20220917.dmnd <span class="at">--outfmt</span> 6 <span class="at">--query-cover</span> 80 <span class="at">--id</span> 80 <span class="at">--max-target-seqs</span> 5 <span class="at">-p</span> 80 <span class="at">-q</span> ../10.geneset/geneset_full.faa <span class="at">-o</span> ARGs.btp <span class="op">&gt;</span> /dev/null <span class="dv">2</span><span class="op">&gt;</span>/dev/null</span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a><span class="fu">perl</span> <span class="at">-ne</span> <span class="st">&#39;chomp;@s=split/\t/;if($s[0] ne $a){print &quot;$s[0]\t$s[1]\n&quot; if $s[10]&lt;1e-10;$a=$s[0]}&#39;</span> ARGs.btp <span class="op">&gt;</span> ARGs.tsv</span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a><span class="co">#perl -e &#39;open I, &quot;ARGs.tsv&quot;; %h; while(&lt;I&gt;){chomp; @s=split/\t/; $h{$s[0]}=$s[1]}; open I, &quot;pigz -dc ../10.geneset/cluster_cluster.tsv.gz |&quot;; while(&lt;I&gt;){chomp; @s=split/\t/; if(exists $h{$s[0]}){print &quot;$_\t$h{$s[0]}\|$s[1]\n&quot;} }&#39; &gt; all_ARG_gene.tsv</span></span>
<span id="cb22-4"><a href="#cb22-4" tabindex="-1"></a><span class="co">#cut -f2 all_ARG_gene.tsv | seqkit grep -f - ../10.geneset/cluster_rep_seq.fasta -o all_ARG_gene.fa</span></span>
<span id="cb22-5"><a href="#cb22-5" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f1</span> ARGs.tsv <span class="kw">|</span> <span class="ex">seqkit</span> grep <span class="at">-f</span> <span class="at">-</span> ../10.geneset/geneset_full.ffn <span class="op">&gt;</span> ARGs.fa</span></code></pre></div>
</div>
<div id="vfg" class="section level2">
<h2>27.VFG</h2>
<div class="sourceCode" id="cb23"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="ex">diamond</span> blastp <span class="at">-d</span> /share/data1/database/VFDB/VFDB_pro.dmnd <span class="at">--outfmt</span> 6 <span class="at">--query-cover</span> 80 <span class="at">--id</span> 80 <span class="at">--max-target-seqs</span> 5 <span class="at">-p</span> 80 <span class="at">-q</span> ../10.geneset/geneset_full.faa <span class="at">-o</span> VFGs.btp <span class="op">&gt;</span> /dev/null <span class="dv">2</span><span class="op">&gt;</span>/dev/null</span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a><span class="fu">perl</span> <span class="at">-ne</span> <span class="st">&#39;chomp;@s=split/\t/;if($s[0] ne $a){$s[1]=~/(VFG\d+)/;print &quot;$s[0]\t$1\n&quot; if $s[10]&lt;1e-10;$a=$s[0]}&#39;</span> VFGs.btp <span class="op">&gt;</span> VFGs.tsv</span>
<span id="cb23-3"><a href="#cb23-3" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f1</span> VFGs.tsv <span class="kw">|</span> <span class="ex">seqkit</span> grep <span class="at">-f</span> <span class="at">-</span> ../10.geneset/geneset_full.faa <span class="at">-o</span> VFGs.fa</span></code></pre></div>
</div>
<div id="pn_ps" class="section level2">
<h2>28.pN_pS</h2>
<div class="sourceCode" id="cb24"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="co">#bowtie2-build -q --large-index --threads 80 geneset_complete.ffn geneset_complete</span></span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a></span>
<span id="cb24-3"><a href="#cb24-3" tabindex="-1"></a><span class="ex">seqkit</span> replace <span class="at">-p</span> <span class="st">&quot;^&quot;</span> <span class="at">-r</span> <span class="st">&quot;ARG|&quot;</span> ../26.ARG/ARGs.fa <span class="op">&gt;</span> gene.fa</span>
<span id="cb24-4"><a href="#cb24-4" tabindex="-1"></a><span class="ex">seqkit</span> replace <span class="at">-p</span> <span class="st">&quot;^&quot;</span> <span class="at">-r</span> <span class="st">&quot;VFG|&quot;</span> ../27.VFG/VFGs.fa <span class="op">&gt;&gt;</span> gene.fa</span>
<span id="cb24-5"><a href="#cb24-5" tabindex="-1"></a><span class="ex">seqkit</span> replace <span class="at">-p</span> <span class="st">&quot;^&quot;</span> <span class="at">-r</span> <span class="st">&quot;MGE|&quot;</span> ../29.MGE/MGEs.fa <span class="op">&gt;&gt;</span> gene.fa</span>
<span id="cb24-6"><a href="#cb24-6" tabindex="-1"></a><span class="ex">bowtie2-build</span> gene.fa gene</span>
<span id="cb24-7"><a href="#cb24-7" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f1,2</span> samples.6574.filelist <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 8 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> run_pnps_analyis.sh {2} gene gene.fa pnps/{1} 14</span></code></pre></div>
</div>
<div id="mge" class="section level2">
<h2>29.MGE</h2>
<div class="sourceCode" id="cb25"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="ex">diamond</span> blastp <span class="at">-d</span> /share/data1/database/mobileOG/mobileOG.dmnd <span class="at">--outfmt</span> 6 <span class="at">--query-cover</span> 80 <span class="at">--id</span> 80 <span class="at">--max-target-seqs</span> 5 <span class="at">-p</span> 110 <span class="at">-q</span> ../10.geneset/geneset_full.faa <span class="at">-o</span> MGEs.btp <span class="op">&gt;</span> /dev/null <span class="dv">2</span><span class="op">&gt;</span>/dev/null</span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a><span class="fu">perl</span> <span class="at">-ne</span> <span class="st">&#39;chomp;@s=split/\t/;if($s[0] ne $a){print &quot;$s[0]\t$s[1]\n&quot; if $s[10] &lt; 1e-10;$a=$s[0]}&#39;</span> MGEs.btp <span class="op">&gt;</span> MGEs.tsv</span>
<span id="cb25-3"><a href="#cb25-3" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f1</span> MGEs.tsv <span class="kw">|</span> <span class="ex">seqkit</span> grep <span class="at">-f</span> <span class="at">-</span> ../10.geneset/geneset_full.ffn <span class="op">&gt;</span> MGEs.fa</span></code></pre></div>
</div>
<div id="mcvg_test" class="section level2">
<h2>mcvg_test</h2>
<div class="sourceCode" id="cb26"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="fu">shuf</span> <span class="at">--random-source</span><span class="op">=&lt;(</span><span class="fu">yes</span> 2024<span class="op">)</span> contigs_3997.name <span class="kw">|</span> <span class="fu">head</span> <span class="at">-50</span> <span class="op">&gt;</span> contigs_3997.name.shuf</span>
<span id="cb26-2"><a href="#cb26-2" tabindex="-1"></a></span>
<span id="cb26-3"><a href="#cb26-3" tabindex="-1"></a><span class="co"># mapping</span></span>
<span id="cb26-4"><a href="#cb26-4" tabindex="-1"></a><span class="fu">cat</span> contigs_3997.name.shuf <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 2 ./run_mapping.sh {} depth</span>
<span id="cb26-5"><a href="#cb26-5" tabindex="-1"></a></span>
<span id="cb26-6"><a href="#cb26-6" tabindex="-1"></a><span class="co"># combine depth</span></span>
<span id="cb26-7"><a href="#cb26-7" tabindex="-1"></a><span class="fu">cat</span> contigs_3997.name.shuf <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 10 echo cp depth/{}_{}.depth depth_01</span>
<span id="cb26-8"><a href="#cb26-8" tabindex="-1"></a><span class="fu">cat</span> contigs_3997.name.shuf <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 10 run_depth_combine.pl {} 5 depth_05/{} <span class="kw">&amp;</span></span>
<span id="cb26-9"><a href="#cb26-9" tabindex="-1"></a><span class="fu">cat</span> contigs_3997.name.shuf <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 10 run_depth_combine.pl {} 10 depth_10/{} <span class="kw">&amp;</span></span>
<span id="cb26-10"><a href="#cb26-10" tabindex="-1"></a><span class="fu">cat</span> contigs_3997.name.shuf <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 10 run_depth_combine.pl {} 15 depth_15/{} <span class="kw">&amp;</span></span>
<span id="cb26-11"><a href="#cb26-11" tabindex="-1"></a><span class="fu">cat</span> contigs_3997.name.shuf <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 10 run_depth_combine.pl {} 20 depth_20/{} <span class="kw">&amp;</span></span>
<span id="cb26-12"><a href="#cb26-12" tabindex="-1"></a><span class="fu">cat</span> contigs_3997.name.shuf <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 10 run_depth_combine.pl {} 25 depth_25/{} <span class="kw">&amp;</span></span>
<span id="cb26-13"><a href="#cb26-13" tabindex="-1"></a><span class="fu">cat</span> contigs_3997.name.shuf <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 10 run_depth_combine.pl {} 30 depth_30/{} <span class="kw">&amp;</span></span>
<span id="cb26-14"><a href="#cb26-14" tabindex="-1"></a><span class="fu">cat</span> contigs_3997.name.shuf <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 10 run_depth_combine.pl {} 40 depth_40/{} <span class="kw">&amp;</span></span>
<span id="cb26-15"><a href="#cb26-15" tabindex="-1"></a><span class="fu">cat</span> contigs_3997.name.shuf <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 10 run_depth_combine.pl {} 50 depth_50/{} <span class="kw">&amp;</span></span>
<span id="cb26-16"><a href="#cb26-16" tabindex="-1"></a></span>
<span id="cb26-17"><a href="#cb26-17" tabindex="-1"></a><span class="co"># binning</span></span>
<span id="cb26-18"><a href="#cb26-18" tabindex="-1"></a><span class="fu">cat</span> contigs_3997.name.shuf.filepath <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 8 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> metabat2 <span class="at">-t</span> 20 <span class="at">-m</span> 1500 <span class="at">-s</span> 200000 <span class="at">--saveCls</span> <span class="at">--unbinned</span> <span class="at">--seed</span> 2024 <span class="at">-i</span> {3} <span class="at">-a</span> depth_01/{2}.depth <span class="at">-o</span> mcvg_01/{2}.bin</span>
<span id="cb26-19"><a href="#cb26-19" tabindex="-1"></a><span class="fu">cat</span> contigs_3997.name.shuf.filepath <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 8 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> metabat2 <span class="at">-t</span> 20 <span class="at">-m</span> 1500 <span class="at">-s</span> 200000 <span class="at">--saveCls</span> <span class="at">--unbinned</span> <span class="at">--seed</span> 2024 <span class="at">-i</span> {3} <span class="at">-a</span> depth_05/{2}.depth <span class="at">-o</span> mcvg_05/{2}.bin</span>
<span id="cb26-20"><a href="#cb26-20" tabindex="-1"></a><span class="fu">cat</span> contigs_3997.name.shuf.filepath <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 8 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> metabat2 <span class="at">-t</span> 20 <span class="at">-m</span> 1500 <span class="at">-s</span> 200000 <span class="at">--saveCls</span> <span class="at">--unbinned</span> <span class="at">--seed</span> 2024 <span class="at">-i</span> {3} <span class="at">-a</span> depth_10/{2}.depth <span class="at">-o</span> mcvg_10/{2}.bin</span>
<span id="cb26-21"><a href="#cb26-21" tabindex="-1"></a><span class="fu">cat</span> contigs_3997.name.shuf.filepath <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 8 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> metabat2 <span class="at">-t</span> 20 <span class="at">-m</span> 1500 <span class="at">-s</span> 200000 <span class="at">--saveCls</span> <span class="at">--unbinned</span> <span class="at">--seed</span> 2024 <span class="at">-i</span> {3} <span class="at">-a</span> depth_15/{2}.depth <span class="at">-o</span> mcvg_15/{2}.bin</span>
<span id="cb26-22"><a href="#cb26-22" tabindex="-1"></a><span class="fu">cat</span> contigs_3997.name.shuf.filepath <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 8 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> metabat2 <span class="at">-t</span> 20 <span class="at">-m</span> 1500 <span class="at">-s</span> 200000 <span class="at">--saveCls</span> <span class="at">--unbinned</span> <span class="at">--seed</span> 2024 <span class="at">-i</span> {3} <span class="at">-a</span> depth_20/{2}.depth <span class="at">-o</span> mcvg_20/{2}.bin</span>
<span id="cb26-23"><a href="#cb26-23" tabindex="-1"></a><span class="fu">cat</span> contigs_3997.name.shuf.filepath <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 8 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> metabat2 <span class="at">-t</span> 20 <span class="at">-m</span> 1500 <span class="at">-s</span> 200000 <span class="at">--saveCls</span> <span class="at">--unbinned</span> <span class="at">--seed</span> 2024 <span class="at">-i</span> {3} <span class="at">-a</span> depth_25/{2}.depth <span class="at">-o</span> mcvg_25/{2}.bin</span>
<span id="cb26-24"><a href="#cb26-24" tabindex="-1"></a><span class="fu">cat</span> contigs_3997.name.shuf.filepath <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 8 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> metabat2 <span class="at">-t</span> 20 <span class="at">-m</span> 1500 <span class="at">-s</span> 200000 <span class="at">--saveCls</span> <span class="at">--unbinned</span> <span class="at">--seed</span> 2024 <span class="at">-i</span> {3} <span class="at">-a</span> depth_30/{2}.depth <span class="at">-o</span> mcvg_30/{2}.bin</span>
<span id="cb26-25"><a href="#cb26-25" tabindex="-1"></a><span class="fu">cat</span> contigs_3997.name.shuf.filepath <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 8 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> metabat2 <span class="at">-t</span> 20 <span class="at">-m</span> 1500 <span class="at">-s</span> 200000 <span class="at">--saveCls</span> <span class="at">--unbinned</span> <span class="at">--seed</span> 2024 <span class="at">-i</span> {3} <span class="at">-a</span> depth_40/{2}.depth <span class="at">-o</span> mcvg_40/{2}.bin</span>
<span id="cb26-26"><a href="#cb26-26" tabindex="-1"></a><span class="fu">cat</span> contigs_3997.name.shuf.filepath <span class="kw">|</span> <span class="ex">parallel</span> <span class="at">-j</span> 8 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> metabat2 <span class="at">-t</span> 20 <span class="at">-m</span> 1500 <span class="at">-s</span> 200000 <span class="at">--saveCls</span> <span class="at">--unbinned</span> <span class="at">--seed</span> 2024 <span class="at">-i</span> {3} <span class="at">-a</span> depth_50/{2}.depth <span class="at">-o</span> mcvg_50/{2}.bin</span>
<span id="cb26-27"><a href="#cb26-27" tabindex="-1"></a><span class="fu">mv</span> mcvg_01/<span class="pp">*</span>.lowDepth.fa mcvg_01/<span class="pp">*</span>.tooShort.fa mcvg_01/<span class="pp">*</span>.unbinned.fa mcvg_01/<span class="pp">*</span>.bin mcvg_tmp_01/</span>
<span id="cb26-28"><a href="#cb26-28" tabindex="-1"></a><span class="fu">mv</span> mcvg_05/<span class="pp">*</span>.lowDepth.fa mcvg_05/<span class="pp">*</span>.tooShort.fa mcvg_05/<span class="pp">*</span>.unbinned.fa mcvg_05/<span class="pp">*</span>.bin mcvg_tmp_05/</span>
<span id="cb26-29"><a href="#cb26-29" tabindex="-1"></a><span class="fu">mv</span> mcvg_10/<span class="pp">*</span>.lowDepth.fa mcvg_10/<span class="pp">*</span>.tooShort.fa mcvg_10/<span class="pp">*</span>.unbinned.fa mcvg_10/<span class="pp">*</span>.bin mcvg_tmp_10/</span>
<span id="cb26-30"><a href="#cb26-30" tabindex="-1"></a><span class="fu">mv</span> mcvg_15/<span class="pp">*</span>.lowDepth.fa mcvg_15/<span class="pp">*</span>.tooShort.fa mcvg_15/<span class="pp">*</span>.unbinned.fa mcvg_15/<span class="pp">*</span>.bin mcvg_tmp_15/</span>
<span id="cb26-31"><a href="#cb26-31" tabindex="-1"></a><span class="fu">mv</span> mcvg_20/<span class="pp">*</span>.lowDepth.fa mcvg_20/<span class="pp">*</span>.tooShort.fa mcvg_20/<span class="pp">*</span>.unbinned.fa mcvg_20/<span class="pp">*</span>.bin mcvg_tmp_20/</span>
<span id="cb26-32"><a href="#cb26-32" tabindex="-1"></a><span class="fu">mv</span> mcvg_25/<span class="pp">*</span>.lowDepth.fa mcvg_25/<span class="pp">*</span>.tooShort.fa mcvg_25/<span class="pp">*</span>.unbinned.fa mcvg_25/<span class="pp">*</span>.bin mcvg_tmp_25/</span>
<span id="cb26-33"><a href="#cb26-33" tabindex="-1"></a><span class="fu">mv</span> mcvg_30/<span class="pp">*</span>.lowDepth.fa mcvg_30/<span class="pp">*</span>.tooShort.fa mcvg_30/<span class="pp">*</span>.unbinned.fa mcvg_30/<span class="pp">*</span>.bin mcvg_tmp_30/</span>
<span id="cb26-34"><a href="#cb26-34" tabindex="-1"></a><span class="fu">mv</span> mcvg_40/<span class="pp">*</span>.lowDepth.fa mcvg_40/<span class="pp">*</span>.tooShort.fa mcvg_40/<span class="pp">*</span>.unbinned.fa mcvg_40/<span class="pp">*</span>.bin mcvg_tmp_40/</span>
<span id="cb26-35"><a href="#cb26-35" tabindex="-1"></a><span class="fu">mv</span> mcvg_50/<span class="pp">*</span>.lowDepth.fa mcvg_50/<span class="pp">*</span>.tooShort.fa mcvg_50/<span class="pp">*</span>.unbinned.fa mcvg_50/<span class="pp">*</span>.bin mcvg_tmp_50/</span>
<span id="cb26-36"><a href="#cb26-36" tabindex="-1"></a></span>
<span id="cb26-37"><a href="#cb26-37" tabindex="-1"></a><span class="co"># ckm2</span></span>
<span id="cb26-38"><a href="#cb26-38" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> 01 05 10 15 20 25 30 40 50<span class="kw">;</span><span class="cf">do</span> <span class="ex">checkm2</span> predict <span class="at">--input</span> mcvg_<span class="va">$i</span> <span class="at">--output-directory</span> ckm2_<span class="va">$i</span> <span class="at">--tmpdir</span> /share/data1/mjx/tmp/tmp_ckm2/ <span class="at">-x</span> .fa <span class="at">--threads</span> 112 <span class="at">--force</span><span class="kw">;</span><span class="cf">done</span></span>
<span id="cb26-39"><a href="#cb26-39" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> 01 05 10 15 20 25 30 40 50<span class="kw">;</span><span class="cf">do</span> <span class="bu">echo</span> gunc run <span class="at">-d</span> ckm2_<span class="va">$i</span>/protein_files/ <span class="at">-t</span> 112 <span class="at">-g</span> <span class="at">-e</span> .faa <span class="at">-o</span> gunc_<span class="va">$i</span>/<span class="kw">;</span><span class="cf">done</span></span>
<span id="cb26-40"><a href="#cb26-40" tabindex="-1"></a><span class="ex">parallel</span> <span class="at">-j</span> 2 dRep compare dRep_95_{} <span class="at">-g</span> mcvg_{}/<span class="pp">*</span> <span class="at">-d</span> <span class="at">-pa</span> 0.9 <span class="at">-sa</span> 0.95  <span class="at">-nc</span> 0.30 <span class="at">-cm</span> larger <span class="at">-p</span> 50 <span class="at">--S_algorithm</span> fastANI ::: 01 05 10 15 20 25 30 40 50 <span class="kw">&amp;</span></span>
<span id="cb26-41"><a href="#cb26-41" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> 01 05 10 15 20 25 30 40 50<span class="kw">;</span> <span class="cf">do</span> <span class="ex">parse_dRep.pl</span> <span class="at">-ckm2</span> ckm2_<span class="va">$i</span>/quality_report.tsv dRep_95_<span class="va">$i</span>/data_tables/Cdb.csv dRep_95_<span class="va">$i</span>.result <span class="kw">;</span><span class="cf">done</span> <span class="kw">&amp;</span></span></code></pre></div>
</div>
<div id="bash-scripts" class="section level2">
<h2>bash scripts</h2>
<div class="sourceCode" id="cb27"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a><span class="co">#### run_mapping.sh ####</span></span>
<span id="cb27-2"><a href="#cb27-2" tabindex="-1"></a><span class="co">#!/usr/bin/bash</span></span>
<span id="cb27-3"><a href="#cb27-3" tabindex="-1"></a></span>
<span id="cb27-4"><a href="#cb27-4" tabindex="-1"></a><span class="kw">(</span> <span class="bu">[</span> <span class="va">$#</span> <span class="ot">-ne</span> 2 <span class="bu">]</span> <span class="kw">)</span> <span class="kw">&amp;&amp;</span> <span class="kw">{</span> <span class="bu">echo</span> <span class="at">-e</span> <span class="st">&quot;Usage: </span><span class="va">$0</span><span class="st"> [sample_name] [out_directory]&quot;</span> <span class="kw">&amp;&amp;</span> <span class="bu">exit</span> 2<span class="kw">;</span> <span class="kw">}</span></span>
<span id="cb27-5"><a href="#cb27-5" tabindex="-1"></a></span>
<span id="cb27-6"><a href="#cb27-6" tabindex="-1"></a><span class="va">smps</span><span class="op">=</span><span class="va">$1</span><span class="kw">;</span> <span class="va">out</span><span class="op">=</span><span class="va">$2</span><span class="kw">;</span></span>
<span id="cb27-7"><a href="#cb27-7" tabindex="-1"></a></span>
<span id="cb27-8"><a href="#cb27-8" tabindex="-1"></a><span class="ex">/share/data1/mjx/proj/07.pigs_binning_mcvg_20230929/01.depth_s1/generate_mapping_cfg.pl</span> <span class="va">$smps</span> 15 <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb27-9"><a href="#cb27-9" tabindex="-1"></a>    <span class="ex">parallel</span> <span class="at">-j</span> 15 <span class="at">--colsep</span><span class="op">=</span><span class="st">&quot;\t&quot;</span> /share/data1/mjx/proj/07.pigs_binning_mcvg_20230929/01.depth_s1/run_minimap2.sh {4} {3} <span class="va">$out</span>/{1}_{2}</span>
<span id="cb27-10"><a href="#cb27-10" tabindex="-1"></a><span class="co"># out prefix: [ref]_[sample].depth</span></span>
<span id="cb27-11"><a href="#cb27-11" tabindex="-1"></a></span>
<span id="cb27-12"><a href="#cb27-12" tabindex="-1"></a><span class="co">#### run_minimap.sh ####</span></span>
<span id="cb27-13"><a href="#cb27-13" tabindex="-1"></a><span class="co">#!/usr/bin/bash</span></span>
<span id="cb27-14"><a href="#cb27-14" tabindex="-1"></a></span>
<span id="cb27-15"><a href="#cb27-15" tabindex="-1"></a><span class="kw">(</span> <span class="bu">[</span> <span class="va">$#</span> <span class="ot">-ne</span> 3 <span class="bu">]</span> <span class="kw">)</span> <span class="kw">&amp;&amp;</span>  <span class="kw">{</span> <span class="bu">echo</span> <span class="at">-e</span> <span class="st">&quot;Usage: </span><span class="va">$0</span><span class="st"> [fq|fq1,fq2] [*.fa] [out_prefix]&quot;</span> <span class="kw">&amp;&amp;</span> <span class="bu">exit</span> 2<span class="kw">;</span> <span class="kw">}</span></span>
<span id="cb27-16"><a href="#cb27-16" tabindex="-1"></a><span class="kw">(</span> <span class="bu">[</span> <span class="ot">-f</span> <span class="va">$3</span>.log <span class="bu">]</span> <span class="kw">)</span> <span class="kw">&amp;&amp;</span> <span class="fu">grep</span> <span class="at">-q</span> <span class="st">&#39;CMD: minimap2&#39;</span> <span class="va">$3</span>.log <span class="kw">&amp;&amp;</span> <span class="kw">{</span> <span class="bu">echo</span> <span class="at">-e</span> <span class="st">&quot;Skip sample: </span><span class="va">${3</span><span class="op">##</span><span class="pp">*</span>/<span class="va">}</span><span class="st"> ..&quot;</span> <span class="kw">&amp;&amp;</span> <span class="bu">exit</span> 0<span class="kw">;</span> <span class="kw">}</span> </span>
<span id="cb27-17"><a href="#cb27-17" tabindex="-1"></a></span>
<span id="cb27-18"><a href="#cb27-18" tabindex="-1"></a><span class="cf">if</span> <span class="kw">[[</span> <span class="va">$1</span> <span class="ot">=~</span> <span class="st">&quot;,&quot;</span> <span class="kw">]];</span><span class="cf">then</span></span>
<span id="cb27-19"><a href="#cb27-19" tabindex="-1"></a>    <span class="va">fq1</span><span class="op">=</span><span class="va">$(</span><span class="bu">echo</span> <span class="va">$1</span> <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-d</span> <span class="st">&quot;,&quot;</span> <span class="at">-f1</span><span class="va">)</span></span>
<span id="cb27-20"><a href="#cb27-20" tabindex="-1"></a>    <span class="va">fq2</span><span class="op">=</span><span class="va">$(</span><span class="bu">echo</span> <span class="va">$1</span> <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-d</span> <span class="st">&quot;,&quot;</span> <span class="at">-f2</span><span class="va">)</span></span>
<span id="cb27-21"><a href="#cb27-21" tabindex="-1"></a>    <span class="ex">minimap2</span> <span class="at">--MD</span> <span class="at">-t</span> 16 <span class="at">-ax</span> sr <span class="va">$2</span> <span class="va">$fq1</span> <span class="va">$fq2</span> <span class="dv">2</span><span class="op">&gt;</span><span class="va">$3</span>.log <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb27-22"><a href="#cb27-22" tabindex="-1"></a>        <span class="ex">/share/data1/mjx/scripts/tools_get_depth_from_sam.single.py</span> <span class="va">$3</span> <span class="op">&gt;</span> <span class="va">$3</span>.depth</span>
<span id="cb27-23"><a href="#cb27-23" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb27-24"><a href="#cb27-24" tabindex="-1"></a>    <span class="ex">minimap2</span> <span class="at">--MD</span> <span class="at">-t</span> 16 <span class="at">-ax</span> sr <span class="va">$2</span> <span class="va">$1</span> <span class="dv">2</span><span class="op">&gt;</span><span class="va">$3</span>.log <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb27-25"><a href="#cb27-25" tabindex="-1"></a>        <span class="ex">/share/data1/mjx/scripts/tools_get_depth_from_sam.single.py</span> <span class="va">$3</span> <span class="op">&gt;</span> <span class="va">$3</span>.depth</span>
<span id="cb27-26"><a href="#cb27-26" tabindex="-1"></a><span class="cf">fi</span></span>
<span id="cb27-27"><a href="#cb27-27" tabindex="-1"></a></span>
<span id="cb27-28"><a href="#cb27-28" tabindex="-1"></a><span class="co">#### run_kraken2.sh ####</span></span>
<span id="cb27-29"><a href="#cb27-29" tabindex="-1"></a><span class="co">#!/usr/bin/bash</span></span>
<span id="cb27-30"><a href="#cb27-30" tabindex="-1"></a><span class="bu">shopt</span> <span class="at">-s</span> expand_aliases</span>
<span id="cb27-31"><a href="#cb27-31" tabindex="-1"></a></span>
<span id="cb27-32"><a href="#cb27-32" tabindex="-1"></a><span class="kw">(</span> <span class="bu">[</span> <span class="va">$#</span> <span class="ot">-lt</span> 3 <span class="bu">]</span><span class="kw">)</span> <span class="kw">&amp;&amp;</span> <span class="kw">{</span> <span class="bu">echo</span> <span class="st">&quot;Usage: </span><span class="va">$0</span><span class="st"> [fq|fq1,fq2] [kk2_db] [out_prefix] [read_len:-150]&quot;</span> <span class="kw">&amp;&amp;</span> <span class="bu">exit</span> 2<span class="kw">;</span> <span class="kw">}</span></span>
<span id="cb27-33"><a href="#cb27-33" tabindex="-1"></a></span>
<span id="cb27-34"><a href="#cb27-34" tabindex="-1"></a><span class="va">fq</span><span class="op">=</span><span class="va">$1</span><span class="kw">;</span> <span class="va">db</span><span class="op">=</span><span class="va">$2</span><span class="kw">;</span> <span class="va">out</span><span class="op">=</span><span class="va">$3</span><span class="kw">;</span> <span class="va">len</span><span class="op">=</span><span class="va">${4</span><span class="op">:-</span>150<span class="va">}</span><span class="kw">;</span> <span class="va">trds</span><span class="op">=</span>16</span>
<span id="cb27-35"><a href="#cb27-35" tabindex="-1"></a><span class="bu">alias</span> kraken2=/share/data1/software/miniconda3/envs/kraken2/bin/kraken2</span>
<span id="cb27-36"><a href="#cb27-36" tabindex="-1"></a><span class="bu">alias</span> bracken=/share/data1/software/miniconda3/envs/kraken2/bin/bracken</span>
<span id="cb27-37"><a href="#cb27-37" tabindex="-1"></a></span>
<span id="cb27-38"><a href="#cb27-38" tabindex="-1"></a><span class="kw">(</span> <span class="bu">[</span> <span class="ot">-f</span> <span class="va">$out</span>.log <span class="bu">]</span> <span class="kw">&amp;&amp;</span> <span class="fu">grep</span> <span class="at">-q</span> <span class="st">&#39;Bracken complete&#39;</span> <span class="va">$out</span>.log <span class="kw">)</span> <span class="kw">&amp;&amp;</span> <span class="dt">\</span></span>
<span id="cb27-39"><a href="#cb27-39" tabindex="-1"></a>    <span class="bu">echo</span> <span class="at">-e</span> <span class="st">&quot;[</span><span class="va">$(</span><span class="fu">date</span> +%Y-%m-%d<span class="dt">\ </span>%H:%M:%S<span class="va">)</span><span class="st">] Skip sample: </span><span class="va">${out</span><span class="op">##</span><span class="pp">*</span>/<span class="va">}</span><span class="st">.&quot;</span> <span class="kw">&amp;&amp;</span> <span class="bu">exit</span> 0</span>
<span id="cb27-40"><a href="#cb27-40" tabindex="-1"></a></span>
<span id="cb27-41"><a href="#cb27-41" tabindex="-1"></a><span class="cf">if</span> <span class="kw">[[</span> <span class="va">$fq</span> <span class="ot">=~</span> <span class="st">&quot;,&quot;</span> <span class="kw">]];</span><span class="cf">then</span></span>
<span id="cb27-42"><a href="#cb27-42" tabindex="-1"></a>    <span class="va">fq1</span><span class="op">=</span><span class="va">$(</span><span class="bu">echo</span> <span class="va">$fq</span> <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-d</span> <span class="st">&quot;,&quot;</span> <span class="at">-f1</span><span class="va">)</span></span>
<span id="cb27-43"><a href="#cb27-43" tabindex="-1"></a>    <span class="va">fq2</span><span class="op">=</span><span class="va">$(</span><span class="bu">echo</span> <span class="va">$fq</span> <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-d</span> <span class="st">&quot;,&quot;</span> <span class="at">-f2</span><span class="va">)</span></span>
<span id="cb27-44"><a href="#cb27-44" tabindex="-1"></a>    <span class="ex">kraken2</span> <span class="at">--db</span> <span class="va">$db</span> <span class="at">--output</span> <span class="va">$out</span>.kk2_out <span class="at">--report</span> <span class="va">$out</span>.kk2_report <span class="at">--threads</span> <span class="va">$trds</span> <span class="at">--use-names</span> <span class="at">--gzip-compressed</span> <span class="va">$fq1</span> <span class="va">$fq2</span> <span class="op">&gt;&gt;</span> <span class="va">$out</span>.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;&amp;</span> <span class="fu">rm</span> <span class="va">$out</span>.kk2_out</span>
<span id="cb27-45"><a href="#cb27-45" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb27-46"><a href="#cb27-46" tabindex="-1"></a>    <span class="ex">kraken2</span> <span class="at">--db</span> <span class="va">$db</span> <span class="at">--output</span> <span class="va">$out</span>.kk2_out <span class="at">--report</span> <span class="va">$out</span>.kk2_report <span class="at">--threads</span> <span class="va">$trds</span> <span class="at">--use-names</span> <span class="at">--gzip-compressed</span> <span class="va">$fq</span> <span class="op">&gt;&gt;</span> <span class="va">$out</span>.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;&amp;</span> <span class="fu">rm</span> <span class="va">$out</span>.kk2_out</span>
<span id="cb27-47"><a href="#cb27-47" tabindex="-1"></a><span class="cf">fi</span></span>
<span id="cb27-48"><a href="#cb27-48" tabindex="-1"></a></span>
<span id="cb27-49"><a href="#cb27-49" tabindex="-1"></a><span class="ex">bracken</span> <span class="at">-l</span> P <span class="at">-r</span> <span class="va">$len</span> <span class="at">-d</span> <span class="va">$db</span> <span class="at">-i</span> <span class="va">$out</span>.kk2_report <span class="at">-o</span> <span class="va">$out</span>.bk_out.p <span class="at">-w</span> <span class="va">$out</span>.bk_report.p <span class="op">&gt;&gt;</span> <span class="va">$out</span>.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;&amp;</span> <span class="fu">rm</span> <span class="va">$out</span>.bk_report.p</span>
<span id="cb27-50"><a href="#cb27-50" tabindex="-1"></a><span class="ex">bracken</span> <span class="at">-l</span> F <span class="at">-r</span> <span class="va">$len</span> <span class="at">-d</span> <span class="va">$db</span> <span class="at">-i</span> <span class="va">$out</span>.kk2_report <span class="at">-o</span> <span class="va">$out</span>.bk_out.f <span class="at">-w</span> <span class="va">$out</span>.bk_report.f <span class="op">&gt;&gt;</span> <span class="va">$out</span>.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;&amp;</span> <span class="fu">rm</span> <span class="va">$out</span>.bk_report.f</span>
<span id="cb27-51"><a href="#cb27-51" tabindex="-1"></a><span class="ex">bracken</span> <span class="at">-l</span> G <span class="at">-r</span> <span class="va">$len</span> <span class="at">-d</span> <span class="va">$db</span> <span class="at">-i</span> <span class="va">$out</span>.kk2_report <span class="at">-o</span> <span class="va">$out</span>.bk_out.g <span class="at">-w</span> <span class="va">$out</span>.bk_report.g <span class="op">&gt;&gt;</span> <span class="va">$out</span>.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;&amp;</span> <span class="fu">rm</span> <span class="va">$out</span>.bk_report.g</span>
<span id="cb27-52"><a href="#cb27-52" tabindex="-1"></a><span class="ex">bracken</span> <span class="at">-l</span> S <span class="at">-r</span> <span class="va">$len</span> <span class="at">-d</span> <span class="va">$db</span> <span class="at">-i</span> <span class="va">$out</span>.kk2_report <span class="at">-o</span> <span class="va">$out</span>.bk_out.s <span class="at">-w</span> <span class="va">$out</span>.bk_report.s <span class="op">&gt;&gt;</span> <span class="va">$out</span>.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;&amp;</span> <span class="fu">rm</span> <span class="va">$out</span>.bk_report.s</span>
<span id="cb27-53"><a href="#cb27-53" tabindex="-1"></a><span class="ex">bracken</span> <span class="at">-l</span> S1 <span class="at">-r</span> <span class="va">$len</span> <span class="at">-d</span> <span class="va">$db</span> <span class="at">-i</span> <span class="va">$out</span>.kk2_report <span class="at">-o</span> <span class="va">$out</span>.bk_out.t <span class="at">-w</span> <span class="va">$out</span>.bk_report.t <span class="op">&gt;&gt;</span> <span class="va">$out</span>.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">&amp;&amp;</span> <span class="fu">rm</span> <span class="va">$out</span>.bk_report.t</span>
<span id="cb27-54"><a href="#cb27-54" tabindex="-1"></a></span>
<span id="cb27-55"><a href="#cb27-55" tabindex="-1"></a><span class="co">#### cat run_bcftools.sh ####</span></span>
<span id="cb27-56"><a href="#cb27-56" tabindex="-1"></a><span class="co">#!/usr/bin/bash</span></span>
<span id="cb27-57"><a href="#cb27-57" tabindex="-1"></a></span>
<span id="cb27-58"><a href="#cb27-58" tabindex="-1"></a><span class="kw">(</span> <span class="bu">[</span> <span class="va">$#</span> <span class="ot">-lt</span> 3 <span class="bu">]</span> <span class="kw">)</span> <span class="kw">&amp;&amp;</span>  <span class="kw">{</span> <span class="bu">echo</span> <span class="at">-e</span> <span class="st">&quot;Usage: </span><span class="va">$0</span><span class="st"> [sort.bam] [ref.fas] [out_prefix] [threads]&quot;</span> <span class="kw">&amp;&amp;</span> <span class="bu">exit</span> 2<span class="kw">;</span> <span class="kw">}</span></span>
<span id="cb27-59"><a href="#cb27-59" tabindex="-1"></a><span class="va">trds</span><span class="op">=</span><span class="va">${4</span><span class="op">:-</span>8<span class="va">}</span></span>
<span id="cb27-60"><a href="#cb27-60" tabindex="-1"></a></span>
<span id="cb27-61"><a href="#cb27-61" tabindex="-1"></a><span class="co"># snp calling, MAPQ &gt; 20, baseQ &gt; 30</span></span>
<span id="cb27-62"><a href="#cb27-62" tabindex="-1"></a><span class="co"># snp filter, only SNP site, QUAL ≥ 30; sum of valid reads ≥ 10.</span></span>
<span id="cb27-63"><a href="#cb27-63" tabindex="-1"></a><span class="ex">bcftools</span> mpileup <span class="at">--threads</span> <span class="va">$trds</span> <span class="at">-q</span> 20 <span class="at">-Q</span> 30 <span class="at">-f</span> <span class="va">$2</span> <span class="va">$1</span> <span class="dv">2</span><span class="op">&gt;</span>/dev/null <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb27-64"><a href="#cb27-64" tabindex="-1"></a>    <span class="ex">bcftools</span> call <span class="at">--threads</span> <span class="va">$trds</span> <span class="at">-mv</span> <span class="at">--ploidy</span> 1 <span class="dv">2</span><span class="op">&gt;</span>/dev/null <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb27-65"><a href="#cb27-65" tabindex="-1"></a>    <span class="ex">bcftools</span> filter <span class="at">--threads</span> <span class="va">$trds</span> <span class="at">-e</span> <span class="st">&#39;QUAL&lt;30 || DP&lt;10 || (INFO/DP4[0]+INFO/DP4[1]+INFO/DP4[2]+INFO/DP4[3]) &lt; 10 || INDEL=1&#39;</span> <span class="at">-Oz</span> <span class="at">-o</span> <span class="va">$3</span>.vcf.gz</span>
<span id="cb27-66"><a href="#cb27-66" tabindex="-1"></a>    </span></code></pre></div>
</div>
<div id="perl-scripts" class="section level2">
<h2>perl scripts</h2>
<div class="sourceCode" id="cb28"><pre class="sourceCode pl"><code class="sourceCode perl"><span id="cb28-1"><a href="#cb28-1" tabindex="-1"></a><span class="co">#### generate_mapping_cfg.pl ####</span></span>
<span id="cb28-2"><a href="#cb28-2" tabindex="-1"></a><span class="kw">#!/usr/bin/perl</span></span>
<span id="cb28-3"><a href="#cb28-3" tabindex="-1"></a><span class="co"># author: Jinxin Meng</span></span>
<span id="cb28-4"><a href="#cb28-4" tabindex="-1"></a><span class="co"># created date: 2024-09-02, 20:48:19</span></span>
<span id="cb28-5"><a href="#cb28-5" tabindex="-1"></a><span class="co"># modified date: 2024-09-02, 20:48:19</span></span>
<span id="cb28-6"><a href="#cb28-6" tabindex="-1"></a></span>
<span id="cb28-7"><a href="#cb28-7" tabindex="-1"></a><span class="fu">use</span> <span class="kw">warnings</span>;</span>
<span id="cb28-8"><a href="#cb28-8" tabindex="-1"></a><span class="fu">use</span> <span class="kw">strict</span>;</span>
<span id="cb28-9"><a href="#cb28-9" tabindex="-1"></a></span>
<span id="cb28-10"><a href="#cb28-10" tabindex="-1"></a><span class="fu">die</span> <span class="ot">&quot;</span><span class="st">Usage: perl </span><span class="wa">$0</span><span class="st"> [sample] [N]</span><span class="ch">\n</span><span class="ot">&quot;</span> <span class="kw">unless</span> <span class="wa">@ARGV</span> <span class="ot">eq</span> <span class="dv">2</span>;</span>
<span id="cb28-11"><a href="#cb28-11" tabindex="-1"></a><span class="kw">my</span> (<span class="dt">$smps</span>, <span class="dt">$n</span>) = <span class="wa">@ARGV</span>; </span>
<span id="cb28-12"><a href="#cb28-12" tabindex="-1"></a><span class="kw">my</span> <span class="dt">$dist_dir</span> = <span class="ot">&quot;</span><span class="st">/share/data1/mjx/proj/07.pigs_binning_mcvg_20230929/00.mash_s1/dist_sort</span><span class="ot">&quot;</span>;</span>
<span id="cb28-13"><a href="#cb28-13" tabindex="-1"></a><span class="kw">my</span> <span class="dt">$filepath</span> = <span class="ot">&quot;</span><span class="st">/share/data1/mjx/proj/07.pigs_binning_mcvg_20230929/00.mash_s1/samples_3997.filelist</span><span class="ot">&quot;</span>;</span>
<span id="cb28-14"><a href="#cb28-14" tabindex="-1"></a></span>
<span id="cb28-15"><a href="#cb28-15" tabindex="-1"></a><span class="kw">my</span> <span class="dt">$h</span>;</span>
<span id="cb28-16"><a href="#cb28-16" tabindex="-1"></a><span class="fu">open</span> I, <span class="ot">&quot;</span><span class="dt">$filepath</span><span class="ot">&quot;</span>;</span>
<span id="cb28-17"><a href="#cb28-17" tabindex="-1"></a><span class="kw">while</span>(<span class="kw">&lt;I&gt;</span>){</span>
<span id="cb28-18"><a href="#cb28-18" tabindex="-1"></a>    <span class="fu">chomp</span>;</span>
<span id="cb28-19"><a href="#cb28-19" tabindex="-1"></a>    <span class="kw">my</span> <span class="dt">@s</span> = <span class="fu">split</span><span class="ot">/\t/</span>;</span>
<span id="cb28-20"><a href="#cb28-20" tabindex="-1"></a>    <span class="dt">$h</span>-&gt;{<span class="dt">$s</span>[<span class="dv">1</span>]}-&gt;{contigs} = <span class="dt">$s</span>[<span class="dv">2</span>];</span>
<span id="cb28-21"><a href="#cb28-21" tabindex="-1"></a>    <span class="dt">$h</span>-&gt;{<span class="dt">$s</span>[<span class="dv">1</span>]}-&gt;{fq} = <span class="dt">$s</span>[<span class="dv">3</span>];</span>
<span id="cb28-22"><a href="#cb28-22" tabindex="-1"></a>}</span>
<span id="cb28-23"><a href="#cb28-23" tabindex="-1"></a><span class="fu">open</span> I, <span class="ot">&quot;</span><span class="dt">$dist_dir</span><span class="st">/</span><span class="dt">$smps</span><span class="st">.dist</span><span class="ot">&quot;</span>;</span>
<span id="cb28-24"><a href="#cb28-24" tabindex="-1"></a><span class="kw">my</span> <span class="dt">$x</span> = <span class="dv">1</span>;</span>
<span id="cb28-25"><a href="#cb28-25" tabindex="-1"></a><span class="kw">while</span>(<span class="kw">&lt;I&gt;</span>){</span>
<span id="cb28-26"><a href="#cb28-26" tabindex="-1"></a>    <span class="fu">chomp</span>;</span>
<span id="cb28-27"><a href="#cb28-27" tabindex="-1"></a>    <span class="kw">last</span> <span class="kw">if</span> <span class="dt">$x</span> &gt; <span class="dt">$n</span>;</span>
<span id="cb28-28"><a href="#cb28-28" tabindex="-1"></a>    <span class="wa">$_</span> =~ <span class="ot">/contigs\/</span><span class="ch">(</span><span class="bn">\S</span><span class="ch">+?)</span><span class="ot">.m1500.</span><span class="ch">*</span><span class="ot">contigs\/</span><span class="ch">(</span><span class="bn">\S</span><span class="ch">+?)</span><span class="ot">.m1500/</span>;</span>
<span id="cb28-29"><a href="#cb28-29" tabindex="-1"></a>    <span class="kw">my</span> (<span class="dt">$ref</span>, <span class="dt">$query</span>) = (<span class="wa">$2</span>, <span class="wa">$1</span>);</span>
<span id="cb28-30"><a href="#cb28-30" tabindex="-1"></a>    <span class="kw">my</span> <span class="dt">$contigs</span> = <span class="dt">$h</span>-&gt;{<span class="dt">$ref</span>}-&gt;{contigs};</span>
<span id="cb28-31"><a href="#cb28-31" tabindex="-1"></a>    <span class="kw">my</span> <span class="dt">$fq</span> = <span class="dt">$h</span>-&gt;{<span class="dt">$query</span>}-&gt;{fq};</span>
<span id="cb28-32"><a href="#cb28-32" tabindex="-1"></a>    <span class="fu">print</span> <span class="ot">&quot;</span><span class="dt">$ref</span><span class="ch">\t</span><span class="dt">$query</span><span class="ch">\t</span><span class="dt">$contigs</span><span class="ch">\t</span><span class="dt">$fq</span><span class="ch">\n</span><span class="ot">&quot;</span>;</span>
<span id="cb28-33"><a href="#cb28-33" tabindex="-1"></a>    <span class="dt">$x</span>++;</span>
<span id="cb28-34"><a href="#cb28-34" tabindex="-1"></a>}</span>
<span id="cb28-35"><a href="#cb28-35" tabindex="-1"></a></span>
<span id="cb28-36"><a href="#cb28-36" tabindex="-1"></a><span class="co">#### run_depth_combine.pl ####</span></span>
<span id="cb28-37"><a href="#cb28-37" tabindex="-1"></a><span class="kw">#!/usr/bin/perl</span></span>
<span id="cb28-38"><a href="#cb28-38" tabindex="-1"></a></span>
<span id="cb28-39"><a href="#cb28-39" tabindex="-1"></a><span class="fu">use</span> <span class="kw">warnings</span>;</span>
<span id="cb28-40"><a href="#cb28-40" tabindex="-1"></a><span class="fu">use</span> <span class="kw">strict</span>;</span>
<span id="cb28-41"><a href="#cb28-41" tabindex="-1"></a></span>
<span id="cb28-42"><a href="#cb28-42" tabindex="-1"></a><span class="fu">die</span> <span class="ot">&quot;</span><span class="st">Usage: perl </span><span class="wa">$0</span><span class="st"> [depth.filelist] [sample] [top_n] [out]</span><span class="ch">\n</span><span class="ot">&quot;</span> <span class="kw">unless</span> <span class="wa">@ARGV</span> <span class="ot">eq</span> <span class="dv">4</span>;</span>
<span id="cb28-43"><a href="#cb28-43" tabindex="-1"></a><span class="kw">my</span> (<span class="dt">$dep</span>, <span class="dt">$smps</span>, <span class="dt">$n</span>, <span class="dt">$out</span>) = <span class="wa">@ARGV</span>;</span>
<span id="cb28-44"><a href="#cb28-44" tabindex="-1"></a></span>
<span id="cb28-45"><a href="#cb28-45" tabindex="-1"></a><span class="kw">my</span> <span class="dt">$h</span>;</span>
<span id="cb28-46"><a href="#cb28-46" tabindex="-1"></a><span class="fu">open</span> I, <span class="ot">&quot;</span><span class="dt">$dep</span><span class="ot">&quot;</span>;</span>
<span id="cb28-47"><a href="#cb28-47" tabindex="-1"></a><span class="kw">while</span>(<span class="kw">&lt;I&gt;</span>){</span>
<span id="cb28-48"><a href="#cb28-48" tabindex="-1"></a>    <span class="fu">chomp</span>;</span>
<span id="cb28-49"><a href="#cb28-49" tabindex="-1"></a>    <span class="kw">my</span> <span class="dt">@s</span> = <span class="fu">split</span><span class="ot">/</span><span class="bn">\s</span><span class="ot">/</span>;</span>
<span id="cb28-50"><a href="#cb28-50" tabindex="-1"></a>    <span class="dt">$h</span>-&gt;{<span class="dt">$s</span>[<span class="dv">0</span>]} = <span class="dt">$s</span>[<span class="dv">1</span>];</span>
<span id="cb28-51"><a href="#cb28-51" tabindex="-1"></a>}</span>
<span id="cb28-52"><a href="#cb28-52" tabindex="-1"></a></span>
<span id="cb28-53"><a href="#cb28-53" tabindex="-1"></a><span class="fu">open</span> I, <span class="ot">&quot;</span><span class="st">/share/data1/mjx/proj/07.pigs_binning_mcvg_20230929/01.depth_s1/generate_mapping_cfg.pl </span><span class="dt">$smps</span><span class="st"> </span><span class="dt">$n</span><span class="st"> |</span><span class="ot">&quot;</span>;</span>
<span id="cb28-54"><a href="#cb28-54" tabindex="-1"></a><span class="kw">my</span> <span class="dt">@depth</span>;</span>
<span id="cb28-55"><a href="#cb28-55" tabindex="-1"></a><span class="kw">my</span> <span class="dt">$contig</span>;</span>
<span id="cb28-56"><a href="#cb28-56" tabindex="-1"></a><span class="kw">while</span>(<span class="kw">&lt;I&gt;</span>){</span>
<span id="cb28-57"><a href="#cb28-57" tabindex="-1"></a>    <span class="fu">chomp</span>;</span>
<span id="cb28-58"><a href="#cb28-58" tabindex="-1"></a>    <span class="kw">my</span> <span class="dt">@s</span> = <span class="fu">split</span> <span class="ot">/\t/</span>;</span>
<span id="cb28-59"><a href="#cb28-59" tabindex="-1"></a>    <span class="dt">$contig</span> = <span class="dt">$s</span>[<span class="dv">2</span>] <span class="kw">if</span> <span class="dt">$s</span>[<span class="dv">0</span>] <span class="ot">eq</span> <span class="dt">$s</span>[<span class="dv">1</span>];</span>
<span id="cb28-60"><a href="#cb28-60" tabindex="-1"></a>    <span class="kw">my</span> <span class="dt">$x</span> = <span class="ot">&quot;</span><span class="dt">$s</span><span class="st">[0]_</span><span class="dt">$s</span><span class="st">[1]</span><span class="ot">&quot;</span>;</span>
<span id="cb28-61"><a href="#cb28-61" tabindex="-1"></a>    <span class="fu">push</span> <span class="dt">@depth</span>, <span class="dt">$h</span>-&gt;{<span class="dt">$x</span>};</span>
<span id="cb28-62"><a href="#cb28-62" tabindex="-1"></a>}</span>
<span id="cb28-63"><a href="#cb28-63" tabindex="-1"></a></span>
<span id="cb28-64"><a href="#cb28-64" tabindex="-1"></a><span class="kw">my</span> <span class="dt">$cmd</span> = <span class="ot">&quot;</span><span class="st">binning_depth_combine.pl </span><span class="dt">$contig</span><span class="st"> </span><span class="ot">&quot;</span>.<span class="fu">join</span>(<span class="ot">&quot;</span><span class="st"> </span><span class="ot">&quot;</span>, <span class="dt">@depth</span>).<span class="ot">&quot;</span><span class="st"> &gt;</span><span class="dt">$out</span><span class="st">.depth</span><span class="ot">&quot;</span>;</span>
<span id="cb28-65"><a href="#cb28-65" tabindex="-1"></a><span class="fu">system</span>(<span class="ot">&quot;</span><span class="dt">$cmd</span><span class="ot">&quot;</span>);</span>
<span id="cb28-66"><a href="#cb28-66" tabindex="-1"></a><span class="co">#### parse_dRep.pl ####</span></span>
<span id="cb28-67"><a href="#cb28-67" tabindex="-1"></a><span class="kw">#!/usr/bin/perl</span></span>
<span id="cb28-68"><a href="#cb28-68" tabindex="-1"></a><span class="co"># Jinxin Meng, jinxmeng@zju.edu.cn</span></span>
<span id="cb28-69"><a href="#cb28-69" tabindex="-1"></a><span class="co"># created date: 2023-10-02, 01:49:47</span></span>
<span id="cb28-70"><a href="#cb28-70" tabindex="-1"></a><span class="co"># modified date: 2025-03-06, 08:32:39</span></span>
<span id="cb28-71"><a href="#cb28-71" tabindex="-1"></a><span class="fu">use</span> <span class="kw">warnings</span>;</span>
<span id="cb28-72"><a href="#cb28-72" tabindex="-1"></a><span class="fu">use</span> <span class="kw">strict</span>;</span>
<span id="cb28-73"><a href="#cb28-73" tabindex="-1"></a><span class="fu">use</span> <span class="fu">Getopt::Long</span>;</span>
<span id="cb28-74"><a href="#cb28-74" tabindex="-1"></a></span>
<span id="cb28-75"><a href="#cb28-75" tabindex="-1"></a><span class="fu">die</span> <span class="ot">&quot;</span><span class="st">Usage: perl </span><span class="wa">$0</span><span class="st"> [Cdb.tsv] [quality_report] [out_file]</span><span class="ch">\n</span><span class="ot">&quot;</span> <span class="kw">unless</span> <span class="wa">@ARGV</span> <span class="ot">eq</span> <span class="dv">3</span>;</span>
<span id="cb28-76"><a href="#cb28-76" tabindex="-1"></a></span>
<span id="cb28-77"><a href="#cb28-77" tabindex="-1"></a><span class="kw">my</span> (<span class="dt">$cdb</span>, <span class="dt">$qs</span>, <span class="dt">$out</span>) = <span class="wa">@ARGV</span>;</span>
<span id="cb28-78"><a href="#cb28-78" tabindex="-1"></a><span class="kw">my</span> (<span class="dt">%clu</span>, <span class="dt">@name</span>, <span class="dt">%value</span>) = (); </span>
<span id="cb28-79"><a href="#cb28-79" tabindex="-1"></a></span>
<span id="cb28-80"><a href="#cb28-80" tabindex="-1"></a><span class="fu">open</span> I, <span class="ot">&quot;</span><span class="st">&lt;</span><span class="dt">$cdb</span><span class="ot">&quot;</span>;</span>
<span id="cb28-81"><a href="#cb28-81" tabindex="-1"></a><span class="kw">while</span> (<span class="kw">&lt;I&gt;</span>) {</span>
<span id="cb28-82"><a href="#cb28-82" tabindex="-1"></a>    <span class="fu">chomp</span>;</span>
<span id="cb28-83"><a href="#cb28-83" tabindex="-1"></a>    <span class="kw">next</span> <span class="kw">if</span> <span class="ot">/primary_cluster/</span>;</span>
<span id="cb28-84"><a href="#cb28-84" tabindex="-1"></a>    <span class="kw">my</span> <span class="dt">@s</span> = <span class="fu">split</span> <span class="ot">/\t/</span>;</span>
<span id="cb28-85"><a href="#cb28-85" tabindex="-1"></a>    <span class="dt">$s</span>[<span class="dv">0</span>] =~ <span class="ot">s/.</span><span class="ch">(</span><span class="ot">fa</span><span class="ch">|</span><span class="ot">fna</span><span class="ch">)</span><span class="ot">//</span>;</span>
<span id="cb28-86"><a href="#cb28-86" tabindex="-1"></a>    <span class="dt">$s</span>[<span class="dv">0</span>] =~ <span class="ot">s/.</span><span class="ch">*</span><span class="ot">\///</span>;</span>
<span id="cb28-87"><a href="#cb28-87" tabindex="-1"></a>    <span class="fu">push</span> <span class="dt">@</span>{<span class="dt">$clu</span>{<span class="dt">$s</span>[<span class="dv">3</span>]}}, <span class="dt">$s</span>[<span class="dv">0</span>];</span>
<span id="cb28-88"><a href="#cb28-88" tabindex="-1"></a>    <span class="fu">push</span> <span class="dt">@name</span>, <span class="dt">$s</span>[<span class="dv">0</span>];</span>
<span id="cb28-89"><a href="#cb28-89" tabindex="-1"></a>}</span>
<span id="cb28-90"><a href="#cb28-90" tabindex="-1"></a></span>
<span id="cb28-91"><a href="#cb28-91" tabindex="-1"></a><span class="fu">open</span> I,<span class="ot">&quot;</span><span class="st">&lt;</span><span class="dt">$qs</span><span class="ot">&quot;</span>;</span>
<span id="cb28-92"><a href="#cb28-92" tabindex="-1"></a><span class="kw">while</span> (<span class="kw">&lt;I&gt;</span>) {</span>
<span id="cb28-93"><a href="#cb28-93" tabindex="-1"></a>    <span class="fu">chomp</span>;</span>
<span id="cb28-94"><a href="#cb28-94" tabindex="-1"></a>    <span class="kw">next</span> <span class="kw">if</span> <span class="ot">/Completeness/</span>;</span>
<span id="cb28-95"><a href="#cb28-95" tabindex="-1"></a>    <span class="kw">my</span> <span class="dt">@s</span> = <span class="fu">split</span> <span class="ot">/\t/</span>;</span>
<span id="cb28-96"><a href="#cb28-96" tabindex="-1"></a>    <span class="dt">$value</span>{<span class="dt">$s</span>[<span class="dv">0</span>]} = <span class="dt">$s</span>[<span class="dv">1</span>] - <span class="dv">5</span> <span class="ot">*</span> <span class="dt">$s</span>[<span class="dv">2</span>] <span class="kw">if</span> (<span class="fu">grep</span> {<span class="wa">$_</span> <span class="ot">eq</span> <span class="dt">$s</span>[<span class="dv">0</span>]} <span class="dt">@name</span>);</span>
<span id="cb28-97"><a href="#cb28-97" tabindex="-1"></a>}</span>
<span id="cb28-98"><a href="#cb28-98" tabindex="-1"></a></span>
<span id="cb28-99"><a href="#cb28-99" tabindex="-1"></a><span class="fu">open</span> O, <span class="ot">&quot;</span><span class="st">&gt;</span><span class="dt">$out</span><span class="ot">&quot;</span>;</span>
<span id="cb28-100"><a href="#cb28-100" tabindex="-1"></a>select_rep(\<span class="dt">%clu</span>, \<span class="dt">%value</span>); <span class="co"># 调用子程序并传递哈希的引用</span></span>
<span id="cb28-101"><a href="#cb28-101" tabindex="-1"></a><span class="fu">close</span> O; </span>
<span id="cb28-102"><a href="#cb28-102" tabindex="-1"></a></span>
<span id="cb28-103"><a href="#cb28-103" tabindex="-1"></a><span class="kw">sub </span><span class="fu">select_rep</span> {</span>
<span id="cb28-104"><a href="#cb28-104" tabindex="-1"></a>    <span class="kw">my</span> (<span class="dt">$clu</span>, <span class="dt">$value</span>) = <span class="dt">@_</span>;  <span class="co"># 获取哈希的引用</span></span>
<span id="cb28-105"><a href="#cb28-105" tabindex="-1"></a>    <span class="kw">my</span> <span class="dt">%clu</span> = <span class="dt">%</span>{<span class="dt">$clu</span>};   <span class="co"># 将哈希引用解引用为哈希</span></span>
<span id="cb28-106"><a href="#cb28-106" tabindex="-1"></a>    <span class="kw">my</span> <span class="dt">%value</span> = <span class="dt">%</span>{<span class="dt">$value</span>};       <span class="co"># 将哈希引用解引用为哈希</span></span>
<span id="cb28-107"><a href="#cb28-107" tabindex="-1"></a>    <span class="kw">for</span> <span class="kw">my</span> <span class="dt">$i</span> (<span class="fu">sort</span> <span class="fu">keys</span> <span class="dt">%clu</span>) {</span>
<span id="cb28-108"><a href="#cb28-108" tabindex="-1"></a>        <span class="kw">my</span> <span class="dt">$count</span> = <span class="fu">scalar</span> <span class="dt">@</span>{<span class="dt">$clu</span>{<span class="dt">$i</span>}};</span>
<span id="cb28-109"><a href="#cb28-109" tabindex="-1"></a>        <span class="kw">if</span> (<span class="dt">$count</span> == <span class="dv">1</span>) {</span>
<span id="cb28-110"><a href="#cb28-110" tabindex="-1"></a>            <span class="fu">print</span> O <span class="ot">&quot;</span><span class="st">clu.</span><span class="dt">$i</span><span class="ch">\t</span><span class="st">1</span><span class="ch">\t</span><span class="dt">@</span><span class="st">{</span><span class="dt">$clu</span><span class="st">{</span><span class="dt">$i</span><span class="st">}}</span><span class="ch">\t</span><span class="dt">@</span><span class="st">{</span><span class="dt">$clu</span><span class="st">{</span><span class="dt">$i</span><span class="st">}}</span><span class="ch">\n</span><span class="ot">&quot;</span>;</span>
<span id="cb28-111"><a href="#cb28-111" tabindex="-1"></a>        } <span class="kw">else</span> {</span>
<span id="cb28-112"><a href="#cb28-112" tabindex="-1"></a>            <span class="kw">my</span> <span class="dt">$max</span> = -<span class="dv">10000</span>;</span>
<span id="cb28-113"><a href="#cb28-113" tabindex="-1"></a>            <span class="kw">my</span> <span class="dt">$rep</span> = <span class="ot">&quot;&quot;</span>;</span>
<span id="cb28-114"><a href="#cb28-114" tabindex="-1"></a>            <span class="kw">for</span> <span class="kw">my</span> <span class="dt">$j</span> (<span class="dt">@</span>{<span class="dt">$clu</span>{<span class="dt">$i</span>}}) {</span>
<span id="cb28-115"><a href="#cb28-115" tabindex="-1"></a>                <span class="kw">if</span> (<span class="dt">$value</span>{<span class="dt">$j</span>} &gt;= <span class="dt">$max</span>) {</span>
<span id="cb28-116"><a href="#cb28-116" tabindex="-1"></a>                    <span class="dt">$max</span> = <span class="dt">$value</span>{<span class="dt">$j</span>};</span>
<span id="cb28-117"><a href="#cb28-117" tabindex="-1"></a>                    <span class="dt">$rep</span> = <span class="dt">$j</span>;</span>
<span id="cb28-118"><a href="#cb28-118" tabindex="-1"></a>                }</span>
<span id="cb28-119"><a href="#cb28-119" tabindex="-1"></a>            }</span>
<span id="cb28-120"><a href="#cb28-120" tabindex="-1"></a>            <span class="fu">print</span> O <span class="ot">&quot;</span><span class="st">clu.</span><span class="dt">$i</span><span class="ch">\t</span><span class="dt">$count</span><span class="ch">\t</span><span class="dt">$rep</span><span class="ch">\t</span><span class="ot">&quot;</span>.<span class="fu">join</span>(<span class="ot">&quot;</span><span class="st">,</span><span class="ot">&quot;</span>, <span class="dt">@</span>{<span class="dt">$clu</span>{<span class="dt">$i</span>}}).<span class="ot">&quot;</span><span class="ch">\n</span><span class="ot">&quot;</span>;</span>
<span id="cb28-121"><a href="#cb28-121" tabindex="-1"></a>        }   </span>
<span id="cb28-122"><a href="#cb28-122" tabindex="-1"></a>    }</span>
<span id="cb28-123"><a href="#cb28-123" tabindex="-1"></a>}</span>
<span id="cb28-124"><a href="#cb28-124" tabindex="-1"></a><span class="co">#### parse_CAZy_blast_for_geneset.pl ####</span></span>
<span id="cb28-125"><a href="#cb28-125" tabindex="-1"></a><span class="kw">#!/usr/bin/perl</span></span>
<span id="cb28-126"><a href="#cb28-126" tabindex="-1"></a><span class="co"># Jinxin Meng, jinxmeng@zju.edu.cn</span></span>
<span id="cb28-127"><a href="#cb28-127" tabindex="-1"></a><span class="co"># created date: 2023-05-06, 12:01:05</span></span>
<span id="cb28-128"><a href="#cb28-128" tabindex="-1"></a><span class="co"># modified date: 2025-08-24, 09:34:17</span></span>
<span id="cb28-129"><a href="#cb28-129" tabindex="-1"></a><span class="fu">use</span> <span class="kw">warnings</span>;</span>
<span id="cb28-130"><a href="#cb28-130" tabindex="-1"></a><span class="fu">use</span> <span class="kw">strict</span>;</span>
<span id="cb28-131"><a href="#cb28-131" tabindex="-1"></a><span class="fu">use</span> POSIX;</span>
<span id="cb28-132"><a href="#cb28-132" tabindex="-1"></a></span>
<span id="cb28-133"><a href="#cb28-133" tabindex="-1"></a><span class="fu">die</span> <span class="ot">&quot;</span><span class="st">Usage: perl </span><span class="wa">$0</span><span class="st"> [CAZymes blast for geneset] [out_prefix]</span><span class="ch">\n</span><span class="ot">&quot;</span> <span class="kw">unless</span> <span class="wa">@ARGV</span> <span class="ot">eq</span> <span class="dv">2</span>;</span>
<span id="cb28-134"><a href="#cb28-134" tabindex="-1"></a><span class="kw">my</span> (<span class="dt">$in</span>, <span class="dt">$out</span>) = <span class="wa">@ARGV</span>;</span>
<span id="cb28-135"><a href="#cb28-135" tabindex="-1"></a><span class="kw">my</span> (<span class="dt">$x</span>, <span class="dt">%all_CAZymes</span>, <span class="dt">%all_CAZyfams</span>) = (<span class="ot">&quot;&quot;</span>);</span>
<span id="cb28-136"><a href="#cb28-136" tabindex="-1"></a></span>
<span id="cb28-137"><a href="#cb28-137" tabindex="-1"></a><span class="fu">open</span> I, <span class="ot">&quot;</span><span class="st">&lt;</span><span class="dt">$in</span><span class="ot">&quot;</span> <span class="ot">or</span> <span class="fu">die</span> <span class="ot">&quot;</span><span class="st">Can&#39;t open </span><span class="dt">$in</span><span class="st">: </span><span class="wa">$!</span><span class="ch">\n</span><span class="ot">&quot;</span>;</span>
<span id="cb28-138"><a href="#cb28-138" tabindex="-1"></a><span class="fu">open</span> O, <span class="ot">&quot;</span><span class="st">&gt;</span><span class="dt">${out}</span><span class="st">.tsv</span><span class="ot">&quot;</span>; </span>
<span id="cb28-139"><a href="#cb28-139" tabindex="-1"></a><span class="kw">while</span> (<span class="kw">&lt;I&gt;</span>) {</span>
<span id="cb28-140"><a href="#cb28-140" tabindex="-1"></a>    <span class="fu">chomp</span>;</span>
<span id="cb28-141"><a href="#cb28-141" tabindex="-1"></a>    <span class="kw">my</span> <span class="dt">@s</span> = <span class="fu">split</span> <span class="ot">/</span><span class="bn">\s</span><span class="ch">+</span><span class="ot">/</span>;</span>
<span id="cb28-142"><a href="#cb28-142" tabindex="-1"></a>    <span class="kw">if</span> (<span class="dt">$x</span> <span class="ot">ne</span> <span class="dt">$s</span>[<span class="dv">0</span>]) {</span>
<span id="cb28-143"><a href="#cb28-143" tabindex="-1"></a>        <span class="wa">$_</span> =~ <span class="ot">/</span><span class="ch">(</span><span class="bn">\S</span><span class="ch">+?)</span><span class="bn">\s\S</span><span class="ch">+?</span><span class="ot">\|</span><span class="ch">(</span><span class="bn">\S</span><span class="ch">+?)</span><span class="bn">\s</span><span class="ot">/</span>;</span>
<span id="cb28-144"><a href="#cb28-144" tabindex="-1"></a>        <span class="kw">my</span> (<span class="dt">$gene</span>, <span class="dt">$CAZyme</span>) = (<span class="wa">$1</span>, <span class="wa">$2</span>);</span>
<span id="cb28-145"><a href="#cb28-145" tabindex="-1"></a>        <span class="fu">print</span> O <span class="ot">&quot;</span><span class="dt">$gene</span><span class="ch">\t</span><span class="dt">$CAZyme</span><span class="ch">\n</span><span class="ot">&quot;</span>;</span>
<span id="cb28-146"><a href="#cb28-146" tabindex="-1"></a>        <span class="kw">my</span> <span class="dt">@s</span> = <span class="fu">split</span><span class="ot">/\|/</span>, <span class="dt">$CAZyme</span>;</span>
<span id="cb28-147"><a href="#cb28-147" tabindex="-1"></a>        <span class="kw">for</span> <span class="kw">my</span> <span class="dt">$i</span> (<span class="dt">@s</span>) {</span>
<span id="cb28-148"><a href="#cb28-148" tabindex="-1"></a>            <span class="dt">$i</span> =~ <span class="ot">s/_</span><span class="bn">\S</span><span class="ch">+</span><span class="ot">//g</span>;</span>
<span id="cb28-149"><a href="#cb28-149" tabindex="-1"></a>            <span class="dt">$all_CAZymes</span>{<span class="dt">$i</span>} += <span class="dv">1</span>/(<span class="dt">$#s</span>+<span class="dv">1</span>);</span>
<span id="cb28-150"><a href="#cb28-150" tabindex="-1"></a>        }</span>
<span id="cb28-151"><a href="#cb28-151" tabindex="-1"></a>    }</span>
<span id="cb28-152"><a href="#cb28-152" tabindex="-1"></a>    <span class="dt">$x</span> = <span class="dt">$s</span>[<span class="dv">0</span>];    </span>
<span id="cb28-153"><a href="#cb28-153" tabindex="-1"></a>}</span>
<span id="cb28-154"><a href="#cb28-154" tabindex="-1"></a><span class="fu">close</span> O;</span>
<span id="cb28-155"><a href="#cb28-155" tabindex="-1"></a></span>
<span id="cb28-156"><a href="#cb28-156" tabindex="-1"></a><span class="fu">open</span> O, <span class="ot">&quot;</span><span class="st">&gt;</span><span class="dt">${out}</span><span class="st">.stat.tsv</span><span class="ot">&quot;</span>;</span>
<span id="cb28-157"><a href="#cb28-157" tabindex="-1"></a><span class="kw">for</span> <span class="kw">my</span> <span class="dt">$i</span> (<span class="fu">sort</span> <span class="fu">keys</span> <span class="dt">%all_CAZymes</span>) { <span class="co"># count all CAZymes</span></span>
<span id="cb28-158"><a href="#cb28-158" tabindex="-1"></a>    <span class="kw">my</span> <span class="dt">$n</span> = <span class="fu">int</span>(<span class="dt">$all_CAZymes</span>{<span class="dt">$i</span>}*<span class="dv">1000</span>)/<span class="dv">1000</span>;</span>
<span id="cb28-159"><a href="#cb28-159" tabindex="-1"></a>    <span class="fu">print</span> O <span class="ot">&quot;</span><span class="dt">$i</span><span class="ch">\t</span><span class="dt">$n</span><span class="ch">\n</span><span class="ot">&quot;</span>;</span>
<span id="cb28-160"><a href="#cb28-160" tabindex="-1"></a>    <span class="kw">if</span> (<span class="dt">$i</span> =~ <span class="ot">/</span><span class="bn">\S</span><span class="ch">+</span><span class="ot">\.</span><span class="bn">\S</span><span class="ch">+</span><span class="ot">\.</span><span class="bn">\S</span><span class="ch">+</span><span class="ot">/</span>) {</span>
<span id="cb28-161"><a href="#cb28-161" tabindex="-1"></a>        <span class="dt">$all_CAZyfams</span>{<span class="ot">&quot;</span><span class="st">Other</span><span class="ot">&quot;</span>} += <span class="dt">$all_CAZymes</span>{<span class="dt">$i</span>};</span>
<span id="cb28-162"><a href="#cb28-162" tabindex="-1"></a>    } <span class="kw">else</span> {</span>
<span id="cb28-163"><a href="#cb28-163" tabindex="-1"></a>        (<span class="kw">my</span> <span class="dt">$x</span> = <span class="dt">$i</span>) =~ <span class="ot">s/</span><span class="bn">\d</span><span class="ch">+</span><span class="ot">//g</span>;</span>
<span id="cb28-164"><a href="#cb28-164" tabindex="-1"></a>        <span class="dt">$all_CAZyfams</span>{<span class="dt">$x</span>} += <span class="dt">$all_CAZymes</span>{<span class="dt">$i</span>};</span>
<span id="cb28-165"><a href="#cb28-165" tabindex="-1"></a>    }</span>
<span id="cb28-166"><a href="#cb28-166" tabindex="-1"></a>}</span>
<span id="cb28-167"><a href="#cb28-167" tabindex="-1"></a><span class="fu">close</span> O;</span>
<span id="cb28-168"><a href="#cb28-168" tabindex="-1"></a></span>
<span id="cb28-169"><a href="#cb28-169" tabindex="-1"></a><span class="fu">open</span> O, <span class="ot">&quot;</span><span class="st">&gt;</span><span class="dt">${out}</span><span class="st">.fams.stat.tsv</span><span class="ot">&quot;</span>;</span>
<span id="cb28-170"><a href="#cb28-170" tabindex="-1"></a><span class="kw">for</span> <span class="kw">my</span> <span class="dt">$i</span> (<span class="fu">sort</span> <span class="fu">keys</span> <span class="dt">%all_CAZyfams</span>) { <span class="co"># count all CAZyme families</span></span>
<span id="cb28-171"><a href="#cb28-171" tabindex="-1"></a>    <span class="kw">my</span> <span class="dt">$n</span> = <span class="fu">int</span>(<span class="dt">$all_CAZyfams</span>{<span class="dt">$i</span>}*<span class="dv">1000</span>)/<span class="dv">1000</span>;</span>
<span id="cb28-172"><a href="#cb28-172" tabindex="-1"></a>    <span class="fu">print</span> O <span class="ot">&quot;</span><span class="dt">$i</span><span class="ch">\t</span><span class="dt">$n</span><span class="ch">\n</span><span class="ot">&quot;</span>;</span>
<span id="cb28-173"><a href="#cb28-173" tabindex="-1"></a>}</span>
<span id="cb28-174"><a href="#cb28-174" tabindex="-1"></a><span class="fu">close</span> O;</span></code></pre></div>
</div>
<div id="python-scripts" class="section level2">
<h2>python scripts</h2>
<div class="sourceCode" id="cb29"><pre class="sourceCode py"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a><span class="co">#### calcu_tpm_for_KO.py ####</span></span>
<span id="cb29-2"><a href="#cb29-2" tabindex="-1"></a><span class="co">#!/share/data1/software/miniconda3/envs/jinxin/bin/python3</span></span>
<span id="cb29-3"><a href="#cb29-3" tabindex="-1"></a><span class="co"># encoding: utf-8</span></span>
<span id="cb29-4"><a href="#cb29-4" tabindex="-1"></a><span class="co"># author: Jinxin Meng</span></span>
<span id="cb29-5"><a href="#cb29-5" tabindex="-1"></a><span class="co"># created date: 2024-03-12, 23:14:49</span></span>
<span id="cb29-6"><a href="#cb29-6" tabindex="-1"></a><span class="co"># modified date: 2024-03-12, 23:14:49</span></span>
<span id="cb29-7"><a href="#cb29-7" tabindex="-1"></a></span>
<span id="cb29-8"><a href="#cb29-8" tabindex="-1"></a><span class="co">&#39;&#39;&#39;</span></span>
<span id="cb29-9"><a href="#cb29-9" tabindex="-1"></a><span class="co">Jinxin Meng, 20240312</span></span>
<span id="cb29-10"><a href="#cb29-10" tabindex="-1"></a><span class="co">Generate the TPM for each KO by summaring the gene abundance.</span></span>
<span id="cb29-11"><a href="#cb29-11" tabindex="-1"></a><span class="co">Required files:</span></span>
<span id="cb29-12"><a href="#cb29-12" tabindex="-1"></a><span class="co">1. KO annotation with field gene name and feature name with &quot;</span><span class="ch">\t</span><span class="co">&quot; or &quot;\s&quot; delimited</span></span>
<span id="cb29-13"><a href="#cb29-13" tabindex="-1"></a><span class="co">2. gene rc and len provided by samtools coverage [rname, startpos, endpos, numreads, covbases, coverage, meandepth, meanbaseq, meanmapq]</span></span>
<span id="cb29-14"><a href="#cb29-14" tabindex="-1"></a><span class="co">&#39;&#39;&#39;</span></span>
<span id="cb29-15"><a href="#cb29-15" tabindex="-1"></a></span>
<span id="cb29-16"><a href="#cb29-16" tabindex="-1"></a><span class="im">import</span> sys</span>
<span id="cb29-17"><a href="#cb29-17" tabindex="-1"></a></span>
<span id="cb29-18"><a href="#cb29-18" tabindex="-1"></a><span class="cf">if</span> <span class="bu">len</span>(sys.argv) <span class="op">!=</span> <span class="dv">4</span>:</span>
<span id="cb29-19"><a href="#cb29-19" tabindex="-1"></a>    sys.exit(<span class="st">&quot;Usage: calcu_tpm_for_KO.py [in_f &lt;gene ID|feature ID&gt;] [cvg file generated from samtools coverage] [out_f]&quot;</span>)</span>
<span id="cb29-20"><a href="#cb29-20" tabindex="-1"></a></span>
<span id="cb29-21"><a href="#cb29-21" tabindex="-1"></a><span class="kw">def</span> main(in_f, cvg, out_f):</span>
<span id="cb29-22"><a href="#cb29-22" tabindex="-1"></a>    KO <span class="op">=</span> {}</span>
<span id="cb29-23"><a href="#cb29-23" tabindex="-1"></a>    <span class="cf">with</span> <span class="bu">open</span>(in_f, <span class="st">&quot;r&quot;</span>) <span class="im">as</span> f:</span>
<span id="cb29-24"><a href="#cb29-24" tabindex="-1"></a>        <span class="cf">for</span> i <span class="kw">in</span> f:</span>
<span id="cb29-25"><a href="#cb29-25" tabindex="-1"></a>            l <span class="op">=</span> i.strip().split()</span>
<span id="cb29-26"><a href="#cb29-26" tabindex="-1"></a>            KO[l[<span class="dv">0</span>]] <span class="op">=</span> l[<span class="dv">1</span>]</span>
<span id="cb29-27"><a href="#cb29-27" tabindex="-1"></a></span>
<span id="cb29-28"><a href="#cb29-28" tabindex="-1"></a>    total <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb29-29"><a href="#cb29-29" tabindex="-1"></a>    <span class="cf">with</span> <span class="bu">open</span>(cvg, <span class="st">&quot;r&quot;</span>) <span class="im">as</span> f:</span>
<span id="cb29-30"><a href="#cb29-30" tabindex="-1"></a>        f.readline()</span>
<span id="cb29-31"><a href="#cb29-31" tabindex="-1"></a>        <span class="cf">for</span> i <span class="kw">in</span> f:</span>
<span id="cb29-32"><a href="#cb29-32" tabindex="-1"></a>            l <span class="op">=</span> i.strip().split()</span>
<span id="cb29-33"><a href="#cb29-33" tabindex="-1"></a>            <span class="cf">if</span> <span class="bu">int</span>(l[<span class="dv">3</span>]) <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb29-34"><a href="#cb29-34" tabindex="-1"></a>                <span class="cf">continue</span></span>
<span id="cb29-35"><a href="#cb29-35" tabindex="-1"></a>            <span class="co"># if float(l[5]) &lt; 10:</span></span>
<span id="cb29-36"><a href="#cb29-36" tabindex="-1"></a>            <span class="co">#    continue</span></span>
<span id="cb29-37"><a href="#cb29-37" tabindex="-1"></a>            total <span class="op">+=</span> <span class="bu">int</span>(l[<span class="dv">3</span>])<span class="op">/</span><span class="bu">int</span>(l[<span class="dv">2</span>])</span>
<span id="cb29-38"><a href="#cb29-38" tabindex="-1"></a>    </span>
<span id="cb29-39"><a href="#cb29-39" tabindex="-1"></a>    res <span class="op">=</span> {}</span>
<span id="cb29-40"><a href="#cb29-40" tabindex="-1"></a>    <span class="cf">with</span> <span class="bu">open</span>(cvg, <span class="st">&quot;r&quot;</span>) <span class="im">as</span> f:</span>
<span id="cb29-41"><a href="#cb29-41" tabindex="-1"></a>        f.readline()</span>
<span id="cb29-42"><a href="#cb29-42" tabindex="-1"></a>        <span class="cf">for</span> i <span class="kw">in</span> f:</span>
<span id="cb29-43"><a href="#cb29-43" tabindex="-1"></a>            l <span class="op">=</span> i.strip().split()</span>
<span id="cb29-44"><a href="#cb29-44" tabindex="-1"></a>            <span class="cf">if</span> <span class="bu">int</span>(l[<span class="dv">3</span>]) <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb29-45"><a href="#cb29-45" tabindex="-1"></a>                <span class="cf">continue</span></span>
<span id="cb29-46"><a href="#cb29-46" tabindex="-1"></a>            <span class="co"># if float(l[5]) &lt; 10:</span></span>
<span id="cb29-47"><a href="#cb29-47" tabindex="-1"></a>            <span class="co">#     continue</span></span>
<span id="cb29-48"><a href="#cb29-48" tabindex="-1"></a>            tpm <span class="op">=</span> (<span class="bu">int</span>(l[<span class="dv">3</span>])<span class="op">/</span><span class="bu">int</span>(l[<span class="dv">2</span>])<span class="op">/</span>total)<span class="op">*</span><span class="dv">1000000</span></span>
<span id="cb29-49"><a href="#cb29-49" tabindex="-1"></a>            <span class="cf">if</span> l[<span class="dv">0</span>] <span class="kw">not</span> <span class="kw">in</span> KO:</span>
<span id="cb29-50"><a href="#cb29-50" tabindex="-1"></a>                <span class="cf">continue</span></span>
<span id="cb29-51"><a href="#cb29-51" tabindex="-1"></a>            <span class="cf">if</span> KO[l[<span class="dv">0</span>]] <span class="kw">not</span> <span class="kw">in</span> res:</span>
<span id="cb29-52"><a href="#cb29-52" tabindex="-1"></a>                res[KO[l[<span class="dv">0</span>]]] <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb29-53"><a href="#cb29-53" tabindex="-1"></a>            res[KO[l[<span class="dv">0</span>]]] <span class="op">+=</span> tpm</span>
<span id="cb29-54"><a href="#cb29-54" tabindex="-1"></a>    out_f <span class="op">=</span> <span class="bu">open</span>(out_f, <span class="st">&quot;w&quot;</span>)</span>
<span id="cb29-55"><a href="#cb29-55" tabindex="-1"></a>    <span class="cf">for</span> k, v <span class="kw">in</span> res.items():</span>
<span id="cb29-56"><a href="#cb29-56" tabindex="-1"></a>        out_f.write(k <span class="op">+</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span> <span class="op">+</span> <span class="bu">str</span>(<span class="bu">round</span>(v, <span class="dv">4</span>)) <span class="op">+</span> <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>)</span>
<span id="cb29-57"><a href="#cb29-57" tabindex="-1"></a></span>
<span id="cb29-58"><a href="#cb29-58" tabindex="-1"></a><span class="cf">if</span> <span class="va">__name__</span> <span class="op">==</span> <span class="st">&quot;__main__&quot;</span>:</span>
<span id="cb29-59"><a href="#cb29-59" tabindex="-1"></a>    main(sys.argv[<span class="dv">1</span>], sys.argv[<span class="dv">2</span>], sys.argv[<span class="dv">3</span>])</span>
<span id="cb29-60"><a href="#cb29-60" tabindex="-1"></a></span>
<span id="cb29-61"><a href="#cb29-61" tabindex="-1"></a><span class="co">#### get_profile_from_cluster.py ####</span></span>
<span id="cb29-62"><a href="#cb29-62" tabindex="-1"></a><span class="co">#!/share/data1/software/miniconda3/envs/jinxin/bin/python3</span></span>
<span id="cb29-63"><a href="#cb29-63" tabindex="-1"></a><span class="co"># encoding: utf-8</span></span>
<span id="cb29-64"><a href="#cb29-64" tabindex="-1"></a><span class="co"># author: Jinxin Meng</span></span>
<span id="cb29-65"><a href="#cb29-65" tabindex="-1"></a><span class="co"># created date: 2024-03-12, 11:55:10</span></span>
<span id="cb29-66"><a href="#cb29-66" tabindex="-1"></a><span class="co"># modified date: 2024-03-12, 18:13:31</span></span>
<span id="cb29-67"><a href="#cb29-67" tabindex="-1"></a></span>
<span id="cb29-68"><a href="#cb29-68" tabindex="-1"></a><span class="co">&#39;&#39;&#39;</span></span>
<span id="cb29-69"><a href="#cb29-69" tabindex="-1"></a><span class="co">Jinxin Meng, 20240312</span></span>
<span id="cb29-70"><a href="#cb29-70" tabindex="-1"></a><span class="co">Generate the profile to determinethe presence or absence of KO in each sample.</span></span>
<span id="cb29-71"><a href="#cb29-71" tabindex="-1"></a><span class="co">All genes in each gene cluster were regarded have same name annotation.</span></span>
<span id="cb29-72"><a href="#cb29-72" tabindex="-1"></a><span class="co">Like KO, CAZyme, COGs ... annotation were used in this script.</span></span>
<span id="cb29-73"><a href="#cb29-73" tabindex="-1"></a><span class="co">Example out:</span></span>
<span id="cb29-74"><a href="#cb29-74" tabindex="-1"></a><span class="co">name    s1  s2  s3 ...</span></span>
<span id="cb29-75"><a href="#cb29-75" tabindex="-1"></a><span class="co">K00001  1   4   2       </span></span>
<span id="cb29-76"><a href="#cb29-76" tabindex="-1"></a><span class="co">K00002  2   1   1</span></span>
<span id="cb29-77"><a href="#cb29-77" tabindex="-1"></a><span class="co">K00003  1   2   3</span></span>
<span id="cb29-78"><a href="#cb29-78" tabindex="-1"></a><span class="co">...</span></span>
<span id="cb29-79"><a href="#cb29-79" tabindex="-1"></a><span class="co">Required files:</span></span>
<span id="cb29-80"><a href="#cb29-80" tabindex="-1"></a><span class="co">1. KO annotation with field gene name and KO name with &quot;</span><span class="ch">\t</span><span class="co">&quot; or &quot;\s&quot; delimited</span></span>
<span id="cb29-81"><a href="#cb29-81" tabindex="-1"></a><span class="co">2. gene cluster relationship provided by mmseqs</span></span>
<span id="cb29-82"><a href="#cb29-82" tabindex="-1"></a><span class="co">&#39;&#39;&#39;</span></span>
<span id="cb29-83"><a href="#cb29-83" tabindex="-1"></a></span>
<span id="cb29-84"><a href="#cb29-84" tabindex="-1"></a><span class="im">import</span> sys</span>
<span id="cb29-85"><a href="#cb29-85" tabindex="-1"></a></span>
<span id="cb29-86"><a href="#cb29-86" tabindex="-1"></a><span class="cf">if</span> <span class="bu">len</span>(sys.argv) <span class="op">!=</span> <span class="dv">4</span>:</span>
<span id="cb29-87"><a href="#cb29-87" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;Usage: get_adjacency_for_samples.py [in_f &lt;gene ID|name ID&gt;] [gene_cluster.tsv &lt;rep. gene ID|gene ID&gt;] [out_f]&quot;</span>)</span>
<span id="cb29-88"><a href="#cb29-88" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;  Reqired files:&quot;</span>)</span>
<span id="cb29-89"><a href="#cb29-89" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;  1. in_f with field gene ID and feature ID, and gene ID with reg. exp. </span><span class="ch">\&quot;</span><span class="st">(.*?_\S+)</span><span class="ch">\&quot;</span><span class="st">, such as s1_K141_100_1&quot;</span>)</span>
<span id="cb29-90"><a href="#cb29-90" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;  2. gene_cluster.tsv with field representative gene ID and all gene ID &quot;</span>)</span>
<span id="cb29-91"><a href="#cb29-91" tabindex="-1"></a>    sys.exit()</span>
<span id="cb29-92"><a href="#cb29-92" tabindex="-1"></a></span>
<span id="cb29-93"><a href="#cb29-93" tabindex="-1"></a><span class="kw">def</span> main(in_f, cls_f, out_f):</span>
<span id="cb29-94"><a href="#cb29-94" tabindex="-1"></a>    name <span class="op">=</span> {}</span>
<span id="cb29-95"><a href="#cb29-95" tabindex="-1"></a>    <span class="cf">with</span> <span class="bu">open</span>(in_f, <span class="st">&quot;r&quot;</span>) <span class="im">as</span> f:</span>
<span id="cb29-96"><a href="#cb29-96" tabindex="-1"></a>        <span class="cf">for</span> i <span class="kw">in</span> f:</span>
<span id="cb29-97"><a href="#cb29-97" tabindex="-1"></a>            l <span class="op">=</span> i.strip().split()</span>
<span id="cb29-98"><a href="#cb29-98" tabindex="-1"></a>            name[l[<span class="dv">0</span>]] <span class="op">=</span> l[<span class="dv">1</span>]</span>
<span id="cb29-99"><a href="#cb29-99" tabindex="-1"></a>    </span>
<span id="cb29-100"><a href="#cb29-100" tabindex="-1"></a>    res <span class="op">=</span> {}</span>
<span id="cb29-101"><a href="#cb29-101" tabindex="-1"></a>    sample <span class="op">=</span> []</span>
<span id="cb29-102"><a href="#cb29-102" tabindex="-1"></a>    <span class="cf">with</span> <span class="bu">open</span>(cls_f, <span class="st">&quot;r&quot;</span>) <span class="im">as</span> f:</span>
<span id="cb29-103"><a href="#cb29-103" tabindex="-1"></a>        <span class="cf">for</span> i <span class="kw">in</span> f:</span>
<span id="cb29-104"><a href="#cb29-104" tabindex="-1"></a>            l <span class="op">=</span> i.strip().split()</span>
<span id="cb29-105"><a href="#cb29-105" tabindex="-1"></a>            <span class="cf">if</span> l[<span class="dv">0</span>] <span class="kw">not</span> <span class="kw">in</span> name:</span>
<span id="cb29-106"><a href="#cb29-106" tabindex="-1"></a>                <span class="cf">continue</span></span>
<span id="cb29-107"><a href="#cb29-107" tabindex="-1"></a>            x <span class="op">=</span> l[<span class="dv">1</span>].split(<span class="st">&quot;_&quot;</span>)[<span class="dv">0</span>]</span>
<span id="cb29-108"><a href="#cb29-108" tabindex="-1"></a>            <span class="cf">if</span> name[l[<span class="dv">0</span>]] <span class="kw">not</span> <span class="kw">in</span> res:</span>
<span id="cb29-109"><a href="#cb29-109" tabindex="-1"></a>                res[name[l[<span class="dv">0</span>]]] <span class="op">=</span> {}</span>
<span id="cb29-110"><a href="#cb29-110" tabindex="-1"></a>            <span class="cf">if</span> x <span class="kw">not</span> <span class="kw">in</span> res[name[l[<span class="dv">0</span>]]]:</span>
<span id="cb29-111"><a href="#cb29-111" tabindex="-1"></a>                res[name[l[<span class="dv">0</span>]]][x] <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb29-112"><a href="#cb29-112" tabindex="-1"></a>            <span class="cf">if</span> x <span class="kw">not</span> <span class="kw">in</span> sample:</span>
<span id="cb29-113"><a href="#cb29-113" tabindex="-1"></a>                sample.append(x)</span>
<span id="cb29-114"><a href="#cb29-114" tabindex="-1"></a>            res[name[l[<span class="dv">0</span>]]][x] <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb29-115"><a href="#cb29-115" tabindex="-1"></a>    </span>
<span id="cb29-116"><a href="#cb29-116" tabindex="-1"></a>    sample <span class="op">=</span> <span class="bu">sorted</span>(sample)</span>
<span id="cb29-117"><a href="#cb29-117" tabindex="-1"></a>    out_f <span class="op">=</span> <span class="bu">open</span>(out_f, <span class="st">&quot;w&quot;</span>)</span>
<span id="cb29-118"><a href="#cb29-118" tabindex="-1"></a>    out_f.write(<span class="st">&quot;name</span><span class="ch">\t</span><span class="st">&quot;</span> <span class="op">+</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>.join(sample) <span class="op">+</span> <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>)</span>
<span id="cb29-119"><a href="#cb29-119" tabindex="-1"></a>    <span class="cf">for</span> k, v <span class="kw">in</span> res.items():</span>
<span id="cb29-120"><a href="#cb29-120" tabindex="-1"></a>        l <span class="op">=</span> [k]</span>
<span id="cb29-121"><a href="#cb29-121" tabindex="-1"></a>        <span class="cf">for</span> i <span class="kw">in</span> sample:</span>
<span id="cb29-122"><a href="#cb29-122" tabindex="-1"></a>            <span class="cf">if</span> i <span class="kw">in</span> v:</span>
<span id="cb29-123"><a href="#cb29-123" tabindex="-1"></a>                l.append(<span class="bu">str</span>(v[i]))</span>
<span id="cb29-124"><a href="#cb29-124" tabindex="-1"></a>                <span class="co"># l.append(&quot;1&quot;)</span></span>
<span id="cb29-125"><a href="#cb29-125" tabindex="-1"></a>            <span class="cf">else</span>:</span>
<span id="cb29-126"><a href="#cb29-126" tabindex="-1"></a>                l.append(<span class="st">&quot;0&quot;</span>)</span>
<span id="cb29-127"><a href="#cb29-127" tabindex="-1"></a>        out_f.write(<span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>.join(l) <span class="op">+</span> <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>)</span>
<span id="cb29-128"><a href="#cb29-128" tabindex="-1"></a></span>
<span id="cb29-129"><a href="#cb29-129" tabindex="-1"></a><span class="cf">if</span> <span class="va">__name__</span> <span class="op">==</span> <span class="st">&quot;__main__&quot;</span>:</span>
<span id="cb29-130"><a href="#cb29-130" tabindex="-1"></a>    main(sys.argv[<span class="dv">1</span>], sys.argv[<span class="dv">2</span>], sys.argv[<span class="dv">3</span>])</span>
<span id="cb29-131"><a href="#cb29-131" tabindex="-1"></a></span>
<span id="cb29-132"><a href="#cb29-132" tabindex="-1"></a><span class="co">#### calcu_pnps_by_gene.py ####</span></span>
<span id="cb29-133"><a href="#cb29-133" tabindex="-1"></a><span class="co">#!/share/data1/software/miniconda3/envs/jinxin/bin/python</span></span>
<span id="cb29-134"><a href="#cb29-134" tabindex="-1"></a><span class="co"># Jinxin Meng, mengjx855@163.com</span></span>
<span id="cb29-135"><a href="#cb29-135" tabindex="-1"></a><span class="co"># created date: 2025-08-28, 21:42:59</span></span>
<span id="cb29-136"><a href="#cb29-136" tabindex="-1"></a><span class="co"># modified date: 2025-09-12, 21:45:41</span></span>
<span id="cb29-137"><a href="#cb29-137" tabindex="-1"></a></span>
<span id="cb29-138"><a href="#cb29-138" tabindex="-1"></a><span class="im">import</span> re, sys, os, datetime</span>
<span id="cb29-139"><a href="#cb29-139" tabindex="-1"></a><span class="im">import</span> argparse, math, statistics</span>
<span id="cb29-140"><a href="#cb29-140" tabindex="-1"></a><span class="im">from</span> Bio <span class="im">import</span> SeqIO</span>
<span id="cb29-141"><a href="#cb29-141" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb29-142"><a href="#cb29-142" tabindex="-1"></a></span>
<span id="cb29-143"><a href="#cb29-143" tabindex="-1"></a><span class="kw">def</span> parse_arguments():</span>
<span id="cb29-144"><a href="#cb29-144" tabindex="-1"></a>    <span class="co">&#39;&#39;&#39;</span></span>
<span id="cb29-145"><a href="#cb29-145" tabindex="-1"></a><span class="co">    Calculate pN/pS for metagenomic genes. </span></span>
<span id="cb29-146"><a href="#cb29-146" tabindex="-1"></a><span class="co">    ref 1. Evolutionary and functional implications of hypervariable loci within the skin virome. 2017. PeerJ.</span></span>
<span id="cb29-147"><a href="#cb29-147" tabindex="-1"></a><span class="co">    ref 2. Genomic variation landscape of the human gut microbiome. 2012. Nature.</span></span>
<span id="cb29-148"><a href="#cb29-148" tabindex="-1"></a><span class="co">    Here some recommend codes for calling single nucleotide variant:</span></span>
<span id="cb29-149"><a href="#cb29-149" tabindex="-1"></a><span class="co">    1. bowtie2-build gene.fa gene &amp;&amp; bowtie2 --very-sensitive-local -k 1 -p 36 --no-unal -x gene -1 xx_1.fq.gz -2 xx_2.fq.gz 2&gt;log |</span><span class="ch">\\</span></span>
<span id="cb29-150"><a href="#cb29-150" tabindex="-1"></a><span class="co">            samtools view -@ 10 -bS | samtools sort -@ 10 -o sample.sort.bam -</span></span>
<span id="cb29-151"><a href="#cb29-151" tabindex="-1"></a><span class="co">    2. bcftools mpileup --threads 8 -d 20 -q 20 -Q 30 -f gene.fa xx.sort.bam |</span><span class="ch">\\</span></span>
<span id="cb29-152"><a href="#cb29-152" tabindex="-1"></a><span class="co">            bcftools call --threads 8 -mv --ploidy 1 |</span><span class="ch">\\</span></span>
<span id="cb29-153"><a href="#cb29-153" tabindex="-1"></a><span class="co">            bcftools filter -e &#39;QUAL&lt;30 || DP&lt;10 || (INFO/DP4[0]+INFO/DP4[1]+INFO/DP4[2]+INFO/DP4[3]) &lt; 10 || INDEL=1&#39; --threads 8 |</span><span class="ch">\\</span></span>
<span id="cb29-154"><a href="#cb29-154" tabindex="-1"></a><span class="co">            bcftools view -H -Oz -o xx.vcf.gz</span></span>
<span id="cb29-155"><a href="#cb29-155" tabindex="-1"></a><span class="co">    3. calcu_pnps_by_gene.py gene.fa xx.vcf.gz xx</span></span>
<span id="cb29-156"><a href="#cb29-156" tabindex="-1"></a><span class="co">    &#39;&#39;&#39;</span></span>
<span id="cb29-157"><a href="#cb29-157" tabindex="-1"></a>    parser <span class="op">=</span> argparse.ArgumentParser(description<span class="op">=</span>parse_arguments.__doc__, formatter_class<span class="op">=</span>argparse.RawTextHelpFormatter)</span>
<span id="cb29-158"><a href="#cb29-158" tabindex="-1"></a>    parser.add_argument(<span class="st">&#39;ref_fasta&#39;</span>, <span class="bu">type</span><span class="op">=</span><span class="bu">str</span>, <span class="bu">help</span><span class="op">=</span><span class="st">&#39;reference gene fasta&#39;</span>)</span>
<span id="cb29-159"><a href="#cb29-159" tabindex="-1"></a>    parser.add_argument(<span class="st">&#39;vcf_file&#39;</span>, <span class="bu">type</span><span class="op">=</span><span class="bu">str</span>, <span class="bu">help</span><span class="op">=</span><span class="st">&#39;vcf file from bcftools&#39;</span>)</span>
<span id="cb29-160"><a href="#cb29-160" tabindex="-1"></a>    parser.add_argument(<span class="st">&#39;out_prefix&#39;</span>, <span class="bu">type</span><span class="op">=</span><span class="bu">str</span>, <span class="bu">help</span><span class="op">=</span><span class="st">&#39;name of output prefix&#39;</span>)</span>
<span id="cb29-161"><a href="#cb29-161" tabindex="-1"></a>    parser.add_argument(<span class="st">&#39;--min-depth&#39;</span>, metavar<span class="op">=</span><span class="st">&#39;&#39;</span>, <span class="bu">type</span><span class="op">=</span><span class="bu">int</span>, default<span class="op">=</span><span class="dv">10</span>, <span class="bu">help</span><span class="op">=</span><span class="st">&#39;SNP minimum depth, corresponding to DP4 in vcf [default: 10]&#39;</span>)</span>
<span id="cb29-162"><a href="#cb29-162" tabindex="-1"></a>    parser.add_argument(<span class="st">&#39;--snp-coverage&#39;</span>, metavar<span class="op">=</span><span class="st">&#39;&#39;</span>, <span class="bu">type</span><span class="op">=</span><span class="bu">float</span>, default<span class="op">=</span><span class="fl">0.05</span>, <span class="bu">help</span><span class="op">=</span><span class="st">&#39;SNP minimum coverage threshold [default: 0.05]&#39;</span>)</span>
<span id="cb29-163"><a href="#cb29-163" tabindex="-1"></a>    parser.add_argument(<span class="st">&#39;--min-snps&#39;</span>, metavar<span class="op">=</span><span class="st">&#39;&#39;</span>, <span class="bu">type</span><span class="op">=</span><span class="bu">int</span>, default<span class="op">=</span><span class="dv">3</span>, <span class="bu">help</span><span class="op">=</span><span class="st">&#39;minimum SNP count within a gene [default: 3]&#39;</span>)</span>
<span id="cb29-164"><a href="#cb29-164" tabindex="-1"></a>    parser.add_argument(<span class="st">&#39;--ns-ratio-file&#39;</span>, metavar<span class="op">=</span><span class="st">&#39;&#39;</span>, <span class="bu">type</span><span class="op">=</span><span class="bu">str</span>, default<span class="op">=</span><span class="st">&#39;/share/data1/mjx/scripts/pNpS_data/sn_tot.list1&#39;</span>, <span class="bu">help</span><span class="op">=</span><span class="st">&#39;expected N/S ratio file [default: </span><span class="sc">%(default)s</span><span class="st">]&#39;</span>)</span>
<span id="cb29-165"><a href="#cb29-165" tabindex="-1"></a>    parser.add_argument(<span class="st">&#39;--quiet&#39;</span>, action<span class="op">=</span><span class="st">&#39;store_true&#39;</span>, <span class="bu">help</span><span class="op">=</span><span class="st">&#39;suppress message [default: False]&#39;</span>)</span>
<span id="cb29-166"><a href="#cb29-166" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">len</span>(sys.argv) <span class="op">==</span> <span class="dv">1</span>:</span>
<span id="cb29-167"><a href="#cb29-167" tabindex="-1"></a>        parser.print_help()</span>
<span id="cb29-168"><a href="#cb29-168" tabindex="-1"></a>        parser.exit()      </span>
<span id="cb29-169"><a href="#cb29-169" tabindex="-1"></a>    args <span class="op">=</span> parser.parse_args()</span>
<span id="cb29-170"><a href="#cb29-170" tabindex="-1"></a>    <span class="cf">return</span> args</span>
<span id="cb29-171"><a href="#cb29-171" tabindex="-1"></a></span>
<span id="cb29-172"><a href="#cb29-172" tabindex="-1"></a><span class="kw">def</span> load_ref_sequences(ref_fasta):</span>
<span id="cb29-173"><a href="#cb29-173" tabindex="-1"></a>    sequences <span class="op">=</span> {}</span>
<span id="cb29-174"><a href="#cb29-174" tabindex="-1"></a>    <span class="cf">for</span> record <span class="kw">in</span> SeqIO.parse(ref_fasta, <span class="st">&#39;fasta&#39;</span>):</span>
<span id="cb29-175"><a href="#cb29-175" tabindex="-1"></a>        sequences[record.<span class="bu">id</span>] <span class="op">=</span> <span class="bu">str</span>(record.seq)</span>
<span id="cb29-176"><a href="#cb29-176" tabindex="-1"></a>    <span class="cf">return</span> sequences</span>
<span id="cb29-177"><a href="#cb29-177" tabindex="-1"></a></span>
<span id="cb29-178"><a href="#cb29-178" tabindex="-1"></a><span class="kw">def</span> process_vcf_file(vcf_file, min_depth, snp_coverage):</span>
<span id="cb29-179"><a href="#cb29-179" tabindex="-1"></a>    vcf_data <span class="op">=</span> pd.read_csv(vcf_file, sep<span class="op">=</span><span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>, header<span class="op">=</span><span class="va">None</span>, comment<span class="op">=</span><span class="st">&#39;#&#39;</span>, dtype<span class="op">=</span><span class="bu">object</span>)</span>
<span id="cb29-180"><a href="#cb29-180" tabindex="-1"></a>    <span class="co"># Extract DP4 values (ref-forward, ref-reverse, alt-forward, alt-reverse)</span></span>
<span id="cb29-181"><a href="#cb29-181" tabindex="-1"></a>    dp4_values <span class="op">=</span> vcf_data[<span class="dv">7</span>].<span class="bu">apply</span>(<span class="kw">lambda</span> x: re.search(<span class="vs">r&#39;DP4=(\S+?);&#39;</span>, x).group(<span class="dv">1</span>))</span>
<span id="cb29-182"><a href="#cb29-182" tabindex="-1"></a>    dp4_values <span class="op">=</span> dp4_values.<span class="bu">apply</span>(<span class="kw">lambda</span> x: <span class="bu">list</span>(<span class="bu">map</span>(<span class="bu">int</span>, x.split(<span class="st">&#39;,&#39;</span>))))</span>
<span id="cb29-183"><a href="#cb29-183" tabindex="-1"></a>    total_depths <span class="op">=</span> dp4_values.<span class="bu">apply</span>(<span class="bu">sum</span>)</span>
<span id="cb29-184"><a href="#cb29-184" tabindex="-1"></a></span>
<span id="cb29-185"><a href="#cb29-185" tabindex="-1"></a>    <span class="co"># Filter by minimum depth</span></span>
<span id="cb29-186"><a href="#cb29-186" tabindex="-1"></a>    depth_filter <span class="op">=</span> (total_depths <span class="op">&gt;=</span> min_depth)</span>
<span id="cb29-187"><a href="#cb29-187" tabindex="-1"></a>    vcf_data <span class="op">=</span> vcf_data.loc[depth_filter].reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb29-188"><a href="#cb29-188" tabindex="-1"></a>    dp4_values<span class="op">=</span> dp4_values.loc[depth_filter].reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb29-189"><a href="#cb29-189" tabindex="-1"></a></span>
<span id="cb29-190"><a href="#cb29-190" tabindex="-1"></a>    <span class="co"># 参考等位基因为出现频率最高的等位基因。若变异频率高，则交换</span></span>
<span id="cb29-191"><a href="#cb29-191" tabindex="-1"></a>    swap_needed <span class="op">=</span> [<span class="bu">sum</span>(x[:<span class="dv">2</span>]) <span class="op">&lt;</span> <span class="bu">sum</span>(x[<span class="dv">2</span>:]) <span class="cf">for</span> x <span class="kw">in</span> dp4_values]</span>
<span id="cb29-192"><a href="#cb29-192" tabindex="-1"></a>    vcf_data.loc[swap_needed, [<span class="dv">3</span>, <span class="dv">4</span>]] <span class="op">=</span> vcf_data.loc[swap_needed, [<span class="dv">4</span>, <span class="dv">3</span>]].values</span>
<span id="cb29-193"><a href="#cb29-193" tabindex="-1"></a>    <span class="cf">for</span> idx, needs_swap <span class="kw">in</span> <span class="bu">enumerate</span>(swap_needed):</span>
<span id="cb29-194"><a href="#cb29-194" tabindex="-1"></a>        <span class="cf">if</span> needs_swap:</span>
<span id="cb29-195"><a href="#cb29-195" tabindex="-1"></a>            dp4_values[idx] <span class="op">=</span> [dp4_values[idx][<span class="dv">2</span>], dp4_values[idx][<span class="dv">3</span>], dp4_values[idx][<span class="dv">0</span>], dp4_values[idx][<span class="dv">1</span>]]</span>
<span id="cb29-196"><a href="#cb29-196" tabindex="-1"></a></span>
<span id="cb29-197"><a href="#cb29-197" tabindex="-1"></a>    <span class="co"># 过滤掉频率过低或过高的变异, 变异频率过低，可能是测序错误而非真正的变异；</span></span>
<span id="cb29-198"><a href="#cb29-198" tabindex="-1"></a>    freq_filter <span class="op">=</span> [(<span class="bu">sum</span>(x[<span class="dv">2</span>:])<span class="op">/</span><span class="bu">sum</span>(x) <span class="op">&gt;=</span> snp_coverage) <span class="op">&amp;</span> (<span class="bu">sum</span>(x[<span class="dv">2</span>:])<span class="op">/</span><span class="bu">sum</span>(x) <span class="op">&lt;</span> (<span class="dv">1</span><span class="op">-</span>snp_coverage)) <span class="cf">for</span> x <span class="kw">in</span> dp4_values]</span>
<span id="cb29-199"><a href="#cb29-199" tabindex="-1"></a>    vcf_data <span class="op">=</span> vcf_data[freq_filter].reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb29-200"><a href="#cb29-200" tabindex="-1"></a>    dp4_values <span class="op">=</span> [dp4_values[i] <span class="cf">for</span> i, keep <span class="kw">in</span> <span class="bu">enumerate</span>(freq_filter) <span class="cf">if</span> keep]</span>
<span id="cb29-201"><a href="#cb29-201" tabindex="-1"></a>        </span>
<span id="cb29-202"><a href="#cb29-202" tabindex="-1"></a>    <span class="co"># Add processed DP4 and alt depth to dataframe</span></span>
<span id="cb29-203"><a href="#cb29-203" tabindex="-1"></a>    vcf_data[<span class="st">&#39;dp4_string&#39;</span>] <span class="op">=</span> [<span class="st">&#39;,&#39;</span>.join(<span class="bu">map</span>(<span class="bu">str</span>, x)) <span class="cf">for</span> x <span class="kw">in</span> dp4_values]</span>
<span id="cb29-204"><a href="#cb29-204" tabindex="-1"></a>    vcf_data[<span class="st">&#39;alt_depth&#39;</span>] <span class="op">=</span> [x[<span class="dv">2</span>] <span class="op">+</span> x[<span class="dv">3</span>] <span class="cf">for</span> x <span class="kw">in</span> dp4_values]</span>
<span id="cb29-205"><a href="#cb29-205" tabindex="-1"></a></span>
<span id="cb29-206"><a href="#cb29-206" tabindex="-1"></a>    <span class="cf">return</span> vcf_data</span>
<span id="cb29-207"><a href="#cb29-207" tabindex="-1"></a></span>
<span id="cb29-208"><a href="#cb29-208" tabindex="-1"></a><span class="kw">def</span> load_ns_ratio_data(ns_ratio_file):</span>
<span id="cb29-209"><a href="#cb29-209" tabindex="-1"></a>    <span class="co"># Load N/S ratio data from file</span></span>
<span id="cb29-210"><a href="#cb29-210" tabindex="-1"></a>    ns_data <span class="op">=</span> pd.read_csv(ns_ratio_file, sep<span class="op">=</span><span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>, header<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb29-211"><a href="#cb29-211" tabindex="-1"></a>    ns_data.columns <span class="op">=</span> [<span class="st">&#39;codon_info&#39;</span>, <span class="st">&#39;n_score&#39;</span>, <span class="st">&#39;s_score&#39;</span>] <span class="co"># N and S represent non-synonymous/synonymous substitution probabilities</span></span>
<span id="cb29-212"><a href="#cb29-212" tabindex="-1"></a>    ns_data[<span class="st">&#39;amino_acid&#39;</span>] <span class="op">=</span> ns_data[<span class="st">&#39;codon_info&#39;</span>].<span class="bu">apply</span>(<span class="kw">lambda</span> x: x[<span class="dv">5</span>])</span>
<span id="cb29-213"><a href="#cb29-213" tabindex="-1"></a>    ns_data[<span class="st">&#39;codon&#39;</span>] <span class="op">=</span> ns_data[<span class="st">&#39;codon_info&#39;</span>].<span class="bu">apply</span>(<span class="kw">lambda</span> x: x[:<span class="dv">3</span>])</span>
<span id="cb29-214"><a href="#cb29-214" tabindex="-1"></a>    <span class="cf">return</span> ns_data</span>
<span id="cb29-215"><a href="#cb29-215" tabindex="-1"></a></span>
<span id="cb29-216"><a href="#cb29-216" tabindex="-1"></a><span class="kw">def</span> analyze_snp_effects(vcf_data, ref_sequences, ns_ratio_file, min_snps_per_gene):</span>
<span id="cb29-217"><a href="#cb29-217" tabindex="-1"></a>    <span class="co"># Analyze the effects of SNPs on codons and amino acids</span></span>
<span id="cb29-218"><a href="#cb29-218" tabindex="-1"></a>    snp_info <span class="op">=</span> pd.DataFrame(columns<span class="op">=</span>[<span class="st">&#39;gene_id&#39;</span>, <span class="st">&#39;ref_base&#39;</span>, <span class="st">&#39;alt_base&#39;</span>, <span class="st">&#39;position&#39;</span>, <span class="st">&#39;alt_depth&#39;</span>, <span class="st">&#39;ref_codon&#39;</span>,</span>
<span id="cb29-219"><a href="#cb29-219" tabindex="-1"></a>                                     <span class="st">&#39;alt_codon&#39;</span>, <span class="st">&#39;codon_position&#39;</span>, <span class="st">&#39;codon_start&#39;</span>, <span class="st">&#39;codon_index&#39;</span>, <span class="st">&#39;ref_aa&#39;</span>, <span class="st">&#39;alt_aa&#39;</span>])</span>
<span id="cb29-220"><a href="#cb29-220" tabindex="-1"></a>    ns_data <span class="op">=</span> load_ns_ratio_data(ns_ratio_file)</span>
<span id="cb29-221"><a href="#cb29-221" tabindex="-1"></a>    codon_to_aa_map <span class="op">=</span> ns_data.set_index(<span class="st">&#39;codon&#39;</span>)[<span class="st">&#39;amino_acid&#39;</span>].to_dict()</span>
<span id="cb29-222"><a href="#cb29-222" tabindex="-1"></a></span>
<span id="cb29-223"><a href="#cb29-223" tabindex="-1"></a>    <span class="cf">for</span> idx <span class="kw">in</span> <span class="bu">range</span>(vcf_data.shape[<span class="dv">0</span>]):</span>
<span id="cb29-224"><a href="#cb29-224" tabindex="-1"></a>        gene_id <span class="op">=</span> vcf_data.iloc[idx, <span class="dv">0</span>] <span class="co"># 基因序列名称</span></span>
<span id="cb29-225"><a href="#cb29-225" tabindex="-1"></a>        position <span class="op">=</span> <span class="bu">int</span>(vcf_data.iloc[idx, <span class="dv">1</span>]) <span class="co"># SNP在基因上的位置</span></span>
<span id="cb29-226"><a href="#cb29-226" tabindex="-1"></a>        ref_base <span class="op">=</span> vcf_data.iloc[idx, <span class="dv">3</span>] <span class="co"># 参考等位基因</span></span>
<span id="cb29-227"><a href="#cb29-227" tabindex="-1"></a>        alt_base <span class="op">=</span> vcf_data.iloc[idx, <span class="dv">4</span>] <span class="co"># 突变等位基因</span></span>
<span id="cb29-228"><a href="#cb29-228" tabindex="-1"></a>        </span>
<span id="cb29-229"><a href="#cb29-229" tabindex="-1"></a>        <span class="co"># Calculate codon information</span></span>
<span id="cb29-230"><a href="#cb29-230" tabindex="-1"></a>        codon_position <span class="op">=</span> (position <span class="op">-</span> <span class="dv">1</span>) <span class="op">%</span> <span class="dv">3</span> <span class="co"># SNP在密码子中的位置</span></span>
<span id="cb29-231"><a href="#cb29-231" tabindex="-1"></a>        codon_start <span class="op">=</span> position <span class="op">-</span> codon_position <span class="co"># 密码子的起始位置</span></span>
<span id="cb29-232"><a href="#cb29-232" tabindex="-1"></a>        codon_index <span class="op">=</span> (codon_start <span class="op">-</span> <span class="dv">1</span>) <span class="op">/</span> <span class="dv">3</span> <span class="co"># 密码子的顺序位置</span></span>
<span id="cb29-233"><a href="#cb29-233" tabindex="-1"></a></span>
<span id="cb29-234"><a href="#cb29-234" tabindex="-1"></a>        <span class="co"># 从参考序列中提取完整的密码子</span></span>
<span id="cb29-235"><a href="#cb29-235" tabindex="-1"></a>        <span class="cf">if</span> gene_id  <span class="kw">in</span> ref_sequences:</span>
<span id="cb29-236"><a href="#cb29-236" tabindex="-1"></a>            raw_codon <span class="op">=</span> ref_sequences[gene_id][codon_start <span class="op">-</span> <span class="dv">1</span>: codon_start <span class="op">+</span> <span class="dv">2</span>]</span>
<span id="cb29-237"><a href="#cb29-237" tabindex="-1"></a>            <span class="cf">if</span> <span class="bu">len</span>(raw_codon) <span class="op">&lt;</span> <span class="dv">3</span>:</span>
<span id="cb29-238"><a href="#cb29-238" tabindex="-1"></a>                <span class="cf">continue</span></span>
<span id="cb29-239"><a href="#cb29-239" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb29-240"><a href="#cb29-240" tabindex="-1"></a>            <span class="bu">print</span>(<span class="ss">f&#39;Warning: Missing reference sequence for </span><span class="sc">{</span>gene_id<span class="sc">}</span><span class="ss"> at position </span><span class="sc">{</span>position<span class="sc">}</span><span class="ss">&#39;</span>)</span>
<span id="cb29-241"><a href="#cb29-241" tabindex="-1"></a>        </span>
<span id="cb29-242"><a href="#cb29-242" tabindex="-1"></a>        <span class="co"># 获取参考和变异密码子</span></span>
<span id="cb29-243"><a href="#cb29-243" tabindex="-1"></a>        <span class="cf">if</span> raw_codon[codon_position] <span class="op">==</span> ref_base:</span>
<span id="cb29-244"><a href="#cb29-244" tabindex="-1"></a>            ref_codon <span class="op">=</span> raw_codon</span>
<span id="cb29-245"><a href="#cb29-245" tabindex="-1"></a>            alt_codon <span class="op">=</span> <span class="bu">list</span>(raw_codon)</span>
<span id="cb29-246"><a href="#cb29-246" tabindex="-1"></a>            alt_codon[codon_position] <span class="op">=</span> alt_base</span>
<span id="cb29-247"><a href="#cb29-247" tabindex="-1"></a>            alt_codon <span class="op">=</span> <span class="st">&#39;&#39;</span>.join(alt_codon)</span>
<span id="cb29-248"><a href="#cb29-248" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb29-249"><a href="#cb29-249" tabindex="-1"></a>            alt_codon <span class="op">=</span> raw_codon</span>
<span id="cb29-250"><a href="#cb29-250" tabindex="-1"></a>            ref_codon <span class="op">=</span> <span class="bu">list</span>(raw_codon)</span>
<span id="cb29-251"><a href="#cb29-251" tabindex="-1"></a>            ref_codon[codon_position] <span class="op">=</span> ref_base</span>
<span id="cb29-252"><a href="#cb29-252" tabindex="-1"></a>            ref_codon <span class="op">=</span> <span class="st">&#39;&#39;</span>.join(ref_codon)</span>
<span id="cb29-253"><a href="#cb29-253" tabindex="-1"></a>        </span>
<span id="cb29-254"><a href="#cb29-254" tabindex="-1"></a>        <span class="co"># 获取氨基酸</span></span>
<span id="cb29-255"><a href="#cb29-255" tabindex="-1"></a>        ref_aa <span class="op">=</span> codon_to_aa_map.get(ref_codon, <span class="st">&#39;X&#39;</span>)</span>
<span id="cb29-256"><a href="#cb29-256" tabindex="-1"></a>        alt_aa <span class="op">=</span> codon_to_aa_map.get(alt_codon, <span class="st">&#39;X&#39;</span>)</span>
<span id="cb29-257"><a href="#cb29-257" tabindex="-1"></a></span>
<span id="cb29-258"><a href="#cb29-258" tabindex="-1"></a>        snp_info.loc[idx] <span class="op">=</span> [gene_id, ref_base, alt_base, position, vcf_data.loc[idx, <span class="st">&#39;alt_depth&#39;</span>], ref_codon, alt_codon,</span>
<span id="cb29-259"><a href="#cb29-259" tabindex="-1"></a>                             codon_position <span class="op">+</span> <span class="dv">1</span>, codon_start, codon_index, ref_aa, alt_aa]</span>
<span id="cb29-260"><a href="#cb29-260" tabindex="-1"></a></span>
<span id="cb29-261"><a href="#cb29-261" tabindex="-1"></a>    <span class="co"># # Filter genes with fewer than min_snps_per_gene SNPs</span></span>
<span id="cb29-262"><a href="#cb29-262" tabindex="-1"></a>    gene_snp_counts <span class="op">=</span> snp_info[<span class="st">&#39;gene_id&#39;</span>].value_counts()</span>
<span id="cb29-263"><a href="#cb29-263" tabindex="-1"></a>    genes_to_keep <span class="op">=</span> gene_snp_counts.index[gene_snp_counts <span class="op">&gt;=</span> min_snps_per_gene]</span>
<span id="cb29-264"><a href="#cb29-264" tabindex="-1"></a>    snp_info <span class="op">=</span> snp_info[snp_info[<span class="st">&#39;gene_id&#39;</span>].isin(genes_to_keep)].reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb29-265"><a href="#cb29-265" tabindex="-1"></a>    </span>
<span id="cb29-266"><a href="#cb29-266" tabindex="-1"></a>    <span class="co"># Add synonymous/non-synonymous classification</span></span>
<span id="cb29-267"><a href="#cb29-267" tabindex="-1"></a>    snp_info[<span class="st">&#39;nonsynonymous&#39;</span>] <span class="op">=</span> (snp_info[<span class="st">&#39;ref_aa&#39;</span>] <span class="op">!=</span> snp_info[<span class="st">&#39;alt_aa&#39;</span>]).astype(<span class="bu">int</span>)</span>
<span id="cb29-268"><a href="#cb29-268" tabindex="-1"></a>    snp_info[<span class="st">&#39;synonymous&#39;</span>] <span class="op">=</span> (snp_info[<span class="st">&#39;ref_aa&#39;</span>] <span class="op">==</span> snp_info[<span class="st">&#39;alt_aa&#39;</span>]).astype(<span class="bu">int</span>)</span>
<span id="cb29-269"><a href="#cb29-269" tabindex="-1"></a>    </span>
<span id="cb29-270"><a href="#cb29-270" tabindex="-1"></a>    <span class="cf">return</span> snp_info</span>
<span id="cb29-271"><a href="#cb29-271" tabindex="-1"></a></span>
<span id="cb29-272"><a href="#cb29-272" tabindex="-1"></a><span class="kw">def</span> replace_base_at_position(sequence, position, new_base):</span>
<span id="cb29-273"><a href="#cb29-273" tabindex="-1"></a>    <span class="co"># 在指定索引位置替换字符</span></span>
<span id="cb29-274"><a href="#cb29-274" tabindex="-1"></a>    <span class="cf">if</span> position <span class="op">&lt;</span> <span class="dv">0</span> <span class="kw">or</span> position <span class="op">&gt;=</span> <span class="bu">len</span>(sequence):</span>
<span id="cb29-275"><a href="#cb29-275" tabindex="-1"></a>        <span class="cf">return</span> sequence <span class="co"># 索引超出范围，返回原字符串</span></span>
<span id="cb29-276"><a href="#cb29-276" tabindex="-1"></a>    <span class="cf">return</span> sequence[:position] <span class="op">+</span> new_base <span class="op">+</span> sequence[position<span class="op">+</span><span class="dv">1</span>:]</span>
<span id="cb29-277"><a href="#cb29-277" tabindex="-1"></a></span>
<span id="cb29-278"><a href="#cb29-278" tabindex="-1"></a><span class="kw">def</span> update_ref_sequences(ref_sequences, snp_info):</span>
<span id="cb29-279"><a href="#cb29-279" tabindex="-1"></a>    <span class="co"># Update reference sequences with reference alleles where needed</span></span>
<span id="cb29-280"><a href="#cb29-280" tabindex="-1"></a>    updated_sequences <span class="op">=</span> {}</span>
<span id="cb29-281"><a href="#cb29-281" tabindex="-1"></a></span>
<span id="cb29-282"><a href="#cb29-282" tabindex="-1"></a>    <span class="co"># Initialize with genes that have SNPs</span></span>
<span id="cb29-283"><a href="#cb29-283" tabindex="-1"></a>    <span class="cf">for</span> gene_id <span class="kw">in</span> snp_info[<span class="st">&#39;gene_id&#39;</span>].unique():</span>
<span id="cb29-284"><a href="#cb29-284" tabindex="-1"></a>        updated_sequences[gene_id] <span class="op">=</span> ref_sequences[gene_id]</span>
<span id="cb29-285"><a href="#cb29-285" tabindex="-1"></a></span>
<span id="cb29-286"><a href="#cb29-286" tabindex="-1"></a>    <span class="co"># Update sequences where necessary</span></span>
<span id="cb29-287"><a href="#cb29-287" tabindex="-1"></a>    <span class="cf">for</span> idx <span class="kw">in</span> <span class="bu">range</span>(snp_info.shape[<span class="dv">0</span>]):</span>
<span id="cb29-288"><a href="#cb29-288" tabindex="-1"></a>        snp <span class="op">=</span> snp_info.iloc[idx]</span>
<span id="cb29-289"><a href="#cb29-289" tabindex="-1"></a>        <span class="cf">if</span> updated_sequences[snp[<span class="st">&#39;gene_id&#39;</span>]][snp[<span class="st">&#39;position&#39;</span>]<span class="op">-</span><span class="dv">1</span>] <span class="op">==</span> snp[<span class="st">&#39;alt_base&#39;</span>]:</span>
<span id="cb29-290"><a href="#cb29-290" tabindex="-1"></a>            updated_sequences[snp[<span class="st">&#39;gene_id&#39;</span>]] <span class="op">=</span> replace_base_at_position(updated_sequences[snp[<span class="st">&#39;gene_id&#39;</span>]], snp[<span class="st">&#39;position&#39;</span>]<span class="op">-</span><span class="dv">1</span>, snp[<span class="st">&#39;ref_base&#39;</span>])</span>
<span id="cb29-291"><a href="#cb29-291" tabindex="-1"></a>    </span>
<span id="cb29-292"><a href="#cb29-292" tabindex="-1"></a>    <span class="cf">return</span> updated_sequences</span>
<span id="cb29-293"><a href="#cb29-293" tabindex="-1"></a></span>
<span id="cb29-294"><a href="#cb29-294" tabindex="-1"></a><span class="kw">def</span> calculate_expected_ns_ratios(sequences, ns_ratio_file):</span>
<span id="cb29-295"><a href="#cb29-295" tabindex="-1"></a>    <span class="co"># Calculate expected N/S ratios for each gene</span></span>
<span id="cb29-296"><a href="#cb29-296" tabindex="-1"></a>    ns_data <span class="op">=</span> load_ns_ratio_data(ns_ratio_file).set_index(<span class="st">&#39;codon&#39;</span>)</span>
<span id="cb29-297"><a href="#cb29-297" tabindex="-1"></a>    expected_ratios <span class="op">=</span> {}</span>
<span id="cb29-298"><a href="#cb29-298" tabindex="-1"></a>    <span class="cf">for</span> gene_id, sequence <span class="kw">in</span> sequences.items():</span>
<span id="cb29-299"><a href="#cb29-299" tabindex="-1"></a>        <span class="co"># Generate codons from sequence</span></span>
<span id="cb29-300"><a href="#cb29-300" tabindex="-1"></a>        codons <span class="op">=</span> [sequence[i:i<span class="op">+</span><span class="dv">3</span>] <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>, <span class="bu">len</span>(sequence)<span class="op">-</span><span class="dv">2</span>, <span class="dv">3</span>) <span class="cf">if</span> <span class="bu">len</span>(sequence[i:i<span class="op">+</span><span class="dv">3</span>]) <span class="op">==</span> <span class="dv">3</span>]</span>
<span id="cb29-301"><a href="#cb29-301" tabindex="-1"></a></span>
<span id="cb29-302"><a href="#cb29-302" tabindex="-1"></a>        <span class="co"># Calculate expected N and S scores</span></span>
<span id="cb29-303"><a href="#cb29-303" tabindex="-1"></a>        expected_n_score <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb29-304"><a href="#cb29-304" tabindex="-1"></a>        expected_s_score <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb29-305"><a href="#cb29-305" tabindex="-1"></a>        <span class="cf">for</span> codon <span class="kw">in</span> codons:</span>
<span id="cb29-306"><a href="#cb29-306" tabindex="-1"></a>            <span class="cf">if</span> codon <span class="kw">in</span> ns_data.index:</span>
<span id="cb29-307"><a href="#cb29-307" tabindex="-1"></a>                expected_n_score <span class="op">+=</span> ns_data.loc[codon][<span class="st">&#39;n_score&#39;</span>]</span>
<span id="cb29-308"><a href="#cb29-308" tabindex="-1"></a>                expected_s_score <span class="op">+=</span> ns_data.loc[codon][<span class="st">&#39;s_score&#39;</span>]</span>
<span id="cb29-309"><a href="#cb29-309" tabindex="-1"></a>                                                        </span>
<span id="cb29-310"><a href="#cb29-310" tabindex="-1"></a>        expected_ratios[gene_id] <span class="op">=</span> [<span class="bu">float</span>(expected_n_score), <span class="bu">float</span>(expected_s_score)]</span>
<span id="cb29-311"><a href="#cb29-311" tabindex="-1"></a>    </span>
<span id="cb29-312"><a href="#cb29-312" tabindex="-1"></a>    <span class="cf">return</span> expected_ratios</span>
<span id="cb29-313"><a href="#cb29-313" tabindex="-1"></a></span>
<span id="cb29-314"><a href="#cb29-314" tabindex="-1"></a><span class="kw">def</span> calculate_pnps_v1(n_counts, expected_n, s_counts, expected_s, pseudo_count<span class="op">=</span><span class="fl">1e-20</span>):</span>
<span id="cb29-315"><a href="#cb29-315" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">sum</span>(n_counts) <span class="op">==</span> <span class="dv">0</span> <span class="kw">and</span> <span class="bu">sum</span>(s_counts) <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb29-316"><a href="#cb29-316" tabindex="-1"></a>        <span class="cf">return</span> <span class="st">&#39;NA&#39;</span></span>
<span id="cb29-317"><a href="#cb29-317" tabindex="-1"></a>    pn <span class="op">=</span> (<span class="bu">sum</span>(n_counts) <span class="op">+</span> pseudo_count) <span class="op">/</span> (expected_n <span class="op">+</span> pseudo_count)</span>
<span id="cb29-318"><a href="#cb29-318" tabindex="-1"></a>    ps <span class="op">=</span> (<span class="bu">sum</span>(s_counts) <span class="op">+</span> pseudo_count) <span class="op">/</span> (expected_s <span class="op">+</span> pseudo_count)</span>
<span id="cb29-319"><a href="#cb29-319" tabindex="-1"></a>    <span class="cf">return</span> <span class="bu">round</span>(pn <span class="op">/</span> ps, <span class="dv">8</span>)</span>
<span id="cb29-320"><a href="#cb29-320" tabindex="-1"></a></span>
<span id="cb29-321"><a href="#cb29-321" tabindex="-1"></a><span class="kw">def</span> calculate_pnps_v2(n_counts, expected_n, s_counts, expected_s, alt_depths):</span>
<span id="cb29-322"><a href="#cb29-322" tabindex="-1"></a>    <span class="co"># Calculate pN/pS with smoothing using median depth</span></span>
<span id="cb29-323"><a href="#cb29-323" tabindex="-1"></a>    <span class="co"># Based on: https://peerj.com/articles/2959</span></span>
<span id="cb29-324"><a href="#cb29-324" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">sum</span>(n_counts) <span class="op">==</span> <span class="dv">0</span> <span class="kw">and</span> <span class="bu">sum</span>(s_counts) <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb29-325"><a href="#cb29-325" tabindex="-1"></a>        <span class="cf">return</span> <span class="st">&#39;NA&#39;</span></span>
<span id="cb29-326"><a href="#cb29-326" tabindex="-1"></a>    <span class="co"># Smoothing factor: sqrt(median depth) / 2</span></span>
<span id="cb29-327"><a href="#cb29-327" tabindex="-1"></a>    pseudo_count <span class="op">=</span> math.sqrt(statistics.median(alt_depths)) <span class="op">/</span> <span class="dv">2</span></span>
<span id="cb29-328"><a href="#cb29-328" tabindex="-1"></a>    pn <span class="op">=</span> (<span class="bu">sum</span>(n_counts) <span class="op">+</span> pseudo_count) <span class="op">/</span> expected_n</span>
<span id="cb29-329"><a href="#cb29-329" tabindex="-1"></a>    ps <span class="op">=</span> (<span class="bu">sum</span>(s_counts) <span class="op">+</span> pseudo_count) <span class="op">/</span> expected_s</span>
<span id="cb29-330"><a href="#cb29-330" tabindex="-1"></a>    <span class="cf">return</span> <span class="bu">round</span>(pn <span class="op">/</span> ps, <span class="dv">8</span>)</span>
<span id="cb29-331"><a href="#cb29-331" tabindex="-1"></a></span>
<span id="cb29-332"><a href="#cb29-332" tabindex="-1"></a><span class="kw">def</span> calculate_gene_pnps_ratios(snp_info, expected_ratios):</span>
<span id="cb29-333"><a href="#cb29-333" tabindex="-1"></a>    <span class="co"># Calculate pN/pS ratios for each gene</span></span>
<span id="cb29-334"><a href="#cb29-334" tabindex="-1"></a>    <span class="co"># Group SNPs by gene</span></span>
<span id="cb29-335"><a href="#cb29-335" tabindex="-1"></a>    gene_groups <span class="op">=</span> snp_info.groupby(<span class="st">&#39;gene_id&#39;</span>)</span>
<span id="cb29-336"><a href="#cb29-336" tabindex="-1"></a>    </span>
<span id="cb29-337"><a href="#cb29-337" tabindex="-1"></a>    pnps_results <span class="op">=</span> {}</span>
<span id="cb29-338"><a href="#cb29-338" tabindex="-1"></a>    <span class="cf">for</span> gene_id, gene_snps <span class="kw">in</span> gene_groups:</span>
<span id="cb29-339"><a href="#cb29-339" tabindex="-1"></a>        pnps <span class="op">=</span> calculate_pnps_v2(gene_snps[<span class="st">&#39;nonsynonymous&#39;</span>], expected_ratios[gene_id][<span class="dv">0</span>],</span>
<span id="cb29-340"><a href="#cb29-340" tabindex="-1"></a>                                 gene_snps[<span class="st">&#39;synonymous&#39;</span>], expected_ratios[gene_id][<span class="dv">1</span>], gene_snps[<span class="st">&#39;alt_depth&#39;</span>])</span>
<span id="cb29-341"><a href="#cb29-341" tabindex="-1"></a>        pnps_results[gene_id] <span class="op">=</span> <span class="bu">round</span>(<span class="bu">float</span>(pnps) <span class="cf">if</span> pnps <span class="op">!=</span> <span class="st">&#39;NA&#39;</span> <span class="cf">else</span> <span class="op">-</span><span class="dv">1</span>, <span class="dv">5</span>)</span>
<span id="cb29-342"><a href="#cb29-342" tabindex="-1"></a>    <span class="co"># Convert to DataFrame for output</span></span>
<span id="cb29-343"><a href="#cb29-343" tabindex="-1"></a>    pnps_df <span class="op">=</span> pd.DataFrame(<span class="bu">list</span>(pnps_results.items()), columns<span class="op">=</span>[<span class="st">&#39;gene_id&#39;</span>, <span class="st">&#39;pn_ps_ratio&#39;</span>])</span>
<span id="cb29-344"><a href="#cb29-344" tabindex="-1"></a>    <span class="cf">return</span> pnps_df</span>
<span id="cb29-345"><a href="#cb29-345" tabindex="-1"></a></span>
<span id="cb29-346"><a href="#cb29-346" tabindex="-1"></a><span class="kw">def</span> print_log(message, quiet<span class="op">=</span><span class="va">False</span>):</span>
<span id="cb29-347"><a href="#cb29-347" tabindex="-1"></a>    <span class="cf">if</span> <span class="kw">not</span> quiet:</span>
<span id="cb29-348"><a href="#cb29-348" tabindex="-1"></a>        timestamp <span class="op">=</span> datetime.datetime.now().strftime(<span class="st">&#39;%Y-%m-</span><span class="sc">%d</span><span class="st"> %H:%M:%S&#39;</span>)</span>
<span id="cb29-349"><a href="#cb29-349" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f&quot;[</span><span class="sc">{</span>timestamp<span class="sc">}</span><span class="ss">] </span><span class="sc">{</span>message<span class="sc">}</span><span class="ss">&quot;</span>)</span>
<span id="cb29-350"><a href="#cb29-350" tabindex="-1"></a></span>
<span id="cb29-351"><a href="#cb29-351" tabindex="-1"></a><span class="kw">def</span> main():</span>
<span id="cb29-352"><a href="#cb29-352" tabindex="-1"></a>    <span class="co"># Main function to run the pN/pS analysis pipeline</span></span>
<span id="cb29-353"><a href="#cb29-353" tabindex="-1"></a>    args <span class="op">=</span> parse_arguments()</span>
<span id="cb29-354"><a href="#cb29-354" tabindex="-1"></a>    print_log(<span class="ss">f&#39;Loading reference sequences from </span><span class="sc">{</span>args<span class="sc">.</span>ref_fasta<span class="sc">}</span><span class="ss">...&#39;</span>, args.quiet)</span>
<span id="cb29-355"><a href="#cb29-355" tabindex="-1"></a>    ref_sequences <span class="op">=</span> load_ref_sequences(args.ref_fasta)</span>
<span id="cb29-356"><a href="#cb29-356" tabindex="-1"></a>    print_log(<span class="ss">f&#39;Processing VCF file </span><span class="sc">{</span>args<span class="sc">.</span>vcf_file<span class="sc">}</span><span class="ss">...&#39;</span>, args.quiet)</span>
<span id="cb29-357"><a href="#cb29-357" tabindex="-1"></a>    vcf_data <span class="op">=</span> process_vcf_file(args.vcf_file, args.min_depth, args.snp_coverage)</span>
<span id="cb29-358"><a href="#cb29-358" tabindex="-1"></a>    print_log(<span class="st">&#39;Analyzing SNP effects on codons and amino acids...&#39;</span>, args.quiet)</span>
<span id="cb29-359"><a href="#cb29-359" tabindex="-1"></a>    snp_info <span class="op">=</span> analyze_snp_effects(vcf_data, ref_sequences, args.ns_ratio_file, args.min_snps)</span>
<span id="cb29-360"><a href="#cb29-360" tabindex="-1"></a>    print_log(<span class="st">&#39;Ensuring reference sequences match reference alleles...&#39;</span>, args.quiet)</span>
<span id="cb29-361"><a href="#cb29-361" tabindex="-1"></a>    updated_sequences <span class="op">=</span> update_ref_sequences(ref_sequences, snp_info)</span>
<span id="cb29-362"><a href="#cb29-362" tabindex="-1"></a>    print_log(<span class="st">&quot;Calculating expected N/S ratios for each gene...&quot;</span>, args.quiet)</span>
<span id="cb29-363"><a href="#cb29-363" tabindex="-1"></a>    expected_ratios <span class="op">=</span> calculate_expected_ns_ratios(updated_sequences, args.ns_ratio_file)</span>
<span id="cb29-364"><a href="#cb29-364" tabindex="-1"></a>    print_log(<span class="st">&#39;Calculating pN/pS ratios...&#39;</span>, args.quiet)</span>
<span id="cb29-365"><a href="#cb29-365" tabindex="-1"></a>    pnps_results <span class="op">=</span> calculate_gene_pnps_ratios(snp_info, expected_ratios)</span>
<span id="cb29-366"><a href="#cb29-366" tabindex="-1"></a>    print_log(<span class="ss">f&#39;Saving results&#39;</span>, args.quiet)</span>
<span id="cb29-367"><a href="#cb29-367" tabindex="-1"></a>    out_path <span class="op">=</span> os.path.dirname(args.out_prefix)</span>
<span id="cb29-368"><a href="#cb29-368" tabindex="-1"></a>    <span class="cf">if</span> out_path <span class="op">!=</span> <span class="st">&#39;&#39;</span>:</span>
<span id="cb29-369"><a href="#cb29-369" tabindex="-1"></a>        <span class="cf">if</span> <span class="kw">not</span> os.path.exists(out_path):</span>
<span id="cb29-370"><a href="#cb29-370" tabindex="-1"></a>            os.makedirs(out_path)</span>
<span id="cb29-371"><a href="#cb29-371" tabindex="-1"></a>    out_snp_file <span class="op">=</span> <span class="ss">f&quot;</span><span class="sc">{</span>args<span class="sc">.</span>out_prefix<span class="sc">}</span><span class="ss">.snp_info.tsv&quot;</span></span>
<span id="cb29-372"><a href="#cb29-372" tabindex="-1"></a>    out_pnps_file <span class="op">=</span> <span class="ss">f&quot;</span><span class="sc">{</span>args<span class="sc">.</span>out_prefix<span class="sc">}</span><span class="ss">.pnps_results.tsv&quot;</span></span>
<span id="cb29-373"><a href="#cb29-373" tabindex="-1"></a>    snp_info.to_csv(out_snp_file, sep<span class="op">=</span><span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>, index<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb29-374"><a href="#cb29-374" tabindex="-1"></a>    pnps_results.to_csv(out_pnps_file, sep<span class="op">=</span><span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>, index<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb29-375"><a href="#cb29-375" tabindex="-1"></a>    </span>
<span id="cb29-376"><a href="#cb29-376" tabindex="-1"></a>    total_genes <span class="op">=</span> <span class="bu">len</span>(pnps_results)</span>
<span id="cb29-377"><a href="#cb29-377" tabindex="-1"></a>    valid_pnps <span class="op">=</span> pnps_results[pnps_results[<span class="st">&#39;pn_ps_ratio&#39;</span>] <span class="op">&gt;</span> <span class="dv">0</span>]</span>
<span id="cb29-378"><a href="#cb29-378" tabindex="-1"></a>            </span>
<span id="cb29-379"><a href="#cb29-379" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f&quot;========= Summary Statistics: =========&quot;</span>)</span>
<span id="cb29-380"><a href="#cb29-380" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f&quot;1. Total genes analyzed: </span><span class="sc">{</span>total_genes<span class="sc">}</span><span class="ss">&quot;</span>)</span>
<span id="cb29-381"><a href="#cb29-381" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f&quot;2. Genes with valid pN/pS: </span><span class="sc">{</span><span class="bu">len</span>(valid_pnps)<span class="sc">}</span><span class="ss">&quot;</span>)</span>
<span id="cb29-382"><a href="#cb29-382" tabindex="-1"></a></span>
<span id="cb29-383"><a href="#cb29-383" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">len</span>(valid_pnps) <span class="op">&gt;</span> <span class="dv">0</span>:</span>
<span id="cb29-384"><a href="#cb29-384" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f&quot;3. Mean pN/pS: </span><span class="sc">{</span>valid_pnps[<span class="st">&#39;pn_ps_ratio&#39;</span>]<span class="sc">.</span>mean()<span class="sc">:.4f}</span><span class="ss">&quot;</span>)</span>
<span id="cb29-385"><a href="#cb29-385" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f&quot;4. Median pN/pS: </span><span class="sc">{</span>valid_pnps[<span class="st">&#39;pn_ps_ratio&#39;</span>]<span class="sc">.</span>median()<span class="sc">:.4f}</span><span class="ss">&quot;</span>)</span>
<span id="cb29-386"><a href="#cb29-386" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f&quot;5. Min pN/pS: </span><span class="sc">{</span>valid_pnps[<span class="st">&#39;pn_ps_ratio&#39;</span>]<span class="sc">.</span><span class="bu">min</span>()<span class="sc">:.4f}</span><span class="ss">&quot;</span>)</span>
<span id="cb29-387"><a href="#cb29-387" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f&quot;6. Max pN/pS: </span><span class="sc">{</span>valid_pnps[<span class="st">&#39;pn_ps_ratio&#39;</span>]<span class="sc">.</span><span class="bu">max</span>()<span class="sc">:.4f}</span><span class="ss">&quot;</span>)</span>
<span id="cb29-388"><a href="#cb29-388" tabindex="-1"></a>        <span class="co"># Count genes with pN/pS &gt; 1 (potential positive selection)</span></span>
<span id="cb29-389"><a href="#cb29-389" tabindex="-1"></a>        positive_selection <span class="op">=</span> <span class="bu">len</span>(valid_pnps[valid_pnps[<span class="st">&#39;pn_ps_ratio&#39;</span>] <span class="op">&gt;</span> <span class="dv">1</span>])</span>
<span id="cb29-390"><a href="#cb29-390" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f&quot;7. Genes with pN/pS &gt; 1 (potential positive selection): </span><span class="sc">{</span>positive_selection<span class="sc">}</span><span class="ss"> (</span><span class="sc">{</span>positive_selection<span class="op">/</span><span class="bu">len</span>(valid_pnps)<span class="op">*</span><span class="dv">100</span><span class="sc">:.1f}</span><span class="ss">%)&quot;</span>)</span>
<span id="cb29-391"><a href="#cb29-391" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f&quot;=======================================&quot;</span>)</span>
<span id="cb29-392"><a href="#cb29-392" tabindex="-1"></a></span>
<span id="cb29-393"><a href="#cb29-393" tabindex="-1"></a><span class="cf">if</span> <span class="va">__name__</span> <span class="op">==</span> <span class="st">&#39;__main__&#39;</span>:</span>
<span id="cb29-394"><a href="#cb29-394" tabindex="-1"></a>    main()</span>
<span id="cb29-395"><a href="#cb29-395" tabindex="-1"></a><span class="co">#### tools_mashANI_cluster.py ####</span></span>
<span id="cb29-396"><a href="#cb29-396" tabindex="-1"></a><span class="co">#!/data/software/miniconda3/envs/jinxin/bin/python</span></span>
<span id="cb29-397"><a href="#cb29-397" tabindex="-1"></a><span class="co"># -*- encoding: utf-8 -*-</span></span>
<span id="cb29-398"><a href="#cb29-398" tabindex="-1"></a><span class="co">##########################################################</span></span>
<span id="cb29-399"><a href="#cb29-399" tabindex="-1"></a><span class="co"># Creater       :  夜下凝月</span></span>
<span id="cb29-400"><a href="#cb29-400" tabindex="-1"></a><span class="co"># Created  date :  2024-03-20, 20:57:37</span></span>
<span id="cb29-401"><a href="#cb29-401" tabindex="-1"></a><span class="co"># Modiffed date :  2024-03-20, 20:57:38</span></span>
<span id="cb29-402"><a href="#cb29-402" tabindex="-1"></a><span class="co">##########################################################</span></span>
<span id="cb29-403"><a href="#cb29-403" tabindex="-1"></a></span>
<span id="cb29-404"><a href="#cb29-404" tabindex="-1"></a><span class="im">import</span> argparse</span>
<span id="cb29-405"><a href="#cb29-405" tabindex="-1"></a><span class="im">import</span> logging</span>
<span id="cb29-406"><a href="#cb29-406" tabindex="-1"></a></span>
<span id="cb29-407"><a href="#cb29-407" tabindex="-1"></a><span class="kw">def</span> get_args():</span>
<span id="cb29-408"><a href="#cb29-408" tabindex="-1"></a>    parser <span class="op">=</span> argparse.ArgumentParser(description <span class="op">=</span> __doc__, formatter_class <span class="op">=</span> argparse.RawTextHelpFormatter)</span>
<span id="cb29-409"><a href="#cb29-409" tabindex="-1"></a>    parser.add_argument(<span class="st">&#39;-i&#39;</span>, required<span class="op">=</span><span class="va">True</span>, <span class="bu">help</span><span class="op">=</span><span class="st">&#39;mash result.&#39;</span>)</span>
<span id="cb29-410"><a href="#cb29-410" tabindex="-1"></a>    parser.add_argument(<span class="st">&#39;-o&#39;</span>, required<span class="op">=</span><span class="va">True</span>, <span class="bu">help</span><span class="op">=</span><span class="st">&quot;output file.&quot;</span>)</span>
<span id="cb29-411"><a href="#cb29-411" tabindex="-1"></a>    parser.add_argument(<span class="st">&#39;-ani&#39;</span>, default<span class="op">=</span><span class="fl">0.95</span>, <span class="bu">type</span><span class="op">=</span><span class="bu">float</span>, <span class="bu">help</span><span class="op">=</span><span class="st">&#39;ANI threshold. [0.95]&#39;</span>)</span>
<span id="cb29-412"><a href="#cb29-412" tabindex="-1"></a>    parser.add_argument(<span class="st">&#39;-linkage_method&#39;</span>, default<span class="op">=</span><span class="st">&quot;average&quot;</span>, <span class="bu">type</span><span class="op">=</span><span class="bu">str</span>, <span class="op">\</span></span>
<span id="cb29-413"><a href="#cb29-413" tabindex="-1"></a>            choices<span class="op">=</span>[<span class="st">&#39;average&#39;</span>,<span class="st">&#39;ward&#39;</span>,<span class="st">&#39;complete&#39;</span>,<span class="st">&#39;single&#39;</span>,<span class="st">&#39;median&#39;</span>,<span class="st">&#39;weighted&#39;</span>], <span class="bu">help</span><span class="op">=</span><span class="st">&#39;linkage method. [average]&#39;</span>)</span>
<span id="cb29-414"><a href="#cb29-414" tabindex="-1"></a>    args <span class="op">=</span> parser.parse_args()</span>
<span id="cb29-415"><a href="#cb29-415" tabindex="-1"></a>    <span class="cf">return</span> args</span>
<span id="cb29-416"><a href="#cb29-416" tabindex="-1"></a></span>
<span id="cb29-417"><a href="#cb29-417" tabindex="-1"></a></span>
<span id="cb29-418"><a href="#cb29-418" tabindex="-1"></a><span class="kw">def</span> main(ani_file, outf, ani_cut, linkage_method<span class="op">=</span><span class="st">&#39;average&#39;</span>):</span>
<span id="cb29-419"><a href="#cb29-419" tabindex="-1"></a></span>
<span id="cb29-420"><a href="#cb29-420" tabindex="-1"></a>    <span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb29-421"><a href="#cb29-421" tabindex="-1"></a>    <span class="im">import</span> scipy</span>
<span id="cb29-422"><a href="#cb29-422" tabindex="-1"></a>    <span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb29-423"><a href="#cb29-423" tabindex="-1"></a>    <span class="im">import</span> scipy.cluster</span>
<span id="cb29-424"><a href="#cb29-424" tabindex="-1"></a>    <span class="im">from</span> scipy.spatial <span class="im">import</span> distance <span class="im">as</span> ssd</span>
<span id="cb29-425"><a href="#cb29-425" tabindex="-1"></a></span>
<span id="cb29-426"><a href="#cb29-426" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f&quot;readfile:</span><span class="ch">\t</span><span class="sc">{</span>ani_file<span class="sc">}</span><span class="ss">&quot;</span>)</span>
<span id="cb29-427"><a href="#cb29-427" tabindex="-1"></a>    df <span class="op">=</span> pd.read_csv(ani_file, sep<span class="op">=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, header<span class="op">=</span><span class="va">None</span>, usecols<span class="op">=</span>[<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">2</span>])</span>
<span id="cb29-428"><a href="#cb29-428" tabindex="-1"></a></span>
<span id="cb29-429"><a href="#cb29-429" tabindex="-1"></a>    df.columns<span class="op">=</span>[<span class="st">&#39;ref&#39;</span>,<span class="st">&#39;query&#39;</span>,<span class="st">&#39;dist&#39;</span>]</span>
<span id="cb29-430"><a href="#cb29-430" tabindex="-1"></a></span>
<span id="cb29-431"><a href="#cb29-431" tabindex="-1"></a>    <span class="co">## 距离矩阵</span></span>
<span id="cb29-432"><a href="#cb29-432" tabindex="-1"></a>    dist <span class="op">=</span> df.pivot(index<span class="op">=</span><span class="st">&#39;ref&#39;</span>,columns<span class="op">=</span><span class="st">&#39;query&#39;</span>,values<span class="op">=</span><span class="st">&#39;dist&#39;</span>).fillna(<span class="dv">1</span>)</span>
<span id="cb29-433"><a href="#cb29-433" tabindex="-1"></a>    <span class="co">#dist = df.pivot(&#39;ref&#39;,&#39;query&#39;,&#39;dist&#39;).fillna(1)</span></span>
<span id="cb29-434"><a href="#cb29-434" tabindex="-1"></a>    genomes_names <span class="op">=</span> <span class="bu">list</span>(dist.columns)</span>
<span id="cb29-435"><a href="#cb29-435" tabindex="-1"></a></span>
<span id="cb29-436"><a href="#cb29-436" tabindex="-1"></a>    arr <span class="op">=</span> ssd.squareform(dist)</span>
<span id="cb29-437"><a href="#cb29-437" tabindex="-1"></a></span>
<span id="cb29-438"><a href="#cb29-438" tabindex="-1"></a>    <span class="co">## 聚类</span></span>
<span id="cb29-439"><a href="#cb29-439" tabindex="-1"></a>    linkage_cutoff <span class="op">=</span> <span class="dv">1</span> <span class="op">-</span> ani_cut</span>
<span id="cb29-440"><a href="#cb29-440" tabindex="-1"></a></span>
<span id="cb29-441"><a href="#cb29-441" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f&quot;Running:</span><span class="ch">\t</span><span class="ss">Cluster&quot;</span>)</span>
<span id="cb29-442"><a href="#cb29-442" tabindex="-1"></a>    linkage <span class="op">=</span> scipy.cluster.hierarchy.linkage(arr, method<span class="op">=</span> linkage_method)</span>
<span id="cb29-443"><a href="#cb29-443" tabindex="-1"></a>    fclust <span class="op">=</span> scipy.cluster.hierarchy.fcluster(linkage, linkage_cutoff, criterion<span class="op">=</span><span class="st">&#39;distance&#39;</span>)</span>
<span id="cb29-444"><a href="#cb29-444" tabindex="-1"></a></span>
<span id="cb29-445"><a href="#cb29-445" tabindex="-1"></a>    <span class="co"># 输出文件</span></span>
<span id="cb29-446"><a href="#cb29-446" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f&quot;Output: </span><span class="ch">\t</span><span class="sc">{</span>outf<span class="sc">}</span><span class="ss">&quot;</span>)</span>
<span id="cb29-447"><a href="#cb29-447" tabindex="-1"></a>    pd.DataFrame({<span class="st">&#39;genomes&#39;</span>:genomes_names,<span class="st">&#39;cluster&#39;</span>:fclust}) <span class="op">\</span></span>
<span id="cb29-448"><a href="#cb29-448" tabindex="-1"></a>        .sort_values(by<span class="op">=</span>[<span class="st">&#39;cluster&#39;</span>], ascending<span class="op">=</span>[<span class="va">True</span>]) <span class="op">\</span></span>
<span id="cb29-449"><a href="#cb29-449" tabindex="-1"></a>        .to_csv(<span class="ss">f&quot;</span><span class="sc">{</span>outf<span class="sc">}</span><span class="ss">&quot;</span>, sep<span class="op">=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, index<span class="op">=</span><span class="va">None</span>)</span>
<span id="cb29-450"><a href="#cb29-450" tabindex="-1"></a></span>
<span id="cb29-451"><a href="#cb29-451" tabindex="-1"></a><span class="cf">if</span> <span class="va">__name__</span> <span class="op">==</span> <span class="st">&quot;__main__&quot;</span>:</span>
<span id="cb29-452"><a href="#cb29-452" tabindex="-1"></a>    args <span class="op">=</span> get_args()</span>
<span id="cb29-453"><a href="#cb29-453" tabindex="-1"></a>    main(args.i, args.o, args.ani, args.linkage_method)</span></code></pre></div>
</div>
</section>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
