---
title: "Pru metagenomic abundance in public datasets"
author: "Jinxin Meng"
date: "2024-09-29"
output:
  prettydoc::html_pretty:
    theme: cayman
    highlight: github
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = F, message = F, eval = F)
```

## Pru metagenomic abundance in public datasets
```{bash}
seqkit replace -p '.*' -r 'Bcoccoides.{nr}' GCF_034355335.1_ASM3435533v1_genomic.fna -o Bcoccoides.fa
seqkit replace -p '.*' -r 'Csporogenes.{nr}' GCF_000155085.1_ASM15508v1_genomic.fna -o Csporogenes.fa
seqkit replace -p '.*' -r 'Prussellii.{nr}' GCF_003012055.1_ASM301205v1_genomic.fna -o Prussellii.fa

minimap2 -d ref_genome.mmi ref_genome.fa
cat clean_fq.filelist | parallel -j 8 --colsep="\t" run_minimap2.sh {5} ref_genome.mmi minimap2/{4}
ls *sort.bam | parallel -j 20 --plus samtools coverage -o {/..}.cvg {}
combine_file_zy_folder_allsample.py -D minimap2/ -suffix .cvg -t 1 -n 1 -v 4 -o minimap2.rc
```

## statistics
```{R}
#### SchirmerM_2018 ####
group <- read.delim('data/group/SchirmerM_2018.PRJNA389280.sample_group') %>% 
  select(sample = run, group = group2)

data <- read.delim('data/map/SchirmerM_2018.PRJNA389280.tsv.gz') %>% 
  mutate(rela_ab = rc / libs * 100) %>% 
  left_join(group, by = 'sample') %>% 
  mutate(group = factor(group, c('CT', 'CD', 'UC')))

comparisons <- calcu_diff(select(data, sample, value = rela_ab), group) %>% 
  filter(pval < 0.05) %>% 
  pull(comparison) %>% 
  strsplit(split = '_vs_')

ggviolin(data, 'group', 'rela_ab', color = 'group', legend = 'none', xlab = '',
         size = 1, width = .7, palette = c('#69a7bc','#E69F00','#f08178'),
         ylab = 'Relative abundance(%)', outlier.shape = NA,
         title = 'SchirmerM_2018') +
  geom_boxplot(aes(color = group), width = .2, size = 1, outlier.shape = NA) + 
  geom_jitter(aes(color = group), width = .2) +
  stat_compare_means(comparisons = comparisons, label = 'p.label', 
                     label.y = .135, tip.length = .03, step.increase = .07, 
                     vjust = .6, size = 3) +
  theme(aspect.ratio = 1.2) +
  theme(plot.title = element_text(hjust = .5, face = 'bold'))
ggsave('SchirmerM_2018.boxplot.pdf', width = 4, height = 4)

wilcox_test(data, rela_ab ~ group, p.adjust.method = 'fdr')

#### LloydPriceJ_2019 ####
group <- read.delim('data/group/ZangY_2022.PRJNA398089.sample_group') %>% 
  select(sample = run, group = group2)

data <- read.delim('data/map/ZangY_2022.PRJNA398089.tsv.gz') %>% 
  mutate(rela_ab = rc / libs * 100) %>% 
  left_join(group, by = 'sample') %>% 
  mutate(group = factor(group, c('CT', 'CD', 'UC')))

comparisons <- calcu_diff(select(data, sample, value = rela_ab), group) %>% 
  filter(pval < 0.05) %>% 
  pull(comparison) %>% 
  strsplit(split = '_vs_')

ggviolin(data, 'group', 'rela_ab', color = 'group', legend = 'none', xlab = '',
         size = 1, width = .7, palette = c('#69a7bc','#E69F00','#f08178'),
         ylab = 'Relative abundance(%)', outlier.shape = NA,
         title = 'ZangY_2022') +
  geom_boxplot(aes(color = group), width = .2, size = 1, outlier.shape = NA) + 
  geom_jitter(aes(color = group), width = .2) +
  lims(y = c(NA, .15)) +
  stat_compare_means(comparisons = comparisons, label = 'p.label', 
                     label.y = .14, tip.length = .012, step.increase = .07, 
                     vjust = .6, size = 3) +
  theme(aspect.ratio = 1.2) +
  theme(plot.title = element_text(hjust = .5, face = 'bold'))
ggsave('ZangY_2022.boxplot.pdf', width = 4, height = 4)

wilcox_test(data, rela_ab ~ group, p.adjust.method = 'fdr')

#### HeQ_2017 ####
group <- read.delim('data/group/HeQ_2017.PRJEB15371.sample_group')

data <- read.delim('data/map/HeQ_2017.PRJEB15371.tsv.gz') %>% 
  mutate(rela_ab = rc / libs * 100) %>% 
  left_join(group, by = 'sample') %>% 
  mutate(group = factor(group, c('Control', 'CD')))

comparisons <- calcu_diff(select(data, sample, value = rela_ab), group) %>% 
  filter(pval < 0.05) %>% 
  pull(comparison) %>% 
  strsplit(split = '_vs_')

ggviolin(data, 'group', 'rela_ab', color = 'group', legend = 'none', xlab = '',
         size = 1, width = .7, palette = c('#69a7bc','#E69F00'),
         ylab = 'Relative abundance(%)', outlier.shape = NA,  title = 'HeQ_2017') +
  geom_boxplot(aes(color = group), width = .2, size = 1, outlier.shape = NA) + 
  geom_jitter(aes(color = group), width = .2) +
  lims(y = c(NA, .2)) +
  stat_compare_means(comparisons = comparisons, label = 'p.label', 
                     label.y = .17, tip.length = .01, step.increase = .07, 
                     vjust = .6, size = 3) +
  theme(aspect.ratio = 1.6) +
  theme(plot.title = element_text(hjust = .5, face = 'bold'))
ggsave('HeQ_2017.boxplot.pdf', width = 4, height = 4)

wilcox_test(data, rela_ab ~ group, p.adjust.method = 'fdr')

#### YanQ_2023c ####
group <- read.delim('data/group/YanQ_2023c.PRJEB67456.sample_group') %>% 
  select(sample = run, group)

data <- read.delim('data/map/YanQ_2023c.PRJEB67456.tsv.gz') %>% 
  mutate(rela_ab = rc / libs * 100) %>% 
  left_join(group, by = 'sample') %>% 
  mutate(group = factor(group, c('healthy', 'CD', 'UC')))

comparisons <- calcu_diff(select(data, sample, value = rela_ab), group) %>% 
  filter(pval < 0.05) %>% 
  pull(comparison) %>% 
  strsplit(split = '_vs_')

ggviolin(data, 'group', 'rela_ab', color = 'group', legend = 'none', xlab = '',
         size = 1, width = .7, palette = c('#69a7bc','#E69F00','#f08178'),
         ylab = 'Relative abundance(%)', outlier.shape = NA, title = 'YanQ_2023c') +
  geom_boxplot(aes(color = group), width = .2, size = 1, outlier.shape = NA) + 
  geom_jitter(aes(color = group), width = .2) +
  lims(y = c(NA, .5)) +
  stat_compare_means(comparisons = comparisons, label = 'p.label', method = 'wilcox',
                     label.y = .46, tip.length = .01, step.increase = .07, 
                     vjust = .6, size = 3) +
  theme(aspect.ratio = 1.2) +
  theme(plot.title = element_text(hjust = .5, face = 'bold'))
ggsave('YanQ_2023c.boxplot.pdf', width = 4, height = 4)

wilcox_test(data, rela_ab ~ group, p.adjust.method = 'fdr')
```